<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 11 Introduction to BioConductor | Master Real-world Bioinformatics analysis in R</title>
  <meta name="description" content="This is an R course teaching you skills to do real-world bioinformatics" />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 11 Introduction to BioConductor | Master Real-world Bioinformatics analysis in R" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is an R course teaching you skills to do real-world bioinformatics" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 11 Introduction to BioConductor | Master Real-world Bioinformatics analysis in R" />
  
  <meta name="twitter:description" content="This is an R course teaching you skills to do real-world bioinformatics" />
  

<meta name="author" content="Ming Tommy Tang" />


<meta name="date" content="2024-07-26" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-visualization-with-ggplot2.html"/>
<link rel="next" href="final-project-analyzing-rnaseq-data-from-geo.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<link href="libs/vembedr-0.1.5/css/vembedr.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Master Real-world Bioinformatics analysis in R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#meet-your-instructor"><i class="fa fa-check"></i><b>2.1</b> Meet your Instructor</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#leveraging-our-online-community"><i class="fa fa-check"></i><b>2.2</b> Leveraging our online community</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#install-r-and-r-studio"><i class="fa fa-check"></i><b>2.3</b> Install R and R studio</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#the-role-of-programming-in-biology"><i class="fa fa-check"></i><b>2.4</b> The role of programming in Biology</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="intro.html"><a href="intro.html#what-are-programming-languages"><i class="fa fa-check"></i><b>2.4.1</b> What are programming languages?</a></li>
<li class="chapter" data-level="2.4.2" data-path="intro.html"><a href="intro.html#why-programming-is-important-in-biology"><i class="fa fa-check"></i><b>2.4.2</b> Why programming is important in biology?</a></li>
<li class="chapter" data-level="2.4.3" data-path="intro.html"><a href="intro.html#what-are-the-most-used-programming-languages-within-biology"><i class="fa fa-check"></i><b>2.4.3</b> What are the most used programming languages within biology?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html"><i class="fa fa-check"></i><b>3</b> Introduction to programming</a>
<ul>
<li class="chapter" data-level="3.1" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#what-is-an-algorithm"><i class="fa fa-check"></i><b>3.1</b> What is an algorithm?</a></li>
<li class="chapter" data-level="3.2" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#why-are-algorithms-important-in-programming"><i class="fa fa-check"></i><b>3.2</b> Why are Algorithms Important in Programming?</a></li>
<li class="chapter" data-level="3.3" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#a-real-world-example"><i class="fa fa-check"></i><b>3.3</b> A Real World Example:</a></li>
<li class="chapter" data-level="3.4" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#variables-data-types-and-expressions"><i class="fa fa-check"></i><b>3.4</b> Variables, Data Types, and Expressions</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#variables"><i class="fa fa-check"></i><b>3.4.1</b> Variables</a></li>
<li class="chapter" data-level="3.4.2" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#data-types"><i class="fa fa-check"></i><b>3.4.2</b> Data Types</a></li>
<li class="chapter" data-level="3.4.3" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#expressions"><i class="fa fa-check"></i><b>3.4.3</b> Expressions</a></li>
<li class="chapter" data-level="3.4.4" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#example-finding-the-average-number-of-bases-appearances-in-a-set-of-dna-strings"><i class="fa fa-check"></i><b>3.4.4</b> Example: Finding the average number of bases appearances in a set of DNA strings</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="getting-started-with-r-and-rstudio.html"><a href="getting-started-with-r-and-rstudio.html"><i class="fa fa-check"></i><b>4</b> Getting started with R and RStudio</a>
<ul>
<li class="chapter" data-level="4.1" data-path="getting-started-with-r-and-rstudio.html"><a href="getting-started-with-r-and-rstudio.html#what-is-r-and-why-is-it-used-in-biology"><i class="fa fa-check"></i><b>4.1</b> What is R and why is it used in Biology?</a></li>
<li class="chapter" data-level="4.2" data-path="getting-started-with-r-and-rstudio.html"><a href="getting-started-with-r-and-rstudio.html#what-is-r"><i class="fa fa-check"></i><b>4.2</b> What is R?</a></li>
<li class="chapter" data-level="4.3" data-path="getting-started-with-r-and-rstudio.html"><a href="getting-started-with-r-and-rstudio.html#why-is-r-used-in-biology"><i class="fa fa-check"></i><b>4.3</b> Why is R Used in Biology?</a></li>
<li class="chapter" data-level="4.4" data-path="getting-started-with-r-and-rstudio.html"><a href="getting-started-with-r-and-rstudio.html#introduction-to-r-studio"><i class="fa fa-check"></i><b>4.4</b> Introduction to R-Studio</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html"><i class="fa fa-check"></i><b>5</b> Introduction to R for biologists</a>
<ul>
<li class="chapter" data-level="5.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#basic-data-types-in-r"><i class="fa fa-check"></i><b>5.1</b> Basic Data Types in R</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#before-starting"><i class="fa fa-check"></i><b>5.1.1</b> Before Starting</a></li>
<li class="chapter" data-level="5.1.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#numeric"><i class="fa fa-check"></i><b>5.1.2</b> 1. Numeric</a></li>
<li class="chapter" data-level="5.1.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#character-string"><i class="fa fa-check"></i><b>5.1.3</b> 2. Character (String)</a></li>
<li class="chapter" data-level="5.1.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#integer"><i class="fa fa-check"></i><b>5.1.4</b> 3. Integer</a></li>
<li class="chapter" data-level="5.1.5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#logical-boolean"><i class="fa fa-check"></i><b>5.1.5</b> 4. Logical (Boolean)</a></li>
<li class="chapter" data-level="5.1.6" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#factor"><i class="fa fa-check"></i><b>5.1.6</b> 5. Factor</a></li>
<li class="chapter" data-level="5.1.7" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#date-and-time"><i class="fa fa-check"></i><b>5.1.7</b> 6. Date and Time</a></li>
<li class="chapter" data-level="5.1.8" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#complex"><i class="fa fa-check"></i><b>5.1.8</b> 7. Complex</a></li>
<li class="chapter" data-level="5.1.9" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#missing-values-na"><i class="fa fa-check"></i><b>5.1.9</b> 8. Missing Values (NA)</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#the-key-concept-of-r-vectors"><i class="fa fa-check"></i><b>5.2</b> The key concept of R: Vectors</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#creating-numeric-and-character-vectors"><i class="fa fa-check"></i><b>5.2.1</b> Creating Numeric and Character Vectors</a></li>
<li class="chapter" data-level="5.2.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#saving-vectors-to-variables"><i class="fa fa-check"></i><b>5.2.2</b> Saving Vectors to Variables</a></li>
<li class="chapter" data-level="5.2.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#vectorized-calculations"><i class="fa fa-check"></i><b>5.2.3</b> Vectorized Calculations</a></li>
<li class="chapter" data-level="5.2.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#performing-vectorized-calculations"><i class="fa fa-check"></i><b>5.2.4</b> Performing Vectorized Calculations</a></li>
<li class="chapter" data-level="5.2.5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#operations-with-character-vectors"><i class="fa fa-check"></i><b>5.2.5</b> Operations with character vectors</a></li>
<li class="chapter" data-level="5.2.6" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#logical-vectors"><i class="fa fa-check"></i><b>5.2.6</b> Logical Vectors</a></li>
<li class="chapter" data-level="5.2.7" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#conclusion"><i class="fa fa-check"></i><b>5.2.7</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#subsetting-and-indexing"><i class="fa fa-check"></i><b>5.3</b> Subsetting and Indexing</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#indexing-in-r"><i class="fa fa-check"></i><b>5.3.1</b> Indexing in R</a></li>
<li class="chapter" data-level="5.3.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#slicing-in-r"><i class="fa fa-check"></i><b>5.3.2</b> Slicing in R</a></li>
<li class="chapter" data-level="5.3.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#negative-indexing"><i class="fa fa-check"></i><b>5.3.3</b> Negative Indexing</a></li>
<li class="chapter" data-level="5.3.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#conclusion-1"><i class="fa fa-check"></i><b>5.3.4</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#understanding-and-manipulating-matrices"><i class="fa fa-check"></i><b>5.4</b> Understanding and Manipulating Matrices</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#creating-a-matrix"><i class="fa fa-check"></i><b>5.4.1</b> Creating a Matrix</a></li>
<li class="chapter" data-level="5.4.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#adding-row-and-column-names"><i class="fa fa-check"></i><b>5.4.2</b> Adding Row and Column Names</a></li>
<li class="chapter" data-level="5.4.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#subsetting-a-matrix"><i class="fa fa-check"></i><b>5.4.3</b> Subsetting a Matrix</a></li>
<li class="chapter" data-level="5.4.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#calculations-with-matrices"><i class="fa fa-check"></i><b>5.4.4</b> Calculations with Matrices</a></li>
<li class="chapter" data-level="5.4.5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#normalization"><i class="fa fa-check"></i><b>5.4.5</b> Normalization</a></li>
<li class="chapter" data-level="5.4.6" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#conclusion-2"><i class="fa fa-check"></i><b>5.4.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#essential-functions-in-r"><i class="fa fa-check"></i><b>5.5</b> Essential Functions in R</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#length-function"><i class="fa fa-check"></i><b>5.5.1</b> <code>length</code> Function</a></li>
<li class="chapter" data-level="5.5.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#unique-function"><i class="fa fa-check"></i><b>5.5.2</b> <code>unique</code> Function</a></li>
<li class="chapter" data-level="5.5.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#sort-function"><i class="fa fa-check"></i><b>5.5.3</b> <code>sort</code> Function</a></li>
<li class="chapter" data-level="5.5.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#cor-function"><i class="fa fa-check"></i><b>5.5.4</b> <code>cor</code> Function</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#functions-organizing-your-code-for-reusability"><i class="fa fa-check"></i><b>5.6</b> Functions: Organizing Your Code for Reusability</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#what-is-a-function"><i class="fa fa-check"></i><b>5.6.1</b> What is a Function?</a></li>
<li class="chapter" data-level="5.6.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#creating-a-basic-function-calculating-the-mean"><i class="fa fa-check"></i><b>5.6.2</b> Creating a Basic Function: Calculating the Mean</a></li>
<li class="chapter" data-level="5.6.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#using-the-custom-function"><i class="fa fa-check"></i><b>5.6.3</b> Using the Custom Function</a></li>
<li class="chapter" data-level="5.6.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#returning-a-value-from-the-function"><i class="fa fa-check"></i><b>5.6.4</b> Returning a Value from the Function</a></li>
<li class="chapter" data-level="5.6.5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#optional-omitting-the-return-statement"><i class="fa fa-check"></i><b>5.6.5</b> Optional: Omitting the return Statement</a></li>
<li class="chapter" data-level="5.6.6" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#using-more-than-one-argument"><i class="fa fa-check"></i><b>5.6.6</b> Using more than one argument</a></li>
<li class="chapter" data-level="5.6.7" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#understanding-missing-values-nas-in-r"><i class="fa fa-check"></i><b>5.6.7</b> Understanding Missing Values (NAs) in R</a></li>
<li class="chapter" data-level="5.6.8" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#initial-attempt-a-function-without-handling-nas"><i class="fa fa-check"></i><b>5.6.8</b> Initial Attempt: A Function Without Handling NAs</a></li>
<li class="chapter" data-level="5.6.9" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#adding-the-remove_na-argument"><i class="fa fa-check"></i><b>5.6.9</b> Adding the <code>remove_na</code> Argument</a></li>
<li class="chapter" data-level="5.6.10" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#practical-application"><i class="fa fa-check"></i><b>5.6.10</b> Practical Application</a></li>
<li class="chapter" data-level="5.6.11" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#getting-help-with-functions"><i class="fa fa-check"></i><b>5.6.11</b> Getting help with functions</a></li>
<li class="chapter" data-level="5.6.12" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#a-real-world-example-1"><i class="fa fa-check"></i><b>5.6.12</b> A Real World Example</a></li>
<li class="chapter" data-level="5.6.13" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#conclusion-3"><i class="fa fa-check"></i><b>5.6.13</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#common-mistakes-to-avoid"><i class="fa fa-check"></i><b>5.7</b> Common Mistakes to avoid</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#mixing-data-types"><i class="fa fa-check"></i><b>5.7.1</b> Mixing Data Types</a></li>
<li class="chapter" data-level="5.7.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#forgetting-function-parentheses"><i class="fa fa-check"></i><b>5.7.2</b> Forgetting Function Parentheses</a></li>
<li class="chapter" data-level="5.7.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#overwriting-built-in-functions"><i class="fa fa-check"></i><b>5.7.3</b> Overwriting Built-in Functions</a></li>
<li class="chapter" data-level="5.7.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#misunderstanding-variable-scoping"><i class="fa fa-check"></i><b>5.7.4</b> Misunderstanding Variable Scoping</a></li>
<li class="chapter" data-level="5.7.5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#best-practices-for-variable-scoping-include"><i class="fa fa-check"></i><b>5.7.5</b> Best practices for variable scoping include</a></li>
<li class="chapter" data-level="5.7.6" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#conclusion-4"><i class="fa fa-check"></i><b>5.7.6</b> Conclusion</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html"><i class="fa fa-check"></i><b>6</b> Controlling the flow of our programs</a>
<ul>
<li class="chapter" data-level="6.1" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#boolean-operators"><i class="fa fa-check"></i><b>6.1</b> Boolean Operators</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#comparison-operators"><i class="fa fa-check"></i><b>6.1.1</b> Comparison Operators</a></li>
<li class="chapter" data-level="6.1.2" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#logical-operators"><i class="fa fa-check"></i><b>6.1.2</b> Logical Operators</a></li>
<li class="chapter" data-level="6.1.3" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#conclusion-5"><i class="fa fa-check"></i><b>6.1.3</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#conditional-statements-if-else"><i class="fa fa-check"></i><b>6.2</b> Conditional statements (<code>if</code>, <code>else</code>)</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#the-if-statement"><i class="fa fa-check"></i><b>6.2.1</b> The <code>if</code> Statement</a></li>
<li class="chapter" data-level="6.2.2" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#else-statement"><i class="fa fa-check"></i><b>6.2.2</b> <code>else</code> statement</a></li>
<li class="chapter" data-level="6.2.3" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#else-if"><i class="fa fa-check"></i><b>6.2.3</b> <code>else if</code></a></li>
<li class="chapter" data-level="6.2.4" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#exercise"><i class="fa fa-check"></i><b>6.2.4</b> Exercise</a></li>
<li class="chapter" data-level="6.2.5" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#conclusion-6"><i class="fa fa-check"></i><b>6.2.5</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#loops"><i class="fa fa-check"></i><b>6.3</b> Loops</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#the-while-loop"><i class="fa fa-check"></i><b>6.3.1</b> The <code>while</code> Loop</a></li>
<li class="chapter" data-level="6.3.2" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#while-loops-in-real-life"><i class="fa fa-check"></i><b>6.3.2</b> While Loops in Real Life</a></li>
<li class="chapter" data-level="6.3.3" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#the-for-loop"><i class="fa fa-check"></i><b>6.3.3</b> The <code>for</code> Loop</a></li>
<li class="chapter" data-level="6.3.4" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#conclusion-7"><i class="fa fa-check"></i><b>6.3.4</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#gene-expression-annotation-using-loops-and-control-structures"><i class="fa fa-check"></i><b>6.4</b> Gene Expression Annotation using Loops and Control Structures</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#the-data"><i class="fa fa-check"></i><b>6.4.1</b> The Data</a></li>
<li class="chapter" data-level="6.4.2" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#creating-annotations"><i class="fa fa-check"></i><b>6.4.2</b> Creating Annotations</a></li>
<li class="chapter" data-level="6.4.3" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#the-loop"><i class="fa fa-check"></i><b>6.4.3</b> The Loop</a></li>
<li class="chapter" data-level="6.4.4" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#output"><i class="fa fa-check"></i><b>6.4.4</b> Output</a></li>
<li class="chapter" data-level="6.4.5" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#putting-it-all-together"><i class="fa fa-check"></i><b>6.4.5</b> Putting it all together</a></li>
<li class="chapter" data-level="6.4.6" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#conclusion-8"><i class="fa fa-check"></i><b>6.4.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#lets-solve-a-challenge"><i class="fa fa-check"></i><b>6.5</b> Let’s solve a Challenge</a></li>
<li class="chapter" data-level="6.6" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#solution"><i class="fa fa-check"></i><b>6.6</b> Solution</a>
<ul>
<li class="chapter" data-level="6.6.1" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#using-a-for-loop"><i class="fa fa-check"></i><b>6.6.1</b> using a for loop</a></li>
<li class="chapter" data-level="6.6.2" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#vectorized-solution"><i class="fa fa-check"></i><b>6.6.2</b> vectorized solution</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#section-complete"><i class="fa fa-check"></i><b>6.7</b> Section complete</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html"><i class="fa fa-check"></i><b>7</b> Going more in-depth with R</a>
<ul>
<li class="chapter" data-level="7.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#handling-missing-values"><i class="fa fa-check"></i><b>7.1</b> Handling Missing Values</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#understanding-na-in-r"><i class="fa fa-check"></i><b>7.1.1</b> Understanding NA in R</a></li>
<li class="chapter" data-level="7.1.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#identifying-missing-values"><i class="fa fa-check"></i><b>7.1.2</b> Identifying Missing Values</a></li>
<li class="chapter" data-level="7.1.3" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#remove-missing-values"><i class="fa fa-check"></i><b>7.1.3</b> remove missing values</a></li>
<li class="chapter" data-level="7.1.4" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#real-life-note"><i class="fa fa-check"></i><b>7.1.4</b> real-life note</a></li>
<li class="chapter" data-level="7.1.5" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#conclusion-9"><i class="fa fa-check"></i><b>7.1.5</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#introduction-to-statistical-tests-and-p-values"><i class="fa fa-check"></i><b>7.2</b> Introduction to Statistical Tests and P-Values</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#practical-application-of-statistical-test-with-r"><i class="fa fa-check"></i><b>7.2.1</b> Practical Application of statistical test with R</a></li>
<li class="chapter" data-level="7.2.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#conclusion-10"><i class="fa fa-check"></i><b>7.2.2</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#understanding-r-packages"><i class="fa fa-check"></i><b>7.3</b> Understanding R Packages</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#what-is-an-r-package"><i class="fa fa-check"></i><b>7.3.1</b> What is an R Package?</a></li>
<li class="chapter" data-level="7.3.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#why-use-r-packages"><i class="fa fa-check"></i><b>7.3.2</b> Why Use R Packages?</a></li>
<li class="chapter" data-level="7.3.3" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#how-to-use-a-package"><i class="fa fa-check"></i><b>7.3.3</b> How to use a package?</a></li>
<li class="chapter" data-level="7.3.4" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#embracing-errors"><i class="fa fa-check"></i><b>7.3.4</b> Embracing Errors</a></li>
<li class="chapter" data-level="7.3.5" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#how-to-look-for-package-docs-within-r-studio"><i class="fa fa-check"></i><b>7.3.5</b> How to look for package docs within R studio?</a></li>
<li class="chapter" data-level="7.3.6" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#exploring-package-vignettes"><i class="fa fa-check"></i><b>7.3.6</b> Exploring Package Vignettes</a></li>
<li class="chapter" data-level="7.3.7" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#practical-tip"><i class="fa fa-check"></i><b>7.3.7</b> Practical Tip</a></li>
<li class="chapter" data-level="7.3.8" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#conclusion-11"><i class="fa fa-check"></i><b>7.3.8</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#exploring-functions-in-different-packages-avoiding-collisions-and-accessing-them"><i class="fa fa-check"></i><b>7.4</b> Exploring Functions in Different Packages: Avoiding Collisions and Accessing Them</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#using-double-colon"><i class="fa fa-check"></i><b>7.4.1</b> Using Double Colon (<code>::</code>)</a></li>
<li class="chapter" data-level="7.4.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#conclusion-12"><i class="fa fa-check"></i><b>7.4.2</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#writing-custom-scripts"><i class="fa fa-check"></i><b>7.5</b> Writing Custom Scripts</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#your-working-directory"><i class="fa fa-check"></i><b>7.5.1</b> Your Working Directory</a></li>
<li class="chapter" data-level="7.5.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#writing-functions-in-scripts"><i class="fa fa-check"></i><b>7.5.2</b> Writing Functions in Scripts</a></li>
<li class="chapter" data-level="7.5.3" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#importing-functions-from-scripts"><i class="fa fa-check"></i><b>7.5.3</b> Importing Functions from Scripts</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#data-io"><i class="fa fa-check"></i><b>7.6</b> Data I/O</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#built-in-functions"><i class="fa fa-check"></i><b>7.6.1</b> Built-In Functions</a></li>
<li class="chapter" data-level="7.6.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#real-world-applications"><i class="fa fa-check"></i><b>7.6.2</b> Real-World Applications</a></li>
<li class="chapter" data-level="7.6.3" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#conclusion-13"><i class="fa fa-check"></i><b>7.6.3</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#best-practices-in-modular-programming-and-project-management"><i class="fa fa-check"></i><b>7.7</b> Best Practices in Modular Programming and Project Management</a>
<ul>
<li class="chapter" data-level="7.7.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#introduction-to-modular-programming"><i class="fa fa-check"></i><b>7.7.1</b> Introduction to Modular Programming</a></li>
<li class="chapter" data-level="7.7.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#breaking-down-your-project"><i class="fa fa-check"></i><b>7.7.2</b> Breaking Down Your Project</a></li>
<li class="chapter" data-level="7.7.3" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#conclusion-14"><i class="fa fa-check"></i><b>7.7.3</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#section-complete-1"><i class="fa fa-check"></i><b>7.8</b> Section complete</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html"><i class="fa fa-check"></i><b>8</b> Fundamental Data Structures in R</a>
<ul>
<li class="chapter" data-level="8.1" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#named-vectors"><i class="fa fa-check"></i><b>8.1</b> Named Vectors</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#what-is-a-named-vector"><i class="fa fa-check"></i><b>8.1.1</b> What is a Named Vector?</a></li>
<li class="chapter" data-level="8.1.2" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#using-names-to-subset-a-vector"><i class="fa fa-check"></i><b>8.1.2</b> Using Names to Subset a Vector</a></li>
<li class="chapter" data-level="8.1.3" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#named-vectors-as-dictionary-like-structures"><i class="fa fa-check"></i><b>8.1.3</b> Named Vectors as Dictionary-Like Structures</a></li>
<li class="chapter" data-level="8.1.4" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#real-life-example"><i class="fa fa-check"></i><b>8.1.4</b> Real-life example</a></li>
<li class="chapter" data-level="8.1.5" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#reordering-and-subsetting-with-names"><i class="fa fa-check"></i><b>8.1.5</b> Reordering and Subsetting with Names</a></li>
<li class="chapter" data-level="8.1.6" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#conclusion-15"><i class="fa fa-check"></i><b>8.1.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#lists"><i class="fa fa-check"></i><b>8.2</b> Lists</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#introduction-to-lists"><i class="fa fa-check"></i><b>8.2.1</b> Introduction to Lists</a></li>
<li class="chapter" data-level="8.2.2" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#creating-a-list"><i class="fa fa-check"></i><b>8.2.2</b> Creating a List</a></li>
<li class="chapter" data-level="8.2.3" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#accessing-list-elements"><i class="fa fa-check"></i><b>8.2.3</b> Accessing List Elements</a></li>
<li class="chapter" data-level="8.2.4" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#conclusions"><i class="fa fa-check"></i><b>8.2.4</b> Conclusions</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#dataframes"><i class="fa fa-check"></i><b>8.3</b> Dataframes</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#what-is-a-data-frame"><i class="fa fa-check"></i><b>8.3.1</b> What is a Data Frame?</a></li>
<li class="chapter" data-level="8.3.2" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#creating-a-data-frame"><i class="fa fa-check"></i><b>8.3.2</b> Creating a Data Frame</a></li>
<li class="chapter" data-level="8.3.3" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#accessing-columns"><i class="fa fa-check"></i><b>8.3.3</b> Accessing Columns</a></li>
<li class="chapter" data-level="8.3.4" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#subsetting-a-data-frame"><i class="fa fa-check"></i><b>8.3.4</b> Subsetting a Data Frame</a></li>
<li class="chapter" data-level="8.3.5" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#conclusion-16"><i class="fa fa-check"></i><b>8.3.5</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#how-to-represent-categorical-data-in-r-understanding-factors"><i class="fa fa-check"></i><b>8.4</b> How to Represent Categorical Data in R? Understanding Factors</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#creating-factors"><i class="fa fa-check"></i><b>8.4.1</b> Creating Factors</a></li>
<li class="chapter" data-level="8.4.2" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#changing-levels"><i class="fa fa-check"></i><b>8.4.2</b> Changing levels</a></li>
<li class="chapter" data-level="8.4.3" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#converting-factors"><i class="fa fa-check"></i><b>8.4.3</b> Converting Factors</a></li>
<li class="chapter" data-level="8.4.4" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#customizing-factor-levels"><i class="fa fa-check"></i><b>8.4.4</b> Customizing Factor Levels</a></li>
<li class="chapter" data-level="8.4.5" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#creating-factors-from-character-vectors"><i class="fa fa-check"></i><b>8.4.5</b> Creating Factors from Character Vectors</a></li>
<li class="chapter" data-level="8.4.6" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#subsetting-factors"><i class="fa fa-check"></i><b>8.4.6</b> Subsetting Factors</a></li>
<li class="chapter" data-level="8.4.7" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#removing-unused-levels"><i class="fa fa-check"></i><b>8.4.7</b> Removing Unused Levels</a></li>
<li class="chapter" data-level="8.4.8" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#conclusion-17"><i class="fa fa-check"></i><b>8.4.8</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#section-complete-2"><i class="fa fa-check"></i><b>8.5</b> Section Complete</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html"><i class="fa fa-check"></i><b>9</b> Introduction to the tidyverse ecosystem</a>
<ul>
<li class="chapter" data-level="9.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#what-is-the-tidyverse"><i class="fa fa-check"></i><b>9.1</b> What is the Tidyverse?</a></li>
<li class="chapter" data-level="9.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#key-tidyverse-packages"><i class="fa fa-check"></i><b>9.2</b> Key Tidyverse Packages</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#dplyr"><i class="fa fa-check"></i><b>9.2.1</b> dplyr</a></li>
<li class="chapter" data-level="9.2.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#a-note-on-the-pipe"><i class="fa fa-check"></i><b>9.2.2</b> A note on the pipe</a></li>
<li class="chapter" data-level="9.2.3" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#ggplot2"><i class="fa fa-check"></i><b>9.2.3</b> ggplot2</a></li>
<li class="chapter" data-level="9.2.4" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#tidyr"><i class="fa fa-check"></i><b>9.2.4</b> tidyr</a></li>
<li class="chapter" data-level="9.2.5" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#readr"><i class="fa fa-check"></i><b>9.2.5</b> readr</a></li>
<li class="chapter" data-level="9.2.6" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#tibble"><i class="fa fa-check"></i><b>9.2.6</b> tibble</a></li>
<li class="chapter" data-level="9.2.7" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#stringr"><i class="fa fa-check"></i><b>9.2.7</b> stringr</a></li>
<li class="chapter" data-level="9.2.8" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#forcats"><i class="fa fa-check"></i><b>9.2.8</b> forcats</a></li>
<li class="chapter" data-level="9.2.9" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#real-world-applications-1"><i class="fa fa-check"></i><b>9.2.9</b> Real-World Applications</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#tibble-and-readr---modern-data-structures-in-r"><i class="fa fa-check"></i><b>9.3</b> Tibble and Readr - Modern Data Structures in R</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#tibbles-a-better-way-to-store-data"><i class="fa fa-check"></i><b>9.3.1</b> Tibbles: A Better Way to Store Data</a></li>
<li class="chapter" data-level="9.3.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#rownames-and-tibbles"><i class="fa fa-check"></i><b>9.3.2</b> Rownames and Tibbles</a></li>
<li class="chapter" data-level="9.3.3" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#reading-data-with-readr"><i class="fa fa-check"></i><b>9.3.3</b> Reading Data with readr</a></li>
<li class="chapter" data-level="9.3.4" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#using-tibbles-with-imported-data"><i class="fa fa-check"></i><b>9.3.4</b> Using Tibbles with Imported Data</a></li>
<li class="chapter" data-level="9.3.5" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#note-on-reading-excel-files"><i class="fa fa-check"></i><b>9.3.5</b> Note on reading Excel files</a></li>
<li class="chapter" data-level="9.3.6" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-18"><i class="fa fa-check"></i><b>9.3.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#the-tidy-data-format"><i class="fa fa-check"></i><b>9.4</b> The tidy data format</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#the-importance-of-tidy-data"><i class="fa fa-check"></i><b>9.4.1</b> The Importance of Tidy Data</a></li>
<li class="chapter" data-level="9.4.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#long-format-vs.-wide-format"><i class="fa fa-check"></i><b>9.4.2</b> Long Format vs. Wide Format</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#introducing-dplyr-your-data-wrangling-toolkit"><i class="fa fa-check"></i><b>9.5</b> Introducing dplyr: Your Data Wrangling Toolkit</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#selecting-specific-columns"><i class="fa fa-check"></i><b>9.5.1</b> Selecting Specific Columns</a></li>
<li class="chapter" data-level="9.5.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#adding-new-columns"><i class="fa fa-check"></i><b>9.5.2</b> Adding New Columns</a></li>
<li class="chapter" data-level="9.5.3" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#reordering-columns"><i class="fa fa-check"></i><b>9.5.3</b> Reordering Columns</a></li>
<li class="chapter" data-level="9.5.4" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#filtering-data"><i class="fa fa-check"></i><b>9.5.4</b> Filtering Data</a></li>
<li class="chapter" data-level="9.5.5" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#summarizing-data"><i class="fa fa-check"></i><b>9.5.5</b> Summarizing Data</a></li>
<li class="chapter" data-level="9.5.6" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#sorting-data"><i class="fa fa-check"></i><b>9.5.6</b> Sorting Data</a></li>
<li class="chapter" data-level="9.5.7" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-19"><i class="fa fa-check"></i><b>9.5.7</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#stringr-your-essential-toolkit-to-manipulate-strings"><i class="fa fa-check"></i><b>9.6</b> stringr: your essential toolkit to manipulate strings</a></li>
<li class="chapter" data-level="9.7" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#purrr-ditch-your-for-loops"><i class="fa fa-check"></i><b>9.7</b> purrr: ditch your for loops</a>
<ul>
<li class="chapter" data-level="9.7.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#nesting-data-with-nest"><i class="fa fa-check"></i><b>9.7.1</b> Nesting Data with nest()</a></li>
<li class="chapter" data-level="9.7.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#map-and-its-variants"><i class="fa fa-check"></i><b>9.7.2</b> <code>map()</code> and Its Variants</a></li>
<li class="chapter" data-level="9.7.3" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#note-on-anonymous-functions"><i class="fa fa-check"></i><b>9.7.3</b> Note on Anonymous functions</a></li>
<li class="chapter" data-level="9.7.4" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-20"><i class="fa fa-check"></i><b>9.7.4</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="9.8" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#tidying-metadata-from-geo."><i class="fa fa-check"></i><b>9.8</b> Tidying metadata from GEO.</a>
<ul>
<li class="chapter" data-level="9.8.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#prerequisites"><i class="fa fa-check"></i><b>9.8.1</b> Prerequisites</a></li>
<li class="chapter" data-level="9.8.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#getting-started"><i class="fa fa-check"></i><b>9.8.2</b> Getting Started</a></li>
<li class="chapter" data-level="9.8.3" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#tidying-the-metadata"><i class="fa fa-check"></i><b>9.8.3</b> Tidying the Metadata</a></li>
<li class="chapter" data-level="9.8.4" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#converting-to-a-tidy-data-frame"><i class="fa fa-check"></i><b>9.8.4</b> Converting to a Tidy Data Frame</a></li>
<li class="chapter" data-level="9.8.5" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#bonus-reading-multiple-tsv-files"><i class="fa fa-check"></i><b>9.8.5</b> Bonus: Reading Multiple TSV Files</a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#section-complete-3"><i class="fa fa-check"></i><b>9.9</b> Section Complete</a>
<ul>
<li class="chapter" data-level="9.9.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-21"><i class="fa fa-check"></i><b>9.9.1</b> Conclusion</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html"><i class="fa fa-check"></i><b>10</b> Data visualization with ggplot2</a>
<ul>
<li class="chapter" data-level="10.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#creating-scatter-plots"><i class="fa fa-check"></i><b>10.1</b> Creating Scatter Plots</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#prerequisites-1"><i class="fa fa-check"></i><b>10.1.1</b> Prerequisites</a></li>
<li class="chapter" data-level="10.1.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#loading-and-exploring-the-data"><i class="fa fa-check"></i><b>10.1.2</b> Loading and Exploring the Data</a></li>
<li class="chapter" data-level="10.1.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#filtering-the-data"><i class="fa fa-check"></i><b>10.1.3</b> Filtering the Data</a></li>
<li class="chapter" data-level="10.1.4" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#creating-a-scatter-plot"><i class="fa fa-check"></i><b>10.1.4</b> Creating a Scatter Plot</a></li>
<li class="chapter" data-level="10.1.5" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#calculating-correlation"><i class="fa fa-check"></i><b>10.1.5</b> Calculating Correlation</a></li>
<li class="chapter" data-level="10.1.6" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#hypothesis-testing"><i class="fa fa-check"></i><b>10.1.6</b> Hypothesis Testing</a></li>
<li class="chapter" data-level="10.1.7" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#adding-a-regression-line"><i class="fa fa-check"></i><b>10.1.7</b> Adding a Regression Line</a></li>
<li class="chapter" data-level="10.1.8" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#conclusion-22"><i class="fa fa-check"></i><b>10.1.8</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#understanding-distributions-with-histograms"><i class="fa fa-check"></i><b>10.2</b> Understanding Distributions with Histograms</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#creating-a-basic-histogram"><i class="fa fa-check"></i><b>10.2.1</b> Creating a Basic Histogram</a></li>
<li class="chapter" data-level="10.2.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#customizing-histogram-bins"><i class="fa fa-check"></i><b>10.2.2</b> Customizing Histogram Bins</a></li>
<li class="chapter" data-level="10.2.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#conclusion-23"><i class="fa fa-check"></i><b>10.2.3</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#visualizing-data-distribution-with-boxplots-and-violin-plots"><i class="fa fa-check"></i><b>10.3</b> Visualizing Data Distribution with Boxplots and Violin Plots</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#creating-a-basic-boxplot"><i class="fa fa-check"></i><b>10.3.1</b> Creating a Basic Boxplot</a></li>
<li class="chapter" data-level="10.3.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#rotating-x-axis-labels"><i class="fa fa-check"></i><b>10.3.2</b> Rotating X-axis Labels</a></li>
<li class="chapter" data-level="10.3.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#adding-color-to-boxplots"><i class="fa fa-check"></i><b>10.3.3</b> Adding Color to Boxplots</a></li>
<li class="chapter" data-level="10.3.4" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#customizing-color-palettes"><i class="fa fa-check"></i><b>10.3.4</b> Customizing Color Palettes</a></li>
<li class="chapter" data-level="10.3.5" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#applying-custom-color-palette"><i class="fa fa-check"></i><b>10.3.5</b> Applying Custom Color Palette</a></li>
<li class="chapter" data-level="10.3.6" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#reordering-boxplots"><i class="fa fa-check"></i><b>10.3.6</b> Reordering Boxplots</a></li>
<li class="chapter" data-level="10.3.7" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#using-violin-plots"><i class="fa fa-check"></i><b>10.3.7</b> Using Violin Plots</a></li>
<li class="chapter" data-level="10.3.8" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#conclusion-24"><i class="fa fa-check"></i><b>10.3.8</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#creating-bar-plots-to-visualize-median-epcam-levels-by-cancer-type"><i class="fa fa-check"></i><b>10.4</b> Creating Bar Plots to Visualize Median EPCAM Levels by Cancer Type</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#calculating-median-epcam-levels-by-cancer-type"><i class="fa fa-check"></i><b>10.4.1</b> Calculating Median EPCAM Levels by Cancer Type</a></li>
<li class="chapter" data-level="10.4.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#creating-a-basic-bar-plot"><i class="fa fa-check"></i><b>10.4.2</b> Creating a Basic Bar Plot</a></li>
<li class="chapter" data-level="10.4.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#reordering-the-bars"><i class="fa fa-check"></i><b>10.4.3</b> Reordering the Bars</a></li>
<li class="chapter" data-level="10.4.4" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#reversing-the-order"><i class="fa fa-check"></i><b>10.4.4</b> Reversing the Order</a></li>
<li class="chapter" data-level="10.4.5" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#note-on-factor-levels"><i class="fa fa-check"></i><b>10.4.5</b> Note on factor levels</a></li>
<li class="chapter" data-level="10.4.6" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#conclusion-25"><i class="fa fa-check"></i><b>10.4.6</b> Conclusion</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html"><i class="fa fa-check"></i><b>11</b> Introduction to BioConductor</a>
<ul>
<li class="chapter" data-level="11.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#introduction-to-biocmanager"><i class="fa fa-check"></i><b>11.1</b> Introduction to BiocManager</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#why-use-biocmanager"><i class="fa fa-check"></i><b>11.1.1</b> Why Use BiocManager?</a></li>
<li class="chapter" data-level="11.1.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>11.1.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="11.1.3" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#installing-bioconductor-packages"><i class="fa fa-check"></i><b>11.1.3</b> Installing Bioconductor Packages</a></li>
<li class="chapter" data-level="11.1.4" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#updating-bioconductor-packages"><i class="fa fa-check"></i><b>11.1.4</b> Updating Bioconductor Packages</a></li>
<li class="chapter" data-level="11.1.5" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#checking-for-valid-bioconductor-versions"><i class="fa fa-check"></i><b>11.1.5</b> Checking for Valid Bioconductor Versions</a></li>
<li class="chapter" data-level="11.1.6" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#conclusion-26"><i class="fa fa-check"></i><b>11.1.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#working-with-genomic-coordinates-and-regions-in-r"><i class="fa fa-check"></i><b>11.2</b> Working with Genomic Coordinates and Regions in R</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#introduction-to-genomicranges"><i class="fa fa-check"></i><b>11.2.1</b> Introduction to GenomicRanges</a></li>
<li class="chapter" data-level="11.2.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#basic-operations-with-genomicranges"><i class="fa fa-check"></i><b>11.2.2</b> Basic Operations with GenomicRanges</a></li>
<li class="chapter" data-level="11.2.3" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#calculating-width-of-each-genomic-interval"><i class="fa fa-check"></i><b>11.2.3</b> Calculating width of Each Genomic Interval</a></li>
<li class="chapter" data-level="11.2.4" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#start-and-end-positions"><i class="fa fa-check"></i><b>11.2.4</b> Start and End Positions</a></li>
<li class="chapter" data-level="11.2.5" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#strand-information"><i class="fa fa-check"></i><b>11.2.5</b> Strand Information</a></li>
<li class="chapter" data-level="11.2.6" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#shifting-the-genomic-range"><i class="fa fa-check"></i><b>11.2.6</b> Shifting the Genomic Range</a></li>
<li class="chapter" data-level="11.2.7" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#subsetting"><i class="fa fa-check"></i><b>11.2.7</b> Subsetting</a></li>
<li class="chapter" data-level="11.2.8" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#flanking-regions"><i class="fa fa-check"></i><b>11.2.8</b> Flanking Regions</a></li>
<li class="chapter" data-level="11.2.9" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#resizing"><i class="fa fa-check"></i><b>11.2.9</b> Resizing</a></li>
<li class="chapter" data-level="11.2.10" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#based-and-1-based-coordinate-system"><i class="fa fa-check"></i><b>11.2.10</b> 0 based and 1 based coordinate system</a></li>
<li class="chapter" data-level="11.2.11" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#other-packages"><i class="fa fa-check"></i><b>11.2.11</b> Other packages</a></li>
<li class="chapter" data-level="11.2.12" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#accessing-genome-annotations"><i class="fa fa-check"></i><b>11.2.12</b> Accessing Genome Annotations</a></li>
<li class="chapter" data-level="11.2.13" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#understanding-genomic-biotypes"><i class="fa fa-check"></i><b>11.2.13</b> Understanding Genomic Biotypes</a></li>
<li class="chapter" data-level="11.2.14" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#creating-a-transcript-database-txdb"><i class="fa fa-check"></i><b>11.2.14</b> Creating a Transcript Database (TxDb)</a></li>
<li class="chapter" data-level="11.2.15" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#extracting-exons-introns-and-intergenic-regions"><i class="fa fa-check"></i><b>11.2.15</b> Extracting Exons, Introns, and Intergenic Regions</a></li>
<li class="chapter" data-level="11.2.16" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#merging-all-exons"><i class="fa fa-check"></i><b>11.2.16</b> Merging All Exons</a></li>
<li class="chapter" data-level="11.2.17" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#introns"><i class="fa fa-check"></i><b>11.2.17</b> Introns</a></li>
<li class="chapter" data-level="11.2.18" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#getting-all-genes"><i class="fa fa-check"></i><b>11.2.18</b> Getting All Genes</a></li>
<li class="chapter" data-level="11.2.19" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#promoters"><i class="fa fa-check"></i><b>11.2.19</b> Promoters</a></li>
<li class="chapter" data-level="11.2.20" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#full-genome"><i class="fa fa-check"></i><b>11.2.20</b> Full Genome</a></li>
<li class="chapter" data-level="11.2.21" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#full-transcriptome"><i class="fa fa-check"></i><b>11.2.21</b> Full Transcriptome</a></li>
<li class="chapter" data-level="11.2.22" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#intergenic-regions"><i class="fa fa-check"></i><b>11.2.22</b> Intergenic Regions</a></li>
<li class="chapter" data-level="11.2.23" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#exploring-untranslated-regions-utrs"><i class="fa fa-check"></i><b>11.2.23</b> Exploring Untranslated Regions (UTRs)</a></li>
<li class="chapter" data-level="11.2.24" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#conclusion-27"><i class="fa fa-check"></i><b>11.2.24</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#exploring-cpg-islands-and-shores-in-genomic-data"><i class="fa fa-check"></i><b>11.3</b> Exploring CpG Islands and Shores in Genomic Data</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#introduction-to-cpg-islands"><i class="fa fa-check"></i><b>11.3.1</b> Introduction to CpG Islands</a></li>
<li class="chapter" data-level="11.3.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#fetching-cpg-island-coordinates"><i class="fa fa-check"></i><b>11.3.2</b> Fetching CpG Island Coordinates</a></li>
<li class="chapter" data-level="11.3.3" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#defining-cpg-shores"><i class="fa fa-check"></i><b>11.3.3</b> Defining CpG Shores</a></li>
<li class="chapter" data-level="11.3.4" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#combining-and-analyzing-cpg-shores"><i class="fa fa-check"></i><b>11.3.4</b> Combining and Analyzing CpG Shores</a></li>
<li class="chapter" data-level="11.3.5" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#conclusion-28"><i class="fa fa-check"></i><b>11.3.5</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#real-world-applications-chip-seq"><i class="fa fa-check"></i><b>11.4</b> Real-World Applications: ChIP-seq</a>
<ul>
<li class="chapter" data-level="11.4.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#read-in-peak-file"><i class="fa fa-check"></i><b>11.4.1</b> read in peak file</a></li>
<li class="chapter" data-level="11.4.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#identify-promoters-that-overlap-with-the-peaks"><i class="fa fa-check"></i><b>11.4.2</b> Identify promoters that overlap with the peaks</a></li>
<li class="chapter" data-level="11.4.3" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#use-the-chipseeker-package"><i class="fa fa-check"></i><b>11.4.3</b> use the ChIPseeker package</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#analyzing-and-visualizing-genomic-data"><i class="fa fa-check"></i><b>11.5</b> Analyzing and Visualizing Genomic Data</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#deseq2"><i class="fa fa-check"></i><b>11.5.1</b> DESeq2</a></li>
<li class="chapter" data-level="11.5.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#geoquery"><i class="fa fa-check"></i><b>11.5.2</b> GEOquery</a></li>
<li class="chapter" data-level="11.5.3" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#converting-gene-ids"><i class="fa fa-check"></i><b>11.5.3</b> Converting Gene IDs</a></li>
<li class="chapter" data-level="11.5.4" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#complexheatmap"><i class="fa fa-check"></i><b>11.5.4</b> ComplexHeatmap</a></li>
<li class="chapter" data-level="11.5.5" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#conclusion-29"><i class="fa fa-check"></i><b>11.5.5</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#real-world-example---tcga-analysis"><i class="fa fa-check"></i><b>11.6</b> Real-World Example - TCGA Analysis</a>
<ul>
<li class="chapter" data-level="11.6.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#introduction-to-tcga"><i class="fa fa-check"></i><b>11.6.1</b> Introduction to TCGA</a></li>
<li class="chapter" data-level="11.6.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#why-analyze-tcga-data"><i class="fa fa-check"></i><b>11.6.2</b> Why Analyze TCGA Data?</a></li>
<li class="chapter" data-level="11.6.3" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#getting-started-1"><i class="fa fa-check"></i><b>11.6.3</b> Getting Started</a></li>
<li class="chapter" data-level="11.6.4" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-1-install-and-load-required-packages"><i class="fa fa-check"></i><b>11.6.4</b> Step 1: Install and Load Required Packages</a></li>
<li class="chapter" data-level="11.6.5" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-2-retrieve-tcga-project-information"><i class="fa fa-check"></i><b>11.6.5</b> Step 2: Retrieve TCGA Project Information</a></li>
<li class="chapter" data-level="11.6.6" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-3-create-a-rangedsummarizedexperiment-object"><i class="fa fa-check"></i><b>11.6.6</b> Step 3: Create a RangedSummarizedExperiment Object</a></li>
<li class="chapter" data-level="11.6.7" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-4-explore-tcga-data"><i class="fa fa-check"></i><b>11.6.7</b> Step 4: Explore TCGA Data</a></li>
<li class="chapter" data-level="11.6.8" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-5-data-transformation---converting-raw-counts-to-tpm"><i class="fa fa-check"></i><b>11.6.8</b> Step 5: Data Transformation - Converting Raw Counts to TPM</a></li>
<li class="chapter" data-level="11.6.9" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-6-combine-data-and-metadata"><i class="fa fa-check"></i><b>11.6.9</b> Step 6: Combine Data and Metadata</a></li>
<li class="chapter" data-level="11.6.10" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-7-renaming-metadata-columns"><i class="fa fa-check"></i><b>11.6.10</b> Step 7: Renaming Metadata Columns</a></li>
<li class="chapter" data-level="11.6.11" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-8-merging-into-a-single-dataframe"><i class="fa fa-check"></i><b>11.6.11</b> Step 8: Merging into a single DataFrame</a></li>
<li class="chapter" data-level="11.6.12" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-9-create-a-gene-expression-heatmap"><i class="fa fa-check"></i><b>11.6.12</b> Step 9: Create a Gene Expression Heatmap</a></li>
<li class="chapter" data-level="11.6.13" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-10-sanity-check-and-scaling-of-gene-expression"><i class="fa fa-check"></i><b>11.6.13</b> Step 10: Sanity Check and Scaling of Gene Expression</a></li>
<li class="chapter" data-level="11.6.14" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#conclusion-30"><i class="fa fa-check"></i><b>11.6.14</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="11.7" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#section-completed"><i class="fa fa-check"></i><b>11.7</b> Section completed</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html"><i class="fa fa-check"></i><b>12</b> Final Project: Analyzing RNAseq Data from GEO</a>
<ul>
<li class="chapter" data-level="12.1" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#final-project-overview"><i class="fa fa-check"></i><b>12.1</b> Final Project Overview</a></li>
<li class="chapter" data-level="12.2" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#how-to-pre-process-rnaseq-data"><i class="fa fa-check"></i><b>12.2</b> How to pre-process RNAseq data</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#rnaseq-pre-processing-steps"><i class="fa fa-check"></i><b>12.2.1</b> RNAseq pre-processing steps</a></li>
<li class="chapter" data-level="12.2.2" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#how-to-use-salmon-to-preprocess-geo-fastq-to-counts"><i class="fa fa-check"></i><b>12.2.2</b> How to use salmon to preprocess GEO fastq to counts</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#download-and-subset-count-matrix"><i class="fa fa-check"></i><b>12.3</b> Download and subset Count Matrix</a>
<ul>
<li class="chapter" data-level="12.3.1" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#downloading-the-count-matrix"><i class="fa fa-check"></i><b>12.3.1</b> Downloading the Count Matrix</a></li>
<li class="chapter" data-level="12.3.2" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#reading-the-count-matrix-in-r"><i class="fa fa-check"></i><b>12.3.2</b> Reading the Count Matrix in R</a></li>
<li class="chapter" data-level="12.3.3" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#examining-the-data"><i class="fa fa-check"></i><b>12.3.3</b> Examining the Data</a></li>
<li class="chapter" data-level="12.3.4" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#subsetting-the-data"><i class="fa fa-check"></i><b>12.3.4</b> Subsetting the Data</a></li>
<li class="chapter" data-level="12.3.5" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#converting-to-a-matrix"><i class="fa fa-check"></i><b>12.3.5</b> Converting to a Matrix</a></li>
<li class="chapter" data-level="12.3.6" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#adding-row-names"><i class="fa fa-check"></i><b>12.3.6</b> Adding Row Names</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#calculate-the-total-exon-length-per-gene"><i class="fa fa-check"></i><b>12.4</b> Calculate the total exon length per gene</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#why-normalize-gene-counts-to-tpm"><i class="fa fa-check"></i><b>12.4.1</b> Why Normalize Gene Counts to TPM?</a></li>
<li class="chapter" data-level="12.4.2" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#creating-a-function-for-normalization"><i class="fa fa-check"></i><b>12.4.2</b> Creating a function for normalization</a></li>
<li class="chapter" data-level="12.4.3" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#join-the-exon-length-table"><i class="fa fa-check"></i><b>12.4.3</b> join the exon length table</a></li>
<li class="chapter" data-level="12.4.4" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#what-genes-are-not-in-the-mapping-table"><i class="fa fa-check"></i><b>12.4.4</b> what genes are not in the mapping table?</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#normalizing-raw-counts-to-transcripts-per-million-tpm"><i class="fa fa-check"></i><b>12.5</b> Normalizing Raw Counts to Transcripts per Million (TPM)</a>
<ul>
<li class="chapter" data-level="12.5.1" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#the-count2tpm-function"><i class="fa fa-check"></i><b>12.5.1</b> The <code>count2tpm</code> Function:</a></li>
<li class="chapter" data-level="12.5.2" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#applying-the-function"><i class="fa fa-check"></i><b>12.5.2</b> Applying the Function</a></li>
<li class="chapter" data-level="12.5.3" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#conclusions-1"><i class="fa fa-check"></i><b>12.5.3</b> Conclusions</a></li>
</ul></li>
<li class="chapter" data-level="12.6" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#analyzing-gene-expression-data-using-t-tests"><i class="fa fa-check"></i><b>12.6</b> Analyzing Gene Expression Data Using t-Tests</a>
<ul>
<li class="chapter" data-level="12.6.1" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#hypothesis-testing-1"><i class="fa fa-check"></i><b>12.6.1</b> Hypothesis Testing</a></li>
<li class="chapter" data-level="12.6.2" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#analyzing-specific-genes"><i class="fa fa-check"></i><b>12.6.2</b> Analyzing Specific Genes</a></li>
<li class="chapter" data-level="12.6.3" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#analyzing-all-genes"><i class="fa fa-check"></i><b>12.6.3</b> Analyzing All Genes</a></li>
<li class="chapter" data-level="12.6.4" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#conclusion-31"><i class="fa fa-check"></i><b>12.6.4</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="12.7" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#analyzing-gene-expression-data-with-ggplot2"><i class="fa fa-check"></i><b>12.7</b> Analyzing Gene Expression Data with ggplot2</a>
<ul>
<li class="chapter" data-level="12.7.1" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#import-libraries-and-load-data"><i class="fa fa-check"></i><b>12.7.1</b> Import Libraries and Load Data</a></li>
<li class="chapter" data-level="12.7.2" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#visualizing-p-value-distribution"><i class="fa fa-check"></i><b>12.7.2</b> Visualizing P-Value Distribution</a></li>
<li class="chapter" data-level="12.7.3" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#identifying-differentially-expressed-genes"><i class="fa fa-check"></i><b>12.7.3</b> Identifying Differentially Expressed Genes</a></li>
<li class="chapter" data-level="12.7.4" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#selecting-differentially-expressed-genes"><i class="fa fa-check"></i><b>12.7.4</b> Selecting Differentially Expressed Genes</a></li>
<li class="chapter" data-level="12.7.5" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#boxplot-to-visualize-gene-expression-changes-between-the-two-conditions."><i class="fa fa-check"></i><b>12.7.5</b> boxplot to visualize gene expression changes between the two conditions.</a></li>
<li class="chapter" data-level="12.7.6" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#creating-the-boxplot"><i class="fa fa-check"></i><b>12.7.6</b> Creating the Boxplot</a></li>
<li class="chapter" data-level="12.7.7" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#visualizing-raw-expression-values"><i class="fa fa-check"></i><b>12.7.7</b> Visualizing Raw Expression Values</a></li>
<li class="chapter" data-level="12.7.8" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#conclusion-32"><i class="fa fa-check"></i><b>12.7.8</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="12.8" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#correcting-for-multiple-comparisons-in-statistical-analysis"><i class="fa fa-check"></i><b>12.8</b> Correcting for Multiple Comparisons in Statistical Analysis</a>
<ul>
<li class="chapter" data-level="12.8.1" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#why-correct-for-multiple-comparisons"><i class="fa fa-check"></i><b>12.8.1</b> Why Correct for Multiple Comparisons?</a></li>
<li class="chapter" data-level="12.8.2" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#the-multiple-comparisons-example"><i class="fa fa-check"></i><b>12.8.2</b> The Multiple Comparisons Example</a></li>
<li class="chapter" data-level="12.8.3" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#correcting-for-multiple-comparisons-false-discovery-rate-fdr"><i class="fa fa-check"></i><b>12.8.3</b> Correcting for Multiple Comparisons: False Discovery Rate (FDR)</a></li>
<li class="chapter" data-level="12.8.4" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#conclusion-33"><i class="fa fa-check"></i><b>12.8.4</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="12.9" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#analyzing-differential-gene-expression-with-deseq2"><i class="fa fa-check"></i><b>12.9</b> Analyzing Differential Gene Expression with DESeq2</a>
<ul>
<li class="chapter" data-level="12.9.1" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#create-a-sample-sheet"><i class="fa fa-check"></i><b>12.9.1</b> Create a Sample Sheet</a></li>
<li class="chapter" data-level="12.9.2" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#create-a-deseq2-object"><i class="fa fa-check"></i><b>12.9.2</b> Create a DESeq2 Object</a></li>
<li class="chapter" data-level="12.9.3" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#get-differential-results"><i class="fa fa-check"></i><b>12.9.3</b> Get Differential Results</a></li>
<li class="chapter" data-level="12.9.4" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#explore-the-deseq2-results-and-create-visualizations."><i class="fa fa-check"></i><b>12.9.4</b> Explore the DESeq2 results and create visualizations.</a></li>
<li class="chapter" data-level="12.9.5" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#examining-the-top-differentially-expressed-genes"><i class="fa fa-check"></i><b>12.9.5</b> Examining the Top Differentially Expressed Genes</a></li>
<li class="chapter" data-level="12.9.6" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#labeling-genes-in-the-volcano-plot"><i class="fa fa-check"></i><b>12.9.6</b> Labeling Genes in the Volcano Plot</a></li>
<li class="chapter" data-level="12.9.7" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#coloring-points-in-the-volcano-plot"><i class="fa fa-check"></i><b>12.9.7</b> Coloring Points in the Volcano Plot</a></li>
<li class="chapter" data-level="12.9.8" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#conclusion-34"><i class="fa fa-check"></i><b>12.9.8</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="12.10" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#principal-component-analysis-pca-using-deseq2"><i class="fa fa-check"></i><b>12.10</b> Principal Component Analysis (PCA) using DESeq2</a>
<ul>
<li class="chapter" data-level="12.10.1" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#what-is-principal-component-analysis-pca"><i class="fa fa-check"></i><b>12.10.1</b> What is Principal Component Analysis (PCA)?</a></li>
<li class="chapter" data-level="12.10.2" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#why-use-pca"><i class="fa fa-check"></i><b>12.10.2</b> Why use PCA?</a></li>
<li class="chapter" data-level="12.10.3" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#using-deseq2-for-pca-analysis"><i class="fa fa-check"></i><b>12.10.3</b> Using DESeq2 for PCA Analysis</a></li>
<li class="chapter" data-level="12.10.4" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#plotting-pca-by-ourselves"><i class="fa fa-check"></i><b>12.10.4</b> Plotting PCA by ourselves</a></li>
<li class="chapter" data-level="12.10.5" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#create-a-dataframe-for-visualization"><i class="fa fa-check"></i><b>12.10.5</b> Create a DataFrame for Visualization</a></li>
<li class="chapter" data-level="12.10.6" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#plot-the-pca"><i class="fa fa-check"></i><b>12.10.6</b> Plot the PCA</a></li>
<li class="chapter" data-level="12.10.7" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#comparison-with-deseq2s-plotpca"><i class="fa fa-check"></i><b>12.10.7</b> Comparison with DESeq2’s plotPCA</a></li>
<li class="chapter" data-level="12.10.8" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#conclusion-35"><i class="fa fa-check"></i><b>12.10.8</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="12.11" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#creating-a-perfect-heatmap"><i class="fa fa-check"></i><b>12.11</b> Creating a Perfect Heatmap</a>
<ul>
<li class="chapter" data-level="12.11.1" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#selecting-significant-genes"><i class="fa fa-check"></i><b>12.11.1</b> Selecting Significant Genes</a></li>
<li class="chapter" data-level="12.11.2" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#creating-the-heatmap"><i class="fa fa-check"></i><b>12.11.2</b> Creating the Heatmap</a></li>
<li class="chapter" data-level="12.11.3" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#why-scaling-is-important"><i class="fa fa-check"></i><b>12.11.3</b> Why Scaling is Important?</a></li>
<li class="chapter" data-level="12.11.4" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#adding-annotations"><i class="fa fa-check"></i><b>12.11.4</b> Adding Annotations</a></li>
<li class="chapter" data-level="12.11.5" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#the-impact-of-scaling"><i class="fa fa-check"></i><b>12.11.5</b> The Impact of Scaling</a></li>
<li class="chapter" data-level="12.11.6" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#conclusion-36"><i class="fa fa-check"></i><b>12.11.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="12.12" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#pathway-analysis-using-over-representation-and-gene-set-enrichment-analysis"><i class="fa fa-check"></i><b>12.12</b> Pathway Analysis Using Over-Representation and Gene Set Enrichment Analysis</a>
<ul>
<li class="chapter" data-level="12.12.1" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#understanding-pathway-analysis"><i class="fa fa-check"></i><b>12.12.1</b> Understanding Pathway Analysis</a></li>
<li class="chapter" data-level="12.12.2" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#over-representation-analysis-ora"><i class="fa fa-check"></i><b>12.12.2</b> Over-Representation Analysis (ORA)</a></li>
<li class="chapter" data-level="12.12.3" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#some-concepts-to-remember"><i class="fa fa-check"></i><b>12.12.3</b> Some concepts to remember</a></li>
<li class="chapter" data-level="12.12.4" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#visualizations"><i class="fa fa-check"></i><b>12.12.4</b> Visualizations</a></li>
<li class="chapter" data-level="12.12.5" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#introduction-to-msigdb"><i class="fa fa-check"></i><b>12.12.5</b> Introduction to MSigDB</a></li>
<li class="chapter" data-level="12.12.6" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#gene-set-enrichment-analysis-gsea"><i class="fa fa-check"></i><b>12.12.6</b> Gene Set Enrichment Analysis (GSEA)</a></li>
<li class="chapter" data-level="12.12.7" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#gsea-in-action"><i class="fa fa-check"></i><b>12.12.7</b> GSEA in action</a></li>
<li class="chapter" data-level="12.12.8" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#visualizing-gsea-results"><i class="fa fa-check"></i><b>12.12.8</b> Visualizing GSEA Results</a></li>
<li class="chapter" data-level="12.12.9" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#conclusion-37"><i class="fa fa-check"></i><b>12.12.9</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="12.13" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#congratulations-for-successfully-completing-this-final-project"><i class="fa fa-check"></i><b>12.13</b> Congratulations for successfully completing this final project!</a>
<ul>
<li class="chapter" data-level="12.13.1" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#final-words"><i class="fa fa-check"></i><b>12.13.1</b> Final words</a></li>
<li class="chapter" data-level="12.13.2" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#materials-to-read-to-uplift-your-r-skills-to-the-next-level"><i class="fa fa-check"></i><b>12.13.2</b> Materials to read to uplift your R skills to the next level</a></li>
<li class="chapter" data-level="12.13.3" data-path="final-project-analyzing-rnaseq-data-from-geo.html"><a href="final-project-analyzing-rnaseq-data-from-geo.html#coding-challenge"><i class="fa fa-check"></i><b>12.13.3</b> Coding Challenge</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Master Real-world Bioinformatics analysis in R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-bioconductor" class="section level1" number="11">
<h1><span class="header-section-number">Chapter 11</span> Introduction to BioConductor</h1>
<p>In the ever-evolving field of bioinformatics and computational biology, researchers and scientists rely on specialized tools and packages to analyze and interpret biological data efficiently. <a href="https://bioconductor.org/"><code>Bioconductor</code></a>, a powerful and comprehensive suite of R packages, plays a pivotal role in addressing the unique challenges posed by genomics, transcriptomics, proteomics, and other biological data domains.</p>
<p>This section, “Introduction to Bioconductor in R,” will serve as your gateway to this remarkable resource. We will explore the fundamentals of Bioconductor, its significance in the realm of life sciences, and how you can harness its capabilities to unlock valuable insights from complex biological datasets. Whether you are a biologist, bioinformatician, or data scientist, understanding Bioconductor is a crucial step towards advancing your research and data analysis endeavors in the field of biology. Let’s embark on this exciting journey into the world of Bioconductor, where data meets discovery.</p>
<div id="introduction-to-biocmanager" class="section level2" number="11.1">
<h2><span class="header-section-number">11.1</span> Introduction to BiocManager</h2>
<p><code>BiocManager</code> is an R package that serves as the primary interface for managing <code>Bioconductor</code> packages, which are extensions of R developed specifically for bioinformatics applications. Bioconductor itself is a project that provides tools for the analysis and comprehension of high-throughput genomic data, including but not limited to next-generation sequencing (NGS), microarrays, and proteomics.</p>
<div id="why-use-biocmanager" class="section level3" number="11.1.1">
<h3><span class="header-section-number">11.1.1</span> Why Use BiocManager?</h3>
<p>The role of BiocManager is ensuring compatibility among Bioconductor packages and between these packages and your version of R. It simplifies the process of installing Bioconductor packages, managing dependencies, and keeping packages up-to-date with the latest releases, thereby fostering a stable and efficient bioinformatics workflow.</p>
</div>
<div id="installing-biocmanager" class="section level3" number="11.1.2">
<h3><span class="header-section-number">11.1.2</span> Installing BiocManager</h3>
<p>To get started with BiocManager, you first need to install it from CRAN (the Comprehensive R Archive Network), which can be done using the following command in R:</p>
<div class="sourceCode" id="cb450"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb450-1"><a href="introduction-to-bioconductor.html#cb450-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span></code></pre></div>
<p>Once installed, you can load <code>BiocManager</code> just like any other R package:</p>
<div class="sourceCode" id="cb451"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb451-1"><a href="introduction-to-bioconductor.html#cb451-1" tabindex="-1"></a><span class="fu">library</span>(BiocManager)</span></code></pre></div>
</div>
<div id="installing-bioconductor-packages" class="section level3" number="11.1.3">
<h3><span class="header-section-number">11.1.3</span> Installing Bioconductor Packages</h3>
<p>With <code>BiocManager</code> loaded, installing Bioconductor packages is straightforward. Suppose you want to install the <code>GenomicRanges</code> package; you can do so with the following command:</p>
<div class="sourceCode" id="cb452"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb452-1"><a href="introduction-to-bioconductor.html#cb452-1" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;GenomicRanges&quot;</span>)</span></code></pre></div>
<p>BiocManager automatically resolves and installs any dependencies, ensuring that all required packages are installed for the <code>GenomicRanges</code> package to function correctly.</p>
</div>
<div id="updating-bioconductor-packages" class="section level3" number="11.1.4">
<h3><span class="header-section-number">11.1.4</span> Updating Bioconductor Packages</h3>
<p>Keeping your Bioconductor packages up-to-date is crucial for accessing the latest features, improvements, and bug fixes. BiocManager facilitates this through the install function, which also checks for and updates any out-of-date packages:</p>
<div class="sourceCode" id="cb453"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb453-1"><a href="introduction-to-bioconductor.html#cb453-1" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>()</span></code></pre></div>
<p>Running this command without specifying a package name updates all installed Bioconductor packages to their latest versions.</p>
</div>
<div id="checking-for-valid-bioconductor-versions" class="section level3" number="11.1.5">
<h3><span class="header-section-number">11.1.5</span> Checking for Valid Bioconductor Versions</h3>
<p>Compatibility between your R version and Bioconductor packages is vital for smooth bioinformatics analyses. BiocManager offers a function to validate this compatibility:</p>
<div class="sourceCode" id="cb454"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb454-1"><a href="introduction-to-bioconductor.html#cb454-1" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">valid</span>()</span></code></pre></div>
<p>This command checks that all installed Bioconductor packages are compatible with each other and with your current version of R, providing a report of any inconsistencies.</p>
</div>
<div id="conclusion-26" class="section level3" number="11.1.6">
<h3><span class="header-section-number">11.1.6</span> Conclusion</h3>
<p><code>BiocManager</code> is an indispensable tool for bioinformatics practitioners working with R and <code>Bioconductor</code>. It simplifies package management, ensuring that researchers can focus on their analyses without being bogged down by software compatibility issues. By leveraging <code>BiocManager</code>, you can maintain a cutting-edge bioinformatics toolkit, fully equipped to tackle the challenges of genomic data analysis.</p>
</div>
</div>
<div id="working-with-genomic-coordinates-and-regions-in-r" class="section level2" number="11.2">
<h2><span class="header-section-number">11.2</span> Working with Genomic Coordinates and Regions in R</h2>
<p>Genomic coordinates are fundamental in the field of genomics. Whether you’re dealing with genes, regulatory elements, ChIP-seq peaks, or mutation calling data, they are all represented as genomic coordinates. In R, you can efficiently handle genomic coordinates and regions using the <code>GenomicRanges</code> package and related tools. In this guide, we’ll explore how to work with genomic ranges, extract relevant information, and perform common operations.</p>
<blockquote>
<p>For a complete overview check docs here: <a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html" class="uri">https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html</a></p>
</blockquote>
<div id="introduction-to-genomicranges" class="section level3" number="11.2.1">
<h3><span class="header-section-number">11.2.1</span> Introduction to GenomicRanges</h3>
<p>The <code>GenomicRanges</code> package in R provides data structures for storing and manipulating genomic ranges. A genomic range typically includes information about the chromosome (seqname), the start and end positions, and the strand of the sequence.</p>
<ol style="list-style-type: decimal">
<li><p>Every gene or regulatory element (promoters, enhancers) in the genome can be represented in chromosome: start-end format.</p></li>
<li><p>You get a peak file from a ChIP-seq experiment. The peak file is usually represented in a at least 3-column bed format: chromosome, start and end.</p></li>
<li><p>You get a mutation calling <a href="https://genome.ucsc.edu/goldenPath/help/vcf.html">VCF</a> file. You will have the chromosome and position of that single variant.</p></li>
</ol>
<p>Let’s start by creating a simple genomic range:</p>
<div class="sourceCode" id="cb455"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb455-1"><a href="introduction-to-bioconductor.html#cb455-1" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb455-2"><a href="introduction-to-bioconductor.html#cb455-2" tabindex="-1"></a></span>
<span id="cb455-3"><a href="introduction-to-bioconductor.html#cb455-3" tabindex="-1"></a><span class="co"># Create a GenomicRanges object</span></span>
<span id="cb455-4"><a href="introduction-to-bioconductor.html#cb455-4" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="st">&quot;chr1&quot;</span>, </span>
<span id="cb455-5"><a href="introduction-to-bioconductor.html#cb455-5" tabindex="-1"></a>              <span class="at">ranges =</span> <span class="fu">IRanges</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">width =</span> <span class="dv">3</span>))</span>
<span id="cb455-6"><a href="introduction-to-bioconductor.html#cb455-6" tabindex="-1"></a></span>
<span id="cb455-7"><a href="introduction-to-bioconductor.html#cb455-7" tabindex="-1"></a>gr</span></code></pre></div>
<pre><code>## GRanges object with 10 ranges and 0 metadata columns:
##        seqnames    ranges strand
##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     chr1       1-3      *
##    [2]     chr1       2-4      *
##    [3]     chr1       3-5      *
##    [4]     chr1       4-6      *
##    [5]     chr1       5-7      *
##    [6]     chr1       6-8      *
##    [7]     chr1       7-9      *
##    [8]     chr1      8-10      *
##    [9]     chr1      9-11      *
##   [10]     chr1     10-12      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>In this example, we’ve created a GenomicRanges object for chromosome 1 with ten intervals of width 3. We did not specify the strand, so it is <code>*</code>. Alternatively, we can
specify the genomic regions are on the <code>+</code> strand.</p>
<div class="sourceCode" id="cb457"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb457-1"><a href="introduction-to-bioconductor.html#cb457-1" tabindex="-1"></a><span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="st">&quot;chr1&quot;</span>, </span>
<span id="cb457-2"><a href="introduction-to-bioconductor.html#cb457-2" tabindex="-1"></a>        <span class="at">ranges =</span> <span class="fu">IRanges</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">width =</span> <span class="dv">3</span>),</span>
<span id="cb457-3"><a href="introduction-to-bioconductor.html#cb457-3" tabindex="-1"></a>        <span class="at">strand =</span> <span class="st">&quot;+&quot;</span>)</span></code></pre></div>
<pre><code>## GRanges object with 10 ranges and 0 metadata columns:
##        seqnames    ranges strand
##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     chr1       1-3      +
##    [2]     chr1       2-4      +
##    [3]     chr1       3-5      +
##    [4]     chr1       4-6      +
##    [5]     chr1       5-7      +
##    [6]     chr1       6-8      +
##    [7]     chr1       7-9      +
##    [8]     chr1      8-10      +
##    [9]     chr1      9-11      +
##   [10]     chr1     10-12      +
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>To understand the standness, read this <a href="https://crazyhottommy.blogspot.com/2014/08/understanding-forward-strand-and.html">blog post</a> by me.</p>
<p><img src="images/strandness.jpeg" /></p>
</div>
<div id="basic-operations-with-genomicranges" class="section level3" number="11.2.2">
<h3><span class="header-section-number">11.2.2</span> Basic Operations with GenomicRanges</h3>
<p>These operations are commonly used in genomics data analysis to perform tasks such as calculating the length of genomic features, extracting specific regions of interest, and modifying intervals for downstream analysis. <code>GenomicRanges</code> provides a flexible and efficient way to work with genomic intervals, which is essential for tasks like annotation, visualization, and statistical analysis of genomics data.</p>
</div>
<div id="calculating-width-of-each-genomic-interval" class="section level3" number="11.2.3">
<h3><span class="header-section-number">11.2.3</span> Calculating width of Each Genomic Interval</h3>
<p>Genomic intervals can represent various features in a genome, such as genes, exons, or regulatory regions. Knowing the width of these intervals is crucial when analyzing genomic data. For example, you might want to calculate the size of a gene or measure the distance between two regulatory elements. The <code>width()</code> function helps you obtain this information quickly.</p>
<div class="sourceCode" id="cb459"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb459-1"><a href="introduction-to-bioconductor.html#cb459-1" tabindex="-1"></a><span class="fu">width</span>(gr)</span></code></pre></div>
<pre><code>##  [1] 3 3 3 3 3 3 3 3 3 3</code></pre>
<p>This function calculates the width (or length) of each genomic interval in the GenomicRanges object <code>gr</code>. In genomics, the width typically represents the number of base pairs or genomic coordinates covered by each interval.</p>
</div>
<div id="start-and-end-positions" class="section level3" number="11.2.4">
<h3><span class="header-section-number">11.2.4</span> Start and End Positions</h3>
<p>Genomic intervals are defined by their start and end positions along a chromosome. These functions allow you to extract these positions, which can be essential for tasks like determining the transcription start site of a gene or identifying the boundaries of a specific genomic region.</p>
<div id="getting-start-position" class="section level4" number="11.2.4.1">
<h4><span class="header-section-number">11.2.4.1</span> Getting start position</h4>
<div class="sourceCode" id="cb461"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb461-1"><a href="introduction-to-bioconductor.html#cb461-1" tabindex="-1"></a><span class="fu">start</span>(gr)</span></code></pre></div>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10</code></pre>
<p>This function retrieves the starting position (or the leftmost coordinate) of each genomic interval in the <code>GenomicRanges</code> object <code>gr</code>. It tells you where each interval begins along the genome.</p>
</div>
<div id="getting-end-position" class="section level4" number="11.2.4.2">
<h4><span class="header-section-number">11.2.4.2</span> Getting end position</h4>
<div class="sourceCode" id="cb463"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb463-1"><a href="introduction-to-bioconductor.html#cb463-1" tabindex="-1"></a><span class="fu">end</span>(gr)</span></code></pre></div>
<pre><code>##  [1]  3  4  5  6  7  8  9 10 11 12</code></pre>
<p>This function retrieves the ending position (or the rightmost coordinate) of each genomic interval in the <code>GenomicRanges</code> object <code>gr</code>. It tells you where each interval ends along the genome.</p>
</div>
</div>
<div id="strand-information" class="section level3" number="11.2.5">
<h3><span class="header-section-number">11.2.5</span> Strand Information</h3>
<p>In genomics, it’s important to know the orientation of genomic features. The strand information (<code>+</code> or <code>-</code>) indicates whether a feature is on the forward (<code>+</code>) or reverse (<code>-</code>) strand of DNA. This can be crucial for understanding gene transcription direction, reading frames, and other biological processes.</p>
<div class="sourceCode" id="cb465"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb465-1"><a href="introduction-to-bioconductor.html#cb465-1" tabindex="-1"></a><span class="fu">strand</span>(gr)</span></code></pre></div>
<pre><code>## factor-Rle of length 10 with 1 run
##   Lengths: 10
##   Values :  *
## Levels(3): + - *</code></pre>
<p>Genomic intervals can be associated with a strand information to represent the directionality of a genomic feature. The strand function retrieves the strand information for each interval. The strand can be either “+” for the forward strand, “-” for the reverse strand, or “*” for strand-agnostic intervals.</p>
</div>
<div id="shifting-the-genomic-range" class="section level3" number="11.2.6">
<h3><span class="header-section-number">11.2.6</span> Shifting the Genomic Range</h3>
<p>Sometimes, you need to shift genomic intervals to examine neighboring regions. For instance, you might want to find regions that overlap with a gene’s promoter, which is typically located upstream of the transcription start site. Shifting intervals allows you to explore nearby genomic areas easily.</p>
<div class="sourceCode" id="cb467"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb467-1"><a href="introduction-to-bioconductor.html#cb467-1" tabindex="-1"></a>gr <span class="sc">+</span> <span class="dv">1</span></span></code></pre></div>
<pre><code>## GRanges object with 10 ranges and 0 metadata columns:
##        seqnames    ranges strand
##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     chr1       0-4      *
##    [2]     chr1       1-5      *
##    [3]     chr1       2-6      *
##    [4]     chr1       3-7      *
##    [5]     chr1       4-8      *
##    [6]     chr1       5-9      *
##    [7]     chr1      6-10      *
##    [8]     chr1      7-11      *
##    [9]     chr1      8-12      *
##   [10]     chr1      9-13      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>This operation demonstrates how you can manipulate the genomic intervals in <code>gr</code>. Here, you are adding 1 to each start <strong>AND</strong> end position in the intervals, effectively expanding them by one base left and right .</p>
<div class="sourceCode" id="cb469"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb469-1"><a href="introduction-to-bioconductor.html#cb469-1" tabindex="-1"></a><span class="fu">width</span>(gr<span class="sc">+</span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>##  [1] 5 5 5 5 5 5 5 5 5 5</code></pre>
</div>
<div id="subsetting" class="section level3" number="11.2.7">
<h3><span class="header-section-number">11.2.7</span> Subsetting</h3>
<p>Genomic data can be extensive, and you often need to focus on specific regions of interest. Subsetting helps you extract only the relevant intervals from a larger dataset. This is especially useful when you want to analyze a particular set of genes or genomic regions.</p>
<div class="sourceCode" id="cb471"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb471-1"><a href="introduction-to-bioconductor.html#cb471-1" tabindex="-1"></a>gr[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1       1-3      *
##   [2]     chr1       2-4      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>This code subset the GenomicRanges object gr to select the first two intervals. It returns a new GenomicRanges object containing only those intervals.</p>
</div>
<div id="flanking-regions" class="section level3" number="11.2.8">
<h3><span class="header-section-number">11.2.8</span> Flanking Regions</h3>
<p>Imagine you have a specific location in the DNA, and you want to study not only that location but also the regions right before and after it. flank lets you do this. For example, if you’re interested in a particular gene, you can use flank to include a bit of the DNA sequence before and after that gene. This helps you see the surrounding context and understand how the gene fits into the bigger picture.</p>
<div class="sourceCode" id="cb473"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb473-1"><a href="introduction-to-bioconductor.html#cb473-1" tabindex="-1"></a><span class="fu">flank</span>(gr, <span class="dv">2</span>)</span></code></pre></div>
<pre><code>## GRanges object with 10 ranges and 0 metadata columns:
##        seqnames    ranges strand
##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     chr1      -1-0      *
##    [2]     chr1       0-1      *
##    [3]     chr1       1-2      *
##    [4]     chr1       2-3      *
##    [5]     chr1       3-4      *
##    [6]     chr1       4-5      *
##    [7]     chr1       5-6      *
##    [8]     chr1       6-7      *
##    [9]     chr1       7-8      *
##   [10]     chr1       8-9      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p><code>flank</code> is useful for tasks like expanding genomic regions of interest to capture nearby regions or creating control regions around known features for downstream analysis. It is commonly used in genomics research to study the context around specific genomic locations.</p>
</div>
<div id="resizing" class="section level3" number="11.2.9">
<h3><span class="header-section-number">11.2.9</span> Resizing</h3>
<p>Think of a situation where you have many different pieces of DNA (intervals), and they’re all different lengths. Maybe you want to compare them or count something in each of them. It’s easier to work with them if they’re all the same size. That’s what <code>resize</code> does. It makes sure that all the pieces of DNA are the same length, so you can compare or analyze them more easily.</p>
<div class="sourceCode" id="cb475"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb475-1"><a href="introduction-to-bioconductor.html#cb475-1" tabindex="-1"></a><span class="fu">resize</span>(gr, <span class="dv">10</span>)</span></code></pre></div>
<pre><code>## GRanges object with 10 ranges and 0 metadata columns:
##        seqnames    ranges strand
##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     chr1      1-10      *
##    [2]     chr1      2-11      *
##    [3]     chr1      3-12      *
##    [4]     chr1      4-13      *
##    [5]     chr1      5-14      *
##    [6]     chr1      6-15      *
##    [7]     chr1      7-16      *
##    [8]     chr1      8-17      *
##    [9]     chr1      9-18      *
##   [10]     chr1     10-19      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p><code>resize</code> is used to standardize the length of genomic intervals, which can be useful for comparing or analyzing regions of interest with consistent sizes. It is often applied to ensure that intervals have the same width, making them suitable for various downstream analyses, such as counting reads or comparing features.</p>
</div>
<div id="based-and-1-based-coordinate-system" class="section level3" number="11.2.10">
<h3><span class="header-section-number">11.2.10</span> 0 based and 1 based coordinate system</h3>
<p>One needs to be aware that there are two genomics coordinate systems: 1 based and 0 based. There is really no mystery between these two. You EITHER count start at 0 OR at 1. However, this can make confusions when analyzing genomic data and one may make mistakes if not keep it in mind. Read <a href="https://www.biostars.org/p/84686/" class="uri">https://www.biostars.org/p/84686/</a>.</p>
<p>The reason that why it matters is that python index starts at 0 while R starts at 1.</p>
<p>Make sure you understand how different bioinformatics format use different coordinate
system.</p>
<p><img src="images/coordinate.png" /></p>
</div>
<div id="other-packages" class="section level3" number="11.2.11">
<h3><span class="header-section-number">11.2.11</span> Other packages</h3>
<p>In genomics research, we often work with genomic data in various formats such as GTF (Gene Transfer Format) and BED files. To facilitate this, we have a few essential packages at our disposal:</p>
<ul>
<li><p><strong>AnnotationHub</strong>: This package provides access to a wide range of genome annotations, including GTF files, which are commonly used to represent gene models. These annotations are invaluable for understanding genomic regions and gene structures.</p></li>
<li><p><strong>GenomicFeatures</strong>: This package is the powerhouse for working with genomic data in R. It provides functions for creating and manipulating genomic feature objects.</p></li>
<li><p><strong>rtracklayer</strong>: This package specializes in reading and handling genomic data files, including BED and GTF files.</p></li>
</ul>
</div>
<div id="accessing-genome-annotations" class="section level3" number="11.2.12">
<h3><span class="header-section-number">11.2.12</span> Accessing Genome Annotations</h3>
<blockquote>
<p>You can check AnnotationHub docs here: <a href="https://bioconductor.org/packages/release/bioc/vignettes/AnnotationHub/inst/doc/AnnotationHub.html" class="uri">https://bioconductor.org/packages/release/bioc/vignettes/AnnotationHub/inst/doc/AnnotationHub.html</a></p>
</blockquote>
<p>Genome annotations provide essential information about the location and structure of genes, which is crucial for understanding how genes function and how they are regulated. For example, knowing the coordinates of exons, introns, and promoters allows us to analyze where specific genetic elements are located in the genome.</p>
<div class="sourceCode" id="cb477"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb477-1"><a href="introduction-to-bioconductor.html#cb477-1" tabindex="-1"></a><span class="fu">library</span>(AnnotationHub)</span>
<span id="cb477-2"><a href="introduction-to-bioconductor.html#cb477-2" tabindex="-1"></a></span>
<span id="cb477-3"><a href="introduction-to-bioconductor.html#cb477-3" tabindex="-1"></a><span class="co"># Initialize the AnnotationHub</span></span>
<span id="cb477-4"><a href="introduction-to-bioconductor.html#cb477-4" tabindex="-1"></a>ah <span class="ot">&lt;-</span> <span class="fu">AnnotationHub</span>()</span>
<span id="cb477-5"><a href="introduction-to-bioconductor.html#cb477-5" tabindex="-1"></a></span>
<span id="cb477-6"><a href="introduction-to-bioconductor.html#cb477-6" tabindex="-1"></a><span class="co"># Query for specific annotations, for example, Homo sapiens (human) in the GRCh37 assembly</span></span>
<span id="cb477-7"><a href="introduction-to-bioconductor.html#cb477-7" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> AnnotationHub<span class="sc">::</span><span class="fu">query</span>(ah, <span class="fu">c</span>(<span class="st">&quot;gtf&quot;</span>, <span class="st">&quot;Homo_sapiens&quot;</span>, <span class="st">&quot;GRCh37&quot;</span>))</span>
<span id="cb477-8"><a href="introduction-to-bioconductor.html#cb477-8" tabindex="-1"></a></span>
<span id="cb477-9"><a href="introduction-to-bioconductor.html#cb477-9" tabindex="-1"></a>annotations</span></code></pre></div>
<pre><code>## AnnotationHub with 7 records
## # snapshotDate(): 2021-10-20
## # $dataprovider: Ensembl
## # $species: Homo sapiens
## # $rdataclass: GRanges
## # additional mcols(): taxonomyid, genome, description,
## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
## #   rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH7558&quot;]]&#39; 
## 
##             title                     
##   AH7558  | Homo_sapiens.GRCh37.70.gtf
##   AH7619  | Homo_sapiens.GRCh37.69.gtf
##   AH7666  | Homo_sapiens.GRCh37.71.gtf
##   AH7726  | Homo_sapiens.GRCh37.72.gtf
##   AH7790  | Homo_sapiens.GRCh37.73.gtf
##   AH8753  | Homo_sapiens.GRCh37.74.gtf
##   AH10684 | Homo_sapiens.GRCh37.75.gtf</code></pre>
<div class="sourceCode" id="cb479"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb479-1"><a href="introduction-to-bioconductor.html#cb479-1" tabindex="-1"></a><span class="co"># Select one of the annotations (e.g., GRCh37.gtf)</span></span>
<span id="cb479-2"><a href="introduction-to-bioconductor.html#cb479-2" tabindex="-1"></a>GRCh37.gtf <span class="ot">&lt;-</span> annotations[[<span class="st">&#39;AH8753&#39;</span>]]</span></code></pre></div>
<p>Now, we have a <code>GenomicRanges</code> object called <code>GRCh37.gtf</code>, which contains genomic features from the GRCh37 assembly of the human genome.</p>
</div>
<div id="understanding-genomic-biotypes" class="section level3" number="11.2.13">
<h3><span class="header-section-number">11.2.13</span> Understanding Genomic Biotypes</h3>
<p>Genes in the genome can have different biotypes, indicating their functional roles. We can filter our genomic features based on biotypes, such as “protein_coding” and “lincRNA.”</p>
<p>Filtering genes by biotype helps us focus on specific classes of genes, such as protein-coding genes, which are involved in producing proteins, or long intergenic non-coding RNAs (lincRNAs), which play regulatory roles.</p>
<div class="sourceCode" id="cb480"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb480-1"><a href="introduction-to-bioconductor.html#cb480-1" tabindex="-1"></a><span class="co"># what are the avaiable biotypes</span></span>
<span id="cb480-2"><a href="introduction-to-bioconductor.html#cb480-2" tabindex="-1"></a><span class="fu">table</span>(GRCh37.gtf<span class="sc">$</span>gene_biotype)</span></code></pre></div>
<pre><code>## 
## 3prime_overlapping_ncrna                antisense                IG_C_gene 
##                       63                    28001                      228 
##          IG_C_pseudogene                IG_D_gene                IG_J_gene 
##                       27                      128                       52 
##          IG_J_pseudogene                IG_V_gene          IG_V_pseudogene 
##                        6                      747                      393 
##                  lincRNA                    miRNA                 misc_RNA 
##                    34236                     3361                     2174 
##                  Mt_rRNA                  Mt_tRNA   polymorphic_pseudogene 
##                        2                       22                     3475 
##     processed_pseudogene     processed_transcript           protein_coding 
##                        1                    12720                  2106659 
##               pseudogene                     rRNA           sense_intronic 
##                    44993                      568                     1662 
##        sense_overlapping                   snoRNA                    snRNA 
##                      841                     1549                     2067 
##                TR_C_gene                TR_D_gene                TR_J_gene 
##                       44                        6                      164 
##          TR_J_pseudogene                TR_V_gene          TR_V_pseudogene 
##                        4                      597                       67</code></pre>
<div class="sourceCode" id="cb482"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb482-1"><a href="introduction-to-bioconductor.html#cb482-1" tabindex="-1"></a><span class="co"># subset </span></span>
<span id="cb482-2"><a href="introduction-to-bioconductor.html#cb482-2" tabindex="-1"></a>GRCh37.gtf <span class="ot">&lt;-</span> GRCh37.gtf[GRCh37.gtf<span class="sc">$</span>gene_biotype <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;protein_coding&quot;</span>, <span class="st">&quot;lincRNA&quot;</span>)]</span></code></pre></div>
</div>
<div id="creating-a-transcript-database-txdb" class="section level3" number="11.2.14">
<h3><span class="header-section-number">11.2.14</span> Creating a Transcript Database (TxDb)</h3>
<blockquote>
<p>You can check GenomicFeatures docs here: <a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicFeatures/inst/doc/GenomicFeatures.html" class="uri">https://bioconductor.org/packages/release/bioc/vignettes/GenomicFeatures/inst/doc/GenomicFeatures.html</a></p>
</blockquote>
<p>To perform more advanced analyses, we’ll create a transcript database (<code>TxDb</code>) from our genomic features. A <code>TxDb</code> is a structured database of transcript information, allowing us to efficiently query and retrieve specific genomic elements for analysis.</p>
<div class="sourceCode" id="cb483"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb483-1"><a href="introduction-to-bioconductor.html#cb483-1" tabindex="-1"></a><span class="fu">library</span>(GenomicFeatures)</span>
<span id="cb483-2"><a href="introduction-to-bioconductor.html#cb483-2" tabindex="-1"></a></span>
<span id="cb483-3"><a href="introduction-to-bioconductor.html#cb483-3" tabindex="-1"></a><span class="co"># Create a TxDb from the filtered genomic features</span></span>
<span id="cb483-4"><a href="introduction-to-bioconductor.html#cb483-4" tabindex="-1"></a>GRCh37.txdb <span class="ot">&lt;-</span> <span class="fu">makeTxDbFromGRanges</span>(GRCh37.gtf)</span>
<span id="cb483-5"><a href="introduction-to-bioconductor.html#cb483-5" tabindex="-1"></a></span>
<span id="cb483-6"><a href="introduction-to-bioconductor.html#cb483-6" tabindex="-1"></a>GRCh37.txdb</span></code></pre></div>
<pre><code>## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Genome: GRCh37
## # Nb of transcripts: 171683
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2024-07-26 08:31:09 -0400 (Fri, 26 Jul 2024)
## # GenomicFeatures version at creation time: 1.46.5
## # RSQLite version at creation time: 2.3.1
## # DBSCHEMAVERSION: 1.2</code></pre>
</div>
<div id="extracting-exons-introns-and-intergenic-regions" class="section level3" number="11.2.15">
<h3><span class="header-section-number">11.2.15</span> Extracting Exons, Introns, and Intergenic Regions</h3>
<p>Now that we have our TxDb, we can extract various genomic elements for further analysis.</p>
<div id="exons-by-gene" class="section level4" number="11.2.15.1">
<h4><span class="header-section-number">11.2.15.1</span> Exons by Gene</h4>
<p>Analyzing exons by gene is essential for understanding the coding regions of genes and their splicing patterns. Let’s retrieve exons grouped by genes.</p>
<div class="sourceCode" id="cb485"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb485-1"><a href="introduction-to-bioconductor.html#cb485-1" tabindex="-1"></a>exonsByGene <span class="ot">&lt;-</span> <span class="fu">exonsBy</span>(GRCh37.txdb, <span class="st">&quot;gene&quot;</span>) </span>
<span id="cb485-2"><a href="introduction-to-bioconductor.html#cb485-2" tabindex="-1"></a></span>
<span id="cb485-3"><a href="introduction-to-bioconductor.html#cb485-3" tabindex="-1"></a><span class="co"># GRangesList object, a list of GRanges </span></span>
<span id="cb485-4"><a href="introduction-to-bioconductor.html#cb485-4" tabindex="-1"></a>exonsByGene</span></code></pre></div>
<pre><code>## GRangesList object of length 30150:
## $ENSG00000000003
## GRanges object with 17 ranges and 2 metadata columns:
##        seqnames            ranges strand |   exon_id       exon_name
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;     &lt;character&gt;
##    [1]        X 99883667-99884983      - |    591006 ENSE00001459322
##    [2]        X 99885756-99885863      - |    591007 ENSE00000868868
##    [3]        X 99887482-99887565      - |    591008 ENSE00000401072
##    [4]        X 99887538-99887565      - |    591009 ENSE00001849132
##    [5]        X 99888402-99888536      - |    591010 ENSE00003554016
##    ...      ...               ...    ... .       ...             ...
##   [13]        X 99890555-99890743      - |    591018 ENSE00003662440
##   [14]        X 99891188-99891686      - |    591019 ENSE00001886883
##   [15]        X 99891605-99891803      - |    591020 ENSE00001855382
##   [16]        X 99891790-99892101      - |    591021 ENSE00001863395
##   [17]        X 99894942-99894988      - |    591022 ENSE00001828996
##   -------
##   seqinfo: 265 sequences (1 circular) from GRCh37 genome
## 
## ...
## &lt;30149 more elements&gt;</code></pre>
</div>
</div>
<div id="merging-all-exons" class="section level3" number="11.2.16">
<h3><span class="header-section-number">11.2.16</span> Merging All Exons</h3>
<p>Merging exons (exons can overlap with each other) helps simplify analysis, such as quantifying the overall exonic content of genes. To get a single range representing all exons, we can reduce them.</p>
<div class="sourceCode" id="cb487"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb487-1"><a href="introduction-to-bioconductor.html#cb487-1" tabindex="-1"></a>allExons <span class="ot">&lt;-</span> <span class="fu">exons</span>(GRCh37.txdb) <span class="sc">%&gt;%</span> </span>
<span id="cb487-2"><a href="introduction-to-bioconductor.html#cb487-2" tabindex="-1"></a>  GenomicRanges<span class="sc">::</span><span class="fu">reduce</span>()</span>
<span id="cb487-3"><a href="introduction-to-bioconductor.html#cb487-3" tabindex="-1"></a></span>
<span id="cb487-4"><a href="introduction-to-bioconductor.html#cb487-4" tabindex="-1"></a>allExons</span></code></pre></div>
<pre><code>## GRanges object with 279054 ranges and 0 metadata columns:
##                 seqnames              ranges strand
##                    &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
##        [1]             1         29554-30039      +
##        [2]             1         30267-30667      +
##        [3]             1         30976-31109      +
##        [4]             1         69091-70008      +
##        [5]             1       160446-160690      +
##        ...           ...                 ...    ...
##   [279050] HSCHR7_1_CTG6 141373867-141374020      -
##   [279051] HSCHR7_1_CTG6 141385280-141385438      -
##   [279052] HSCHR7_1_CTG6 141386361-141386460      -
##   [279053] HSCHR7_1_CTG6 141401359-141401418      -
##   [279054] HSCHR7_1_CTG6 141401689-141401956      -
##   -------
##   seqinfo: 265 sequences (1 circular) from GRCh37 genome</code></pre>
</div>
<div id="introns" class="section level3" number="11.2.17">
<h3><span class="header-section-number">11.2.17</span> Introns</h3>
<p>Identifying introns is crucial for studying gene splicing and understanding the non-coding regions within genes. To find intronic regions, we can use the <code>intronsByTranscript</code> function.</p>
<div class="sourceCode" id="cb489"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb489-1"><a href="introduction-to-bioconductor.html#cb489-1" tabindex="-1"></a>introns <span class="ot">&lt;-</span> <span class="fu">intronsByTranscript</span>(GRCh37.txdb) <span class="sc">%&gt;%</span> </span>
<span id="cb489-2"><a href="introduction-to-bioconductor.html#cb489-2" tabindex="-1"></a>  <span class="fu">unlist</span>() <span class="sc">%&gt;%</span></span>
<span id="cb489-3"><a href="introduction-to-bioconductor.html#cb489-3" tabindex="-1"></a>  GenomicRanges<span class="sc">::</span><span class="fu">reduce</span>()</span>
<span id="cb489-4"><a href="introduction-to-bioconductor.html#cb489-4" tabindex="-1"></a></span>
<span id="cb489-5"><a href="introduction-to-bioconductor.html#cb489-5" tabindex="-1"></a>introns</span></code></pre></div>
<pre><code>## GRanges object with 185379 ranges and 0 metadata columns:
##                 seqnames              ranges strand
##                    &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
##        [1]             1         30040-30563      +
##        [2]             1         30668-30975      +
##        [3]             1       160691-161313      +
##        [4]             1       317782-334128      +
##        [5]             1       334298-439466      +
##        ...           ...                 ...    ...
##   [185375] HSCHR7_1_CTG6 141365119-141366086      -
##   [185376] HSCHR7_1_CTG6 141366187-141373866      -
##   [185377] HSCHR7_1_CTG6 141374021-141385279      -
##   [185378] HSCHR7_1_CTG6 141385439-141386360      -
##   [185379] HSCHR7_1_CTG6 141386461-141401688      -
##   -------
##   seqinfo: 265 sequences (1 circular) from GRCh37 genome</code></pre>
</div>
<div id="getting-all-genes" class="section level3" number="11.2.18">
<h3><span class="header-section-number">11.2.18</span> Getting All Genes</h3>
<p>Having a complete list of genes is essential for various genomics analyses, including differential gene expression studies. Obtaining all genes is straightforward.</p>
<div class="sourceCode" id="cb491"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb491-1"><a href="introduction-to-bioconductor.html#cb491-1" tabindex="-1"></a>allGenes <span class="ot">&lt;-</span> <span class="fu">genes</span>(GRCh37.txdb)</span>
<span id="cb491-2"><a href="introduction-to-bioconductor.html#cb491-2" tabindex="-1"></a></span>
<span id="cb491-3"><a href="introduction-to-bioconductor.html#cb491-3" tabindex="-1"></a>allGenes</span></code></pre></div>
<pre><code>## GRanges object with 30150 ranges and 1 metadata column:
##                                seqnames              ranges strand |
##                                   &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |
##   ENSG00000000003                     X   99883667-99894988      - |
##   ENSG00000000005                     X   99839799-99854882      + |
##   ENSG00000000419                    20   49551404-49575092      - |
##   ENSG00000000457                     1 169818772-169863408      - |
##   ENSG00000000460                     1 169631245-169823221      + |
##               ...                   ...                 ...    ... .
##   ENSG00000273488                     3 100080031-100080481      + |
##   ENSG00000273490 HSCHR19LRC_LRC_J_CTG1   54693789-54697585      + |
##   ENSG00000273491          HG1308_PATCH 130600118-130603315      + |
##   ENSG00000273492                    21   27543189-27589700      + |
##   ENSG00000273493                     3   58315692-58315845      + |
##                           gene_id
##                       &lt;character&gt;
##   ENSG00000000003 ENSG00000000003
##   ENSG00000000005 ENSG00000000005
##   ENSG00000000419 ENSG00000000419
##   ENSG00000000457 ENSG00000000457
##   ENSG00000000460 ENSG00000000460
##               ...             ...
##   ENSG00000273488 ENSG00000273488
##   ENSG00000273490 ENSG00000273490
##   ENSG00000273491 ENSG00000273491
##   ENSG00000273492 ENSG00000273492
##   ENSG00000273493 ENSG00000273493
##   -------
##   seqinfo: 265 sequences (1 circular) from GRCh37 genome</code></pre>
</div>
<div id="promoters" class="section level3" number="11.2.19">
<h3><span class="header-section-number">11.2.19</span> Promoters</h3>
<p>Promoter regions are critical for understanding gene regulation and identifying potential binding sites for transcription factors. To find promoter regions, typically defined as the region from <code>-1kb</code> to <code>+500bp</code> around the transcription start site (<code>TSS</code>), we can use the promoters function.</p>
<div class="sourceCode" id="cb493"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb493-1"><a href="introduction-to-bioconductor.html#cb493-1" tabindex="-1"></a>promoterRegions <span class="ot">&lt;-</span> <span class="fu">promoters</span>(<span class="fu">genes</span>(GRCh37.txdb), </span>
<span id="cb493-2"><a href="introduction-to-bioconductor.html#cb493-2" tabindex="-1"></a>                             <span class="at">upstream =</span> <span class="dv">1000</span>, </span>
<span id="cb493-3"><a href="introduction-to-bioconductor.html#cb493-3" tabindex="-1"></a>                             <span class="at">downstream =</span> <span class="dv">500</span>)</span>
<span id="cb493-4"><a href="introduction-to-bioconductor.html#cb493-4" tabindex="-1"></a></span>
<span id="cb493-5"><a href="introduction-to-bioconductor.html#cb493-5" tabindex="-1"></a>promoterRegions</span></code></pre></div>
<pre><code>## GRanges object with 30150 ranges and 1 metadata column:
##                                seqnames              ranges strand |
##                                   &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |
##   ENSG00000000003                     X   99894489-99895988      - |
##   ENSG00000000005                     X   99838799-99840298      + |
##   ENSG00000000419                    20   49574593-49576092      - |
##   ENSG00000000457                     1 169862909-169864408      - |
##   ENSG00000000460                     1 169630245-169631744      + |
##               ...                   ...                 ...    ... .
##   ENSG00000273488                     3 100079031-100080530      + |
##   ENSG00000273490 HSCHR19LRC_LRC_J_CTG1   54692789-54694288      + |
##   ENSG00000273491          HG1308_PATCH 130599118-130600617      + |
##   ENSG00000273492                    21   27542189-27543688      + |
##   ENSG00000273493                     3   58314692-58316191      + |
##                           gene_id
##                       &lt;character&gt;
##   ENSG00000000003 ENSG00000000003
##   ENSG00000000005 ENSG00000000005
##   ENSG00000000419 ENSG00000000419
##   ENSG00000000457 ENSG00000000457
##   ENSG00000000460 ENSG00000000460
##               ...             ...
##   ENSG00000273488 ENSG00000273488
##   ENSG00000273490 ENSG00000273490
##   ENSG00000273491 ENSG00000273491
##   ENSG00000273492 ENSG00000273492
##   ENSG00000273493 ENSG00000273493
##   -------
##   seqinfo: 265 sequences (1 circular) from GRCh37 genome</code></pre>
</div>
<div id="full-genome" class="section level3" number="11.2.20">
<h3><span class="header-section-number">11.2.20</span> Full Genome</h3>
<p>Having the entire genome as a single object is useful for genome-wide analyses and visualizations. To represent the entire genome as a GRanges object:</p>
<div class="sourceCode" id="cb495"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb495-1"><a href="introduction-to-bioconductor.html#cb495-1" tabindex="-1"></a>chrom_granges <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">seqinfo</span>(GRCh37.txdb), <span class="st">&quot;GRanges&quot;</span>)</span>
<span id="cb495-2"><a href="introduction-to-bioconductor.html#cb495-2" tabindex="-1"></a>chrom_granges</span></code></pre></div>
<pre><code>## GRanges object with 265 ranges and 0 metadata columns:
##                        seqnames      ranges strand
##                           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
##                1              1 1-249250621      *
##                2              2 1-243199373      *
##                3              3 1-198022430      *
##                4              4 1-191154276      *
##                5              5 1-180915260      *
##              ...            ...         ...    ...
##    HSCHR7_1_CTG6  HSCHR7_1_CTG6 1-159144671      *
##    HSCHR9_1_CTG1  HSCHR9_1_CTG1 1-141228243      *
##   HSCHR9_1_CTG35 HSCHR9_1_CTG35 1-141221627      *
##   HSCHR9_2_CTG35 HSCHR9_2_CTG35 1-141219511      *
##   HSCHR9_3_CTG35 HSCHR9_3_CTG35 1-141224529      *
##   -------
##   seqinfo: 265 sequences (1 circular) from GRCh37 genome</code></pre>
</div>
<div id="full-transcriptome" class="section level3" number="11.2.21">
<h3><span class="header-section-number">11.2.21</span> Full Transcriptome</h3>
<p>Merging overlapping transcripts simplifies transcript-level analyses and helps identify the full extent of genes. To represent the entire transcriptome, we can merge overlapping features.</p>
<div class="sourceCode" id="cb497"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb497-1"><a href="introduction-to-bioconductor.html#cb497-1" tabindex="-1"></a>collapsed_tx <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">reduce</span>(<span class="fu">transcripts</span>(GRCh37.txdb))</span>
<span id="cb497-2"><a href="introduction-to-bioconductor.html#cb497-2" tabindex="-1"></a></span>
<span id="cb497-3"><a href="introduction-to-bioconductor.html#cb497-3" tabindex="-1"></a><span class="co"># Set strand information to &#39;*&#39;</span></span>
<span id="cb497-4"><a href="introduction-to-bioconductor.html#cb497-4" tabindex="-1"></a><span class="fu">strand</span>(collapsed_tx) <span class="ot">&lt;-</span> <span class="st">&quot;*&quot;</span></span></code></pre></div>
</div>
<div id="intergenic-regions" class="section level3" number="11.2.22">
<h3><span class="header-section-number">11.2.22</span> Intergenic Regions</h3>
<p>Intergenic regions often contain important regulatory elements, and identifying them can provide insights into gene regulation. To find regions that are not within any annotated genes, we can use the setdiff function.</p>
<div class="sourceCode" id="cb498"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb498-1"><a href="introduction-to-bioconductor.html#cb498-1" tabindex="-1"></a>intergenicRegions <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">setdiff</span>(chrom_granges, collapsed_tx)</span>
<span id="cb498-2"><a href="introduction-to-bioconductor.html#cb498-2" tabindex="-1"></a></span>
<span id="cb498-3"><a href="introduction-to-bioconductor.html#cb498-3" tabindex="-1"></a>intergenicRegions </span></code></pre></div>
<pre><code>## GRanges object with 24100 ranges and 0 metadata columns:
##                 seqnames              ranges strand
##                    &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
##       [1]              1             1-29553      *
##       [2]              1         31110-34553      *
##       [3]              1         36082-69090      *
##       [4]              1         70009-89294      *
##       [5]              1       133567-134900      *
##       ...            ...                 ...    ...
##   [24096]  HSCHR7_1_CTG6 141493731-159144671      *
##   [24097]  HSCHR9_1_CTG1         1-141228243      *
##   [24098] HSCHR9_1_CTG35         1-141221627      *
##   [24099] HSCHR9_2_CTG35         1-141219511      *
##   [24100] HSCHR9_3_CTG35         1-141224529      *
##   -------
##   seqinfo: 265 sequences (1 circular) from GRCh37 genome</code></pre>
</div>
<div id="exploring-untranslated-regions-utrs" class="section level3" number="11.2.23">
<h3><span class="header-section-number">11.2.23</span> Exploring Untranslated Regions (UTRs)</h3>
<p>UTRs play crucial roles in post-transcriptional regulation, and analyzing them can provide insights into gene regulation mechanisms. If you’re interested in untranslated regions (UTRs) of genes, you can use functions like <code>fiveUTRsByTranscript</code> and <code>threeUTRsByTranscript</code> provided by the <code>GenomicFeatures</code> package.</p>
<div class="sourceCode" id="cb500"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb500-1"><a href="introduction-to-bioconductor.html#cb500-1" tabindex="-1"></a><span class="co"># To get 5&#39; UTRs by transcript</span></span>
<span id="cb500-2"><a href="introduction-to-bioconductor.html#cb500-2" tabindex="-1"></a>fiveUTRs <span class="ot">&lt;-</span> <span class="fu">fiveUTRsByTranscript</span>(GRCh37.txdb)</span>
<span id="cb500-3"><a href="introduction-to-bioconductor.html#cb500-3" tabindex="-1"></a></span>
<span id="cb500-4"><a href="introduction-to-bioconductor.html#cb500-4" tabindex="-1"></a><span class="co"># To get 3&#39; UTRs by transcript</span></span>
<span id="cb500-5"><a href="introduction-to-bioconductor.html#cb500-5" tabindex="-1"></a>threeUTRs <span class="ot">&lt;-</span> <span class="fu">threeUTRsByTranscript</span>(GRCh37.txdb)</span></code></pre></div>
</div>
<div id="conclusion-27" class="section level3" number="11.2.24">
<h3><span class="header-section-number">11.2.24</span> Conclusion</h3>
<p>In this lesson, we’ve explored various genomic features and their manipulation using R packages such as <code>GenomicFeatures</code>, <code>AnnotationHub</code>, and <code>rtracklayer</code>. These tools are invaluable for genomics research, allowing you to analyze and interpret genomic data effectively. Whether you’re working with ChIP-seq, RNA-seq, or genome annotation, understanding genomic features is essential to uncover the secrets of the genome.</p>
</div>
</div>
<div id="exploring-cpg-islands-and-shores-in-genomic-data" class="section level2" number="11.3">
<h2><span class="header-section-number">11.3</span> Exploring CpG Islands and Shores in Genomic Data</h2>
<p>In this lesson, we will delve into the fascinating world of genomics and learn how to manipulate and analyze genomic data using the powerful tools available in the Bioconductor package. We will specifically focus on CpG islands and their shores, exploring how to extract and analyze these critical genomic features.</p>
<div id="introduction-to-cpg-islands" class="section level3" number="11.3.1">
<h3><span class="header-section-number">11.3.1</span> Introduction to CpG Islands</h3>
<p>CpG islands are regions of DNA that contain a high frequency of cytosine-guanine (CpG) dinucleotide pairs. These regions are essential for regulating gene expression and have critical roles in various biological processes, including DNA methylation and epigenetic modifications. Analyzing CpG islands can provide valuable insights into gene regulation and genome function.</p>
<p>In this lesson, we will use Bioconductor to fetch CpG island coordinates from the UCSC Genome Browser, extract CpG shores, and perform various genomic operations.</p>
</div>
<div id="fetching-cpg-island-coordinates" class="section level3" number="11.3.2">
<h3><span class="header-section-number">11.3.2</span> Fetching CpG Island Coordinates</h3>
<p>CpG islands are critical genomic regions involved in gene regulation and epigenetic modifications. Accessing their coordinates is the first step in understanding their distribution across the genome and their potential functional roles.</p>
<p>Researchers often use CpG island coordinates to investigate gene promoters, identify potential regulatory elements, and study the epigenetic regulation of specific genes in various diseases, including cancer.</p>
<p>To begin, we will retrieve the CpG island coordinates from the UCSC Genome Browser using the <code>AnnotationHub</code> package. CpG islands are available for various species, and in this example, we are using Homo sapiens (human) data.</p>
<div class="sourceCode" id="cb501"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb501-1"><a href="introduction-to-bioconductor.html#cb501-1" tabindex="-1"></a><span class="co"># Fetching CpG island coordinates from UCSC Genome Browser</span></span>
<span id="cb501-2"><a href="introduction-to-bioconductor.html#cb501-2" tabindex="-1"></a><span class="fu">library</span>(AnnotationHub)</span>
<span id="cb501-3"><a href="introduction-to-bioconductor.html#cb501-3" tabindex="-1"></a>ah <span class="ot">&lt;-</span> <span class="fu">AnnotationHub</span>()</span>
<span id="cb501-4"><a href="introduction-to-bioconductor.html#cb501-4" tabindex="-1"></a></span>
<span id="cb501-5"><a href="introduction-to-bioconductor.html#cb501-5" tabindex="-1"></a>AnnotationHub<span class="sc">::</span><span class="fu">query</span>(ah, <span class="fu">c</span>(<span class="st">&quot;cpg&quot;</span>, <span class="st">&quot;UCSC&quot;</span>))</span></code></pre></div>
<pre><code>## AnnotationHub with 59 records
## # snapshotDate(): 2021-10-20
## # $dataprovider: UCSC
## # $species: Homo sapiens, Bos taurus, Pan troglodytes, Felis catus, Rattus n...
## # $rdataclass: GRanges
## # additional mcols(): taxonomyid, genome, description,
## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
## #   rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., &#39;object[[&quot;AH5086&quot;]]&#39; 
## 
##            title      
##   AH5086 | CpG Islands
##   AH5096 | Evo Cpg    
##   AH5204 | CpG Islands
##   AH5227 | Evo Cpg    
##   AH5344 | CpG Islands
##   ...      ...        
##   AH7109 | CpG Islands
##   AH7116 | CpG Islands
##   AH7135 | CpG Islands
##   AH7168 | CpG Islands
##   AH7203 | CpG Islands</code></pre>
<p>use the first entry</p>
<div class="sourceCode" id="cb503"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb503-1"><a href="introduction-to-bioconductor.html#cb503-1" tabindex="-1"></a>cgi <span class="ot">&lt;-</span> ah[[<span class="st">&quot;AH5086&quot;</span>]]</span>
<span id="cb503-2"><a href="introduction-to-bioconductor.html#cb503-2" tabindex="-1"></a>cgi</span></code></pre></div>
<pre><code>## GRanges object with 28691 ranges and 1 metadata column:
##                        seqnames        ranges strand |        name
##                           &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       [1]                  chr1   28736-29810      * |    CpG:_116
##       [2]                  chr1 135125-135563      * |     CpG:_30
##       [3]                  chr1 327791-328229      * |     CpG:_29
##       [4]                  chr1 437152-438164      * |     CpG:_84
##       [5]                  chr1 449274-450544      * |     CpG:_99
##       ...                   ...           ...    ... .         ...
##   [28687]  chr9_gl000201_random   15651-15909      * |     CpG:_30
##   [28688]  chr9_gl000201_random   26397-26873      * |     CpG:_43
##   [28689] chr11_gl000202_random   16284-16540      * |     CpG:_23
##   [28690] chr17_gl000204_random   54686-57368      * |    CpG:_228
##   [28691] chr17_gl000205_random 117501-117801      * |     CpG:_23
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>We now have the CpG island coordinates stored in the <code>cgi</code> GenomicRanges object.</p>
</div>
<div id="defining-cpg-shores" class="section level3" number="11.3.3">
<h3><span class="header-section-number">11.3.3</span> Defining CpG Shores</h3>
<p>CpG shores are regions located near CpG islands, and they play a crucial role in gene regulation. Defining these shores allows us to explore the regulatory landscape around CpG islands and identify regions of potential interest.</p>
<p>By analyzing CpG shores, researchers can gain insights into how epigenetic modifications in these regions affect gene expression. This knowledge is vital for understanding diseases that involve aberrant gene regulation.</p>
<p>CpG shores are regions located 2000 base pairs upstream and 2000 base pairs downstream of CpG islands. We can use Bioconductor to extract these shores.</p>
<div class="sourceCode" id="cb505"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb505-1"><a href="introduction-to-bioconductor.html#cb505-1" tabindex="-1"></a><span class="co"># Extract the shore defined by 2000 bp upstream of CpG islands</span></span>
<span id="cb505-2"><a href="introduction-to-bioconductor.html#cb505-2" tabindex="-1"></a>shore1 <span class="ot">&lt;-</span> <span class="fu">trim</span>(<span class="fu">flank</span>(cgi, <span class="at">width =</span> <span class="dv">2000</span>, <span class="at">start =</span> <span class="cn">TRUE</span>))</span>
<span id="cb505-3"><a href="introduction-to-bioconductor.html#cb505-3" tabindex="-1"></a></span>
<span id="cb505-4"><a href="introduction-to-bioconductor.html#cb505-4" tabindex="-1"></a><span class="co"># Extract the shore defined by 2000 bp downstream of CpG islands</span></span>
<span id="cb505-5"><a href="introduction-to-bioconductor.html#cb505-5" tabindex="-1"></a>shore2 <span class="ot">&lt;-</span> <span class="fu">trim</span>(<span class="fu">flank</span>(cgi, <span class="at">width =</span> <span class="dv">2000</span>, <span class="at">start =</span> <span class="cn">FALSE</span>))</span></code></pre></div>
<p><code>trim</code> will trim off the bases that exceed the chromosome ends since we extend 2000 bp
upstream and downstream of the CpG sites. Some CpG sites can be very close to the ends of the chromosomes.</p>
</div>
<div id="combining-and-analyzing-cpg-shores" class="section level3" number="11.3.4">
<h3><span class="header-section-number">11.3.4</span> Combining and Analyzing CpG Shores</h3>
<p>Combining the upstream and downstream shores and analyzing their overlap with CpG islands helps identify regions with unique genomic characteristics. This step allows researchers to pinpoint areas of interest for further investigation.</p>
<p>Researchers often use this analysis to identify differentially methylated regions (DMRs) associated with specific diseases or conditions. DMRs can serve as biomarkers or potential therapeutic targets.</p>
<p>Now, let’s perform some genomic operations on these CpG shores. We’ll combine the upstream and downstream shores and identify the features that are present in shores but not in CpG islands (i.e., shores not overlapping with islands).</p>
<div class="sourceCode" id="cb506"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb506-1"><a href="introduction-to-bioconductor.html#cb506-1" tabindex="-1"></a><span class="co"># Combine the shores where they overlap</span></span>
<span id="cb506-2"><a href="introduction-to-bioconductor.html#cb506-2" tabindex="-1"></a>shore1_2 <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">reduce</span>(<span class="fu">c</span>(shore1, shore2))</span>
<span id="cb506-3"><a href="introduction-to-bioconductor.html#cb506-3" tabindex="-1"></a></span>
<span id="cb506-4"><a href="introduction-to-bioconductor.html#cb506-4" tabindex="-1"></a><span class="co"># Extract the features (ranges) that are present in shores only and not in CpG islands</span></span>
<span id="cb506-5"><a href="introduction-to-bioconductor.html#cb506-5" tabindex="-1"></a>cpgi_shores <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">setdiff</span>(shore1_2, cgi)</span>
<span id="cb506-6"><a href="introduction-to-bioconductor.html#cb506-6" tabindex="-1"></a>cpgi_shores<span class="sc">$</span>name <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;shore&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(cpgi_shores), <span class="at">sep =</span> <span class="st">&quot;_&quot;</span>)</span>
<span id="cb506-7"><a href="introduction-to-bioconductor.html#cb506-7" tabindex="-1"></a></span>
<span id="cb506-8"><a href="introduction-to-bioconductor.html#cb506-8" tabindex="-1"></a>cpgi_shores</span></code></pre></div>
<pre><code>## GRanges object with 51914 ranges and 1 metadata column:
##                 seqnames        ranges strand |        name
##                    &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       [1]           chr1   26736-28735      * |     shore_1
##       [2]           chr1   29811-31810      * |     shore_2
##       [3]           chr1 133125-135124      * |     shore_3
##       [4]           chr1 135564-137563      * |     shore_4
##       [5]           chr1 325791-327790      * |     shore_5
##       ...            ...           ...    ... .         ...
##   [51910] chrUn_gl000241   37274-39273      * | shore_51910
##   [51911] chrUn_gl000242   10843-12842      * | shore_51911
##   [51912] chrUn_gl000242   13100-15099      * | shore_51912
##   [51913] chrUn_gl000243   28420-30419      * | shore_51913
##   [51914] chrUn_gl000243   30716-32715      * | shore_51914
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>Now, <code>cpgi_shores</code> contains the <code>GenomicRanges</code> object representing CpG shores, and each shore is labeled with a unique name.</p>
</div>
<div id="conclusion-28" class="section level3" number="11.3.5">
<h3><span class="header-section-number">11.3.5</span> Conclusion</h3>
<p>In this lesson, we’ve explored the powerful capabilities of the Bioconductor package for working with genomic data. We’ve fetched CpG island coordinates, extracted CpG shores, and performed genomic operations to identify regions of interest. These techniques are fundamental for researchers and bioinformaticians working with genomics data to unravel the mysteries of the genome.</p>
<p>If you’re interested in diving deeper into genomics analysis, consider exploring the tutorials provided by Bioconductor on their website. They offer a wealth of knowledge and resources to help you harness the full potential of genomic data analysis.</p>
</div>
</div>
<div id="real-world-applications-chip-seq" class="section level2" number="11.4">
<h2><span class="header-section-number">11.4</span> Real-World Applications: ChIP-seq</h2>
<p>Understanding genomic features is crucial for various genomics tasks, including:</p>
<ul>
<li><p>ChIP-seq Analysis: You can use these genomic ranges to determine how many ChIP-seq peaks fall into promoters, exons, introns, or intergenic regions, helping you interpret the functional significance of your data.</p></li>
<li><p>RNA-seq Analysis: Identifying which exons are covered by RNA-seq reads and counting reads in each exon allows you to quantify gene expression accurately.</p></li>
<li><p>Functional Genomics: Investigating the genomic context of genes helps in understanding their regulatory elements, including promoters and enhancers.</p></li>
<li><p>Genome Annotation: These tools are essential for creating comprehensive annotations of genomes, enabling researchers to understand gene structures and functions.</p></li>
</ul>
<div id="read-in-peak-file" class="section level3" number="11.4.1">
<h3><span class="header-section-number">11.4.1</span> read in peak file</h3>
</div>
<div id="identify-promoters-that-overlap-with-the-peaks" class="section level3" number="11.4.2">
<h3><span class="header-section-number">11.4.2</span> Identify promoters that overlap with the peaks</h3>
</div>
<div id="use-the-chipseeker-package" class="section level3" number="11.4.3">
<h3><span class="header-section-number">11.4.3</span> use the ChIPseeker package</h3>
</div>
</div>
<div id="analyzing-and-visualizing-genomic-data" class="section level2" number="11.5">
<h2><span class="header-section-number">11.5</span> Analyzing and Visualizing Genomic Data</h2>
<p>In this lesson, we will explore several essential tools and techniques used in genomics research, including <code>GEOquery</code> for data retrieval, gene ID conversion using <code>biomaRt</code> and <code>org.Hs.eg.db</code>, and visualization with <code>ComplexHeatmap</code>. These tools are commonly used in genomics to analyze and visualize gene expression data. We will briefly mention <code>DESeq2</code> for differential expression analysis.</p>
<div id="deseq2" class="section level3" number="11.5.1">
<h3><span class="header-section-number">11.5.1</span> DESeq2</h3>
<blockquote>
<p>You can explore docs here: <a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html" class="uri">https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html</a></p>
</blockquote>
<p><code>DESeq2</code> is a powerful Bioconductor package used for differential expression analysis. It is particularly helpful when working with RNA-seq data. This tool helps identify genes that are differentially expressed under different conditions. The tutorial on Bioconductor is very comprehensive and I will leave the students to read by themselves. We will use it in our final project.</p>
</div>
<div id="geoquery" class="section level3" number="11.5.2">
<h3><span class="header-section-number">11.5.2</span> GEOquery</h3>
<blockquote>
<p>You can explore docs here: <a href="https://bioconductor.org/packages/release/bioc/vignettes/GEOquery/inst/doc/GEOquery.html" class="uri">https://bioconductor.org/packages/release/bioc/vignettes/GEOquery/inst/doc/GEOquery.html</a></p>
</blockquote>
<p><code>GEOquery</code> is a valuable R package for downloading and importing gene expression data directly from public repositories such as the Gene Expression Omnibus (GEO).</p>
<div class="sourceCode" id="cb508"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb508-1"><a href="introduction-to-bioconductor.html#cb508-1" tabindex="-1"></a><span class="co"># Loading the GEOquery library</span></span>
<span id="cb508-2"><a href="introduction-to-bioconductor.html#cb508-2" tabindex="-1"></a><span class="fu">library</span>(GEOquery)</span>
<span id="cb508-3"><a href="introduction-to-bioconductor.html#cb508-3" tabindex="-1"></a></span>
<span id="cb508-4"><a href="introduction-to-bioconductor.html#cb508-4" tabindex="-1"></a><span class="co"># Downloading and importing data from GEO</span></span>
<span id="cb508-5"><a href="introduction-to-bioconductor.html#cb508-5" tabindex="-1"></a>GSE197576 <span class="ot">&lt;-</span> <span class="fu">getGEO</span>(<span class="at">GEO =</span> <span class="st">&quot;GSE197576&quot;</span>, <span class="at">GSEMatrix =</span> <span class="cn">TRUE</span>, <span class="at">destdir =</span> <span class="st">&quot;~/Downloads&quot;</span>)</span>
<span id="cb508-6"><a href="introduction-to-bioconductor.html#cb508-6" tabindex="-1"></a></span>
<span id="cb508-7"><a href="introduction-to-bioconductor.html#cb508-7" tabindex="-1"></a>GSE197576</span></code></pre></div>
<pre><code>## $GSE197576_series_matrix.txt.gz
## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 0 features, 12 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: GSM5920759 GSM5920760 ... GSM5920770 (12 total)
##   varLabels: title geo_accession ... tissue:ch1 (43 total)
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
##   pubMedIds: 35487218 
## Annotation: GPL18573</code></pre>
<div class="sourceCode" id="cb510"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb510-1"><a href="introduction-to-bioconductor.html#cb510-1" tabindex="-1"></a><span class="co"># Accessing expression data</span></span>
<span id="cb510-2"><a href="introduction-to-bioconductor.html#cb510-2" tabindex="-1"></a><span class="fu">exprs</span>(GSE197576<span class="sc">$</span>GSE197576_series_matrix.txt.gz)</span></code></pre></div>
<pre><code>##      GSM5920759 GSM5920760 GSM5920761 GSM5920762 GSM5920763 GSM5920764
##      GSM5920765 GSM5920766 GSM5920767 GSM5920768 GSM5920769 GSM5920770</code></pre>
<p>In this example, it returns an empty <code>ExpressionSet</code> object. We used <code>exprs</code> to return
the matrix but it has no features. You may try a different GEO accession.</p>
<p><code>GEOquery</code> simplifies the process of retrieving gene expression data from public repositories like GEO. Researchers use it to access valuable datasets for their studies.</p>
</div>
<div id="converting-gene-ids" class="section level3" number="11.5.3">
<h3><span class="header-section-number">11.5.3</span> Converting Gene IDs</h3>
<p>In genomics research, integrating data from various sources often involves working with different gene identifier systems. Converting gene IDs is a crucial step to harmonize and standardize data for downstream analysis. Here, we discuss two commonly used methods, biomaRt and org.Hs.eg.db, and explain why these tasks are essential in a researcher’s environment.</p>
<div id="using-biomart" class="section level4" number="11.5.3.1">
<h4><span class="header-section-number">11.5.3.1</span> Using biomaRt</h4>
<blockquote>
<p>You can explore docs here: <a href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html" class="uri">https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html</a></p>
</blockquote>
<p>Converting gene IDs using <code>biomaRt</code> is essential for mapping Ensembl gene IDs to more recognizable gene symbols and associated information.</p>
<div class="sourceCode" id="cb512"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb512-1"><a href="introduction-to-bioconductor.html#cb512-1" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb512-2"><a href="introduction-to-bioconductor.html#cb512-2" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> <span class="fu">useMart</span>(<span class="st">&quot;ensembl&quot;</span>, <span class="at">dataset =</span> <span class="st">&quot;hsapiens_gene_ensembl&quot;</span>)</span>
<span id="cb512-3"><a href="introduction-to-bioconductor.html#cb512-3" tabindex="-1"></a></span>
<span id="cb512-4"><a href="introduction-to-bioconductor.html#cb512-4" tabindex="-1"></a>gene_info <span class="ot">&lt;-</span> <span class="fu">getBM</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;gene_biotype&quot;</span>,</span>
<span id="cb512-5"><a href="introduction-to-bioconductor.html#cb512-5" tabindex="-1"></a>                                  <span class="st">&quot;chromosome_name&quot;</span>, <span class="st">&quot;hgnc_symbol&quot;</span>),</span>
<span id="cb512-6"><a href="introduction-to-bioconductor.html#cb512-6" tabindex="-1"></a>                   <span class="at">filters =</span> <span class="st">&quot;ensembl_gene_id&quot;</span>, </span>
<span id="cb512-7"><a href="introduction-to-bioconductor.html#cb512-7" tabindex="-1"></a>                   <span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;ENSG00000164307&quot;</span>), </span>
<span id="cb512-8"><a href="introduction-to-bioconductor.html#cb512-8" tabindex="-1"></a>                   <span class="at">mart =</span> ensembl)</span>
<span id="cb512-9"><a href="introduction-to-bioconductor.html#cb512-9" tabindex="-1"></a><span class="fu">print</span>(gene_info)</span></code></pre></div>
<pre><code>##   ensembl_gene_id   gene_biotype chromosome_name hgnc_symbol
## 1 ENSG00000164307 protein_coding               5       ERAP1</code></pre>
</div>
<div id="using-org.hs.eg.db" class="section level4" number="11.5.3.2">
<h4><span class="header-section-number">11.5.3.2</span> Using org.Hs.eg.db</h4>
<blockquote>
<p>You can explore docs here: <a href="https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html" class="uri">https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html</a></p>
</blockquote>
<p>Mapping ENTREZIDs to official gene symbols using <code>org.Hs.eg.db</code> is crucial for researchers working with gene expression data.</p>
<p>Why Researchers Do This:</p>
<ul>
<li><p>Consistency in Annotation: ENTREZIDs are a widely accepted and consistent gene identifier system. Mapping other identifiers to ENTREZIDs ensures that gene information is uniform and can be compared across different studies.</p></li>
<li><p>Integration with Other Databases: Many databases and tools use ENTREZIDs (e.g., the KEGG pathway database) as a standard for gene annotation. Converting identifiers to ENTREZIDs facilitates seamless integration with these resources.</p></li>
<li><p>Gene Symbol Mapping: Once converted to ENTREZIDs, researchers can efficiently map these to official gene symbols, providing meaningful and interpretable gene names for further analysis and reporting.</p></li>
</ul>
<div class="sourceCode" id="cb514"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb514-1"><a href="introduction-to-bioconductor.html#cb514-1" tabindex="-1"></a><span class="co"># Loading the org.Hs.eg.db library</span></span>
<span id="cb514-2"><a href="introduction-to-bioconductor.html#cb514-2" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb514-3"><a href="introduction-to-bioconductor.html#cb514-3" tabindex="-1"></a><span class="fu">library</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)</span>
<span id="cb514-4"><a href="introduction-to-bioconductor.html#cb514-4" tabindex="-1"></a></span>
<span id="cb514-5"><a href="introduction-to-bioconductor.html#cb514-5" tabindex="-1"></a><span class="co"># Accessing gene information</span></span>
<span id="cb514-6"><a href="introduction-to-bioconductor.html#cb514-6" tabindex="-1"></a>hg19_genes <span class="ot">&lt;-</span> <span class="fu">genes</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)</span>
<span id="cb514-7"><a href="introduction-to-bioconductor.html#cb514-7" tabindex="-1"></a></span>
<span id="cb514-8"><a href="introduction-to-bioconductor.html#cb514-8" tabindex="-1"></a><span class="co"># Mapping ENTREZIDs to official gene symbols</span></span>
<span id="cb514-9"><a href="introduction-to-bioconductor.html#cb514-9" tabindex="-1"></a>map <span class="ot">&lt;-</span> AnnotationDbi<span class="sc">::</span><span class="fu">select</span>(org.Hs.eg.db, <span class="at">keys =</span> hg19_genes<span class="sc">$</span>gene_id, </span>
<span id="cb514-10"><a href="introduction-to-bioconductor.html#cb514-10" tabindex="-1"></a>                            <span class="at">columns =</span> <span class="st">&quot;SYMBOL&quot;</span>, <span class="at">keytype =</span> <span class="st">&quot;ENTREZID&quot;</span>)</span>
<span id="cb514-11"><a href="introduction-to-bioconductor.html#cb514-11" tabindex="-1"></a></span>
<span id="cb514-12"><a href="introduction-to-bioconductor.html#cb514-12" tabindex="-1"></a><span class="fu">head</span>(map, <span class="at">n =</span> <span class="dv">10</span>)</span></code></pre></div>
<pre><code>##     ENTREZID     SYMBOL
## 1          1       A1BG
## 2         10       NAT2
## 3        100        ADA
## 4       1000       CDH2
## 5      10000       AKT3
## 6  100008586    GAGE12F
## 7  100009676 ZBTB11-AS1
## 8      10001       MED6
## 9      10002      NR2E3
## 10     10003    NAALAD2</code></pre>
<p>In genomics research, this type of mapping is essential when you have data that uses ENTREZIDs, and you want to work with more interpretable gene symbols for analysis or visualization. It allows you to associate gene identifiers with their official names, making the data more understandable and facilitating downstream analyses and interpretation.</p>
</div>
</div>
<div id="complexheatmap" class="section level3" number="11.5.4">
<h3><span class="header-section-number">11.5.4</span> ComplexHeatmap</h3>
<blockquote>
<p>You can explore docs here: <a href="https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html" class="uri">https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html</a></p>
</blockquote>
<p>Visualizing gene expression data is a crucial step in genomics research because it helps researchers gain insights into how genes are expressed under different conditions or in various samples. <code>ComplexHeatmap</code> is a versatile R package that serves as an artistic palette for creating intricate and informative heatmaps. Here’s why visualizing gene expression data using <code>ComplexHeatmap</code> is essential:</p>
<ol style="list-style-type: decimal">
<li><p>Identify Expression Patterns: Gene expression data often involve a large number of genes and samples. Heatmaps allow you to quickly identify patterns in gene expression, such as clusters of genes with similar expression profiles. This can reveal groups of genes that are co-regulated or have similar functions.</p></li>
<li><p>Visualize Differential Expression: When comparing gene expression between different conditions or treatments (e.g., healthy vs. diseased tissue), heatmaps can highlight genes that are significantly upregulated or downregulated. This visualization aids in pinpointing genes of interest for further investigation.</p></li>
<li><p>Sample Relationships: Heatmaps also help in understanding the relationships between samples. For example, you can identify outliers, detect batch effects, or confirm the consistency of replicates by examining how samples cluster based on their expression profiles.</p></li>
<li><p>Publication-Ready Figures: ComplexHeatmap generates high-quality heatmap images that are suitable for inclusion in research papers and presentations. It provides the ability to export heatmaps in various formats (e.g., PDF, PNG) for easy sharing and publication.</p></li>
<li><p>Interactive Exploration: In addition to static heatmaps, <a href="https://www.bioconductor.org/packages/release/bioc/html/InteractiveComplexHeatmap.html">InteractiveComplexHeatmap</a> from the same author supports interactive exploration. You can zoom in on specific sections of the heatmap, hover over cells to view gene names or expression values, and provide interactive tools for your audience to explore the data themselves.</p></li>
</ol>
<p>Le’s go through an example by simulating a count matrix from a gene expression experiment:</p>
<div class="sourceCode" id="cb516"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb516-1"><a href="introduction-to-bioconductor.html#cb516-1" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb516-2"><a href="introduction-to-bioconductor.html#cb516-2" tabindex="-1"></a></span>
<span id="cb516-3"><a href="introduction-to-bioconductor.html#cb516-3" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb516-4"><a href="introduction-to-bioconductor.html#cb516-4" tabindex="-1"></a></span>
<span id="cb516-5"><a href="introduction-to-bioconductor.html#cb516-5" tabindex="-1"></a><span class="co"># Parameters for the negative binomial distribution</span></span>
<span id="cb516-6"><a href="introduction-to-bioconductor.html#cb516-6" tabindex="-1"></a>size_parameter <span class="ot">&lt;-</span> <span class="dv">3</span> <span class="co"># Size parameter (dispersion)</span></span>
<span id="cb516-7"><a href="introduction-to-bioconductor.html#cb516-7" tabindex="-1"></a>mean_parameter <span class="ot">&lt;-</span> <span class="dv">10</span>  <span class="co"># Mean parameter</span></span>
<span id="cb516-8"><a href="introduction-to-bioconductor.html#cb516-8" tabindex="-1"></a></span>
<span id="cb516-9"><a href="introduction-to-bioconductor.html#cb516-9" tabindex="-1"></a><span class="co"># Number of genes (rows) and samples (columns)</span></span>
<span id="cb516-10"><a href="introduction-to-bioconductor.html#cb516-10" tabindex="-1"></a>num_genes <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb516-11"><a href="introduction-to-bioconductor.html#cb516-11" tabindex="-1"></a>num_samples <span class="ot">&lt;-</span> <span class="dv">20</span></span>
<span id="cb516-12"><a href="introduction-to-bioconductor.html#cb516-12" tabindex="-1"></a></span>
<span id="cb516-13"><a href="introduction-to-bioconductor.html#cb516-13" tabindex="-1"></a><span class="co"># Simulate a count table with negative binomial distribution</span></span>
<span id="cb516-14"><a href="introduction-to-bioconductor.html#cb516-14" tabindex="-1"></a>expr_data <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnbinom</span>(num_genes <span class="sc">*</span> num_samples, </span>
<span id="cb516-15"><a href="introduction-to-bioconductor.html#cb516-15" tabindex="-1"></a>                            <span class="at">size =</span> size_parameter, </span>
<span id="cb516-16"><a href="introduction-to-bioconductor.html#cb516-16" tabindex="-1"></a>                            <span class="at">mu =</span> mean_parameter), </span>
<span id="cb516-17"><a href="introduction-to-bioconductor.html#cb516-17" tabindex="-1"></a>                    <span class="at">nrow =</span> num_genes, </span>
<span id="cb516-18"><a href="introduction-to-bioconductor.html#cb516-18" tabindex="-1"></a>                    <span class="at">ncol =</span> num_samples)</span></code></pre></div>
<p>check the quantiles of the data. This is important because we need to know the
ranges of the data so we can map the values to color.</p>
<div class="sourceCode" id="cb517"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb517-1"><a href="introduction-to-bioconductor.html#cb517-1" tabindex="-1"></a><span class="fu">quantile</span>(expr_data, <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.2</span>, <span class="fl">0.5</span>, <span class="fl">0.8</span>, <span class="dv">1</span>))</span></code></pre></div>
<pre><code>##   0%  20%  50%  80% 100% 
##    0    4    9   14   31</code></pre>
<p>Map the color to the values</p>
<div class="sourceCode" id="cb519"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb519-1"><a href="introduction-to-bioconductor.html#cb519-1" tabindex="-1"></a>color_map<span class="ot">&lt;-</span> circlize<span class="sc">::</span><span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">9</span>, <span class="dv">25</span>), <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>))</span></code></pre></div>
<p>Any value beyond 25 will be mapped to the same intensity of redness.</p>
<p>Make the heatmap:</p>
<div class="sourceCode" id="cb520"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb520-1"><a href="introduction-to-bioconductor.html#cb520-1" tabindex="-1"></a><span class="fu">Heatmap</span>(expr_data, <span class="at">col=</span>color_map, <span class="at">name =</span> <span class="st">&quot;gene_expression&quot;</span>)</span></code></pre></div>
<p><img src="10_Introduction_to_BioConductor_files/figure-html/unnamed-chunk-39-1.png" width="672" /></p>
<p>By default, it will cluster both the rows and columns.</p>
<p>Let’s see how it looks if we change the color mapping</p>
<div class="sourceCode" id="cb521"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb521-1"><a href="introduction-to-bioconductor.html#cb521-1" tabindex="-1"></a>color_map2<span class="ot">&lt;-</span> circlize<span class="sc">::</span><span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">25</span>, <span class="dv">31</span>), <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>))</span>
<span id="cb521-2"><a href="introduction-to-bioconductor.html#cb521-2" tabindex="-1"></a><span class="fu">Heatmap</span>(expr_data, <span class="at">col=</span>color_map2, <span class="at">name =</span> <span class="st">&quot;gene_expression&quot;</span>)</span></code></pre></div>
<p><img src="10_Introduction_to_BioConductor_files/figure-html/unnamed-chunk-40-1.png" width="672" />
Now, we see very fewer red cells, but the underlying data is the same! How you map the values to color makes a big difference on how the heatmap look. Note the legend reflects our changes in the color mapping.</p>
<p>We will use it again in the next section.</p>
</div>
<div id="conclusion-29" class="section level3" number="11.5.5">
<h3><span class="header-section-number">11.5.5</span> Conclusion</h3>
<p>This lesson has provided an overview of essential tools and techniques used in genomics research. We have explored the following key topics:</p>
<ul>
<li><p><code>DESeq2</code>: We learned about DESeq2 for differential expression analysis, which is crucial for identifying genes that are differentially expressed under different conditions, such as in disease versus healthy states.</p></li>
<li><p><code>GEOquery</code>: We discussed how to retrieve gene expression data from public repositories like GEO using the GEOquery package. Accessing publicly available datasets is a valuable resource for genomics research.</p></li>
<li><p>Gene ID Conversion: We covered two methods, <code>biomaRt</code> and <code>org.Hs.eg.db</code>, for converting gene IDs. This step is essential for integrating data from various sources and ensuring consistent gene identification.</p></li>
<li><p><code>ComplexHeatmap</code>: We explored the <code>ComplexHeatmap</code> package, a powerful tool for visualizing gene expression data through heatmaps. Visualization aids in identifying patterns and trends in large genomic datasets.</p></li>
</ul>
<p>These tools and techniques are indispensable for genomics researchers, enabling them to analyze, integrate, and visualize genomic data effectively. By mastering these skills, researchers can gain valuable insights into gene expression patterns, biological processes, and potential biomarkers for various conditions and diseases.</p>
</div>
</div>
<div id="real-world-example---tcga-analysis" class="section level2" number="11.6">
<h2><span class="header-section-number">11.6</span> Real-World Example - TCGA Analysis</h2>
<p>In this lesson, we will explore a real-world example of analyzing cancer genomics data from <a href="https://www.cancer.gov/ccg/research/genome-sequencing/tcga">The Cancer Genome Atlas (TCGA)</a> project. TCGA is one of the largest and most renowned cancer sequencing projects, providing access to a wealth of genomic data from various cancer types. We will use R and several bioinformatics packages to download raw RNA-seq counts for 33 different cancer types, convert them to TPM (transcripts per million), and visualize the data in a heatmap.</p>
<div id="introduction-to-tcga" class="section level3" number="11.6.1">
<h3><span class="header-section-number">11.6.1</span> Introduction to TCGA</h3>
<p>The Cancer Genome Atlas (TCGA) project is a groundbreaking initiative that has sequenced approximately 10,000 treatment-naive tumors across 33 different cancer types. It has generated a diverse range of data types, including whole-exome sequencing, whole-genome sequencing, copy-number variation analysis (SNP arrays), bulk RNA-seq, protein expression data (Reverse-Phase Protein Array), and DNA methylation profiles. TCGA has significantly contributed to our understanding of cancer biology and has opened up new avenues for cancer research.</p>
</div>
<div id="why-analyze-tcga-data" class="section level3" number="11.6.2">
<h3><span class="header-section-number">11.6.2</span> Why Analyze TCGA Data?</h3>
<p>Analyzing TCGA data can provide valuable insights into the molecular basis of cancer. Researchers can use this data to identify potential biomarkers, discover novel therapeutic targets, and gain a deeper understanding of the genetic alterations associated with specific cancer types. Moreover, TCGA data is freely accessible, making it a valuable resource for the scientific community.</p>
</div>
<div id="getting-started-1" class="section level3" number="11.6.3">
<h3><span class="header-section-number">11.6.3</span> Getting Started</h3>
<p>We will use the <a href="https://bioconductor.org/packages/release/bioc/html/recount3.html"><code>recount3</code></a> package to access TCGA data. <code>recount3</code> is an online resource that provides RNA-seq gene, exon, and exon-exon junction counts, along with coverage bigWig files for thousands of studies in both human and mouse. It represents the third generation of the <a href="https://rna.recount.bio/">ReCount</a> project.</p>
</div>
<div id="step-1-install-and-load-required-packages" class="section level3" number="11.6.4">
<h3><span class="header-section-number">11.6.4</span> Step 1: Install and Load Required Packages</h3>
<div class="sourceCode" id="cb522"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb522-1"><a href="introduction-to-bioconductor.html#cb522-1" tabindex="-1"></a><span class="co"># Install the recount3 package if not already installed</span></span>
<span id="cb522-2"><a href="introduction-to-bioconductor.html#cb522-2" tabindex="-1"></a><span class="co"># BiocManager::install(&quot;recount3&quot;)</span></span>
<span id="cb522-3"><a href="introduction-to-bioconductor.html#cb522-3" tabindex="-1"></a></span>
<span id="cb522-4"><a href="introduction-to-bioconductor.html#cb522-4" tabindex="-1"></a><span class="co"># Load necessary libraries</span></span>
<span id="cb522-5"><a href="introduction-to-bioconductor.html#cb522-5" tabindex="-1"></a><span class="fu">library</span>(recount3)</span>
<span id="cb522-6"><a href="introduction-to-bioconductor.html#cb522-6" tabindex="-1"></a><span class="fu">library</span>(purrr)</span>
<span id="cb522-7"><a href="introduction-to-bioconductor.html#cb522-7" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb522-8"><a href="introduction-to-bioconductor.html#cb522-8" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code></pre></div>
<ul>
<li><p>recount3: This package is crucial for accessing TCGA data. It allows us to retrieve RNA-seq gene counts and other genomic information from the TCGA project.</p></li>
<li><p>purrr: The purrr package is used for functional programming in R. We use it to apply functions to elements of a list, which is particularly useful for handling multiple datasets or projects.</p></li>
<li><p>dplyr: dplyr is a powerful package for data manipulation. It helps us filter and process data efficiently, making it easier to work with large datasets like TCGA.</p></li>
<li><p>ggplot2: For data visualization, we use the ggplot2 package. It allows us to create high-quality graphs and plots, which can be essential for presenting our findings.</p></li>
</ul>
<p>By loading these packages, we ensure that we have access to the tools needed to analyze and visualize TCGA data effectively.</p>
</div>
<div id="step-2-retrieve-tcga-project-information" class="section level3" number="11.6.5">
<h3><span class="header-section-number">11.6.5</span> Step 2: Retrieve TCGA Project Information</h3>
<p>Let’s fetch information about available TCGA projects. TCGA encompasses a wide range of cancer types and studies. We filter the projects to focus only on those that originate from TCGA data sources.</p>
<div class="sourceCode" id="cb523"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb523-1"><a href="introduction-to-bioconductor.html#cb523-1" tabindex="-1"></a><span class="co"># Get information about available TCGA projects</span></span>
<span id="cb523-2"><a href="introduction-to-bioconductor.html#cb523-2" tabindex="-1"></a>human_projects <span class="ot">&lt;-</span> <span class="fu">available_projects</span>()</span>
<span id="cb523-3"><a href="introduction-to-bioconductor.html#cb523-3" tabindex="-1"></a></span>
<span id="cb523-4"><a href="introduction-to-bioconductor.html#cb523-4" tabindex="-1"></a><span class="co"># Filter projects that are from TCGA data sources</span></span>
<span id="cb523-5"><a href="introduction-to-bioconductor.html#cb523-5" tabindex="-1"></a>tcga_info <span class="ot">&lt;-</span> <span class="fu">subset</span>(</span>
<span id="cb523-6"><a href="introduction-to-bioconductor.html#cb523-6" tabindex="-1"></a>    human_projects,</span>
<span id="cb523-7"><a href="introduction-to-bioconductor.html#cb523-7" tabindex="-1"></a>    file_source <span class="sc">==</span> <span class="st">&quot;tcga&quot;</span> <span class="sc">&amp;</span> project_type <span class="sc">==</span> <span class="st">&quot;data_sources&quot;</span></span>
<span id="cb523-8"><a href="introduction-to-bioconductor.html#cb523-8" tabindex="-1"></a>)</span>
<span id="cb523-9"><a href="introduction-to-bioconductor.html#cb523-9" tabindex="-1"></a></span>
<span id="cb523-10"><a href="introduction-to-bioconductor.html#cb523-10" tabindex="-1"></a></span>
<span id="cb523-11"><a href="introduction-to-bioconductor.html#cb523-11" tabindex="-1"></a><span class="fu">head</span>(tcga_info)</span></code></pre></div>
<pre><code>##      project organism file_source      project_home project_type n_samples
## 8710     ACC    human        tcga data_sources/tcga data_sources        79
## 8711    BLCA    human        tcga data_sources/tcga data_sources       433
## 8712    BRCA    human        tcga data_sources/tcga data_sources      1256
## 8713    CESC    human        tcga data_sources/tcga data_sources       309
## 8714    CHOL    human        tcga data_sources/tcga data_sources        45
## 8715    COAD    human        tcga data_sources/tcga data_sources       546</code></pre>
<p>This step is essential to identify the relevant projects and data sources within TCGA. By narrowing our focus to TCGA data sources, we ensure that we are working with the specific datasets we need for our analysis.</p>
</div>
<div id="step-3-create-a-rangedsummarizedexperiment-object" class="section level3" number="11.6.6">
<h3><span class="header-section-number">11.6.6</span> Step 3: Create a RangedSummarizedExperiment Object</h3>
<p>Now that we have identified the TCGA data sources of interest, we proceed to create a data structure called a <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html"><code>RangedSummarizedExperiment</code></a> object. This object allows us to efficiently organize and work with genomic data, including RNA-seq counts.</p>
<div class="sourceCode" id="cb525"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb525-1"><a href="introduction-to-bioconductor.html#cb525-1" tabindex="-1"></a><span class="co"># Create a list of RangedSummarizedExperiment objects</span></span>
<span id="cb525-2"><a href="introduction-to-bioconductor.html#cb525-2" tabindex="-1"></a>proj_info <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map</span>(<span class="fu">seq</span>(<span class="fu">nrow</span>(tcga_info)), <span class="sc">~</span>tcga_info[.x, ])</span>
<span id="cb525-3"><a href="introduction-to-bioconductor.html#cb525-3" tabindex="-1"></a>rse_tcga <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map</span>(proj_info, <span class="sc">~</span><span class="fu">create_rse</span>(.x))</span></code></pre></div>
<p>Here, we create a list of <code>RangedSummarizedExperiment</code> objects, one for each project within TCGA. These objects will serve as containers for the gene expression data, making it easier to manipulate and analyze the information.</p>
<p>The first time you use <code>recount3</code>, it will ask:</p>
<blockquote>
<p>/Users/tommytang/Library/Caches/org.R-project.R/R/recount3 does not exist, create directory? (yes/no): yes</p>
</blockquote>
<p><code>rse_tcga</code> is a list of <code>RangedSummarizedExperimentobject</code>, let’s take a look at one of them. <code>RangedSummarizedExperiment</code> is a child of the <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb526"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb526-1"><a href="introduction-to-bioconductor.html#cb526-1" tabindex="-1"></a>rse_tcga[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 63856 79 
## metadata(8): time_created recount3_version ... annotation recount3_url
## assays(1): raw_counts
## rownames(63856): ENSG00000278704.1 ENSG00000277400.1 ...
##   ENSG00000182484.15_PAR_Y ENSG00000227159.8_PAR_Y
## rowData names(10): source type ... havana_gene tag
## colnames(79): 43e715bf-28d9-4b5e-b762-8cd1b69a430e
##   1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872 ...
##   a08b85ea-d1e7-4b77-8dec-36294305b9f7
##   aa2d53e5-d389-4332-9dd5-a736052e48f8
## colData names(937): rail_id external_id ... recount_seq_qc.errq
##   BigWigURL</code></pre>
</div>
<div id="step-4-explore-tcga-data" class="section level3" number="11.6.7">
<h3><span class="header-section-number">11.6.7</span> Step 4: Explore TCGA Data</h3>
<p>With our <code>RangedSummarizedExperiment</code> objects in place, we can now explore the TCGA data in more detail. Let’s look at three aspects: gene counts, gene information, and metadata.</p>
<div id="gene-counts" class="section level4" number="11.6.7.1">
<h4><span class="header-section-number">11.6.7.1</span> Gene Counts</h4>
<p>We start by examining the raw RNA-seq counts, which represent how many times each gene was sequenced in each sample.</p>
<div class="sourceCode" id="cb528"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb528-1"><a href="introduction-to-bioconductor.html#cb528-1" tabindex="-1"></a><span class="co"># Access raw gene counts from one RangedSummarizedExperiment object</span></span>
<span id="cb528-2"><a href="introduction-to-bioconductor.html#cb528-2" tabindex="-1"></a>raw_counts <span class="ot">&lt;-</span> rse_tcga[[<span class="dv">1</span>]]<span class="sc">@</span>assays<span class="sc">@</span>data<span class="sc">$</span>raw_counts[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb528-3"><a href="introduction-to-bioconductor.html#cb528-3" tabindex="-1"></a></span>
<span id="cb528-4"><a href="introduction-to-bioconductor.html#cb528-4" tabindex="-1"></a>raw_counts</span></code></pre></div>
<pre><code>##                   43e715bf-28d9-4b5e-b762-8cd1b69a430e
## ENSG00000278704.1                                    0
## ENSG00000277400.1                                    0
## ENSG00000274847.1                                    0
## ENSG00000277428.1                                    0
## ENSG00000276256.1                                    0
##                   1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872
## ENSG00000278704.1                                    0
## ENSG00000277400.1                                    0
## ENSG00000274847.1                                    0
## ENSG00000277428.1                                    0
## ENSG00000276256.1                                    0
##                   93b382e4-9c9a-43f5-bd3b-502cc260b886
## ENSG00000278704.1                                    0
## ENSG00000277400.1                                    0
## ENSG00000274847.1                                    0
## ENSG00000277428.1                                    0
## ENSG00000276256.1                                    0
##                   1f39dadd-3655-474e-ba4c-a5bd32c97a8b
## ENSG00000278704.1                                    0
## ENSG00000277400.1                                    0
## ENSG00000274847.1                                    0
## ENSG00000277428.1                                    0
## ENSG00000276256.1                                    0
##                   8c8c09b9-ec83-45ec-bc4c-0ba92de60acb
## ENSG00000278704.1                                    0
## ENSG00000277400.1                                    0
## ENSG00000274847.1                                    0
## ENSG00000277428.1                                    0
## ENSG00000276256.1                                    0</code></pre>
<p>This line extracts a subset of raw gene counts from the first dataset in the <code>RangedSummarizedExperiment</code> object <code>rse_tcga</code>. It takes the first 5 genes and the first 5 samples, providing a small portion of the data for analysis or exploration.</p>
<p>The <code>raw_counts</code> matrix provides a glimpse into the data, allowing us to see the sequencing counts for a subset of genes and samples.</p>
</div>
<div id="gene-information" class="section level4" number="11.6.7.2">
<h4><span class="header-section-number">11.6.7.2</span> Gene Information</h4>
<p>Next, we retrieve information about the genes included in the dataset. This information includes details such as genomic location, gene type, and gene names.</p>
<div class="sourceCode" id="cb530"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb530-1"><a href="introduction-to-bioconductor.html#cb530-1" tabindex="-1"></a><span class="co"># Access gene information</span></span>
<span id="cb530-2"><a href="introduction-to-bioconductor.html#cb530-2" tabindex="-1"></a>gene_info <span class="ot">&lt;-</span> rse_tcga[[<span class="dv">1</span>]]<span class="sc">@</span>rowRanges</span>
<span id="cb530-3"><a href="introduction-to-bioconductor.html#cb530-3" tabindex="-1"></a></span>
<span id="cb530-4"><a href="introduction-to-bioconductor.html#cb530-4" tabindex="-1"></a>gene_info</span></code></pre></div>
<pre><code>## GRanges object with 63856 ranges and 10 metadata columns:
##                              seqnames            ranges strand |   source
##                                 &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
##          ENSG00000278704.1 GL000009.2       56140-58376      - |  ENSEMBL
##          ENSG00000277400.1 GL000194.1      53590-115018      - |  ENSEMBL
##          ENSG00000274847.1 GL000194.1      53594-115055      - |  ENSEMBL
##          ENSG00000277428.1 GL000195.1       37434-37534      - |  ENSEMBL
##          ENSG00000276256.1 GL000195.1       42939-49164      - |  ENSEMBL
##                        ...        ...               ...    ... .      ...
##   ENSG00000124334.17_PAR_Y       chrY 57184101-57197337      + |   HAVANA
##   ENSG00000185203.12_PAR_Y       chrY 57201143-57203357      - |   HAVANA
##    ENSG00000270726.6_PAR_Y       chrY 57190738-57208756      + |   HAVANA
##   ENSG00000182484.15_PAR_Y       chrY 57207346-57212230      + |   HAVANA
##    ENSG00000227159.8_PAR_Y       chrY 57212184-57214397      - |   HAVANA
##                                type bp_length     phase                gene_id
##                            &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;            &lt;character&gt;
##          ENSG00000278704.1     gene      2237      &lt;NA&gt;      ENSG00000278704.1
##          ENSG00000277400.1     gene      2179      &lt;NA&gt;      ENSG00000277400.1
##          ENSG00000274847.1     gene      1599      &lt;NA&gt;      ENSG00000274847.1
##          ENSG00000277428.1     gene       101      &lt;NA&gt;      ENSG00000277428.1
##          ENSG00000276256.1     gene      2195      &lt;NA&gt;      ENSG00000276256.1
##                        ...      ...       ...       ...                    ...
##   ENSG00000124334.17_PAR_Y     gene      2504      &lt;NA&gt; ENSG00000124334.17_P..
##   ENSG00000185203.12_PAR_Y     gene      1054      &lt;NA&gt; ENSG00000185203.12_P..
##    ENSG00000270726.6_PAR_Y     gene       773      &lt;NA&gt; ENSG00000270726.6_PA..
##   ENSG00000182484.15_PAR_Y     gene      4618      &lt;NA&gt; ENSG00000182484.15_P..
##    ENSG00000227159.8_PAR_Y     gene      1306      &lt;NA&gt; ENSG00000227159.8_PA..
##                                         gene_type   gene_name       level
##                                       &lt;character&gt; &lt;character&gt; &lt;character&gt;
##          ENSG00000278704.1         protein_coding  BX004987.1           3
##          ENSG00000277400.1         protein_coding  AC145212.2           3
##          ENSG00000274847.1         protein_coding  AC145212.1           3
##          ENSG00000277428.1               misc_RNA       Y_RNA           3
##          ENSG00000276256.1         protein_coding  AC011043.1           3
##                        ...                    ...         ...         ...
##   ENSG00000124334.17_PAR_Y         protein_coding        IL9R           2
##   ENSG00000185203.12_PAR_Y              antisense      WASIR1           2
##    ENSG00000270726.6_PAR_Y   processed_transcript AJ271736.10           2
##   ENSG00000182484.15_PAR_Y transcribed_unproces..      WASH6P           2
##    ENSG00000227159.8_PAR_Y unprocessed_pseudogene    DDX11L16           2
##                                     havana_gene         tag
##                                     &lt;character&gt; &lt;character&gt;
##          ENSG00000278704.1                 &lt;NA&gt;        &lt;NA&gt;
##          ENSG00000277400.1                 &lt;NA&gt;        &lt;NA&gt;
##          ENSG00000274847.1                 &lt;NA&gt;        &lt;NA&gt;
##          ENSG00000277428.1                 &lt;NA&gt;        &lt;NA&gt;
##          ENSG00000276256.1                 &lt;NA&gt;        &lt;NA&gt;
##                        ...                  ...         ...
##   ENSG00000124334.17_PAR_Y OTTHUMG00000022720.1         PAR
##   ENSG00000185203.12_PAR_Y OTTHUMG00000022676.3         PAR
##    ENSG00000270726.6_PAR_Y OTTHUMG00000184987.2         PAR
##   ENSG00000182484.15_PAR_Y OTTHUMG00000022677.5         PAR
##    ENSG00000227159.8_PAR_Y OTTHUMG00000022678.1         PAR
##   -------
##   seqinfo: 374 sequences from an unspecified genome; no seqlengths</code></pre>
<p>The <code>gene_info</code> object provides essential context about the genes being studied, enabling us to interpret the gene expression data more effectively.</p>
</div>
<div id="metadata-information" class="section level4" number="11.6.7.3">
<h4><span class="header-section-number">11.6.7.3</span> Metadata Information</h4>
<p>Lastly, we access metadata information associated with the samples. Metadata contains additional details about each sample, such as unique identifiers, external IDs, and the TCGA study to which each sample belongs.</p>
<div class="sourceCode" id="cb532"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb532-1"><a href="introduction-to-bioconductor.html#cb532-1" tabindex="-1"></a><span class="co"># Access metadata information</span></span>
<span id="cb532-2"><a href="introduction-to-bioconductor.html#cb532-2" tabindex="-1"></a>metadata_info<span class="ot">&lt;-</span> rse_tcga[[<span class="dv">1</span>]]<span class="sc">@</span>colData<span class="sc">@</span>listData <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> <span class="st">`</span><span class="at">[</span><span class="st">`</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)</span>
<span id="cb532-3"><a href="introduction-to-bioconductor.html#cb532-3" tabindex="-1"></a></span>
<span id="cb532-4"><a href="introduction-to-bioconductor.html#cb532-4" tabindex="-1"></a>metadata_info </span></code></pre></div>
<pre><code>##   rail_id                          external_id study
## 1  106797 43e715bf-28d9-4b5e-b762-8cd1b69a430e   ACC
## 2  110230 1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872   ACC
## 3  110773 93b382e4-9c9a-43f5-bd3b-502cc260b886   ACC
## 4  110869 1f39dadd-3655-474e-ba4c-a5bd32c97a8b   ACC
## 5  116503 8c8c09b9-ec83-45ec-bc4c-0ba92de60acb   ACC
##              tcga.tcga_barcode                     tcga.gdc_file_id
## 1 TCGA-OR-A5KU-01A-11R-A29S-07 43e715bf-28d9-4b5e-b762-8cd1b69a430e
## 2 TCGA-P6-A5OG-01A-22R-A29S-07 1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872
## 3 TCGA-OR-A5K5-01A-11R-A29S-07 93b382e4-9c9a-43f5-bd3b-502cc260b886
## 4 TCGA-OR-A5K4-01A-11R-A29S-07 1f39dadd-3655-474e-ba4c-a5bd32c97a8b
## 5 TCGA-OR-A5LP-01A-11R-A29S-07 8c8c09b9-ec83-45ec-bc4c-0ba92de60acb</code></pre>
<blockquote>
<p>NOTE: <code>[</code> is a function itself for subsetting.</p>
</blockquote>
<p>This line creates a dataframe (metadata_info) containing metadata information from the first dataset in the <code>RangedSummarizedExperiment</code> object <code>rse_tcga</code>. It selects the first 5 rows and first 5 columns of this metadata for examination or analysis. Metadata often includes details about the samples, helping researchers understand their characteristics and context within the dataset.</p>
<p>This metadata is crucial for tracking and organizing the samples and understanding their context within the TCGA project.</p>
</div>
</div>
<div id="step-5-data-transformation---converting-raw-counts-to-tpm" class="section level3" number="11.6.8">
<h3><span class="header-section-number">11.6.8</span> Step 5: Data Transformation - Converting Raw Counts to TPM</h3>
<p>In this step, we will convert the raw RNA-seq counts into TPM (Transcripts Per Million) values. This transformation normalizes the data, making it comparable across samples and genes. TPM accounts for both the length of genes and the total number of reads in each sample.</p>
<p>First, we define a function called <code>count2tpm</code> that takes a <code>RangedSummarizedExperiment</code> object (rse) as input and performs the TPM conversion. Here’s what each part of the function does:</p>
<ol style="list-style-type: decimal">
<li><p>We extract the raw gene count matrix from the rse object.</p></li>
<li><p>We retrieve the effective gene length for each gene, which is needed for TPM calculation. In this example, we use the gene length information provided in the TCGA data.</p></li>
<li><p>We calculate the Reads Per Kilobase (RPK) for each gene by dividing the raw counts by the gene length.</p></li>
<li><p>We calculate the sum of RPK values for each sample (column) and divide it by 1,000,000 (1e6) to scale to a per-million basis.</p></li>
<li><p>Finally, we calculate TPM values by dividing the RPK values for each gene by the per-million scale factor.</p></li>
</ol>
<div class="sourceCode" id="cb534"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb534-1"><a href="introduction-to-bioconductor.html#cb534-1" tabindex="-1"></a>genes_of_interest<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;MSLN&quot;</span>, <span class="st">&quot;EGFR&quot;</span>, <span class="st">&quot;ERBB2&quot;</span>, <span class="st">&quot;CEACAM5&quot;</span>, <span class="st">&quot;NECTIN4&quot;</span>, <span class="st">&quot;EPCAM&quot;</span>, </span>
<span id="cb534-2"><a href="introduction-to-bioconductor.html#cb534-2" tabindex="-1"></a>                      <span class="st">&quot;MUC16&quot;</span>, <span class="st">&quot;MUC1&quot;</span>, <span class="st">&quot;CD276&quot;</span>, <span class="st">&quot;FOLH1&quot;</span>, <span class="st">&quot;DLL3&quot;</span>, <span class="st">&quot;VTCN1&quot;</span>, </span>
<span id="cb534-3"><a href="introduction-to-bioconductor.html#cb534-3" tabindex="-1"></a>                      <span class="st">&quot;PROM1&quot;</span>, <span class="st">&quot;PVR&quot;</span>, <span class="st">&quot;CLDN6&quot;</span>, <span class="st">&quot;MET&quot;</span>, <span class="st">&quot;FOLR1&quot;</span>, <span class="st">&quot;TNFRSF10B&quot;</span>, </span>
<span id="cb534-4"><a href="introduction-to-bioconductor.html#cb534-4" tabindex="-1"></a>                      <span class="st">&quot;TACSTD2&quot;</span>, <span class="st">&quot;CD24&quot;</span>)</span>
<span id="cb534-5"><a href="introduction-to-bioconductor.html#cb534-5" tabindex="-1"></a></span>
<span id="cb534-6"><a href="introduction-to-bioconductor.html#cb534-6" tabindex="-1"></a>count2tpm <span class="ot">&lt;-</span> <span class="cf">function</span>(rse) {</span>
<span id="cb534-7"><a href="introduction-to-bioconductor.html#cb534-7" tabindex="-1"></a>    count_matrix <span class="ot">&lt;-</span> rse<span class="sc">@</span>assays<span class="sc">@</span>data<span class="sc">$</span>raw_counts</span>
<span id="cb534-8"><a href="introduction-to-bioconductor.html#cb534-8" tabindex="-1"></a>    gene_length <span class="ot">&lt;-</span> rse<span class="sc">@</span>rowRanges<span class="sc">$</span>bp_length</span>
<span id="cb534-9"><a href="introduction-to-bioconductor.html#cb534-9" tabindex="-1"></a>    reads_per_rpk <span class="ot">&lt;-</span> count_matrix <span class="sc">/</span> gene_length</span>
<span id="cb534-10"><a href="introduction-to-bioconductor.html#cb534-10" tabindex="-1"></a>    per_mil_scale <span class="ot">&lt;-</span> <span class="fu">colSums</span>(reads_per_rpk) <span class="sc">/</span> <span class="fl">1e6</span></span>
<span id="cb534-11"><a href="introduction-to-bioconductor.html#cb534-11" tabindex="-1"></a>    tpm_matrix <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">t</span>(reads_per_rpk) <span class="sc">/</span> per_mil_scale)</span>
<span id="cb534-12"><a href="introduction-to-bioconductor.html#cb534-12" tabindex="-1"></a>    </span>
<span id="cb534-13"><a href="introduction-to-bioconductor.html#cb534-13" tabindex="-1"></a>    <span class="co"># Make sure they match the ENSG and gene order</span></span>
<span id="cb534-14"><a href="introduction-to-bioconductor.html#cb534-14" tabindex="-1"></a>    gene_ind <span class="ot">&lt;-</span> rse<span class="sc">@</span>rowRanges<span class="sc">$</span>gene_name <span class="sc">%in%</span> genes_of_interest</span>
<span id="cb534-15"><a href="introduction-to-bioconductor.html#cb534-15" tabindex="-1"></a>    tpm_submatrix <span class="ot">&lt;-</span> tpm_matrix[gene_ind,]</span>
<span id="cb534-16"><a href="introduction-to-bioconductor.html#cb534-16" tabindex="-1"></a>    <span class="fu">rownames</span>(tpm_submatrix) <span class="ot">&lt;-</span> rse<span class="sc">@</span>rowRanges[gene_ind, ]<span class="sc">$</span>gene_name</span>
<span id="cb534-17"><a href="introduction-to-bioconductor.html#cb534-17" tabindex="-1"></a>    </span>
<span id="cb534-18"><a href="introduction-to-bioconductor.html#cb534-18" tabindex="-1"></a>    <span class="fu">return</span>(tpm_submatrix)</span>
<span id="cb534-19"><a href="introduction-to-bioconductor.html#cb534-19" tabindex="-1"></a>}</span></code></pre></div>
<p>After defining the <code>count2tpm</code> function, we apply it to each of our <code>RangedSummarizedExperiment</code> objects stored in <code>rse_tcga</code>. This step converts the raw counts to TPM values and subsets the data to include only the genes of interest (specified in the genes_of_interest vector). The resulting <code>tpm_data</code> is a list of TPM matrices for each TCGA project.</p>
</div>
<div id="step-6-combine-data-and-metadata" class="section level3" number="11.6.9">
<h3><span class="header-section-number">11.6.9</span> Step 6: Combine Data and Metadata</h3>
<p>In this step, we will combine the TPM data matrices from different TCGA projects and merge the associated metadata. This process is essential for creating a comprehensive dataset for further analysis and visualization.</p>
<div id="combining-tpm-data-matrices" class="section level4" number="11.6.9.1">
<h4><span class="header-section-number">11.6.9.1</span> Combining TPM Data Matrices</h4>
<p>We have already obtained TPM values for each TCGA project and stored them in the <code>tpm_data</code> list, where each element represents a TPM matrix for one project. Now, we will combine these matrices into a single matrix, <code>tpm_data2</code>, which will contain TPM values for all samples across all projects.</p>
<div class="sourceCode" id="cb535"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb535-1"><a href="introduction-to-bioconductor.html#cb535-1" tabindex="-1"></a><span class="co"># Convert raw count matrix per cancer type to TPM and subset to only the genes of interest</span></span>
<span id="cb535-2"><a href="introduction-to-bioconductor.html#cb535-2" tabindex="-1"></a>tpm_data <span class="ot">&lt;-</span> <span class="fu">map</span>(rse_tcga, count2tpm)</span>
<span id="cb535-3"><a href="introduction-to-bioconductor.html#cb535-3" tabindex="-1"></a></span>
<span id="cb535-4"><a href="introduction-to-bioconductor.html#cb535-4" tabindex="-1"></a><span class="co"># Combine the TPM data matrices into one matrix</span></span>
<span id="cb535-5"><a href="introduction-to-bioconductor.html#cb535-5" tabindex="-1"></a>tpm_data2 <span class="ot">&lt;-</span> <span class="fu">do.call</span>(cbind, tpm_data)</span></code></pre></div>
<p>In the code above, we use the <code>map()</code> function to apply the <code>count2tpm</code> function to each <code>RangedSummarizedExperiment</code> object in rse_tcga. The <code>do.call()</code> function then combines the resulting TPM matrices horizontally (column-wise), creating the <code>tpm_data2</code> matrix.</p>
</div>
<div id="combining-metadata" class="section level4" number="11.6.9.2">
<h4><span class="header-section-number">11.6.9.2</span> Combining Metadata</h4>
<p>Next, we need to combine the metadata associated with each TCGA project into a single metadata table. This metadata contains information about the samples, such as sample type, study, and unique identifiers.</p>
<div class="sourceCode" id="cb536"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb536-1"><a href="introduction-to-bioconductor.html#cb536-1" tabindex="-1"></a><span class="co"># Get the metadata columns for each project and convert them to data frames</span></span>
<span id="cb536-2"><a href="introduction-to-bioconductor.html#cb536-2" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">map</span>(rse_tcga, <span class="sc">~</span>.x<span class="sc">@</span>colData<span class="sc">@</span>listData <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>())</span>
<span id="cb536-3"><a href="introduction-to-bioconductor.html#cb536-3" tabindex="-1"></a></span>
<span id="cb536-4"><a href="introduction-to-bioconductor.html#cb536-4" tabindex="-1"></a><span class="co"># Combine the metadata data frames into one data frame</span></span>
<span id="cb536-5"><a href="introduction-to-bioconductor.html#cb536-5" tabindex="-1"></a>metadata2 <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, metadata)</span></code></pre></div>
<p>In this code, we again use the <code>map()</code> function to extract metadata columns from each <code>RangedSummarizedExperiment</code> object. We convert these columns into data frames and then use <code>bind_rows()</code> to vertically stack them into the <code>metadata2</code> data frame.
<code>do.call</code> is a function from base R. <code>do.call(cbind)</code> is similar to <code>dplyr::bind_rows</code>, and <code>do.call(rbind</code>) is similar to <code>dplyr::bind_cols()</code>.</p>
</div>
<div id="checking-the-dimensions" class="section level4" number="11.6.9.3">
<h4><span class="header-section-number">11.6.9.3</span> Checking the Dimensions</h4>
<p>Finally, let’s check the dimensions of our combined data matrix and metadata data frame to ensure that the combination was successful.</p>
<div class="sourceCode" id="cb537"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb537-1"><a href="introduction-to-bioconductor.html#cb537-1" tabindex="-1"></a><span class="co"># Check the dimensions of the combined TPM data matrix</span></span>
<span id="cb537-2"><a href="introduction-to-bioconductor.html#cb537-2" tabindex="-1"></a><span class="fu">dim</span>(tpm_data2)</span></code></pre></div>
<pre><code>## [1]    20 11348</code></pre>
<div class="sourceCode" id="cb539"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb539-1"><a href="introduction-to-bioconductor.html#cb539-1" tabindex="-1"></a><span class="co"># Check the dimensions of the combined metadata data frame</span></span>
<span id="cb539-2"><a href="introduction-to-bioconductor.html#cb539-2" tabindex="-1"></a><span class="fu">dim</span>(metadata2)</span></code></pre></div>
<pre><code>## [1] 11348   937</code></pre>
<p>Running the <code>dim()</code> function on <code>tpm_data2</code> will give you the dimensions of the TPM data matrix, which should indicate the number of genes and samples in the combined dataset. Similarly, checking the dimensions of <code>metadata2</code> will confirm the number of metadata columns and rows, which correspond to the samples and their associated information.</p>
<p>With this combined dataset, you can proceed with various analyses, visualizations, and explorations of gene expression patterns and relationships with clinical metadata across different TCGA projects.</p>
</div>
</div>
<div id="step-7-renaming-metadata-columns" class="section level3" number="11.6.10">
<h3><span class="header-section-number">11.6.10</span> Step 7: Renaming Metadata Columns</h3>
<p>Now, let’s proceed with renaming some of the metadata columns for clarity and convenience. We will select specific columns from the metadata and create a new column called <code>sample_type</code> based on the <code>tcga.cgc_sample_sample_typ</code>e column’s values. This new column will categorize samples into “cancer,” “metastatic,” or “normal” based on their sample type.</p>
<div class="sourceCode" id="cb541"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb541-1"><a href="introduction-to-bioconductor.html#cb541-1" tabindex="-1"></a>metadata2 <span class="ot">&lt;-</span> metadata2 <span class="sc">%&gt;%</span></span>
<span id="cb541-2"><a href="introduction-to-bioconductor.html#cb541-2" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(tcga.tcga_barcode, tcga.cgc_sample_sample_type, study) <span class="sc">%&gt;%</span></span>
<span id="cb541-3"><a href="introduction-to-bioconductor.html#cb541-3" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sample_type =</span> <span class="fu">case_when</span>(</span>
<span id="cb541-4"><a href="introduction-to-bioconductor.html#cb541-4" tabindex="-1"></a>    tcga.cgc_sample_sample_type <span class="sc">==</span> <span class="st">&quot;Additional - New Primary&quot;</span> <span class="sc">~</span> <span class="st">&quot;cancer&quot;</span>,</span>
<span id="cb541-5"><a href="introduction-to-bioconductor.html#cb541-5" tabindex="-1"></a>    tcga.cgc_sample_sample_type <span class="sc">==</span> <span class="st">&quot;Additional Metastatic&quot;</span> <span class="sc">~</span> <span class="st">&quot;metastatic&quot;</span>,</span>
<span id="cb541-6"><a href="introduction-to-bioconductor.html#cb541-6" tabindex="-1"></a>    tcga.cgc_sample_sample_type <span class="sc">==</span> <span class="st">&quot;Metastatic&quot;</span> <span class="sc">~</span> <span class="st">&quot;metastatic&quot;</span>,</span>
<span id="cb541-7"><a href="introduction-to-bioconductor.html#cb541-7" tabindex="-1"></a>    tcga.cgc_sample_sample_type <span class="sc">==</span> <span class="st">&quot;Primary Blood Derived Cancer - Peripheral Blood &quot;</span> <span class="sc">~</span> <span class="st">&quot;cancer&quot;</span>,</span>
<span id="cb541-8"><a href="introduction-to-bioconductor.html#cb541-8" tabindex="-1"></a>    tcga.cgc_sample_sample_type <span class="sc">==</span> <span class="st">&quot;Primary Tumor&quot;</span> <span class="sc">~</span> <span class="st">&quot;cancer&quot;</span>,</span>
<span id="cb541-9"><a href="introduction-to-bioconductor.html#cb541-9" tabindex="-1"></a>    tcga.cgc_sample_sample_type <span class="sc">==</span> <span class="st">&quot;Recurrent Tumor&quot;</span> <span class="sc">~</span> <span class="st">&quot;cancer&quot;</span>,</span>
<span id="cb541-10"><a href="introduction-to-bioconductor.html#cb541-10" tabindex="-1"></a>    tcga.cgc_sample_sample_type <span class="sc">==</span> <span class="st">&quot;Solid Tissue Normal&quot;</span> <span class="sc">~</span> <span class="st">&quot;normal&quot;</span></span>
<span id="cb541-11"><a href="introduction-to-bioconductor.html#cb541-11" tabindex="-1"></a>  ))</span></code></pre></div>
<p>In the code above, we use the <code>select()</code> function to choose specific columns from the metadata that we want to retain. We then create a new <code>sample_type</code> column using <code>mutate()</code> and <code>case_when()</code> based on the values of <code>tcga.cgc_sample_sample_type</code>.</p>
</div>
<div id="step-8-merging-into-a-single-dataframe" class="section level3" number="11.6.11">
<h3><span class="header-section-number">11.6.11</span> Step 8: Merging into a single DataFrame</h3>
<p>With the TPM data matrix and the updated metadata, we can combine them into a single dataframe named <code>final_df</code>. This dataframe will have TPM values for the selected genes along with associated metadata columns.</p>
<div class="sourceCode" id="cb542"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb542-1"><a href="introduction-to-bioconductor.html#cb542-1" tabindex="-1"></a><span class="co"># Combine the TPM data matrix and metadata into a single dataframe</span></span>
<span id="cb542-2"><a href="introduction-to-bioconductor.html#cb542-2" tabindex="-1"></a>final_df <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">t</span>(tpm_data2), metadata2)</span></code></pre></div>
<p>In this code, we use <code>cbind()</code> to merge the transposed TPM data matrix <code>(t(tpm_data2))</code> with the metadata <code>(metadata2)</code> to create the <code>final_df</code> dataframe.</p>
<div id="displaying-the-head-of-the-combined-dataframe" class="section level4" number="11.6.11.1">
<h4><span class="header-section-number">11.6.11.1</span> Displaying the Head of the Combined Dataframe</h4>
<p>Finally, let’s check the first few rows of the combined dataframe to ensure that the merging process was successful:</p>
<div class="sourceCode" id="cb543"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb543-1"><a href="introduction-to-bioconductor.html#cb543-1" tabindex="-1"></a><span class="co"># Display the first few rows of the combined dataframe</span></span>
<span id="cb543-2"><a href="introduction-to-bioconductor.html#cb543-2" tabindex="-1"></a><span class="fu">head</span>(final_df)</span></code></pre></div>
<pre><code>##                                         TACSTD2      VTCN1       MUC1
## 43e715bf-28d9-4b5e-b762-8cd1b69a430e  0.7035937 0.00000000 0.67502205
## 1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872 25.4360736 0.00000000 2.01525394
## 93b382e4-9c9a-43f5-bd3b-502cc260b886  1.5756197 0.00000000 0.90784666
## 1f39dadd-3655-474e-ba4c-a5bd32c97a8b  0.2702156 0.09099681 0.04293345
## 8c8c09b9-ec83-45ec-bc4c-0ba92de60acb  0.4122814 0.00000000 0.11484380
## 85a86b91-4f24-4e77-ae2d-520f8e205efc  4.5469193 4.85973690 0.04208195
##                                         NECTIN4     FOLH1      FOLR1     CD276
## 43e715bf-28d9-4b5e-b762-8cd1b69a430e 0.08620727  7.213342 0.00000000  52.75981
## 1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872 0.07279804 23.552286 0.12154673  78.78551
## 93b382e4-9c9a-43f5-bd3b-502cc260b886 0.69905270  2.853812 1.01000271 145.84399
## 1f39dadd-3655-474e-ba4c-a5bd32c97a8b 0.01652257  1.157070 0.27942068  48.45022
## 8c8c09b9-ec83-45ec-bc4c-0ba92de60acb 0.03168398  2.408137 0.04922458  42.25592
## 85a86b91-4f24-4e77-ae2d-520f8e205efc 0.06828305  1.010411 0.02248965  20.63795
##                                            MSLN      CLDN6     ERBB2
## 43e715bf-28d9-4b5e-b762-8cd1b69a430e 0.06674445 0.09704962  1.879518
## 1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872 0.95554610 0.25458796  7.777976
## 93b382e4-9c9a-43f5-bd3b-502cc260b886 0.04563568 0.25701910  2.905926
## 1f39dadd-3655-474e-ba4c-a5bd32c97a8b 0.03154912 0.24746913  4.914280
## 8c8c09b9-ec83-45ec-bc4c-0ba92de60acb 0.26968788 0.12576720  1.494744
## 85a86b91-4f24-4e77-ae2d-520f8e205efc 0.01336404 0.01823883 13.474689
##                                             MUC16       DLL3 CEACAM5      PVR
## 43e715bf-28d9-4b5e-b762-8cd1b69a430e 0.0011479879 0.49589978       0 52.08113
## 1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872 0.0008049670 2.52244014       0 40.87926
## 93b382e4-9c9a-43f5-bd3b-502cc260b886 0.0026190288 0.77074712       0 33.26727
## 1f39dadd-3655-474e-ba4c-a5bd32c97a8b 0.0051705741 0.10636402       0 28.26457
## 8c8c09b9-ec83-45ec-bc4c-0ba92de60acb 0.0004894306 0.04483123       0 41.66776
## 85a86b91-4f24-4e77-ae2d-520f8e205efc 0.0000000000 0.01184285       0 30.18711
##                                          EPCAM       PROM1       CD24      EGFR
## 43e715bf-28d9-4b5e-b762-8cd1b69a430e  4.521984 0.025311008 0.55036003  1.286481
## 1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872  9.530414 0.023576862 9.67272890  5.373307
## 93b382e4-9c9a-43f5-bd3b-502cc260b886 42.358567 0.000000000 0.06939934  4.600918
## 1f39dadd-3655-474e-ba4c-a5bd32c97a8b 16.316524 0.007783431 0.84522244  3.010374
## 8c8c09b9-ec83-45ec-bc4c-0ba92de60acb 12.529742 0.019204339 0.21369023 16.476552
## 85a86b91-4f24-4e77-ae2d-520f8e205efc  2.430109 0.043719865 4.95506593  2.010338
##                                             MET TNFRSF10B
## 43e715bf-28d9-4b5e-b762-8cd1b69a430e  0.9320235  12.80547
## 1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872  8.0610999  31.46289
## 93b382e4-9c9a-43f5-bd3b-502cc260b886  0.1295387  65.57967
## 1f39dadd-3655-474e-ba4c-a5bd32c97a8b  2.9728030  24.31636
## 8c8c09b9-ec83-45ec-bc4c-0ba92de60acb 19.7360055  21.11014
## 85a86b91-4f24-4e77-ae2d-520f8e205efc  8.6087283  37.91574
##                                                 tcga.tcga_barcode
## 43e715bf-28d9-4b5e-b762-8cd1b69a430e TCGA-OR-A5KU-01A-11R-A29S-07
## 1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872 TCGA-P6-A5OG-01A-22R-A29S-07
## 93b382e4-9c9a-43f5-bd3b-502cc260b886 TCGA-OR-A5K5-01A-11R-A29S-07
## 1f39dadd-3655-474e-ba4c-a5bd32c97a8b TCGA-OR-A5K4-01A-11R-A29S-07
## 8c8c09b9-ec83-45ec-bc4c-0ba92de60acb TCGA-OR-A5LP-01A-11R-A29S-07
## 85a86b91-4f24-4e77-ae2d-520f8e205efc TCGA-PK-A5H9-01A-11R-A29S-07
##                                      tcga.cgc_sample_sample_type study
## 43e715bf-28d9-4b5e-b762-8cd1b69a430e               Primary Tumor   ACC
## 1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872               Primary Tumor   ACC
## 93b382e4-9c9a-43f5-bd3b-502cc260b886               Primary Tumor   ACC
## 1f39dadd-3655-474e-ba4c-a5bd32c97a8b               Primary Tumor   ACC
## 8c8c09b9-ec83-45ec-bc4c-0ba92de60acb               Primary Tumor   ACC
## 85a86b91-4f24-4e77-ae2d-520f8e205efc               Primary Tumor   ACC
##                                      sample_type
## 43e715bf-28d9-4b5e-b762-8cd1b69a430e      cancer
## 1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872      cancer
## 93b382e4-9c9a-43f5-bd3b-502cc260b886      cancer
## 1f39dadd-3655-474e-ba4c-a5bd32c97a8b      cancer
## 8c8c09b9-ec83-45ec-bc4c-0ba92de60acb      cancer
## 85a86b91-4f24-4e77-ae2d-520f8e205efc      cancer</code></pre>
<p>With everything in a single dataframe, we are ready to do anything you want:) You will notice that this is the dataset we used in previous lectures and I just showed you how to get it from the source.</p>
<p>With everything now in a single dataframe, you can proceed with various data analysis tasks, such as identifying patterns, conducting differential expression analysis, or generating visualizations to gain insights from this comprehensive TCGA dataset.</p>
</div>
</div>
<div id="step-9-create-a-gene-expression-heatmap" class="section level3" number="11.6.12">
<h3><span class="header-section-number">11.6.12</span> Step 9: Create a Gene Expression Heatmap</h3>
<p>To create the gene expression heatmap, we will follow these steps:</p>
<div class="sourceCode" id="cb545"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb545-1"><a href="introduction-to-bioconductor.html#cb545-1" tabindex="-1"></a><span class="co"># import ComplexHeatmap</span></span>
<span id="cb545-2"><a href="introduction-to-bioconductor.html#cb545-2" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span></code></pre></div>
<div id="filter-and-transform-data" class="section level4" number="11.6.12.1">
<h4><span class="header-section-number">11.6.12.1</span> Filter and Transform Data:</h4>
<p>First, we’ll filter the data to include only cancer samples, calculate the median expression for each gene within the same cancer type, and take the logarithm of the expression values to create a <code>log2</code> transformation. This makes the data suitable for heatmap visualization.</p>
<div class="sourceCode" id="cb546"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb546-1"><a href="introduction-to-bioconductor.html#cb546-1" tabindex="-1"></a>tcga_df <span class="ot">&lt;-</span> final_df <span class="sc">%&gt;%</span></span>
<span id="cb546-2"><a href="introduction-to-bioconductor.html#cb546-2" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(sample_type <span class="sc">==</span> <span class="st">&quot;cancer&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb546-3"><a href="introduction-to-bioconductor.html#cb546-3" tabindex="-1"></a>  <span class="fu">group_by</span>(sample_type, study) <span class="sc">%&gt;%</span></span>
<span id="cb546-4"><a href="introduction-to-bioconductor.html#cb546-4" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="sc">~</span><span class="fu">log2</span>(.x <span class="sc">+</span> <span class="dv">1</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb546-5"><a href="introduction-to-bioconductor.html#cb546-5" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, median)) <span class="sc">%&gt;%</span></span>
<span id="cb546-6"><a href="introduction-to-bioconductor.html#cb546-6" tabindex="-1"></a>  <span class="fu">arrange</span>(study) <span class="sc">%&gt;%</span></span>
<span id="cb546-7"><a href="introduction-to-bioconductor.html#cb546-7" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(sample_type))</span></code></pre></div>
</div>
<div id="create-a-gene-expression-matrix" class="section level4" number="11.6.12.2">
<h4><span class="header-section-number">11.6.12.2</span> Create a Gene Expression Matrix:</h4>
<p>Let’s convert the summarized data into a matrix format. Rows represent cancer types, columns represent genes, and the matrix values are the median log2-transformed expression levels.</p>
<div class="sourceCode" id="cb547"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb547-1"><a href="introduction-to-bioconductor.html#cb547-1" tabindex="-1"></a><span class="co"># Create a matrix &#39;tcga_mat&#39; from &#39;tcga_df&#39; excluding the first two columns and convert it to a matrix format.</span></span>
<span id="cb547-2"><a href="introduction-to-bioconductor.html#cb547-2" tabindex="-1"></a>tcga_mat <span class="ot">&lt;-</span> tcga_df[, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>)] <span class="sc">%&gt;%</span> <span class="fu">as.matrix</span>()</span>
<span id="cb547-3"><a href="introduction-to-bioconductor.html#cb547-3" tabindex="-1"></a><span class="fu">rownames</span>(tcga_mat) <span class="ot">&lt;-</span> tcga_df <span class="sc">%&gt;%</span> <span class="fu">pull</span>(study)</span>
<span id="cb547-4"><a href="introduction-to-bioconductor.html#cb547-4" tabindex="-1"></a></span>
<span id="cb547-5"><a href="introduction-to-bioconductor.html#cb547-5" tabindex="-1"></a><span class="fu">dim</span>(tcga_mat)</span></code></pre></div>
<pre><code>## [1] 32 20</code></pre>
<p><code>dim(tcga_mat)</code> tells us that we have 32 rows (cancer types) and 20 columns (genes) in the resulting matrix. Let’s print the matrix:</p>
<div class="sourceCode" id="cb549"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb549-1"><a href="introduction-to-bioconductor.html#cb549-1" tabindex="-1"></a>tcga_mat</span></code></pre></div>
<pre><code>##        TACSTD2       VTCN1      MUC1    NECTIN4     FOLH1      FOLR1    CD276
## ACC  0.4807693 0.000000000 0.2410192 0.04359167 1.6264037 0.02425151 5.297494
## BLCA 9.3725643 2.220041940 5.5991868 6.57397122 0.8918716 0.30467605 5.381322
## BRCA 8.0335879 5.770754248 7.5184905 5.51512285 1.4233695 1.39552103 5.784310
## CESC 9.7498568 1.914881813 6.5567883 6.43148129 0.7041431 0.87034258 4.965822
## CHOL 5.5273588 6.051437587 5.1893312 4.24539532 1.5456811 3.20878730 5.811681
## COAD 3.3928422 0.074410401 5.4975137 3.38347931 0.8062397 1.13574612 5.090426
## DLBC 0.6352793 0.002451433 1.1990101 1.01033523 0.2645891 0.03749739 3.266925
## ESCA 8.2368067 1.499102374 5.9425749 4.73125564 1.1132842 0.50368752 5.398227
## GBM  0.5207019 0.204680580 2.7142306 0.21242126 2.3137440 2.37227202 5.916635
## HNSC 9.0661567 0.694964205 4.0868590 6.08414302 1.1091992 0.38462756 5.830814
## KICH 2.7171000 0.145383937 5.5732785 0.38002326 0.9940703 1.89521752 4.194597
## KIRC 2.7574839 0.307661347 4.6082145 0.28522181 3.8988559 5.87646625 4.802535
## KIRP 6.5632546 2.287412718 4.9965286 0.17556715 0.4862326 5.84513870 4.926838
## LGG  0.1926966 0.341667316 1.7542645 0.07059579 2.5460983 1.14655021 4.305272
## LIHC 0.6768282 0.097613683 0.4391886 0.14510725 2.4297942 0.24022185 4.218155
## LUAD 7.8276755 1.355902742 8.4823052 5.09395755 1.1373591 6.45229154 5.201273
## LUSC 8.4101169 3.040923288 5.2575659 5.71623462 1.9452955 2.82656424 5.812421
## MESO 1.2488321 0.053038967 4.5448185 0.57555047 0.8397168 0.21716736 5.864723
## OV   7.3244772 5.414729025 7.7623819 3.37780261 1.3177417 8.16821960 5.019641
## PAAD 8.0663095 3.093104563 8.1649247 5.04149982 1.1797304 3.50409020 5.631054
## PCPG 1.0556914 0.020899233 0.2936292 0.13453283 1.4682181 0.81474148 4.896633
## PRAD 8.7594024 1.127776607 2.9181243 4.41268848 7.9254012 0.68870009 5.627149
## READ 3.4642158 0.056164097 5.6436544 3.24900294 0.7618605 1.61010910 5.087184
## SARC 1.2359343 0.139256548 2.6489240 0.50282829 1.5561093 0.25330451 6.600442
## SKCM 2.2675734 0.097902971 1.4645616 1.38994449 0.6182104 0.07875303 6.006916
## STAD 6.6014640 0.525928740 7.3745946 2.94416228 0.8882602 1.23634263 4.948302
## TGCT 2.0988218 0.338403710 1.0458394 2.27417760 0.7553944 2.78491077 5.465165
## THCA 8.2074834 0.569348321 5.1203104 4.15113608 1.3056050 5.00909622 4.848942
## THYM 3.4738470 0.017066392 1.1953473 1.10766635 0.2974423 0.83108801 3.581791
## UCEC 7.8256710 5.512091113 7.5435743 4.42551293 2.9885794 5.72164350 5.635153
## UCS  4.6713276 2.773715527 4.5590560 2.71970946 2.1128680 3.81105575 5.946331
## UVM  0.1193770 0.015265577 0.5954239 0.02765640 0.4841875 0.08134510 4.948453
##            MSLN       CLDN6    ERBB2        MUC16        DLL3      CEACAM5
## ACC  0.04481252 0.100146747 1.671858 0.0012205866 0.272796764  0.000000000
## BLCA 0.68716010 0.070442282 5.919806 0.0336637234 0.208256653  1.778875643
## BRCA 0.44071809 0.102271454 6.271432 0.1706249992 0.127821241  2.086550224
## CESC 5.62058497 0.093658539 5.440120 1.5432603634 0.829009483  5.852352010
## CHOL 0.60189628 0.327171872 5.643004 0.0085288100 0.039092934  0.621102898
## COAD 4.45458393 0.074751017 5.451245 0.0076928548 0.187729160 10.878821129
## DLBC 0.35785162 0.003087473 2.287948 0.0233179585 0.993488458  0.013941585
## ESCA 2.47373379 0.181754311 5.399585 0.2162955799 0.201342375  7.197142265
## GBM  0.92682384 0.149746286 3.975576 0.0013188121 4.821499549  0.000000000
## HNSC 1.96099164 0.095296810 4.968078 0.2377308618 0.339048581  4.297622737
## KICH 0.05250724 0.043960617 4.687870 0.0007542286 0.010856365  0.000000000
## KIRC 0.62610191 0.065242760 4.640322 0.0136965478 0.022333527  0.012439869
## KIRP 3.31497246 0.135890060 5.756473 0.0041568708 0.132753240  0.007485503
## LGG  0.72084397 0.159661042 3.215578 0.0011538098 7.342314792  0.000000000
## LIHC 0.03555045 0.053806519 3.900610 0.0014097586 0.002397106  0.005960490
## LUAD 6.40328757 0.515735694 5.797269 1.1197883023 0.342515956  6.795916804
## LUSC 2.66588838 0.148816023 4.807536 0.2162983747 0.544306196  4.197498762
## MESO 9.10430072 0.423970579 4.832141 1.6452238522 0.233216970  0.000000000
## OV   8.95238588 5.140635778 5.072491 4.6261895707 0.971504155  0.021736955
## PAAD 7.51409138 0.599000931 5.573353 1.3005505006 0.112664555  7.670442709
## PCPG 0.04768524 0.141276749 2.243807 0.0006859763 0.176422947  0.000000000
## PRAD 0.69895768 0.047194319 5.811768 0.0312615665 0.304708642  0.605374582
## READ 4.71118780 0.071464799 5.539358 0.0059694499 0.273579598 11.213512117
## SARC 0.18614478 0.128843904 4.080999 0.0034064531 0.166898040  0.000000000
## SKCM 0.10666438 0.063647636 4.131557 0.0052637362 2.811811044  0.132423575
## STAD 5.24125577 0.150099973 5.290161 0.0541710205 0.117172504  7.877833103
## TGCT 0.51413379 7.520373820 4.118593 0.1542327692 3.401171404  0.070429833
## THCA 0.33791649 0.397080016 6.005224 0.0813373551 0.183221365  0.019130524
## THYM 0.22949846 0.157873855 3.329512 0.0079995046 0.096684822  0.000000000
## UCEC 5.00045864 0.679964295 5.431228 2.5515510076 0.244214559  0.775156469
## UCS  3.32976790 3.507766256 5.372831 0.5394212880 2.112957592  0.177189912
## UVM  0.04446844 0.000000000 3.891753 0.0000000000 1.581155709  0.000000000
##           PVR      EPCAM      PROM1       CD24      EGFR      MET TNFRSF10B
## ACC  5.236533 2.54354419 0.01273144  0.3505085 2.5059929 1.224786  4.334822
## BLCA 4.616088 6.36016443 0.20996061  8.1322461 4.0402929 4.300953  4.400105
## BRCA 3.834349 7.34815989 2.38468060  9.2653878 2.3867813 2.572547  4.026348
## CESC 4.730008 6.26554142 0.16395043  8.4236687 4.6460625 4.552453  4.979278
## CHOL 5.344069 7.97991762 4.37980601  8.4224757 3.9911158 4.949958  5.485029
## COAD 5.252634 9.60464201 4.47048763  9.3837212 3.2617795 5.377500  5.022903
## DLBC 2.163426 0.65838277 0.04557316  4.9255259 0.7242894 1.279921  4.059292
## ESCA 5.327993 7.57273004 2.16415031  8.6033079 5.3580321 5.291490  4.893869
## GBM  4.081546 0.38659163 2.68647622  5.1230942 6.7923898 2.005737  4.332658
## HNSC 4.922199 4.68395428 0.11238413  8.1208121 5.7435940 4.889075  4.895662
## KICH 5.132339 6.56645497 0.65202457  8.2691455 3.2130325 6.078532  4.564581
## KIRC 4.517746 4.65016367 2.21935887 10.0347534 5.6655666 6.309732  5.241214
## KIRP 4.655545 6.17909091 4.99464557 10.6188107 3.9479224 6.983211  5.494021
## LGG  3.433609 0.74956575 1.74465625  6.1547041 5.9290943 0.682397  3.480682
## LIHC 4.980117 0.78819179 0.06097059  5.6470623 3.6541165 4.775692  3.945549
## LUAD 4.743134 8.26018569 1.72616107  7.4193353 4.3624974 5.558406  4.989287
## LUSC 4.620441 7.12538734 0.97017148  7.5103897 5.2339820 5.074190  4.589941
## MESO 4.577427 1.10033694 0.05612307  2.2708816 4.6625905 5.780901  5.282852
## OV   3.750990 7.69797342 1.18943790  8.8827993 2.3811636 2.427727  3.382974
## PAAD 5.080733 8.06801056 4.61622807  8.5264790 3.7740009 5.247906  5.201323
## PCPG 4.892586 2.84560470 0.07544391  8.2692372 0.8566280 1.003257  2.480249
## PRAD 4.821357 7.61044397 0.75473658  7.6350131 4.1718705 2.386331  3.680519
## READ 5.301009 9.66106282 4.36896752  9.3307734 3.3160883 5.297970  4.827250
## SARC 4.520111 0.28448581 0.15045652  3.3333123 3.7024832 2.351284  4.271224
## SKCM 4.571221 0.26562350 0.58196240  1.8289097 0.9286357 3.559768  4.674391
## STAD 4.997742 8.32728127 3.93302612  8.9745717 3.8687001 4.725972  4.507678
## TGCT 5.091183 6.87150918 4.57983863  5.9294673 1.7305245 1.975280  3.881551
## THCA 3.901983 8.42405535 0.27570756  9.8733054 4.3838020 7.102301  4.882152
## THYM 2.833464 3.87134280 0.23738095  1.6799685 4.2379247 3.469279  4.077410
## UCEC 4.354987 8.44878219 4.28734222  9.3528813 2.6807882 3.738476  4.634779
## UCS  4.348322 6.58161996 3.48922500  7.8918205 3.3461966 3.826455  4.321049
## UVM  3.877644 0.08178735 0.13753182  0.1211111 0.4593655 6.347510  3.523352</code></pre>
</div>
<div id="define-a-cell-function-for-grid-lines" class="section level4" number="11.6.12.3">
<h4><span class="header-section-number">11.6.12.3</span> Define a Cell Function for Grid Lines</h4>
<div class="sourceCode" id="cb551"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb551-1"><a href="introduction-to-bioconductor.html#cb551-1" tabindex="-1"></a>cell_fun <span class="ot">=</span> <span class="cf">function</span>(j, i, x, y, w, h, fill) {</span>
<span id="cb551-2"><a href="introduction-to-bioconductor.html#cb551-2" tabindex="-1"></a>  <span class="fu">grid.rect</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">width =</span> w, <span class="at">height =</span> h, <span class="at">gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> <span class="st">&quot;black&quot;</span>, <span class="at">fill =</span> <span class="cn">NA</span>))</span>
<span id="cb551-3"><a href="introduction-to-bioconductor.html#cb551-3" tabindex="-1"></a>}</span></code></pre></div>
<p>We define a custom cell function (<code>cell_fun</code>) to add black grid lines to the heatmap for better visualization.</p>
</div>
<div id="create-the-heatmap" class="section level4" number="11.6.12.4">
<h4><span class="header-section-number">11.6.12.4</span> Create the Heatmap</h4>
<div class="sourceCode" id="cb552"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb552-1"><a href="introduction-to-bioconductor.html#cb552-1" tabindex="-1"></a><span class="fu">Heatmap</span>(tcga_mat, <span class="at">cluster_columns =</span> <span class="cn">TRUE</span>, <span class="at">cell_fun =</span> cell_fun, <span class="at">name =</span> <span class="st">&quot;log2TPM&quot;</span>)</span></code></pre></div>
<p><img src="10_Introduction_to_BioConductor_files/figure-html/unnamed-chunk-61-1.png" width="672" /></p>
<p>Now, we create the heatmap using the <code>Heatmap</code> function from the <code>ComplexHeatmap</code> package. We specify <code>cluster_columns = TRUE</code> to cluster the columns (genes) for better visualization. The <code>cell_fun</code> parameter is set to our custom function for adding grid lines, and we name the data as “log2TPM.”</p>
<p>The resulting heatmap will show the median gene expression levels for each gene across different cancer types.</p>
<p>Grid lines will help in distinguishing individual cells within the heatmap so we added them.
This visualization can provide insights into gene expression patterns in various cancer types, which may be useful for further analysis and interpretation.</p>
</div>
</div>
<div id="step-10-sanity-check-and-scaling-of-gene-expression" class="section level3" number="11.6.13">
<h3><span class="header-section-number">11.6.13</span> Step 10: Sanity Check and Scaling of Gene Expression</h3>
<div id="sanity-check" class="section level4" number="11.6.13.1">
<h4><span class="header-section-number">11.6.13.1</span> Sanity Check</h4>
<p>Before proceeding with scaling, we will conduct a sanity check of the gene expression heatmap to see if the results make biological sense. For example, we can observe if certain genes are highly expressed in specific cancer types, which could indicate potential biomarkers or interesting biological phenomena.</p>
<div class="sourceCode" id="cb553"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb553-1"><a href="introduction-to-bioconductor.html#cb553-1" tabindex="-1"></a><span class="co"># Sanity check</span></span>
<span id="cb553-2"><a href="introduction-to-bioconductor.html#cb553-2" tabindex="-1"></a>sanity_check_genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;MSLN&quot;</span>, <span class="st">&quot;FOLH1&quot;</span>)</span>
<span id="cb553-3"><a href="introduction-to-bioconductor.html#cb553-3" tabindex="-1"></a></span>
<span id="cb553-4"><a href="introduction-to-bioconductor.html#cb553-4" tabindex="-1"></a><span class="co"># Extract the rows (genes) corresponding to sanity check genes</span></span>
<span id="cb553-5"><a href="introduction-to-bioconductor.html#cb553-5" tabindex="-1"></a>sanity_check_data <span class="ot">&lt;-</span> tcga_mat[, <span class="fu">colnames</span>(tcga_mat) <span class="sc">%in%</span> sanity_check_genes]</span>
<span id="cb553-6"><a href="introduction-to-bioconductor.html#cb553-6" tabindex="-1"></a></span>
<span id="cb553-7"><a href="introduction-to-bioconductor.html#cb553-7" tabindex="-1"></a><span class="co"># Print the results</span></span>
<span id="cb553-8"><a href="introduction-to-bioconductor.html#cb553-8" tabindex="-1"></a><span class="fu">print</span>(sanity_check_data)</span></code></pre></div>
<pre><code>##          FOLH1       MSLN
## ACC  1.6264037 0.04481252
## BLCA 0.8918716 0.68716010
## BRCA 1.4233695 0.44071809
## CESC 0.7041431 5.62058497
## CHOL 1.5456811 0.60189628
## COAD 0.8062397 4.45458393
## DLBC 0.2645891 0.35785162
## ESCA 1.1132842 2.47373379
## GBM  2.3137440 0.92682384
## HNSC 1.1091992 1.96099164
## KICH 0.9940703 0.05250724
## KIRC 3.8988559 0.62610191
## KIRP 0.4862326 3.31497246
## LGG  2.5460983 0.72084397
## LIHC 2.4297942 0.03555045
## LUAD 1.1373591 6.40328757
## LUSC 1.9452955 2.66588838
## MESO 0.8397168 9.10430072
## OV   1.3177417 8.95238588
## PAAD 1.1797304 7.51409138
## PCPG 1.4682181 0.04768524
## PRAD 7.9254012 0.69895768
## READ 0.7618605 4.71118780
## SARC 1.5561093 0.18614478
## SKCM 0.6182104 0.10666438
## STAD 0.8882602 5.24125577
## TGCT 0.7553944 0.51413379
## THCA 1.3056050 0.33791649
## THYM 0.2974423 0.22949846
## UCEC 2.9885794 5.00045864
## UCS  2.1128680 3.32976790
## UVM  0.4841875 0.04446844</code></pre>
<p>We see <code>MSLN</code> is high in <code>MESO</code>, <code>FOLH1</code> is high in prostate cancer (<code>PRAD</code>). We are probably on the right track!</p>
</div>
<div id="scaling-the-data" class="section level4" number="11.6.13.2">
<h4><span class="header-section-number">11.6.13.2</span> Scaling the data</h4>
<p>To visualize gene expression in a more comparative manner, we can scale the expression values for each gene across the cancer types. Scaling standardizes the data, making it easier to identify relative expression levels.</p>
<div class="sourceCode" id="cb555"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb555-1"><a href="introduction-to-bioconductor.html#cb555-1" tabindex="-1"></a><span class="co"># Scale the expression data</span></span>
<span id="cb555-2"><a href="introduction-to-bioconductor.html#cb555-2" tabindex="-1"></a>scaled_tcga_mat <span class="ot">&lt;-</span> <span class="fu">scale</span>(tcga_mat)</span>
<span id="cb555-3"><a href="introduction-to-bioconductor.html#cb555-3" tabindex="-1"></a></span>
<span id="cb555-4"><a href="introduction-to-bioconductor.html#cb555-4" tabindex="-1"></a><span class="co"># Create a scaled heatmap</span></span>
<span id="cb555-5"><a href="introduction-to-bioconductor.html#cb555-5" tabindex="-1"></a><span class="fu">Heatmap</span>(scaled_tcga_mat, </span>
<span id="cb555-6"><a href="introduction-to-bioconductor.html#cb555-6" tabindex="-1"></a>        <span class="at">cluster_columns =</span> <span class="cn">TRUE</span>, </span>
<span id="cb555-7"><a href="introduction-to-bioconductor.html#cb555-7" tabindex="-1"></a>        <span class="at">cell_fun =</span> cell_fun, </span>
<span id="cb555-8"><a href="introduction-to-bioconductor.html#cb555-8" tabindex="-1"></a>        <span class="at">name =</span> <span class="st">&quot;scaled</span><span class="sc">\n</span><span class="st">log2TPM&quot;</span>)</span></code></pre></div>
<p><img src="10_Introduction_to_BioConductor_files/figure-html/unnamed-chunk-63-1.png" width="672" /></p>
<p>Here, we use the <code>scale()</code> function to standardize the expression values across cancer types for each gene. We then create a new heatmap using the scaled data.</p>
<p>By comparing the original and scaled heatmaps, we can gain insights into how the expression of genes varies relative to each other across different cancer types. This scaling helps us focus on the relative expression patterns rather than the absolute values.</p>
</div>
</div>
<div id="conclusion-30" class="section level3" number="11.6.14">
<h3><span class="header-section-number">11.6.14</span> Conclusion</h3>
<p>In this comprehensive example, we have walked through various essential steps for performing gene expression analysis using R, focusing on the analysis of The Cancer Genome Atlas (TCGA) data as an illustrative example. Gene expression analysis is a fundamental component of genomics research, enabling us to uncover insights into the molecular mechanisms underlying complex biological processes, such as cancer.</p>
</div>
</div>
<div id="section-completed" class="section level2" number="11.7">
<h2><span class="header-section-number">11.7</span> Section completed</h2>
<p>Congratulations on completing this section!</p>
<p>You’ve learned crucial genomics data handling and analysis techniques, like using the <code>GenomicRanges</code> package, analyzing CpG islands, converting gene IDs, and creating visualizations with <code>ComplexHeatmap</code>. These skills are vital for bioinformatics, giving you the confidence to navigate genomic data complexities.</p>
<p>Don’t hesitate to interact with your peers and instructors in the Q&amp;A section and comments. Share your experiences, ask questions, and offer support. Your engagement enriches everyone’s learning journey.</p>
<p>As you approach the final project, remember that what you’ve learned is more than just theory—it’s practical knowledge you can apply. Use this opportunity to showcase your skills in a real-world scenario. Dive into the challenge with enthusiasm and curiosity.</p>
<p>Good luck!
Let’s make this final project a success together!</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-visualization-with-ggplot2.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="final-project-analyzing-rnaseq-data-from-geo.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/10_Introduction_to_BioConductor.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
