<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Introduction to programming | Master Real-world Bioinformatics analysis in R</title>
  <meta name="description" content="This is an R course teaching you skills to do real-world bioinformatics" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Introduction to programming | Master Real-world Bioinformatics analysis in R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is an R course teaching you skills to do real-world bioinformatics" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Introduction to programming | Master Real-world Bioinformatics analysis in R" />
  
  <meta name="twitter:description" content="This is an R course teaching you skills to do real-world bioinformatics" />
  

<meta name="author" content="Ming Tommy Tang" />


<meta name="date" content="2024-05-25" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="intro.html"/>
<link rel="next" href="getting-started-with-r-and-rstudio.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<link href="libs/vembedr-0.1.5/css/vembedr.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li><a href="index.html#preface" id="toc-preface"><span class="toc-section-number">1</span> Preface</a></li>
<li><a href="intro.html#intro" id="toc-intro"><span class="toc-section-number">2</span> Introduction</a>
<ul>
<li><a href="intro.html#meet-your-instructor" id="toc-meet-your-instructor"><span class="toc-section-number">2.1</span> Meet your Instructor</a></li>
<li><a href="intro.html#leveraging-our-online-community" id="toc-leveraging-our-online-community"><span class="toc-section-number">2.2</span> Leveraging our online community</a></li>
<li><a href="intro.html#install-r-and-r-studio" id="toc-install-r-and-r-studio"><span class="toc-section-number">2.3</span> Install R and R studio</a></li>
<li><a href="intro.html#the-role-of-programming-in-biology" id="toc-the-role-of-programming-in-biology"><span class="toc-section-number">2.4</span> The role of programming in Biology</a>
<ul>
<li><a href="intro.html#what-are-programming-languages" id="toc-what-are-programming-languages"><span class="toc-section-number">2.4.1</span> What are programming languages?</a></li>
<li><a href="intro.html#why-programming-is-important-in-biology" id="toc-why-programming-is-important-in-biology"><span class="toc-section-number">2.4.2</span> Why programming is important in biology?</a></li>
<li><a href="intro.html#what-are-the-most-used-programming-languages-within-biology" id="toc-what-are-the-most-used-programming-languages-within-biology"><span class="toc-section-number">2.4.3</span> What are the most used programming languages within biology?</a></li>
</ul></li>
</ul></li>
<li><a href="introduction-to-programming.html#introduction-to-programming" id="toc-introduction-to-programming"><span class="toc-section-number">3</span> Introduction to programming</a>
<ul>
<li><a href="introduction-to-programming.html#what-is-an-algorithm" id="toc-what-is-an-algorithm"><span class="toc-section-number">3.1</span> What is an algorithm?</a></li>
<li><a href="introduction-to-programming.html#why-are-algorithms-important-in-programming" id="toc-why-are-algorithms-important-in-programming"><span class="toc-section-number">3.2</span> Why are Algorithms Important in Programming?</a></li>
<li><a href="introduction-to-programming.html#a-real-world-example" id="toc-a-real-world-example"><span class="toc-section-number">3.3</span> A Real World Example:</a></li>
<li><a href="introduction-to-programming.html#variables-data-types-and-expressions" id="toc-variables-data-types-and-expressions"><span class="toc-section-number">3.4</span> Variables, Data Types, and Expressions</a>
<ul>
<li><a href="introduction-to-programming.html#variables" id="toc-variables"><span class="toc-section-number">3.4.1</span> Variables</a></li>
<li><a href="introduction-to-programming.html#data-types" id="toc-data-types"><span class="toc-section-number">3.4.2</span> Data Types</a></li>
<li><a href="introduction-to-programming.html#expressions" id="toc-expressions"><span class="toc-section-number">3.4.3</span> Expressions</a></li>
<li><a href="introduction-to-programming.html#example-finding-the-average-number-of-bases-appearances-in-a-set-of-dna-strings" id="toc-example-finding-the-average-number-of-bases-appearances-in-a-set-of-dna-strings"><span class="toc-section-number">3.4.4</span> Example: Finding the average number of bases appearances in a set of DNA strings</a></li>
</ul></li>
</ul></li>
<li><a href="getting-started-with-r-and-rstudio.html#getting-started-with-r-and-rstudio" id="toc-getting-started-with-r-and-rstudio"><span class="toc-section-number">4</span> Getting started with R and RStudio</a>
<ul>
<li><a href="getting-started-with-r-and-rstudio.html#what-is-r-and-why-is-it-used-in-biology" id="toc-what-is-r-and-why-is-it-used-in-biology"><span class="toc-section-number">4.1</span> What is R and why is it used in Biology?</a></li>
<li><a href="getting-started-with-r-and-rstudio.html#what-is-r" id="toc-what-is-r"><span class="toc-section-number">4.2</span> What is R?</a></li>
<li><a href="getting-started-with-r-and-rstudio.html#why-is-r-used-in-biology" id="toc-why-is-r-used-in-biology"><span class="toc-section-number">4.3</span> Why is R Used in Biology?</a></li>
<li><a href="getting-started-with-r-and-rstudio.html#introduction-to-r-studio" id="toc-introduction-to-r-studio"><span class="toc-section-number">4.4</span> Introduction to R-Studio</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#introduction-to-r-for-biologists" id="toc-introduction-to-r-for-biologists"><span class="toc-section-number">5</span> Introduction to R for biologists</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#basic-data-types-in-r" id="toc-basic-data-types-in-r"><span class="toc-section-number">5.1</span> Basic Data Types in R</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#before-starting" id="toc-before-starting"><span class="toc-section-number">5.1.1</span> Before Starting</a></li>
<li><a href="introduction-to-r-for-biologists.html#numeric" id="toc-numeric"><span class="toc-section-number">5.1.2</span> 1. Numeric</a></li>
<li><a href="introduction-to-r-for-biologists.html#character-string" id="toc-character-string"><span class="toc-section-number">5.1.3</span> 2. Character (String)</a></li>
<li><a href="introduction-to-r-for-biologists.html#integer" id="toc-integer"><span class="toc-section-number">5.1.4</span> 3. Integer</a></li>
<li><a href="introduction-to-r-for-biologists.html#logical-boolean" id="toc-logical-boolean"><span class="toc-section-number">5.1.5</span> 4. Logical (Boolean)</a></li>
<li><a href="introduction-to-r-for-biologists.html#factor" id="toc-factor"><span class="toc-section-number">5.1.6</span> 5. Factor</a></li>
<li><a href="introduction-to-r-for-biologists.html#date-and-time" id="toc-date-and-time"><span class="toc-section-number">5.1.7</span> 6. Date and Time</a></li>
<li><a href="introduction-to-r-for-biologists.html#complex" id="toc-complex"><span class="toc-section-number">5.1.8</span> 7. Complex</a></li>
<li><a href="introduction-to-r-for-biologists.html#missing-values-na" id="toc-missing-values-na"><span class="toc-section-number">5.1.9</span> 8. Missing Values (NA)</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#the-key-concept-of-r-vectors" id="toc-the-key-concept-of-r-vectors"><span class="toc-section-number">5.2</span> The key concept of R: Vectors</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#creating-numeric-and-character-vectors" id="toc-creating-numeric-and-character-vectors"><span class="toc-section-number">5.2.1</span> Creating Numeric and Character Vectors</a></li>
<li><a href="introduction-to-r-for-biologists.html#saving-vectors-to-variables" id="toc-saving-vectors-to-variables"><span class="toc-section-number">5.2.2</span> Saving Vectors to Variables</a></li>
<li><a href="introduction-to-r-for-biologists.html#vectorized-calculations" id="toc-vectorized-calculations"><span class="toc-section-number">5.2.3</span> Vectorized Calculations</a></li>
<li><a href="introduction-to-r-for-biologists.html#performing-vectorized-calculations" id="toc-performing-vectorized-calculations"><span class="toc-section-number">5.2.4</span> Performing Vectorized Calculations</a></li>
<li><a href="introduction-to-r-for-biologists.html#operations-with-character-vectors" id="toc-operations-with-character-vectors"><span class="toc-section-number">5.2.5</span> Operations with character vectors</a></li>
<li><a href="introduction-to-r-for-biologists.html#logical-vectors" id="toc-logical-vectors"><span class="toc-section-number">5.2.6</span> Logical Vectors</a></li>
<li><a href="introduction-to-r-for-biologists.html#conclusion" id="toc-conclusion"><span class="toc-section-number">5.2.7</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#subsetting-and-indexing" id="toc-subsetting-and-indexing"><span class="toc-section-number">5.3</span> Subsetting and Indexing</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#indexing-in-r" id="toc-indexing-in-r"><span class="toc-section-number">5.3.1</span> Indexing in R</a></li>
<li><a href="introduction-to-r-for-biologists.html#slicing-in-r" id="toc-slicing-in-r"><span class="toc-section-number">5.3.2</span> Slicing in R</a></li>
<li><a href="introduction-to-r-for-biologists.html#negative-indexing" id="toc-negative-indexing"><span class="toc-section-number">5.3.3</span> Negative Indexing</a></li>
<li><a href="introduction-to-r-for-biologists.html#conclusion-1" id="toc-conclusion-1"><span class="toc-section-number">5.3.4</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#understanding-and-manipulating-matrices" id="toc-understanding-and-manipulating-matrices"><span class="toc-section-number">5.4</span> Understanding and Manipulating Matrices</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#creating-a-matrix" id="toc-creating-a-matrix"><span class="toc-section-number">5.4.1</span> Creating a Matrix</a></li>
<li><a href="introduction-to-r-for-biologists.html#adding-row-and-column-names" id="toc-adding-row-and-column-names"><span class="toc-section-number">5.4.2</span> Adding Row and Column Names</a></li>
<li><a href="introduction-to-r-for-biologists.html#subsetting-a-matrix" id="toc-subsetting-a-matrix"><span class="toc-section-number">5.4.3</span> Subsetting a Matrix</a></li>
<li><a href="introduction-to-r-for-biologists.html#calculations-with-matrices" id="toc-calculations-with-matrices"><span class="toc-section-number">5.4.4</span> Calculations with Matrices</a></li>
<li><a href="introduction-to-r-for-biologists.html#normalization" id="toc-normalization"><span class="toc-section-number">5.4.5</span> Normalization</a></li>
<li><a href="introduction-to-r-for-biologists.html#conclusion-2" id="toc-conclusion-2"><span class="toc-section-number">5.4.6</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#essential-functions-in-r" id="toc-essential-functions-in-r"><span class="toc-section-number">5.5</span> Essential Functions in R</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#length-function" id="toc-length-function"><span class="toc-section-number">5.5.1</span> <code>length</code> Function</a></li>
<li><a href="introduction-to-r-for-biologists.html#unique-function" id="toc-unique-function"><span class="toc-section-number">5.5.2</span> <code>unique</code> Function</a></li>
<li><a href="introduction-to-r-for-biologists.html#sort-function" id="toc-sort-function"><span class="toc-section-number">5.5.3</span> <code>sort</code> Function</a></li>
<li><a href="introduction-to-r-for-biologists.html#cor-function" id="toc-cor-function"><span class="toc-section-number">5.5.4</span> <code>cor</code> Function</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#functions-organizing-your-code-for-reusability" id="toc-functions-organizing-your-code-for-reusability"><span class="toc-section-number">5.6</span> Functions: Organizing Your Code for Reusability</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#what-is-a-function" id="toc-what-is-a-function"><span class="toc-section-number">5.6.1</span> What is a Function?</a></li>
<li><a href="introduction-to-r-for-biologists.html#creating-a-basic-function-calculating-the-mean" id="toc-creating-a-basic-function-calculating-the-mean"><span class="toc-section-number">5.6.2</span> Creating a Basic Function: Calculating the Mean</a></li>
<li><a href="introduction-to-r-for-biologists.html#using-the-custom-function" id="toc-using-the-custom-function"><span class="toc-section-number">5.6.3</span> Using the Custom Function</a></li>
<li><a href="introduction-to-r-for-biologists.html#returning-a-value-from-the-function" id="toc-returning-a-value-from-the-function"><span class="toc-section-number">5.6.4</span> Returning a Value from the Function</a></li>
<li><a href="introduction-to-r-for-biologists.html#optional-omitting-the-return-statement" id="toc-optional-omitting-the-return-statement"><span class="toc-section-number">5.6.5</span> Optional: Omitting the return Statement</a></li>
<li><a href="introduction-to-r-for-biologists.html#using-more-than-one-argument" id="toc-using-more-than-one-argument"><span class="toc-section-number">5.6.6</span> Using more than one argument</a></li>
<li><a href="introduction-to-r-for-biologists.html#understanding-missing-values-nas-in-r" id="toc-understanding-missing-values-nas-in-r"><span class="toc-section-number">5.6.7</span> Understanding Missing Values (NAs) in R</a></li>
<li><a href="introduction-to-r-for-biologists.html#initial-attempt-a-function-without-handling-nas" id="toc-initial-attempt-a-function-without-handling-nas"><span class="toc-section-number">5.6.8</span> Initial Attempt: A Function Without Handling NAs</a></li>
<li><a href="introduction-to-r-for-biologists.html#adding-the-remove_na-argument" id="toc-adding-the-remove_na-argument"><span class="toc-section-number">5.6.9</span> Adding the <code>remove_na</code> Argument</a></li>
<li><a href="introduction-to-r-for-biologists.html#practical-application" id="toc-practical-application"><span class="toc-section-number">5.6.10</span> Practical Application</a></li>
<li><a href="introduction-to-r-for-biologists.html#getting-help-with-functions" id="toc-getting-help-with-functions"><span class="toc-section-number">5.6.11</span> Getting help with functions</a></li>
<li><a href="introduction-to-r-for-biologists.html#a-real-world-example-1" id="toc-a-real-world-example-1"><span class="toc-section-number">5.6.12</span> A Real World Example</a></li>
<li><a href="introduction-to-r-for-biologists.html#conclusion-3" id="toc-conclusion-3"><span class="toc-section-number">5.6.13</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#common-mistakes-to-avoid" id="toc-common-mistakes-to-avoid"><span class="toc-section-number">5.7</span> Common Mistakes to avoid</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#mixing-data-types" id="toc-mixing-data-types"><span class="toc-section-number">5.7.1</span> Mixing Data Types</a></li>
<li><a href="introduction-to-r-for-biologists.html#forgetting-function-parentheses" id="toc-forgetting-function-parentheses"><span class="toc-section-number">5.7.2</span> Forgetting Function Parentheses</a></li>
<li><a href="introduction-to-r-for-biologists.html#overwriting-built-in-functions" id="toc-overwriting-built-in-functions"><span class="toc-section-number">5.7.3</span> Overwriting Built-in Functions</a></li>
<li><a href="introduction-to-r-for-biologists.html#misunderstanding-variable-scoping" id="toc-misunderstanding-variable-scoping"><span class="toc-section-number">5.7.4</span> Misunderstanding Variable Scoping</a></li>
<li><a href="introduction-to-r-for-biologists.html#best-practices-for-variable-scoping-include" id="toc-best-practices-for-variable-scoping-include"><span class="toc-section-number">5.7.5</span> Best practices for variable scoping include</a></li>
<li><a href="introduction-to-r-for-biologists.html#conclusion-4" id="toc-conclusion-4"><span class="toc-section-number">5.7.6</span> Conclusion</a></li>
</ul></li>
</ul></li>
<li><a href="controlling-the-flow-of-our-programs.html#controlling-the-flow-of-our-programs" id="toc-controlling-the-flow-of-our-programs"><span class="toc-section-number">6</span> Controlling the flow of our programs</a>
<ul>
<li><a href="controlling-the-flow-of-our-programs.html#boolean-operators" id="toc-boolean-operators"><span class="toc-section-number">6.1</span> Boolean Operators</a>
<ul>
<li><a href="controlling-the-flow-of-our-programs.html#comparison-operators" id="toc-comparison-operators"><span class="toc-section-number">6.1.1</span> Comparison Operators</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#logical-operators" id="toc-logical-operators"><span class="toc-section-number">6.1.2</span> Logical Operators</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#conclusion-5" id="toc-conclusion-5"><span class="toc-section-number">6.1.3</span> Conclusion</a></li>
</ul></li>
<li><a href="controlling-the-flow-of-our-programs.html#conditional-statements-if-else" id="toc-conditional-statements-if-else"><span class="toc-section-number">6.2</span> Conditional statements (<code>if</code>, <code>else</code>)</a>
<ul>
<li><a href="controlling-the-flow-of-our-programs.html#the-if-statement" id="toc-the-if-statement"><span class="toc-section-number">6.2.1</span> The <code>if</code> Statement</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#else-statement" id="toc-else-statement"><span class="toc-section-number">6.2.2</span> <code>else</code> statement</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#else-if" id="toc-else-if"><span class="toc-section-number">6.2.3</span> <code>else if</code></a></li>
<li><a href="controlling-the-flow-of-our-programs.html#exercise" id="toc-exercise"><span class="toc-section-number">6.2.4</span> Exercise</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#conclusion-6" id="toc-conclusion-6"><span class="toc-section-number">6.2.5</span> Conclusion</a></li>
</ul></li>
<li><a href="controlling-the-flow-of-our-programs.html#loops" id="toc-loops"><span class="toc-section-number">6.3</span> Loops</a>
<ul>
<li><a href="controlling-the-flow-of-our-programs.html#the-while-loop" id="toc-the-while-loop"><span class="toc-section-number">6.3.1</span> The <code>while</code> Loop</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#while-loops-in-real-life" id="toc-while-loops-in-real-life"><span class="toc-section-number">6.3.2</span> While Loops in Real Life</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#the-for-loop" id="toc-the-for-loop"><span class="toc-section-number">6.3.3</span> The <code>for</code> Loop</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#conclusion-7" id="toc-conclusion-7"><span class="toc-section-number">6.3.4</span> Conclusion</a></li>
</ul></li>
<li><a href="controlling-the-flow-of-our-programs.html#gene-expression-annotation-using-loops-and-control-structures" id="toc-gene-expression-annotation-using-loops-and-control-structures"><span class="toc-section-number">6.4</span> Gene Expression Annotation using Loops and Control Structures</a>
<ul>
<li><a href="controlling-the-flow-of-our-programs.html#the-data" id="toc-the-data"><span class="toc-section-number">6.4.1</span> The Data</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#creating-annotations" id="toc-creating-annotations"><span class="toc-section-number">6.4.2</span> Creating Annotations</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#the-loop" id="toc-the-loop"><span class="toc-section-number">6.4.3</span> The Loop</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#output" id="toc-output"><span class="toc-section-number">6.4.4</span> Output</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#putting-it-all-together" id="toc-putting-it-all-together"><span class="toc-section-number">6.4.5</span> Putting it all together</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#conclusion-8" id="toc-conclusion-8"><span class="toc-section-number">6.4.6</span> Conclusion</a></li>
</ul></li>
<li><a href="controlling-the-flow-of-our-programs.html#lets-solve-a-challenge" id="toc-lets-solve-a-challenge"><span class="toc-section-number">6.5</span> Let’s solve a Challenge</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#solution" id="toc-solution"><span class="toc-section-number">6.6</span> Solution</a>
<ul>
<li><a href="controlling-the-flow-of-our-programs.html#using-a-for-loop" id="toc-using-a-for-loop"><span class="toc-section-number">6.6.1</span> using a for loop</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#vectorized-solution" id="toc-vectorized-solution"><span class="toc-section-number">6.6.2</span> vectorized solution</a></li>
</ul></li>
<li><a href="controlling-the-flow-of-our-programs.html#section-complete" id="toc-section-complete"><span class="toc-section-number">6.7</span> Section complete</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#going-more-in-depth-with-r" id="toc-going-more-in-depth-with-r"><span class="toc-section-number">7</span> Going more in-depth with R</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#handling-missing-values" id="toc-handling-missing-values"><span class="toc-section-number">7.1</span> Handling Missing Values</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#understanding-na-in-r" id="toc-understanding-na-in-r"><span class="toc-section-number">7.1.1</span> Understanding NA in R</a></li>
<li><a href="going-more-in-depth-with-r.html#identifying-missing-values" id="toc-identifying-missing-values"><span class="toc-section-number">7.1.2</span> Identifying Missing Values</a></li>
<li><a href="going-more-in-depth-with-r.html#remove-missing-values" id="toc-remove-missing-values"><span class="toc-section-number">7.1.3</span> remove missing values</a></li>
<li><a href="going-more-in-depth-with-r.html#real-life-note" id="toc-real-life-note"><span class="toc-section-number">7.1.4</span> real-life note</a></li>
<li><a href="going-more-in-depth-with-r.html#conclusion-9" id="toc-conclusion-9"><span class="toc-section-number">7.1.5</span> Conclusion</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#introduction-to-statistical-tests-and-p-values" id="toc-introduction-to-statistical-tests-and-p-values"><span class="toc-section-number">7.2</span> Introduction to Statistical Tests and P-Values</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#practical-application-of-statistical-test-with-r" id="toc-practical-application-of-statistical-test-with-r"><span class="toc-section-number">7.2.1</span> Practical Application of statistical test with R</a></li>
<li><a href="going-more-in-depth-with-r.html#conclusion-10" id="toc-conclusion-10"><span class="toc-section-number">7.2.2</span> Conclusion</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#understanding-r-packages" id="toc-understanding-r-packages"><span class="toc-section-number">7.3</span> Understanding R Packages</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#what-is-an-r-package" id="toc-what-is-an-r-package"><span class="toc-section-number">7.3.1</span> What is an R Package?</a></li>
<li><a href="going-more-in-depth-with-r.html#why-use-r-packages" id="toc-why-use-r-packages"><span class="toc-section-number">7.3.2</span> Why Use R Packages?</a></li>
<li><a href="going-more-in-depth-with-r.html#how-to-use-a-package" id="toc-how-to-use-a-package"><span class="toc-section-number">7.3.3</span> How to use a package?</a></li>
<li><a href="going-more-in-depth-with-r.html#embracing-errors" id="toc-embracing-errors"><span class="toc-section-number">7.3.4</span> Embracing Errors</a></li>
<li><a href="going-more-in-depth-with-r.html#how-to-look-for-package-docs-within-r-studio" id="toc-how-to-look-for-package-docs-within-r-studio"><span class="toc-section-number">7.3.5</span> How to look for package docs within R studio?</a></li>
<li><a href="going-more-in-depth-with-r.html#exploring-package-vignettes" id="toc-exploring-package-vignettes"><span class="toc-section-number">7.3.6</span> Exploring Package Vignettes</a></li>
<li><a href="going-more-in-depth-with-r.html#practical-tip" id="toc-practical-tip"><span class="toc-section-number">7.3.7</span> Practical Tip</a></li>
<li><a href="going-more-in-depth-with-r.html#conclusion-11" id="toc-conclusion-11"><span class="toc-section-number">7.3.8</span> Conclusion</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#exploring-functions-in-different-packages-avoiding-collisions-and-accessing-them" id="toc-exploring-functions-in-different-packages-avoiding-collisions-and-accessing-them"><span class="toc-section-number">7.4</span> Exploring Functions in Different Packages: Avoiding Collisions and Accessing Them</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#using-double-colon" id="toc-using-double-colon"><span class="toc-section-number">7.4.1</span> Using Double Colon (<code>::</code>)</a></li>
<li><a href="going-more-in-depth-with-r.html#conclusion-12" id="toc-conclusion-12"><span class="toc-section-number">7.4.2</span> Conclusion</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#writing-custom-scripts" id="toc-writing-custom-scripts"><span class="toc-section-number">7.5</span> Writing Custom Scripts</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#your-working-directory" id="toc-your-working-directory"><span class="toc-section-number">7.5.1</span> Your Working Directory</a></li>
<li><a href="going-more-in-depth-with-r.html#writing-functions-in-scripts" id="toc-writing-functions-in-scripts"><span class="toc-section-number">7.5.2</span> Writing Functions in Scripts</a></li>
<li><a href="going-more-in-depth-with-r.html#importing-functions-from-scripts" id="toc-importing-functions-from-scripts"><span class="toc-section-number">7.5.3</span> Importing Functions from Scripts</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#data-io" id="toc-data-io"><span class="toc-section-number">7.6</span> Data I/O</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#built-in-functions" id="toc-built-in-functions"><span class="toc-section-number">7.6.1</span> Built-In Functions</a></li>
<li><a href="going-more-in-depth-with-r.html#real-world-applications" id="toc-real-world-applications"><span class="toc-section-number">7.6.2</span> Real-World Applications</a></li>
<li><a href="going-more-in-depth-with-r.html#conclusion-13" id="toc-conclusion-13"><span class="toc-section-number">7.6.3</span> Conclusion</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#best-practices-in-modular-programming-and-project-management" id="toc-best-practices-in-modular-programming-and-project-management"><span class="toc-section-number">7.7</span> Best Practices in Modular Programming and Project Management</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#introduction-to-modular-programming" id="toc-introduction-to-modular-programming"><span class="toc-section-number">7.7.1</span> Introduction to Modular Programming</a></li>
<li><a href="going-more-in-depth-with-r.html#breaking-down-your-project" id="toc-breaking-down-your-project"><span class="toc-section-number">7.7.2</span> Breaking Down Your Project</a></li>
<li><a href="going-more-in-depth-with-r.html#conclusion-14" id="toc-conclusion-14"><span class="toc-section-number">7.7.3</span> Conclusion</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#section-complete-1" id="toc-section-complete-1"><span class="toc-section-number">7.8</span> Section complete</a></li>
</ul></li>
<li><a href="fundamental-data-structures-in-r.html#fundamental-data-structures-in-r" id="toc-fundamental-data-structures-in-r"><span class="toc-section-number">8</span> Fundamental Data Structures in R</a>
<ul>
<li><a href="fundamental-data-structures-in-r.html#named-vectors" id="toc-named-vectors"><span class="toc-section-number">8.1</span> Named Vectors</a>
<ul>
<li><a href="fundamental-data-structures-in-r.html#what-is-a-named-vector" id="toc-what-is-a-named-vector"><span class="toc-section-number">8.1.1</span> What is a Named Vector?</a></li>
<li><a href="fundamental-data-structures-in-r.html#using-names-to-subset-a-vector" id="toc-using-names-to-subset-a-vector"><span class="toc-section-number">8.1.2</span> Using Names to Subset a Vector</a></li>
<li><a href="fundamental-data-structures-in-r.html#named-vectors-as-dictionary-like-structures" id="toc-named-vectors-as-dictionary-like-structures"><span class="toc-section-number">8.1.3</span> Named Vectors as Dictionary-Like Structures</a></li>
<li><a href="fundamental-data-structures-in-r.html#real-life-example" id="toc-real-life-example"><span class="toc-section-number">8.1.4</span> Real-life example</a></li>
<li><a href="fundamental-data-structures-in-r.html#reordering-and-subsetting-with-names" id="toc-reordering-and-subsetting-with-names"><span class="toc-section-number">8.1.5</span> Reordering and Subsetting with Names</a></li>
<li><a href="fundamental-data-structures-in-r.html#conclusion-15" id="toc-conclusion-15"><span class="toc-section-number">8.1.6</span> Conclusion</a></li>
</ul></li>
<li><a href="fundamental-data-structures-in-r.html#lists" id="toc-lists"><span class="toc-section-number">8.2</span> Lists</a>
<ul>
<li><a href="fundamental-data-structures-in-r.html#introduction-to-lists" id="toc-introduction-to-lists"><span class="toc-section-number">8.2.1</span> Introduction to Lists</a></li>
<li><a href="fundamental-data-structures-in-r.html#creating-a-list" id="toc-creating-a-list"><span class="toc-section-number">8.2.2</span> Creating a List</a></li>
<li><a href="fundamental-data-structures-in-r.html#accessing-list-elements" id="toc-accessing-list-elements"><span class="toc-section-number">8.2.3</span> Accessing List Elements</a></li>
<li><a href="fundamental-data-structures-in-r.html#conclusions" id="toc-conclusions"><span class="toc-section-number">8.2.4</span> Conclusions</a></li>
</ul></li>
<li><a href="fundamental-data-structures-in-r.html#dataframes" id="toc-dataframes"><span class="toc-section-number">8.3</span> Dataframes</a>
<ul>
<li><a href="fundamental-data-structures-in-r.html#what-is-a-data-frame" id="toc-what-is-a-data-frame"><span class="toc-section-number">8.3.1</span> What is a Data Frame?</a></li>
<li><a href="fundamental-data-structures-in-r.html#creating-a-data-frame" id="toc-creating-a-data-frame"><span class="toc-section-number">8.3.2</span> Creating a Data Frame</a></li>
<li><a href="fundamental-data-structures-in-r.html#accessing-columns" id="toc-accessing-columns"><span class="toc-section-number">8.3.3</span> Accessing Columns</a></li>
<li><a href="fundamental-data-structures-in-r.html#subsetting-a-data-frame" id="toc-subsetting-a-data-frame"><span class="toc-section-number">8.3.4</span> Subsetting a Data Frame</a></li>
<li><a href="fundamental-data-structures-in-r.html#conclusion-16" id="toc-conclusion-16"><span class="toc-section-number">8.3.5</span> Conclusion</a></li>
</ul></li>
<li><a href="fundamental-data-structures-in-r.html#how-to-represent-categorical-data-in-r-understanding-factors" id="toc-how-to-represent-categorical-data-in-r-understanding-factors"><span class="toc-section-number">8.4</span> How to Represent Categorical Data in R? Understanding Factors</a>
<ul>
<li><a href="fundamental-data-structures-in-r.html#creating-factors" id="toc-creating-factors"><span class="toc-section-number">8.4.1</span> Creating Factors</a></li>
<li><a href="fundamental-data-structures-in-r.html#changing-levels" id="toc-changing-levels"><span class="toc-section-number">8.4.2</span> Changing levels</a></li>
<li><a href="fundamental-data-structures-in-r.html#converting-factors" id="toc-converting-factors"><span class="toc-section-number">8.4.3</span> Converting Factors</a></li>
<li><a href="fundamental-data-structures-in-r.html#customizing-factor-levels" id="toc-customizing-factor-levels"><span class="toc-section-number">8.4.4</span> Customizing Factor Levels</a></li>
<li><a href="fundamental-data-structures-in-r.html#creating-factors-from-character-vectors" id="toc-creating-factors-from-character-vectors"><span class="toc-section-number">8.4.5</span> Creating Factors from Character Vectors</a></li>
<li><a href="fundamental-data-structures-in-r.html#subsetting-factors" id="toc-subsetting-factors"><span class="toc-section-number">8.4.6</span> Subsetting Factors</a></li>
<li><a href="fundamental-data-structures-in-r.html#removing-unused-levels" id="toc-removing-unused-levels"><span class="toc-section-number">8.4.7</span> Removing Unused Levels</a></li>
<li><a href="fundamental-data-structures-in-r.html#conclusion-17" id="toc-conclusion-17"><span class="toc-section-number">8.4.8</span> Conclusion</a></li>
</ul></li>
<li><a href="fundamental-data-structures-in-r.html#section-complete-2" id="toc-section-complete-2"><span class="toc-section-number">8.5</span> Section Complete</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#introduction-to-the-tidyverse-ecosystem" id="toc-introduction-to-the-tidyverse-ecosystem"><span class="toc-section-number">9</span> Introduction to the tidyverse ecosystem</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#what-is-the-tidyverse" id="toc-what-is-the-tidyverse"><span class="toc-section-number">9.1</span> What is the Tidyverse?</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#key-tidyverse-packages" id="toc-key-tidyverse-packages"><span class="toc-section-number">9.2</span> Key Tidyverse Packages</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#dplyr" id="toc-dplyr"><span class="toc-section-number">9.2.1</span> dplyr</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#a-note-on-the-pipe" id="toc-a-note-on-the-pipe"><span class="toc-section-number">9.2.2</span> A note on the pipe</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#ggplot2" id="toc-ggplot2"><span class="toc-section-number">9.2.3</span> ggplot2</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#tidyr" id="toc-tidyr"><span class="toc-section-number">9.2.4</span> tidyr</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#readr" id="toc-readr"><span class="toc-section-number">9.2.5</span> readr</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#tibble" id="toc-tibble"><span class="toc-section-number">9.2.6</span> tibble</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#stringr" id="toc-stringr"><span class="toc-section-number">9.2.7</span> stringr</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#forcats" id="toc-forcats"><span class="toc-section-number">9.2.8</span> forcats</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#real-world-applications-1" id="toc-real-world-applications-1"><span class="toc-section-number">9.2.9</span> Real-World Applications</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#tibble-and-readr---modern-data-structures-in-r" id="toc-tibble-and-readr---modern-data-structures-in-r"><span class="toc-section-number">9.3</span> Tibble and Readr - Modern Data Structures in R</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#tibbles-a-better-way-to-store-data" id="toc-tibbles-a-better-way-to-store-data"><span class="toc-section-number">9.3.1</span> Tibbles: A Better Way to Store Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#rownames-and-tibbles" id="toc-rownames-and-tibbles"><span class="toc-section-number">9.3.2</span> Rownames and Tibbles</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#reading-data-with-readr" id="toc-reading-data-with-readr"><span class="toc-section-number">9.3.3</span> Reading Data with readr</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#using-tibbles-with-imported-data" id="toc-using-tibbles-with-imported-data"><span class="toc-section-number">9.3.4</span> Using Tibbles with Imported Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#note-on-reading-excel-files" id="toc-note-on-reading-excel-files"><span class="toc-section-number">9.3.5</span> Note on reading Excel files</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-18" id="toc-conclusion-18"><span class="toc-section-number">9.3.6</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#the-tidy-data-format" id="toc-the-tidy-data-format"><span class="toc-section-number">9.4</span> The tidy data format</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#the-importance-of-tidy-data" id="toc-the-importance-of-tidy-data"><span class="toc-section-number">9.4.1</span> The Importance of Tidy Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#long-format-vs.-wide-format" id="toc-long-format-vs.-wide-format"><span class="toc-section-number">9.4.2</span> Long Format vs. Wide Format</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#introducing-dplyr-your-data-wrangling-toolkit" id="toc-introducing-dplyr-your-data-wrangling-toolkit"><span class="toc-section-number">9.5</span> Introducing dplyr: Your Data Wrangling Toolkit</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#selecting-specific-columns" id="toc-selecting-specific-columns"><span class="toc-section-number">9.5.1</span> Selecting Specific Columns</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#adding-new-columns" id="toc-adding-new-columns"><span class="toc-section-number">9.5.2</span> Adding New Columns</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#reordering-columns" id="toc-reordering-columns"><span class="toc-section-number">9.5.3</span> Reordering Columns</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#filtering-data" id="toc-filtering-data"><span class="toc-section-number">9.5.4</span> Filtering Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#summarizing-data" id="toc-summarizing-data"><span class="toc-section-number">9.5.5</span> Summarizing Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#sorting-data" id="toc-sorting-data"><span class="toc-section-number">9.5.6</span> Sorting Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-19" id="toc-conclusion-19"><span class="toc-section-number">9.5.7</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#stringr-your-essential-toolkit-to-manipulate-strings" id="toc-stringr-your-essential-toolkit-to-manipulate-strings"><span class="toc-section-number">9.6</span> stringr: your essential toolkit to manipulate strings</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#purrr-ditch-your-for-loops" id="toc-purrr-ditch-your-for-loops"><span class="toc-section-number">9.7</span> purrr: ditch your for loops</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#nesting-data-with-nest" id="toc-nesting-data-with-nest"><span class="toc-section-number">9.7.1</span> Nesting Data with nest()</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#map-and-its-variants" id="toc-map-and-its-variants"><span class="toc-section-number">9.7.2</span> <code>map()</code> and Its Variants</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#note-on-anonymous-functions" id="toc-note-on-anonymous-functions"><span class="toc-section-number">9.7.3</span> Note on Anonymous functions</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-20" id="toc-conclusion-20"><span class="toc-section-number">9.7.4</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#tidying-metadata-from-geo." id="toc-tidying-metadata-from-geo."><span class="toc-section-number">9.8</span> Tidying metadata from GEO.</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#prerequisites" id="toc-prerequisites"><span class="toc-section-number">9.8.1</span> Prerequisites</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#getting-started" id="toc-getting-started"><span class="toc-section-number">9.8.2</span> Getting Started</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#tidying-the-metadata" id="toc-tidying-the-metadata"><span class="toc-section-number">9.8.3</span> Tidying the Metadata</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#converting-to-a-tidy-data-frame" id="toc-converting-to-a-tidy-data-frame"><span class="toc-section-number">9.8.4</span> Converting to a Tidy Data Frame</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#bonus-reading-multiple-tsv-files" id="toc-bonus-reading-multiple-tsv-files"><span class="toc-section-number">9.8.5</span> Bonus: Reading Multiple TSV Files</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#section-complete-3" id="toc-section-complete-3"><span class="toc-section-number">9.9</span> Section Complete</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-21" id="toc-conclusion-21"><span class="toc-section-number">9.9.1</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#data-visualization-with-ggplot2" id="toc-data-visualization-with-ggplot2"><span class="toc-section-number">9.10</span> Data visualization with ggplot2</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#creating-scatter-plots" id="toc-creating-scatter-plots"><span class="toc-section-number">9.11</span> Creating Scatter Plots</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#prerequisites-1" id="toc-prerequisites-1"><span class="toc-section-number">9.11.1</span> Prerequisites</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#loading-and-exploring-the-data" id="toc-loading-and-exploring-the-data"><span class="toc-section-number">9.11.2</span> Loading and Exploring the Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#filtering-the-data" id="toc-filtering-the-data"><span class="toc-section-number">9.11.3</span> Filtering the Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#creating-a-scatter-plot" id="toc-creating-a-scatter-plot"><span class="toc-section-number">9.11.4</span> Creating a Scatter Plot</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#calculating-correlation" id="toc-calculating-correlation"><span class="toc-section-number">9.11.5</span> Calculating Correlation</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#hypothesis-testing" id="toc-hypothesis-testing"><span class="toc-section-number">9.11.6</span> Hypothesis Testing</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#adding-a-regression-line" id="toc-adding-a-regression-line"><span class="toc-section-number">9.11.7</span> Adding a Regression Line</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-22" id="toc-conclusion-22"><span class="toc-section-number">9.11.8</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#understanding-distributions-with-histograms" id="toc-understanding-distributions-with-histograms"><span class="toc-section-number">9.12</span> Understanding Distributions with Histograms</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#creating-a-basic-histogram" id="toc-creating-a-basic-histogram"><span class="toc-section-number">9.12.1</span> Creating a Basic Histogram</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#customizing-histogram-bins" id="toc-customizing-histogram-bins"><span class="toc-section-number">9.12.2</span> Customizing Histogram Bins</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-23" id="toc-conclusion-23"><span class="toc-section-number">9.12.3</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#visualizing-data-distribution-with-boxplots-and-violin-plots" id="toc-visualizing-data-distribution-with-boxplots-and-violin-plots"><span class="toc-section-number">9.13</span> Visualizing Data Distribution with Boxplots and Violin Plots</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#creating-a-basic-boxplot" id="toc-creating-a-basic-boxplot"><span class="toc-section-number">9.13.1</span> Creating a Basic Boxplot</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#rotating-x-axis-labels" id="toc-rotating-x-axis-labels"><span class="toc-section-number">9.13.2</span> Rotating X-axis Labels</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#adding-color-to-boxplots" id="toc-adding-color-to-boxplots"><span class="toc-section-number">9.13.3</span> Adding Color to Boxplots</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#customizing-color-palettes" id="toc-customizing-color-palettes"><span class="toc-section-number">9.13.4</span> Customizing Color Palettes</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#applying-custom-color-palette" id="toc-applying-custom-color-palette"><span class="toc-section-number">9.13.5</span> Applying Custom Color Palette</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#reordering-boxplots" id="toc-reordering-boxplots"><span class="toc-section-number">9.13.6</span> Reordering Boxplots</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#using-violin-plots" id="toc-using-violin-plots"><span class="toc-section-number">9.13.7</span> Using Violin Plots</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-24" id="toc-conclusion-24"><span class="toc-section-number">9.13.8</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#creating-bar-plots-to-visualize-median-epcam-levels-by-cancer-type" id="toc-creating-bar-plots-to-visualize-median-epcam-levels-by-cancer-type"><span class="toc-section-number">9.14</span> Creating Bar Plots to Visualize Median EPCAM Levels by Cancer Type</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#calculating-median-epcam-levels-by-cancer-type" id="toc-calculating-median-epcam-levels-by-cancer-type"><span class="toc-section-number">9.14.1</span> Calculating Median EPCAM Levels by Cancer Type</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#creating-a-basic-bar-plot" id="toc-creating-a-basic-bar-plot"><span class="toc-section-number">9.14.2</span> Creating a Basic Bar Plot</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#reordering-the-bars" id="toc-reordering-the-bars"><span class="toc-section-number">9.14.3</span> Reordering the Bars</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#reversing-the-order" id="toc-reversing-the-order"><span class="toc-section-number">9.14.4</span> Reversing the Order</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#note-on-factor-levels" id="toc-note-on-factor-levels"><span class="toc-section-number">9.14.5</span> Note on factor levels</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-25" id="toc-conclusion-25"><span class="toc-section-number">9.14.6</span> Conclusion</a></li>
</ul></li>
</ul></li>
<li><a href="introduction-to-bioconductor.html#introduction-to-bioconductor" id="toc-introduction-to-bioconductor"><span class="toc-section-number">10</span> Introduction to BioConductor</a>
<ul>
<li><a href="introduction-to-bioconductor.html#introduction-to-biocmanager" id="toc-introduction-to-biocmanager"><span class="toc-section-number">10.1</span> Introduction to BiocManager</a>
<ul>
<li><a href="introduction-to-bioconductor.html#why-use-biocmanager" id="toc-why-use-biocmanager"><span class="toc-section-number">10.1.1</span> Why Use BiocManager?</a></li>
<li><a href="introduction-to-bioconductor.html#installing-biocmanager" id="toc-installing-biocmanager"><span class="toc-section-number">10.1.2</span> Installing BiocManager</a></li>
<li><a href="introduction-to-bioconductor.html#installing-bioconductor-packages" id="toc-installing-bioconductor-packages"><span class="toc-section-number">10.1.3</span> Installing Bioconductor Packages</a></li>
<li><a href="introduction-to-bioconductor.html#updating-bioconductor-packages" id="toc-updating-bioconductor-packages"><span class="toc-section-number">10.1.4</span> Updating Bioconductor Packages</a></li>
<li><a href="introduction-to-bioconductor.html#checking-for-valid-bioconductor-versions" id="toc-checking-for-valid-bioconductor-versions"><span class="toc-section-number">10.1.5</span> Checking for Valid Bioconductor Versions</a></li>
<li><a href="introduction-to-bioconductor.html#conclusion-26" id="toc-conclusion-26"><span class="toc-section-number">10.1.6</span> Conclusion</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Master Real-world Bioinformatics analysis in R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-programming" class="section level1" number="3">
<h1><span class="header-section-number">Chapter 3</span> Introduction to programming</h1>
<div id="what-is-an-algorithm" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> What is an algorithm?</h2>
<p>Welcome to the first lesson of our course! Today, we’re going to explore two fundamental concepts in programming: algorithms and flowcharts. Don’t worry if these terms sound a bit technical; we’ll break them down into simple ideas.</p>
<p>Imagine you’re following a recipe to bake a cake. The recipe gives you step-by-step instructions on what to do, right? An algorithm is similar. It’s a set of instructions or steps designed to perform a specific task. In programming, we use algorithms to tell the computer exactly what we want it to do.</p>
</div>
<div id="why-are-algorithms-important-in-programming" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Why are Algorithms Important in Programming?</h2>
<ul>
<li><p>Clarity: Algorithms serve as an essential tool for strategizing and planning our code. They provide us with a clear roadmap of the steps we need to follow before we start writing the actual code. This pre-coding stage can help us avoid potential issues and ensure that our solutions are well thought out.</p></li>
<li><p>Problem-Solving: Algorithms play a crucial role in problem-solving. They allow us to break down complex tasks into a series of simpler steps, making them easier to manage and understand. By using algorithms, we can tackle large problems by solving each small part one at a time, thus making the overall problem-solving process more efficient and manageable.</p></li>
<li><p>Efficiency: A well-designed algorithm can save significant time and resources. It can help us optimize our code to perform tasks in the fastest and most efficient way possible. By improving the efficiency of our code, we can ensure that it runs smoothly and quickly, thus enhancing the performance of our software or application.</p></li>
</ul>
</div>
<div id="a-real-world-example" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> A Real World Example:</h2>
<p>In this algorithm, we’ll learn how to find the length of a DNA sequence. Knowing the length of a DNA sequence is important for various biological analyses.</p>
<ol style="list-style-type: decimal">
<li><p>Start: Begin the algorithm.</p></li>
<li><p>Input DNA Sequence: Ask the user to input a DNA sequence. For example, the sequence could be a string of letters like “ATCGATGCTA.”</p></li>
<li><p>Initialize Length: Set a variable called “Length” to 0. This variable will be used to keep track of the length of the sequence.</p></li>
<li><p>For Each Base in DNA Sequence:</p></li>
</ol>
<ul>
<li><p>Start a loop that goes through each base in the DNA sequence, one by one.</p></li>
<li><p>For the input “ATCGATGCTA,” the loop will start with “A.”</p></li>
</ul>
<ol start="5" style="list-style-type: decimal">
<li><p>Increase Length by 1: For each base you encounter in the sequence, add 1 to the “Length” variable. This counts the number of bases in the sequence.</p></li>
<li><p>Repeat: Continue the loop until you have processed all the bases in the DNA sequence.</p></li>
<li><p>Output Length: Once the loop is finished, the “Length” variable will contain the length of the DNA sequence. Display this value as the output.</p></li>
<li><p>End: End the algorithm.</p></li>
</ol>
</div>
<div id="variables-data-types-and-expressions" class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> Variables, Data Types, and Expressions</h2>
<p>In this lesson, we’re going to explore some fundamental concepts of programming that are crucial in understanding how to write code for computational biology. We’ll focus on three key ideas: variables, data types, and expressions. Think of these as the building blocks for creating a language that your computer can understand and use to solve biological problems.</p>
<div id="variables" class="section level3" number="3.4.1">
<h3><span class="header-section-number">3.4.1</span> Variables</h3>
<p>A variable is quite similar to a labeled box. It’s a container where you can store information or data. Once you have this box (variable), you can do various things with it: you can put things into it, take things out of it, or even change what’s inside it to something else. This flexibility is immensely useful when dealing with large volumes of data or when you’re accessing often to a specific piece of that data, a common occurrence in the field of computational biology.</p>
<p><img src="images/variable.png" /></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="introduction-to-programming.html#cb1-1" tabindex="-1"></a><span class="co"># Step 1: Define a variable and store the DNA sequence</span></span>
<span id="cb1-2"><a href="introduction-to-programming.html#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="introduction-to-programming.html#cb1-3" tabindex="-1"></a>gene_sequence <span class="ot">=</span> <span class="st">&quot;ATCGAGCTAGCTGCTAGCTAGCTAGCT&quot;</span></span>
<span id="cb1-4"><a href="introduction-to-programming.html#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="introduction-to-programming.html#cb1-5" tabindex="-1"></a><span class="co"># Step 2: Print the stored DNA sequence</span></span>
<span id="cb1-6"><a href="introduction-to-programming.html#cb1-6" tabindex="-1"></a></span>
<span id="cb1-7"><a href="introduction-to-programming.html#cb1-7" tabindex="-1"></a><span class="fu">print</span>(gene_sequence)</span></code></pre></div>
<pre><code>## [1] &quot;ATCGAGCTAGCTGCTAGCTAGCTAGCT&quot;</code></pre>
<p>For instance, consider a situation where you’re working with gene sequences. You can use a variable like ‘gene_sequence’ to store a particular gene’s sequence. Later, you can access this stored sequence, manipulate it, or compare it with other sequences as needed in your computational biology tasks.</p>
</div>
<div id="data-types" class="section level3" number="3.4.2">
<h3><span class="header-section-number">3.4.2</span> Data Types</h3>
<p>In programming, similar to the real world, we come across various data types that help us to structure and understand information. These data types include numbers, which could be whole numbers or numbers with decimal points; text, often referred to as strings in programming terms; booleans, representing True or False values; and lists, which are utilized to hold collections of items.</p>
<p><img src="images/datatype.jpeg" /></p>
<p>Each of these data types serve a unique purpose and are used in different contexts, playing an integral role in how we write and interpret code.</p>
<ul>
<li><p>Numbers: In programming, numbers can be whole or decimal. They represent measurements or quantities in computational biology, like DNA length or chemical concentration.</p></li>
<li><p>Text (Strings): Strings are sequences of characters, used for gene names, DNA sequences, or any textual information in computational biology.</p></li>
<li><p>Lists/Arrays: Lists group items together, useful in handling multiple genes, proteins, or biological elements. They can store different gene names or DNA sequences.</p></li>
<li><p>Booleans: Booleans represent true or false values and are used to express conditions or decisions based on a yes/no scenario, like determining if a specific gene is present.</p></li>
</ul>
<p>Understanding data types is crucial because it helps you work with biological data accurately and efficiently. When you’re coding in computational biology, knowing whether you’re dealing with numbers, text, or lists allows you to use the right tools and operations for each type of data.</p>
<p>For instance, if you need to perform calculations on gene lengths (numbers), you’ll use mathematical operations. If you want to search for a specific gene name (text), you’ll use string manipulation techniques. And when you’re handling multiple genes (lists), you’ll employ list-related functions to process them effectively.</p>
</div>
<div id="expressions" class="section level3" number="3.4.3">
<h3><span class="header-section-number">3.4.3</span> Expressions</h3>
<p><img src="images/expression.png" /></p>
<p>In the world of programming, expressions are like simple puzzles or equations that you can use to do things with data. These expressions are created by putting together a few essential elements:</p>
<ul>
<li><p>Values: Think of these as numbers, like 2 or 3, that you want to work with. In computational biology, these could be things like the length of a gene or the number of amino acids in a protein.</p></li>
<li><p>Variables: These are like labeled containers where you can store information. For example, you might have a variable called ‘gene_length’ that holds the length of a specific gene sequence.</p></li>
<li><p>Operators: Operators are special symbols like +, -, *, or / that you use to perform operations on values and variables. They tell the computer what kind of action to take.</p></li>
</ul>
<p>Now, let’s dive into some beginner-level examples in computational biology:</p>
</div>
<div id="example-finding-the-average-number-of-bases-appearances-in-a-set-of-dna-strings" class="section level3" number="3.4.4">
<h3><span class="header-section-number">3.4.4</span> Example: Finding the average number of bases appearances in a set of DNA strings</h3>
<p>Let’s suppose we want to calculate the average number of bases in a DNA string. Let’s assume we already processed the DNA string and we know the counts for each one.</p>
<p>Given counts of each base:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="introduction-to-programming.html#cb3-1" tabindex="-1"></a>count_A <span class="ot">=</span> <span class="dv">120</span></span>
<span id="cb3-2"><a href="introduction-to-programming.html#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="introduction-to-programming.html#cb3-3" tabindex="-1"></a>count_T <span class="ot">=</span> <span class="dv">90</span></span>
<span id="cb3-4"><a href="introduction-to-programming.html#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="introduction-to-programming.html#cb3-5" tabindex="-1"></a>count_C <span class="ot">=</span> <span class="dv">80</span></span>
<span id="cb3-6"><a href="introduction-to-programming.html#cb3-6" tabindex="-1"></a></span>
<span id="cb3-7"><a href="introduction-to-programming.html#cb3-7" tabindex="-1"></a>count_G <span class="ot">=</span> <span class="dv">110</span></span></code></pre></div>
<p>Calculate the total number of bases</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="introduction-to-programming.html#cb4-1" tabindex="-1"></a>total_bases <span class="ot">=</span> count_A <span class="sc">+</span> count_T <span class="sc">+</span> count_C <span class="sc">+</span> count_G</span></code></pre></div>
<p>Calculate the average number of bases</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="introduction-to-programming.html#cb5-1" tabindex="-1"></a>average_bases <span class="ot">=</span> total_bases <span class="sc">/</span> <span class="dv">4</span></span></code></pre></div>
<p>Print the result</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="introduction-to-programming.html#cb6-1" tabindex="-1"></a><span class="fu">print</span>(average_bases) </span></code></pre></div>
<pre><code>## [1] 100</code></pre>
<p>In this code:</p>
<ol style="list-style-type: decimal">
<li><p>We start by declaring the counts of each base using variables (number_a, number_t, number_c, number_g).</p></li>
<li><p>We calculate the total number of bases by summing up the counts.</p></li>
<li><p>Then, we compute the average number of bases by dividing the total by the number of different bases (4 in this case, representing A, T, C, and G).</p></li>
<li><p>Finally, we print the result, which gives us the average number of bases in the DNA string.</p></li>
</ol>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="intro.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="getting-started-with-r-and-rstudio.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/02-Introduction_to_programming.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
