<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Introduction to R for biologists | Master Real-world Bioinformatics analysis in R</title>
  <meta name="description" content="This is an R course teaching you skills to do real-world bioinformatics" />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Introduction to R for biologists | Master Real-world Bioinformatics analysis in R" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is an R course teaching you skills to do real-world bioinformatics" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Introduction to R for biologists | Master Real-world Bioinformatics analysis in R" />
  
  <meta name="twitter:description" content="This is an R course teaching you skills to do real-world bioinformatics" />
  

<meta name="author" content="Ming Tommy Tang" />


<meta name="date" content="2024-06-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="getting-started-with-r-and-rstudio.html"/>
<link rel="next" href="controlling-the-flow-of-our-programs.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<link href="libs/vembedr-0.1.5/css/vembedr.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#meet-your-instructor"><i class="fa fa-check"></i><b>2.1</b> Meet your Instructor</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#leveraging-our-online-community"><i class="fa fa-check"></i><b>2.2</b> Leveraging our online community</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#install-r-and-r-studio"><i class="fa fa-check"></i><b>2.3</b> Install R and R studio</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#the-role-of-programming-in-biology"><i class="fa fa-check"></i><b>2.4</b> The role of programming in Biology</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="intro.html"><a href="intro.html#what-are-programming-languages"><i class="fa fa-check"></i><b>2.4.1</b> What are programming languages?</a></li>
<li class="chapter" data-level="2.4.2" data-path="intro.html"><a href="intro.html#why-programming-is-important-in-biology"><i class="fa fa-check"></i><b>2.4.2</b> Why programming is important in biology?</a></li>
<li class="chapter" data-level="2.4.3" data-path="intro.html"><a href="intro.html#what-are-the-most-used-programming-languages-within-biology"><i class="fa fa-check"></i><b>2.4.3</b> What are the most used programming languages within biology?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html"><i class="fa fa-check"></i><b>3</b> Introduction to programming</a>
<ul>
<li class="chapter" data-level="3.1" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#what-is-an-algorithm"><i class="fa fa-check"></i><b>3.1</b> What is an algorithm?</a></li>
<li class="chapter" data-level="3.2" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#why-are-algorithms-important-in-programming"><i class="fa fa-check"></i><b>3.2</b> Why are Algorithms Important in Programming?</a></li>
<li class="chapter" data-level="3.3" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#a-real-world-example"><i class="fa fa-check"></i><b>3.3</b> A Real World Example:</a></li>
<li class="chapter" data-level="3.4" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#variables-data-types-and-expressions"><i class="fa fa-check"></i><b>3.4</b> Variables, Data Types, and Expressions</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#variables"><i class="fa fa-check"></i><b>3.4.1</b> Variables</a></li>
<li class="chapter" data-level="3.4.2" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#data-types"><i class="fa fa-check"></i><b>3.4.2</b> Data Types</a></li>
<li class="chapter" data-level="3.4.3" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#expressions"><i class="fa fa-check"></i><b>3.4.3</b> Expressions</a></li>
<li class="chapter" data-level="3.4.4" data-path="introduction-to-programming.html"><a href="introduction-to-programming.html#example-finding-the-average-number-of-bases-appearances-in-a-set-of-dna-strings"><i class="fa fa-check"></i><b>3.4.4</b> Example: Finding the average number of bases appearances in a set of DNA strings</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="getting-started-with-r-and-rstudio.html"><a href="getting-started-with-r-and-rstudio.html"><i class="fa fa-check"></i><b>4</b> Getting started with R and RStudio</a>
<ul>
<li class="chapter" data-level="4.1" data-path="getting-started-with-r-and-rstudio.html"><a href="getting-started-with-r-and-rstudio.html#what-is-r-and-why-is-it-used-in-biology"><i class="fa fa-check"></i><b>4.1</b> What is R and why is it used in Biology?</a></li>
<li class="chapter" data-level="4.2" data-path="getting-started-with-r-and-rstudio.html"><a href="getting-started-with-r-and-rstudio.html#what-is-r"><i class="fa fa-check"></i><b>4.2</b> What is R?</a></li>
<li class="chapter" data-level="4.3" data-path="getting-started-with-r-and-rstudio.html"><a href="getting-started-with-r-and-rstudio.html#why-is-r-used-in-biology"><i class="fa fa-check"></i><b>4.3</b> Why is R Used in Biology?</a></li>
<li class="chapter" data-level="4.4" data-path="getting-started-with-r-and-rstudio.html"><a href="getting-started-with-r-and-rstudio.html#introduction-to-r-studio"><i class="fa fa-check"></i><b>4.4</b> Introduction to R-Studio</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html"><i class="fa fa-check"></i><b>5</b> Introduction to R for biologists</a>
<ul>
<li class="chapter" data-level="5.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#basic-data-types-in-r"><i class="fa fa-check"></i><b>5.1</b> Basic Data Types in R</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#before-starting"><i class="fa fa-check"></i><b>5.1.1</b> Before Starting</a></li>
<li class="chapter" data-level="5.1.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#numeric"><i class="fa fa-check"></i><b>5.1.2</b> 1. Numeric</a></li>
<li class="chapter" data-level="5.1.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#character-string"><i class="fa fa-check"></i><b>5.1.3</b> 2. Character (String)</a></li>
<li class="chapter" data-level="5.1.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#integer"><i class="fa fa-check"></i><b>5.1.4</b> 3. Integer</a></li>
<li class="chapter" data-level="5.1.5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#logical-boolean"><i class="fa fa-check"></i><b>5.1.5</b> 4. Logical (Boolean)</a></li>
<li class="chapter" data-level="5.1.6" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#factor"><i class="fa fa-check"></i><b>5.1.6</b> 5. Factor</a></li>
<li class="chapter" data-level="5.1.7" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#date-and-time"><i class="fa fa-check"></i><b>5.1.7</b> 6. Date and Time</a></li>
<li class="chapter" data-level="5.1.8" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#complex"><i class="fa fa-check"></i><b>5.1.8</b> 7. Complex</a></li>
<li class="chapter" data-level="5.1.9" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#missing-values-na"><i class="fa fa-check"></i><b>5.1.9</b> 8. Missing Values (NA)</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#the-key-concept-of-r-vectors"><i class="fa fa-check"></i><b>5.2</b> The key concept of R: Vectors</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#creating-numeric-and-character-vectors"><i class="fa fa-check"></i><b>5.2.1</b> Creating Numeric and Character Vectors</a></li>
<li class="chapter" data-level="5.2.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#saving-vectors-to-variables"><i class="fa fa-check"></i><b>5.2.2</b> Saving Vectors to Variables</a></li>
<li class="chapter" data-level="5.2.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#vectorized-calculations"><i class="fa fa-check"></i><b>5.2.3</b> Vectorized Calculations</a></li>
<li class="chapter" data-level="5.2.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#performing-vectorized-calculations"><i class="fa fa-check"></i><b>5.2.4</b> Performing Vectorized Calculations</a></li>
<li class="chapter" data-level="5.2.5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#operations-with-character-vectors"><i class="fa fa-check"></i><b>5.2.5</b> Operations with character vectors</a></li>
<li class="chapter" data-level="5.2.6" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#logical-vectors"><i class="fa fa-check"></i><b>5.2.6</b> Logical Vectors</a></li>
<li class="chapter" data-level="5.2.7" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#conclusion"><i class="fa fa-check"></i><b>5.2.7</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#subsetting-and-indexing"><i class="fa fa-check"></i><b>5.3</b> Subsetting and Indexing</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#indexing-in-r"><i class="fa fa-check"></i><b>5.3.1</b> Indexing in R</a></li>
<li class="chapter" data-level="5.3.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#slicing-in-r"><i class="fa fa-check"></i><b>5.3.2</b> Slicing in R</a></li>
<li class="chapter" data-level="5.3.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#negative-indexing"><i class="fa fa-check"></i><b>5.3.3</b> Negative Indexing</a></li>
<li class="chapter" data-level="5.3.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#conclusion-1"><i class="fa fa-check"></i><b>5.3.4</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#understanding-and-manipulating-matrices"><i class="fa fa-check"></i><b>5.4</b> Understanding and Manipulating Matrices</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#creating-a-matrix"><i class="fa fa-check"></i><b>5.4.1</b> Creating a Matrix</a></li>
<li class="chapter" data-level="5.4.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#adding-row-and-column-names"><i class="fa fa-check"></i><b>5.4.2</b> Adding Row and Column Names</a></li>
<li class="chapter" data-level="5.4.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#subsetting-a-matrix"><i class="fa fa-check"></i><b>5.4.3</b> Subsetting a Matrix</a></li>
<li class="chapter" data-level="5.4.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#calculations-with-matrices"><i class="fa fa-check"></i><b>5.4.4</b> Calculations with Matrices</a></li>
<li class="chapter" data-level="5.4.5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#normalization"><i class="fa fa-check"></i><b>5.4.5</b> Normalization</a></li>
<li class="chapter" data-level="5.4.6" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#conclusion-2"><i class="fa fa-check"></i><b>5.4.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#essential-functions-in-r"><i class="fa fa-check"></i><b>5.5</b> Essential Functions in R</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#length-function"><i class="fa fa-check"></i><b>5.5.1</b> <code>length</code> Function</a></li>
<li class="chapter" data-level="5.5.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#unique-function"><i class="fa fa-check"></i><b>5.5.2</b> <code>unique</code> Function</a></li>
<li class="chapter" data-level="5.5.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#sort-function"><i class="fa fa-check"></i><b>5.5.3</b> <code>sort</code> Function</a></li>
<li class="chapter" data-level="5.5.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#cor-function"><i class="fa fa-check"></i><b>5.5.4</b> <code>cor</code> Function</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#functions-organizing-your-code-for-reusability"><i class="fa fa-check"></i><b>5.6</b> Functions: Organizing Your Code for Reusability</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#what-is-a-function"><i class="fa fa-check"></i><b>5.6.1</b> What is a Function?</a></li>
<li class="chapter" data-level="5.6.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#creating-a-basic-function-calculating-the-mean"><i class="fa fa-check"></i><b>5.6.2</b> Creating a Basic Function: Calculating the Mean</a></li>
<li class="chapter" data-level="5.6.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#using-the-custom-function"><i class="fa fa-check"></i><b>5.6.3</b> Using the Custom Function</a></li>
<li class="chapter" data-level="5.6.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#returning-a-value-from-the-function"><i class="fa fa-check"></i><b>5.6.4</b> Returning a Value from the Function</a></li>
<li class="chapter" data-level="5.6.5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#optional-omitting-the-return-statement"><i class="fa fa-check"></i><b>5.6.5</b> Optional: Omitting the return Statement</a></li>
<li class="chapter" data-level="5.6.6" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#using-more-than-one-argument"><i class="fa fa-check"></i><b>5.6.6</b> Using more than one argument</a></li>
<li class="chapter" data-level="5.6.7" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#understanding-missing-values-nas-in-r"><i class="fa fa-check"></i><b>5.6.7</b> Understanding Missing Values (NAs) in R</a></li>
<li class="chapter" data-level="5.6.8" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#initial-attempt-a-function-without-handling-nas"><i class="fa fa-check"></i><b>5.6.8</b> Initial Attempt: A Function Without Handling NAs</a></li>
<li class="chapter" data-level="5.6.9" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#adding-the-remove_na-argument"><i class="fa fa-check"></i><b>5.6.9</b> Adding the <code>remove_na</code> Argument</a></li>
<li class="chapter" data-level="5.6.10" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#practical-application"><i class="fa fa-check"></i><b>5.6.10</b> Practical Application</a></li>
<li class="chapter" data-level="5.6.11" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#getting-help-with-functions"><i class="fa fa-check"></i><b>5.6.11</b> Getting help with functions</a></li>
<li class="chapter" data-level="5.6.12" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#a-real-world-example-1"><i class="fa fa-check"></i><b>5.6.12</b> A Real World Example</a></li>
<li class="chapter" data-level="5.6.13" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#conclusion-3"><i class="fa fa-check"></i><b>5.6.13</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#common-mistakes-to-avoid"><i class="fa fa-check"></i><b>5.7</b> Common Mistakes to avoid</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#mixing-data-types"><i class="fa fa-check"></i><b>5.7.1</b> Mixing Data Types</a></li>
<li class="chapter" data-level="5.7.2" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#forgetting-function-parentheses"><i class="fa fa-check"></i><b>5.7.2</b> Forgetting Function Parentheses</a></li>
<li class="chapter" data-level="5.7.3" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#overwriting-built-in-functions"><i class="fa fa-check"></i><b>5.7.3</b> Overwriting Built-in Functions</a></li>
<li class="chapter" data-level="5.7.4" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#misunderstanding-variable-scoping"><i class="fa fa-check"></i><b>5.7.4</b> Misunderstanding Variable Scoping</a></li>
<li class="chapter" data-level="5.7.5" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#best-practices-for-variable-scoping-include"><i class="fa fa-check"></i><b>5.7.5</b> Best practices for variable scoping include</a></li>
<li class="chapter" data-level="5.7.6" data-path="introduction-to-r-for-biologists.html"><a href="introduction-to-r-for-biologists.html#conclusion-4"><i class="fa fa-check"></i><b>5.7.6</b> Conclusion</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html"><i class="fa fa-check"></i><b>6</b> Controlling the flow of our programs</a>
<ul>
<li class="chapter" data-level="6.1" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#boolean-operators"><i class="fa fa-check"></i><b>6.1</b> Boolean Operators</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#comparison-operators"><i class="fa fa-check"></i><b>6.1.1</b> Comparison Operators</a></li>
<li class="chapter" data-level="6.1.2" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#logical-operators"><i class="fa fa-check"></i><b>6.1.2</b> Logical Operators</a></li>
<li class="chapter" data-level="6.1.3" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#conclusion-5"><i class="fa fa-check"></i><b>6.1.3</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#conditional-statements-if-else"><i class="fa fa-check"></i><b>6.2</b> Conditional statements (<code>if</code>, <code>else</code>)</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#the-if-statement"><i class="fa fa-check"></i><b>6.2.1</b> The <code>if</code> Statement</a></li>
<li class="chapter" data-level="6.2.2" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#else-statement"><i class="fa fa-check"></i><b>6.2.2</b> <code>else</code> statement</a></li>
<li class="chapter" data-level="6.2.3" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#else-if"><i class="fa fa-check"></i><b>6.2.3</b> <code>else if</code></a></li>
<li class="chapter" data-level="6.2.4" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#exercise"><i class="fa fa-check"></i><b>6.2.4</b> Exercise</a></li>
<li class="chapter" data-level="6.2.5" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#conclusion-6"><i class="fa fa-check"></i><b>6.2.5</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#loops"><i class="fa fa-check"></i><b>6.3</b> Loops</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#the-while-loop"><i class="fa fa-check"></i><b>6.3.1</b> The <code>while</code> Loop</a></li>
<li class="chapter" data-level="6.3.2" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#while-loops-in-real-life"><i class="fa fa-check"></i><b>6.3.2</b> While Loops in Real Life</a></li>
<li class="chapter" data-level="6.3.3" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#the-for-loop"><i class="fa fa-check"></i><b>6.3.3</b> The <code>for</code> Loop</a></li>
<li class="chapter" data-level="6.3.4" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#conclusion-7"><i class="fa fa-check"></i><b>6.3.4</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#gene-expression-annotation-using-loops-and-control-structures"><i class="fa fa-check"></i><b>6.4</b> Gene Expression Annotation using Loops and Control Structures</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#the-data"><i class="fa fa-check"></i><b>6.4.1</b> The Data</a></li>
<li class="chapter" data-level="6.4.2" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#creating-annotations"><i class="fa fa-check"></i><b>6.4.2</b> Creating Annotations</a></li>
<li class="chapter" data-level="6.4.3" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#the-loop"><i class="fa fa-check"></i><b>6.4.3</b> The Loop</a></li>
<li class="chapter" data-level="6.4.4" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#output"><i class="fa fa-check"></i><b>6.4.4</b> Output</a></li>
<li class="chapter" data-level="6.4.5" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#putting-it-all-together"><i class="fa fa-check"></i><b>6.4.5</b> Putting it all together</a></li>
<li class="chapter" data-level="6.4.6" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#conclusion-8"><i class="fa fa-check"></i><b>6.4.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#lets-solve-a-challenge"><i class="fa fa-check"></i><b>6.5</b> Letâ€™s solve a Challenge</a></li>
<li class="chapter" data-level="6.6" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#solution"><i class="fa fa-check"></i><b>6.6</b> Solution</a>
<ul>
<li class="chapter" data-level="6.6.1" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#using-a-for-loop"><i class="fa fa-check"></i><b>6.6.1</b> using a for loop</a></li>
<li class="chapter" data-level="6.6.2" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#vectorized-solution"><i class="fa fa-check"></i><b>6.6.2</b> vectorized solution</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="controlling-the-flow-of-our-programs.html"><a href="controlling-the-flow-of-our-programs.html#section-complete"><i class="fa fa-check"></i><b>6.7</b> Section complete</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html"><i class="fa fa-check"></i><b>7</b> Going more in-depth with R</a>
<ul>
<li class="chapter" data-level="7.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#handling-missing-values"><i class="fa fa-check"></i><b>7.1</b> Handling Missing Values</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#understanding-na-in-r"><i class="fa fa-check"></i><b>7.1.1</b> Understanding NA in R</a></li>
<li class="chapter" data-level="7.1.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#identifying-missing-values"><i class="fa fa-check"></i><b>7.1.2</b> Identifying Missing Values</a></li>
<li class="chapter" data-level="7.1.3" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#remove-missing-values"><i class="fa fa-check"></i><b>7.1.3</b> remove missing values</a></li>
<li class="chapter" data-level="7.1.4" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#real-life-note"><i class="fa fa-check"></i><b>7.1.4</b> real-life note</a></li>
<li class="chapter" data-level="7.1.5" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#conclusion-9"><i class="fa fa-check"></i><b>7.1.5</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#introduction-to-statistical-tests-and-p-values"><i class="fa fa-check"></i><b>7.2</b> Introduction to Statistical Tests and P-Values</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#practical-application-of-statistical-test-with-r"><i class="fa fa-check"></i><b>7.2.1</b> Practical Application of statistical test with R</a></li>
<li class="chapter" data-level="7.2.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#conclusion-10"><i class="fa fa-check"></i><b>7.2.2</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#understanding-r-packages"><i class="fa fa-check"></i><b>7.3</b> Understanding R Packages</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#what-is-an-r-package"><i class="fa fa-check"></i><b>7.3.1</b> What is an R Package?</a></li>
<li class="chapter" data-level="7.3.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#why-use-r-packages"><i class="fa fa-check"></i><b>7.3.2</b> Why Use R Packages?</a></li>
<li class="chapter" data-level="7.3.3" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#how-to-use-a-package"><i class="fa fa-check"></i><b>7.3.3</b> How to use a package?</a></li>
<li class="chapter" data-level="7.3.4" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#embracing-errors"><i class="fa fa-check"></i><b>7.3.4</b> Embracing Errors</a></li>
<li class="chapter" data-level="7.3.5" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#how-to-look-for-package-docs-within-r-studio"><i class="fa fa-check"></i><b>7.3.5</b> How to look for package docs within R studio?</a></li>
<li class="chapter" data-level="7.3.6" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#exploring-package-vignettes"><i class="fa fa-check"></i><b>7.3.6</b> Exploring Package Vignettes</a></li>
<li class="chapter" data-level="7.3.7" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#practical-tip"><i class="fa fa-check"></i><b>7.3.7</b> Practical Tip</a></li>
<li class="chapter" data-level="7.3.8" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#conclusion-11"><i class="fa fa-check"></i><b>7.3.8</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#exploring-functions-in-different-packages-avoiding-collisions-and-accessing-them"><i class="fa fa-check"></i><b>7.4</b> Exploring Functions in Different Packages: Avoiding Collisions and Accessing Them</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#using-double-colon"><i class="fa fa-check"></i><b>7.4.1</b> Using Double Colon (<code>::</code>)</a></li>
<li class="chapter" data-level="7.4.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#conclusion-12"><i class="fa fa-check"></i><b>7.4.2</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#writing-custom-scripts"><i class="fa fa-check"></i><b>7.5</b> Writing Custom Scripts</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#your-working-directory"><i class="fa fa-check"></i><b>7.5.1</b> Your Working Directory</a></li>
<li class="chapter" data-level="7.5.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#writing-functions-in-scripts"><i class="fa fa-check"></i><b>7.5.2</b> Writing Functions in Scripts</a></li>
<li class="chapter" data-level="7.5.3" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#importing-functions-from-scripts"><i class="fa fa-check"></i><b>7.5.3</b> Importing Functions from Scripts</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#data-io"><i class="fa fa-check"></i><b>7.6</b> Data I/O</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#built-in-functions"><i class="fa fa-check"></i><b>7.6.1</b> Built-In Functions</a></li>
<li class="chapter" data-level="7.6.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#real-world-applications"><i class="fa fa-check"></i><b>7.6.2</b> Real-World Applications</a></li>
<li class="chapter" data-level="7.6.3" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#conclusion-13"><i class="fa fa-check"></i><b>7.6.3</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#best-practices-in-modular-programming-and-project-management"><i class="fa fa-check"></i><b>7.7</b> Best Practices in Modular Programming and Project Management</a>
<ul>
<li class="chapter" data-level="7.7.1" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#introduction-to-modular-programming"><i class="fa fa-check"></i><b>7.7.1</b> Introduction to Modular Programming</a></li>
<li class="chapter" data-level="7.7.2" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#breaking-down-your-project"><i class="fa fa-check"></i><b>7.7.2</b> Breaking Down Your Project</a></li>
<li class="chapter" data-level="7.7.3" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#conclusion-14"><i class="fa fa-check"></i><b>7.7.3</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="going-more-in-depth-with-r.html"><a href="going-more-in-depth-with-r.html#section-complete-1"><i class="fa fa-check"></i><b>7.8</b> Section complete</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html"><i class="fa fa-check"></i><b>8</b> Fundamental Data Structures in R</a>
<ul>
<li class="chapter" data-level="8.1" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#named-vectors"><i class="fa fa-check"></i><b>8.1</b> Named Vectors</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#what-is-a-named-vector"><i class="fa fa-check"></i><b>8.1.1</b> What is a Named Vector?</a></li>
<li class="chapter" data-level="8.1.2" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#using-names-to-subset-a-vector"><i class="fa fa-check"></i><b>8.1.2</b> Using Names to Subset a Vector</a></li>
<li class="chapter" data-level="8.1.3" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#named-vectors-as-dictionary-like-structures"><i class="fa fa-check"></i><b>8.1.3</b> Named Vectors as Dictionary-Like Structures</a></li>
<li class="chapter" data-level="8.1.4" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#real-life-example"><i class="fa fa-check"></i><b>8.1.4</b> Real-life example</a></li>
<li class="chapter" data-level="8.1.5" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#reordering-and-subsetting-with-names"><i class="fa fa-check"></i><b>8.1.5</b> Reordering and Subsetting with Names</a></li>
<li class="chapter" data-level="8.1.6" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#conclusion-15"><i class="fa fa-check"></i><b>8.1.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#lists"><i class="fa fa-check"></i><b>8.2</b> Lists</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#introduction-to-lists"><i class="fa fa-check"></i><b>8.2.1</b> Introduction to Lists</a></li>
<li class="chapter" data-level="8.2.2" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#creating-a-list"><i class="fa fa-check"></i><b>8.2.2</b> Creating a List</a></li>
<li class="chapter" data-level="8.2.3" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#accessing-list-elements"><i class="fa fa-check"></i><b>8.2.3</b> Accessing List Elements</a></li>
<li class="chapter" data-level="8.2.4" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#conclusions"><i class="fa fa-check"></i><b>8.2.4</b> Conclusions</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#dataframes"><i class="fa fa-check"></i><b>8.3</b> Dataframes</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#what-is-a-data-frame"><i class="fa fa-check"></i><b>8.3.1</b> What is a Data Frame?</a></li>
<li class="chapter" data-level="8.3.2" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#creating-a-data-frame"><i class="fa fa-check"></i><b>8.3.2</b> Creating a Data Frame</a></li>
<li class="chapter" data-level="8.3.3" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#accessing-columns"><i class="fa fa-check"></i><b>8.3.3</b> Accessing Columns</a></li>
<li class="chapter" data-level="8.3.4" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#subsetting-a-data-frame"><i class="fa fa-check"></i><b>8.3.4</b> Subsetting a Data Frame</a></li>
<li class="chapter" data-level="8.3.5" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#conclusion-16"><i class="fa fa-check"></i><b>8.3.5</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#how-to-represent-categorical-data-in-r-understanding-factors"><i class="fa fa-check"></i><b>8.4</b> How to Represent Categorical Data in R? Understanding Factors</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#creating-factors"><i class="fa fa-check"></i><b>8.4.1</b> Creating Factors</a></li>
<li class="chapter" data-level="8.4.2" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#changing-levels"><i class="fa fa-check"></i><b>8.4.2</b> Changing levels</a></li>
<li class="chapter" data-level="8.4.3" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#converting-factors"><i class="fa fa-check"></i><b>8.4.3</b> Converting Factors</a></li>
<li class="chapter" data-level="8.4.4" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#customizing-factor-levels"><i class="fa fa-check"></i><b>8.4.4</b> Customizing Factor Levels</a></li>
<li class="chapter" data-level="8.4.5" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#creating-factors-from-character-vectors"><i class="fa fa-check"></i><b>8.4.5</b> Creating Factors from Character Vectors</a></li>
<li class="chapter" data-level="8.4.6" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#subsetting-factors"><i class="fa fa-check"></i><b>8.4.6</b> Subsetting Factors</a></li>
<li class="chapter" data-level="8.4.7" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#removing-unused-levels"><i class="fa fa-check"></i><b>8.4.7</b> Removing Unused Levels</a></li>
<li class="chapter" data-level="8.4.8" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#conclusion-17"><i class="fa fa-check"></i><b>8.4.8</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="fundamental-data-structures-in-r.html"><a href="fundamental-data-structures-in-r.html#section-complete-2"><i class="fa fa-check"></i><b>8.5</b> Section Complete</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html"><i class="fa fa-check"></i><b>9</b> Introduction to the tidyverse ecosystem</a>
<ul>
<li class="chapter" data-level="9.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#what-is-the-tidyverse"><i class="fa fa-check"></i><b>9.1</b> What is the Tidyverse?</a></li>
<li class="chapter" data-level="9.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#key-tidyverse-packages"><i class="fa fa-check"></i><b>9.2</b> Key Tidyverse Packages</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#dplyr"><i class="fa fa-check"></i><b>9.2.1</b> dplyr</a></li>
<li class="chapter" data-level="9.2.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#a-note-on-the-pipe"><i class="fa fa-check"></i><b>9.2.2</b> A note on the pipe</a></li>
<li class="chapter" data-level="9.2.3" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#ggplot2"><i class="fa fa-check"></i><b>9.2.3</b> ggplot2</a></li>
<li class="chapter" data-level="9.2.4" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#tidyr"><i class="fa fa-check"></i><b>9.2.4</b> tidyr</a></li>
<li class="chapter" data-level="9.2.5" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#readr"><i class="fa fa-check"></i><b>9.2.5</b> readr</a></li>
<li class="chapter" data-level="9.2.6" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#tibble"><i class="fa fa-check"></i><b>9.2.6</b> tibble</a></li>
<li class="chapter" data-level="9.2.7" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#stringr"><i class="fa fa-check"></i><b>9.2.7</b> stringr</a></li>
<li class="chapter" data-level="9.2.8" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#forcats"><i class="fa fa-check"></i><b>9.2.8</b> forcats</a></li>
<li class="chapter" data-level="9.2.9" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#real-world-applications-1"><i class="fa fa-check"></i><b>9.2.9</b> Real-World Applications</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#tibble-and-readr---modern-data-structures-in-r"><i class="fa fa-check"></i><b>9.3</b> Tibble and Readr - Modern Data Structures in R</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#tibbles-a-better-way-to-store-data"><i class="fa fa-check"></i><b>9.3.1</b> Tibbles: A Better Way to Store Data</a></li>
<li class="chapter" data-level="9.3.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#rownames-and-tibbles"><i class="fa fa-check"></i><b>9.3.2</b> Rownames and Tibbles</a></li>
<li class="chapter" data-level="9.3.3" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#reading-data-with-readr"><i class="fa fa-check"></i><b>9.3.3</b> Reading Data with readr</a></li>
<li class="chapter" data-level="9.3.4" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#using-tibbles-with-imported-data"><i class="fa fa-check"></i><b>9.3.4</b> Using Tibbles with Imported Data</a></li>
<li class="chapter" data-level="9.3.5" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#note-on-reading-excel-files"><i class="fa fa-check"></i><b>9.3.5</b> Note on reading Excel files</a></li>
<li class="chapter" data-level="9.3.6" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-18"><i class="fa fa-check"></i><b>9.3.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#the-tidy-data-format"><i class="fa fa-check"></i><b>9.4</b> The tidy data format</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#the-importance-of-tidy-data"><i class="fa fa-check"></i><b>9.4.1</b> The Importance of Tidy Data</a></li>
<li class="chapter" data-level="9.4.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#long-format-vs.-wide-format"><i class="fa fa-check"></i><b>9.4.2</b> Long Format vs.Â Wide Format</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#introducing-dplyr-your-data-wrangling-toolkit"><i class="fa fa-check"></i><b>9.5</b> Introducing dplyr: Your Data Wrangling Toolkit</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#selecting-specific-columns"><i class="fa fa-check"></i><b>9.5.1</b> Selecting Specific Columns</a></li>
<li class="chapter" data-level="9.5.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#adding-new-columns"><i class="fa fa-check"></i><b>9.5.2</b> Adding New Columns</a></li>
<li class="chapter" data-level="9.5.3" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#reordering-columns"><i class="fa fa-check"></i><b>9.5.3</b> Reordering Columns</a></li>
<li class="chapter" data-level="9.5.4" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#filtering-data"><i class="fa fa-check"></i><b>9.5.4</b> Filtering Data</a></li>
<li class="chapter" data-level="9.5.5" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#summarizing-data"><i class="fa fa-check"></i><b>9.5.5</b> Summarizing Data</a></li>
<li class="chapter" data-level="9.5.6" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#sorting-data"><i class="fa fa-check"></i><b>9.5.6</b> Sorting Data</a></li>
<li class="chapter" data-level="9.5.7" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-19"><i class="fa fa-check"></i><b>9.5.7</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#stringr-your-essential-toolkit-to-manipulate-strings"><i class="fa fa-check"></i><b>9.6</b> stringr: your essential toolkit to manipulate strings</a></li>
<li class="chapter" data-level="9.7" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#purrr-ditch-your-for-loops"><i class="fa fa-check"></i><b>9.7</b> purrr: ditch your for loops</a>
<ul>
<li class="chapter" data-level="9.7.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#nesting-data-with-nest"><i class="fa fa-check"></i><b>9.7.1</b> Nesting Data with nest()</a></li>
<li class="chapter" data-level="9.7.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#map-and-its-variants"><i class="fa fa-check"></i><b>9.7.2</b> <code>map()</code> and Its Variants</a></li>
<li class="chapter" data-level="9.7.3" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#note-on-anonymous-functions"><i class="fa fa-check"></i><b>9.7.3</b> Note on Anonymous functions</a></li>
<li class="chapter" data-level="9.7.4" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-20"><i class="fa fa-check"></i><b>9.7.4</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="9.8" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#tidying-metadata-from-geo."><i class="fa fa-check"></i><b>9.8</b> Tidying metadata from GEO.</a>
<ul>
<li class="chapter" data-level="9.8.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#prerequisites"><i class="fa fa-check"></i><b>9.8.1</b> Prerequisites</a></li>
<li class="chapter" data-level="9.8.2" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#getting-started"><i class="fa fa-check"></i><b>9.8.2</b> Getting Started</a></li>
<li class="chapter" data-level="9.8.3" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#tidying-the-metadata"><i class="fa fa-check"></i><b>9.8.3</b> Tidying the Metadata</a></li>
<li class="chapter" data-level="9.8.4" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#converting-to-a-tidy-data-frame"><i class="fa fa-check"></i><b>9.8.4</b> Converting to a Tidy Data Frame</a></li>
<li class="chapter" data-level="9.8.5" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#bonus-reading-multiple-tsv-files"><i class="fa fa-check"></i><b>9.8.5</b> Bonus: Reading Multiple TSV Files</a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#section-complete-3"><i class="fa fa-check"></i><b>9.9</b> Section Complete</a>
<ul>
<li class="chapter" data-level="9.9.1" data-path="introduction-to-the-tidyverse-ecosystem.html"><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-21"><i class="fa fa-check"></i><b>9.9.1</b> Conclusion</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html"><i class="fa fa-check"></i><b>10</b> Data visualization with ggplot2</a>
<ul>
<li class="chapter" data-level="10.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#creating-scatter-plots"><i class="fa fa-check"></i><b>10.1</b> Creating Scatter Plots</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#prerequisites-1"><i class="fa fa-check"></i><b>10.1.1</b> Prerequisites</a></li>
<li class="chapter" data-level="10.1.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#loading-and-exploring-the-data"><i class="fa fa-check"></i><b>10.1.2</b> Loading and Exploring the Data</a></li>
<li class="chapter" data-level="10.1.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#filtering-the-data"><i class="fa fa-check"></i><b>10.1.3</b> Filtering the Data</a></li>
<li class="chapter" data-level="10.1.4" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#creating-a-scatter-plot"><i class="fa fa-check"></i><b>10.1.4</b> Creating a Scatter Plot</a></li>
<li class="chapter" data-level="10.1.5" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#calculating-correlation"><i class="fa fa-check"></i><b>10.1.5</b> Calculating Correlation</a></li>
<li class="chapter" data-level="10.1.6" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#hypothesis-testing"><i class="fa fa-check"></i><b>10.1.6</b> Hypothesis Testing</a></li>
<li class="chapter" data-level="10.1.7" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#adding-a-regression-line"><i class="fa fa-check"></i><b>10.1.7</b> Adding a Regression Line</a></li>
<li class="chapter" data-level="10.1.8" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#conclusion-22"><i class="fa fa-check"></i><b>10.1.8</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#understanding-distributions-with-histograms"><i class="fa fa-check"></i><b>10.2</b> Understanding Distributions with Histograms</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#creating-a-basic-histogram"><i class="fa fa-check"></i><b>10.2.1</b> Creating a Basic Histogram</a></li>
<li class="chapter" data-level="10.2.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#customizing-histogram-bins"><i class="fa fa-check"></i><b>10.2.2</b> Customizing Histogram Bins</a></li>
<li class="chapter" data-level="10.2.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#conclusion-23"><i class="fa fa-check"></i><b>10.2.3</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#visualizing-data-distribution-with-boxplots-and-violin-plots"><i class="fa fa-check"></i><b>10.3</b> Visualizing Data Distribution with Boxplots and Violin Plots</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#creating-a-basic-boxplot"><i class="fa fa-check"></i><b>10.3.1</b> Creating a Basic Boxplot</a></li>
<li class="chapter" data-level="10.3.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#rotating-x-axis-labels"><i class="fa fa-check"></i><b>10.3.2</b> Rotating X-axis Labels</a></li>
<li class="chapter" data-level="10.3.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#adding-color-to-boxplots"><i class="fa fa-check"></i><b>10.3.3</b> Adding Color to Boxplots</a></li>
<li class="chapter" data-level="10.3.4" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#customizing-color-palettes"><i class="fa fa-check"></i><b>10.3.4</b> Customizing Color Palettes</a></li>
<li class="chapter" data-level="10.3.5" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#applying-custom-color-palette"><i class="fa fa-check"></i><b>10.3.5</b> Applying Custom Color Palette</a></li>
<li class="chapter" data-level="10.3.6" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#reordering-boxplots"><i class="fa fa-check"></i><b>10.3.6</b> Reordering Boxplots</a></li>
<li class="chapter" data-level="10.3.7" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#using-violin-plots"><i class="fa fa-check"></i><b>10.3.7</b> Using Violin Plots</a></li>
<li class="chapter" data-level="10.3.8" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#conclusion-24"><i class="fa fa-check"></i><b>10.3.8</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#creating-bar-plots-to-visualize-median-epcam-levels-by-cancer-type"><i class="fa fa-check"></i><b>10.4</b> Creating Bar Plots to Visualize Median EPCAM Levels by Cancer Type</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#calculating-median-epcam-levels-by-cancer-type"><i class="fa fa-check"></i><b>10.4.1</b> Calculating Median EPCAM Levels by Cancer Type</a></li>
<li class="chapter" data-level="10.4.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#creating-a-basic-bar-plot"><i class="fa fa-check"></i><b>10.4.2</b> Creating a Basic Bar Plot</a></li>
<li class="chapter" data-level="10.4.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#reordering-the-bars"><i class="fa fa-check"></i><b>10.4.3</b> Reordering the Bars</a></li>
<li class="chapter" data-level="10.4.4" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#reversing-the-order"><i class="fa fa-check"></i><b>10.4.4</b> Reversing the Order</a></li>
<li class="chapter" data-level="10.4.5" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#note-on-factor-levels"><i class="fa fa-check"></i><b>10.4.5</b> Note on factor levels</a></li>
<li class="chapter" data-level="10.4.6" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#conclusion-25"><i class="fa fa-check"></i><b>10.4.6</b> Conclusion</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html"><i class="fa fa-check"></i><b>11</b> Introduction to BioConductor</a>
<ul>
<li class="chapter" data-level="11.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#introduction-to-biocmanager"><i class="fa fa-check"></i><b>11.1</b> Introduction to BiocManager</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#why-use-biocmanager"><i class="fa fa-check"></i><b>11.1.1</b> Why Use BiocManager?</a></li>
<li class="chapter" data-level="11.1.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>11.1.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="11.1.3" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#installing-bioconductor-packages"><i class="fa fa-check"></i><b>11.1.3</b> Installing Bioconductor Packages</a></li>
<li class="chapter" data-level="11.1.4" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#updating-bioconductor-packages"><i class="fa fa-check"></i><b>11.1.4</b> Updating Bioconductor Packages</a></li>
<li class="chapter" data-level="11.1.5" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#checking-for-valid-bioconductor-versions"><i class="fa fa-check"></i><b>11.1.5</b> Checking for Valid Bioconductor Versions</a></li>
<li class="chapter" data-level="11.1.6" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#conclusion-26"><i class="fa fa-check"></i><b>11.1.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#working-with-genomic-coordinates-and-regions-in-r"><i class="fa fa-check"></i><b>11.2</b> Working with Genomic Coordinates and Regions in R</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#introduction-to-genomicranges"><i class="fa fa-check"></i><b>11.2.1</b> Introduction to GenomicRanges</a></li>
<li class="chapter" data-level="11.2.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#basic-operations-with-genomicranges"><i class="fa fa-check"></i><b>11.2.2</b> Basic Operations with GenomicRanges</a></li>
<li class="chapter" data-level="11.2.3" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#calculating-width-of-each-genomic-interval"><i class="fa fa-check"></i><b>11.2.3</b> Calculating width of Each Genomic Interval</a></li>
<li class="chapter" data-level="11.2.4" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#start-and-end-positions"><i class="fa fa-check"></i><b>11.2.4</b> Start and End Positions</a></li>
<li class="chapter" data-level="11.2.5" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#strand-information"><i class="fa fa-check"></i><b>11.2.5</b> Strand Information</a></li>
<li class="chapter" data-level="11.2.6" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#shifting-the-genomic-range"><i class="fa fa-check"></i><b>11.2.6</b> Shifting the Genomic Range</a></li>
<li class="chapter" data-level="11.2.7" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#subsetting"><i class="fa fa-check"></i><b>11.2.7</b> Subsetting</a></li>
<li class="chapter" data-level="11.2.8" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#flanking-regions"><i class="fa fa-check"></i><b>11.2.8</b> Flanking Regions</a></li>
<li class="chapter" data-level="11.2.9" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#resizing"><i class="fa fa-check"></i><b>11.2.9</b> Resizing</a></li>
<li class="chapter" data-level="11.2.10" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#based-and-1-based-coordinate-system"><i class="fa fa-check"></i><b>11.2.10</b> 0 based and 1 based coordinate system</a></li>
<li class="chapter" data-level="11.2.11" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#other-packages"><i class="fa fa-check"></i><b>11.2.11</b> Other packages</a></li>
<li class="chapter" data-level="11.2.12" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#accessing-genome-annotations"><i class="fa fa-check"></i><b>11.2.12</b> Accessing Genome Annotations</a></li>
<li class="chapter" data-level="11.2.13" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#understanding-genomic-biotypes"><i class="fa fa-check"></i><b>11.2.13</b> Understanding Genomic Biotypes</a></li>
<li class="chapter" data-level="11.2.14" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#creating-a-transcript-database-txdb"><i class="fa fa-check"></i><b>11.2.14</b> Creating a Transcript Database (TxDb)</a></li>
<li class="chapter" data-level="11.2.15" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#extracting-exons-introns-and-intergenic-regions"><i class="fa fa-check"></i><b>11.2.15</b> Extracting Exons, Introns, and Intergenic Regions</a></li>
<li class="chapter" data-level="11.2.16" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#merging-all-exons"><i class="fa fa-check"></i><b>11.2.16</b> Merging All Exons</a></li>
<li class="chapter" data-level="11.2.17" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#introns"><i class="fa fa-check"></i><b>11.2.17</b> Introns</a></li>
<li class="chapter" data-level="11.2.18" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#getting-all-genes"><i class="fa fa-check"></i><b>11.2.18</b> Getting All Genes</a></li>
<li class="chapter" data-level="11.2.19" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#promoters"><i class="fa fa-check"></i><b>11.2.19</b> Promoters</a></li>
<li class="chapter" data-level="11.2.20" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#full-genome"><i class="fa fa-check"></i><b>11.2.20</b> Full Genome</a></li>
<li class="chapter" data-level="11.2.21" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#full-transcriptome"><i class="fa fa-check"></i><b>11.2.21</b> Full Transcriptome</a></li>
<li class="chapter" data-level="11.2.22" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#intergenic-regions"><i class="fa fa-check"></i><b>11.2.22</b> Intergenic Regions</a></li>
<li class="chapter" data-level="11.2.23" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#exploring-untranslated-regions-utrs"><i class="fa fa-check"></i><b>11.2.23</b> Exploring Untranslated Regions (UTRs)</a></li>
<li class="chapter" data-level="11.2.24" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#conclusion-27"><i class="fa fa-check"></i><b>11.2.24</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#exploring-cpg-islands-and-shores-in-genomic-data"><i class="fa fa-check"></i><b>11.3</b> Exploring CpG Islands and Shores in Genomic Data</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#introduction-to-cpg-islands"><i class="fa fa-check"></i><b>11.3.1</b> Introduction to CpG Islands</a></li>
<li class="chapter" data-level="11.3.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#fetching-cpg-island-coordinates"><i class="fa fa-check"></i><b>11.3.2</b> Fetching CpG Island Coordinates</a></li>
<li class="chapter" data-level="11.3.3" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#defining-cpg-shores"><i class="fa fa-check"></i><b>11.3.3</b> Defining CpG Shores</a></li>
<li class="chapter" data-level="11.3.4" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#combining-and-analyzing-cpg-shores"><i class="fa fa-check"></i><b>11.3.4</b> Combining and Analyzing CpG Shores</a></li>
<li class="chapter" data-level="11.3.5" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#conclusion-28"><i class="fa fa-check"></i><b>11.3.5</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#real-world-applications-chip-seq"><i class="fa fa-check"></i><b>11.4</b> Real-World Applications: ChIP-seq</a>
<ul>
<li class="chapter" data-level="11.4.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#identify-promoters-that-overlap-with-the-peaks"><i class="fa fa-check"></i><b>11.4.1</b> Identify promoters that overlap with the peaks</a></li>
<li class="chapter" data-level="11.4.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#use-the-chipseeker-package"><i class="fa fa-check"></i><b>11.4.2</b> use the ChIPseeker package</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#analyzing-and-visualizing-genomic-data"><i class="fa fa-check"></i><b>11.5</b> Analyzing and Visualizing Genomic Data</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#deseq2"><i class="fa fa-check"></i><b>11.5.1</b> DESeq2</a></li>
<li class="chapter" data-level="11.5.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#geoquery"><i class="fa fa-check"></i><b>11.5.2</b> GEOquery</a></li>
<li class="chapter" data-level="11.5.3" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#converting-gene-ids"><i class="fa fa-check"></i><b>11.5.3</b> Converting Gene IDs</a></li>
<li class="chapter" data-level="11.5.4" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#complexheatmap"><i class="fa fa-check"></i><b>11.5.4</b> ComplexHeatmap</a></li>
<li class="chapter" data-level="11.5.5" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#conclusion-29"><i class="fa fa-check"></i><b>11.5.5</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#real-world-example---tcga-analysis"><i class="fa fa-check"></i><b>11.6</b> Real-World Example - TCGA Analysis</a>
<ul>
<li class="chapter" data-level="11.6.1" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#introduction-to-tcga"><i class="fa fa-check"></i><b>11.6.1</b> Introduction to TCGA</a></li>
<li class="chapter" data-level="11.6.2" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#why-analyze-tcga-data"><i class="fa fa-check"></i><b>11.6.2</b> Why Analyze TCGA Data?</a></li>
<li class="chapter" data-level="11.6.3" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#getting-started-1"><i class="fa fa-check"></i><b>11.6.3</b> Getting Started</a></li>
<li class="chapter" data-level="11.6.4" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-1-install-and-load-required-packages"><i class="fa fa-check"></i><b>11.6.4</b> Step 1: Install and Load Required Packages</a></li>
<li class="chapter" data-level="11.6.5" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-2-retrieve-tcga-project-information"><i class="fa fa-check"></i><b>11.6.5</b> Step 2: Retrieve TCGA Project Information</a></li>
<li class="chapter" data-level="11.6.6" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-3-create-a-rangedsummarizedexperiment-object"><i class="fa fa-check"></i><b>11.6.6</b> Step 3: Create a RangedSummarizedExperiment Object</a></li>
<li class="chapter" data-level="11.6.7" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-4-explore-tcga-data"><i class="fa fa-check"></i><b>11.6.7</b> Step 4: Explore TCGA Data</a></li>
<li class="chapter" data-level="11.6.8" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-5-data-transformation---converting-raw-counts-to-tpm"><i class="fa fa-check"></i><b>11.6.8</b> Step 5: Data Transformation - Converting Raw Counts to TPM</a></li>
<li class="chapter" data-level="11.6.9" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-6-combine-data-and-metadata"><i class="fa fa-check"></i><b>11.6.9</b> Step 6: Combine Data and Metadata</a></li>
<li class="chapter" data-level="11.6.10" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-7-renaming-metadata-columns"><i class="fa fa-check"></i><b>11.6.10</b> Step 7: Renaming Metadata Columns</a></li>
<li class="chapter" data-level="11.6.11" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-8-merging-into-a-single-dataframe"><i class="fa fa-check"></i><b>11.6.11</b> Step 8: Merging into a single DataFrame</a></li>
<li class="chapter" data-level="11.6.12" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-9-create-a-gene-expression-heatmap"><i class="fa fa-check"></i><b>11.6.12</b> Step 9: Create a Gene Expression Heatmap</a></li>
<li class="chapter" data-level="11.6.13" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#step-10-sanity-check-and-scaling-of-gene-expression"><i class="fa fa-check"></i><b>11.6.13</b> Step 10: Sanity Check and Scaling of Gene Expression</a></li>
<li class="chapter" data-level="11.6.14" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#conclusion-30"><i class="fa fa-check"></i><b>11.6.14</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="11.7" data-path="introduction-to-bioconductor.html"><a href="introduction-to-bioconductor.html#section-completed"><i class="fa fa-check"></i><b>11.7</b> Section completed</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Master Real-world Bioinformatics analysis in R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-r-for-biologists" class="section level1" number="5">
<h1><span class="header-section-number">Chapter 5</span> Introduction to R for biologists</h1>
<div id="basic-data-types-in-r" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> Basic Data Types in R</h2>
<div id="before-starting" class="section level3" number="5.1.1">
<h3><span class="header-section-number">5.1.1</span> Before Starting</h3>
<p>When youâ€™re working with R, itâ€™s crucial to name your variables properly. Here are some simple rules to follow:</p>
<ol style="list-style-type: decimal">
<li><p>Allowed Characters: Variable names can include letters, numbers, underscores (_), and periods (.).</p></li>
<li><p>Starting Characters: A variable name must start with a letter or a period. It canâ€™t start with a number.</p></li>
<li><p>Case Sensitivity: R treats variable names as case-sensitive. For example, myvar is different from MyVar.</p></li>
<li><p>Dots in Names: While dots are allowed in variable names, itâ€™s best to avoid them except at the beginning.</p></li>
<li><p>Avoiding Conflicts: Donâ€™t use names that match existing functions in R, like mean or c.</p></li>
</ol>
<div id="examples" class="section level4" number="5.1.1.1">
<h4><span class="header-section-number">5.1.1.1</span> Examples:</h4>
<p>Valid:</p>
<ul>
<li><p>myvar</p></li>
<li><p>my.var</p></li>
<li><p>var1</p></li>
<li><p>var_1</p></li>
</ul>
<p>Invalid:</p>
<ul>
<li><p>1var (canâ€™t start with a number)</p></li>
<li><p><em>temp (canâ€™t start with a </em>)</p></li>
<li><p>c (matches existing function)</p></li>
<li><p>my-var (hyphens arenâ€™t allowed)</p></li>
<li><p>my var (spaces arenâ€™t allowed)</p></li>
</ul>
</div>
<div id="tips-for-naming" class="section level4" number="5.1.1.2">
<h4><span class="header-section-number">5.1.1.2</span> Tips for Naming:</h4>
<ul>
<li><p>Meaningful Names: Choose descriptive names like patient_data instead of generic ones like x, y, or z.</p></li>
<li><p>Short and Clear: Keep your names short but make sure they clearly represent what the variable contains.</p></li>
<li><p>Naming Style: You can use camelCase or underscores_between_words for multi-word names. Make sure to stick to one style and be consistent.</p></li>
</ul>
<p>Case Consistency: Decide whether you want all lowercase names or CapWords (also known as PascalCase), and stick to it throughout your code.</p>
<p>By following these naming conventions, your code will be easier to understand, and youâ€™ll avoid unexpected errors. Consistency is key when naming variables across your R scripts.</p>
<p>In R, data can be classified into several fundamental types, each serving specific purposes in data analysis and manipulation. Understanding these data types is crucial for effective data handling. Letâ€™s explore the primary data types in R:</p>
</div>
</div>
<div id="numeric" class="section level3" number="5.1.2">
<h3><span class="header-section-number">5.1.2</span> 1. Numeric</h3>
<p>Numeric data represents continuous numerical values. These can be integers or real numbers (floating-point). Numeric data is used for mathematical calculations and statistical analysis.</p>
<p>Example:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="introduction-to-r-for-biologists.html#cb8-1" tabindex="-1"></a><span class="co"># Numeric data</span></span>
<span id="cb8-2"><a href="introduction-to-r-for-biologists.html#cb8-2" tabindex="-1"></a>age <span class="ot">&lt;-</span> <span class="dv">28</span></span>
<span id="cb8-3"><a href="introduction-to-r-for-biologists.html#cb8-3" tabindex="-1"></a>height <span class="ot">&lt;-</span> <span class="fl">1.75</span></span></code></pre></div>
<p>The <code>&lt;-</code> operator and the <code>=</code> operator in R are both used for assignment but have some key differences.</p>
<p>The <code>&lt;-</code> operator is the standard assignment operator in R. It assigns values to objects.</p>
<ul>
<li><p>The arrow can be read as â€œgetsâ€. So age gets the value of 28.</p></li>
<li><p>All R objects should be created using the <code>&lt;-</code> operator.</p></li>
</ul>
<p>The <code>=</code> Operator can also be used for assignments.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="introduction-to-r-for-biologists.html#cb9-1" tabindex="-1"></a>age <span class="ot">=</span> <span class="dv">28</span></span></code></pre></div>
<ul>
<li><p>This also assigns 28 to age.</p></li>
<li><p>The <code>=</code> should be read as â€œis equal toâ€. This is mainly used to specify function arguments.</p></li>
</ul>
<p>So in a function call like:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="introduction-to-r-for-biologists.html#cb10-1" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> mydata)</span></code></pre></div>
<p>We are specifying that the x argument is equal to mydata.</p>
<p>In summary, <code>&lt;-</code> is the operator you should use when creating R objects, while
<code>=</code> is reserved for function arguments. For assignments, both <code>&lt;-</code> and <code>=</code> can be used.
If you want to read more differences, take a look at <a href="https://stat.ethz.ch/R-manual/R-patched/library/base/html/assignOps.html" class="uri">https://stat.ethz.ch/R-manual/R-patched/library/base/html/assignOps.html</a>.</p>
</div>
<div id="character-string" class="section level3" number="5.1.3">
<h3><span class="header-section-number">5.1.3</span> 2. Character (String)</h3>
<p>Character data represents text or strings of characters. You use character data for storing and manipulating text-based information, such as names, descriptions, and labels.</p>
<p>Example:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="introduction-to-r-for-biologists.html#cb11-1" tabindex="-1"></a><span class="co"># Character data</span></span>
<span id="cb11-2"><a href="introduction-to-r-for-biologists.html#cb11-2" tabindex="-1"></a>name <span class="ot">&lt;-</span> <span class="st">&quot;John Doe&quot;</span></span>
<span id="cb11-3"><a href="introduction-to-r-for-biologists.html#cb11-3" tabindex="-1"></a>city <span class="ot">&lt;-</span> <span class="st">&quot;New York&quot;</span></span></code></pre></div>
</div>
<div id="integer" class="section level3" number="5.1.4">
<h3><span class="header-section-number">5.1.4</span> 3. Integer</h3>
<p>Integer data represents whole numbers. Unlike numeric data, which includes decimal points, integer data includes only whole numbers. Itâ€™s commonly used when dealing with counts or discrete quantities.</p>
<p>Example:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="introduction-to-r-for-biologists.html#cb12-1" tabindex="-1"></a><span class="co"># Integer data</span></span>
<span id="cb12-2"><a href="introduction-to-r-for-biologists.html#cb12-2" tabindex="-1"></a>count <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb12-3"><a href="introduction-to-r-for-biologists.html#cb12-3" tabindex="-1"></a>students <span class="ot">&lt;-</span> <span class="dv">42</span></span></code></pre></div>
</div>
<div id="logical-boolean" class="section level3" number="5.1.5">
<h3><span class="header-section-number">5.1.5</span> 4. Logical (Boolean)</h3>
<p>Logical data consists of two possible values: <code>TRUE</code> or <code>FALSE</code>. These values are used for binary decisions, conditions, and logical operations.</p>
<p>Example:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="introduction-to-r-for-biologists.html#cb13-1" tabindex="-1"></a><span class="co"># Logical data</span></span>
<span id="cb13-2"><a href="introduction-to-r-for-biologists.html#cb13-2" tabindex="-1"></a>is_student <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb13-3"><a href="introduction-to-r-for-biologists.html#cb13-3" tabindex="-1"></a>has_permission <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span></code></pre></div>
</div>
<div id="factor" class="section level3" number="5.1.6">
<h3><span class="header-section-number">5.1.6</span> 5. Factor</h3>
<p>Factor data represents categorical variables with predefined levels or categories. Factors are used when you have data that can be divided into distinct categories, such as â€œHigh,â€ â€œMedium,â€ and â€œLow.â€</p>
<p>Example:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="introduction-to-r-for-biologists.html#cb14-1" tabindex="-1"></a><span class="co"># Factor data</span></span>
<span id="cb14-2"><a href="introduction-to-r-for-biologists.html#cb14-2" tabindex="-1"></a>grade <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;A&quot;</span>))</span></code></pre></div>
</div>
<div id="date-and-time" class="section level3" number="5.1.7">
<h3><span class="header-section-number">5.1.7</span> 6. Date and Time</h3>
<p>Date and time data types are used for representing dates, times, or both. These data types are crucial when dealing with time series data or conducting temporal analysis.</p>
<p>Example:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="introduction-to-r-for-biologists.html#cb15-1" tabindex="-1"></a><span class="co"># Date and time data</span></span>
<span id="cb15-2"><a href="introduction-to-r-for-biologists.html#cb15-2" tabindex="-1"></a>birth_date <span class="ot">&lt;-</span> <span class="fu">as.Date</span>(<span class="st">&quot;1990-05-15&quot;</span>)</span>
<span id="cb15-3"><a href="introduction-to-r-for-biologists.html#cb15-3" tabindex="-1"></a>timestamp <span class="ot">&lt;-</span> <span class="fu">as.POSIXct</span>(<span class="st">&quot;2023-01-09 14:30:00&quot;</span>)</span></code></pre></div>
</div>
<div id="complex" class="section level3" number="5.1.8">
<h3><span class="header-section-number">5.1.8</span> 7. Complex</h3>
<p>Complex data types represent complex numbers, which have both real and imaginary parts. Complex numbers are used in advanced mathematical and engineering applications.</p>
<p>Example:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="introduction-to-r-for-biologists.html#cb16-1" tabindex="-1"></a><span class="co"># Complex data</span></span>
<span id="cb16-2"><a href="introduction-to-r-for-biologists.html#cb16-2" tabindex="-1"></a>z <span class="ot">&lt;-</span> <span class="dv">3</span> <span class="sc">+</span> 2i</span></code></pre></div>
</div>
<div id="missing-values-na" class="section level3" number="5.1.9">
<h3><span class="header-section-number">5.1.9</span> 8. Missing Values (NA)</h3>
<p>In R, missing values are represented as NA. These values indicate the absence of data or an undefined value. Handling missing data is essential in data analysis.</p>
<p>Example:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="introduction-to-r-for-biologists.html#cb17-1" tabindex="-1"></a><span class="co"># Missing value</span></span>
<span id="cb17-2"><a href="introduction-to-r-for-biologists.html#cb17-2" tabindex="-1"></a>missing_data <span class="ot">&lt;-</span> <span class="cn">NA</span></span></code></pre></div>
<p>Understanding these data types and their characteristics is fundamental to effective data manipulation and analysis in R. Different operations and functions may behave differently depending on the data type, so being familiar with these types will help you work with data effectively.</p>
</div>
</div>
<div id="the-key-concept-of-r-vectors" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> The key concept of R: Vectors</h2>
<p>In the world of R, vectors are the building blocks of data. In fact, everything in R is a vector. Whether youâ€™re dealing with numbers or characters, R treats them all as vectors, which are simply collections of these elements. There is no concept of a scalar in R; even a single value is considered a one-element vector. To create a vector, youâ€™ll use the c function, which stands for â€œcombineâ€ or â€œconcatenate.â€</p>
<div id="creating-numeric-and-character-vectors" class="section level3" number="5.2.1">
<h3><span class="header-section-number">5.2.1</span> Creating Numeric and Character Vectors</h3>
<p>Letâ€™s dive right in with examples:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="introduction-to-r-for-biologists.html#cb18-1" tabindex="-1"></a><span class="co"># A numeric vector</span></span>
<span id="cb18-2"><a href="introduction-to-r-for-biologists.html#cb18-2" tabindex="-1"></a><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>)</span></code></pre></div>
<pre><code>## [1] 1 2 3</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="introduction-to-r-for-biologists.html#cb20-1" tabindex="-1"></a><span class="co"># Output: [1] 1 2 3</span></span>
<span id="cb20-2"><a href="introduction-to-r-for-biologists.html#cb20-2" tabindex="-1"></a></span>
<span id="cb20-3"><a href="introduction-to-r-for-biologists.html#cb20-3" tabindex="-1"></a><span class="co"># A character vector</span></span>
<span id="cb20-4"><a href="introduction-to-r-for-biologists.html#cb20-4" tabindex="-1"></a><span class="fu">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;G&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;A&quot; &quot;T&quot; &quot;C&quot; &quot;G&quot;</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="introduction-to-r-for-biologists.html#cb22-1" tabindex="-1"></a><span class="co"># Output: [1] &quot;A&quot; &quot;T&quot; &quot;C&quot; &quot;G&quot;</span></span></code></pre></div>
<p>Notice how <code>c()</code> is used to combine values into vectors. Even a single element, such as â€œAâ€, is a vector in R. Similarly, numeric values like 5 are considered one-element vectors.</p>
</div>
<div id="saving-vectors-to-variables" class="section level3" number="5.2.2">
<h3><span class="header-section-number">5.2.2</span> Saving Vectors to Variables</h3>
<p>Now, letâ€™s save these vectors into variables with meaningful names:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="introduction-to-r-for-biologists.html#cb23-1" tabindex="-1"></a>number_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>)</span>
<span id="cb23-2"><a href="introduction-to-r-for-biologists.html#cb23-2" tabindex="-1"></a>dna_base <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;G&quot;</span>) </span>
<span id="cb23-3"><a href="introduction-to-r-for-biologists.html#cb23-3" tabindex="-1"></a>gene_ids <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;PAX6&quot;</span>, <span class="st">&quot;TP53&quot;</span>, <span class="st">&quot;MYC&quot;</span>)</span></code></pre></div>
<p>Remember, itâ€™s good practice to use informative variable names. Variable names cannot start with a number, so opt for names like <code>number_vec</code> and <code>dna_base</code>.</p>
</div>
<div id="vectorized-calculations" class="section level3" number="5.2.3">
<h3><span class="header-section-number">5.2.3</span> Vectorized Calculations</h3>
<p>One of the powerful features of R is vectorization, where operations are automatically applied element-wise to vectors. For example:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="introduction-to-r-for-biologists.html#cb24-1" tabindex="-1"></a>number_vec <span class="sc">+</span> <span class="dv">1</span> </span></code></pre></div>
<pre><code>## [1] 2 3 4</code></pre>
<p>Here, weâ€™ve added 1 to each element of <code>number_vec</code>. This vectorized behavior simplifies many calculations.</p>
<div id="understanding-indexing" class="section level4" number="5.2.3.1">
<h4><span class="header-section-number">5.2.3.1</span> Understanding Indexing</h4>
<p>You may have noticed the <code>[1]</code> that appears in the output. In R, itâ€™s called an index, and it indicates the position of each element in the result. For instance:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="introduction-to-r-for-biologists.html#cb26-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">100</span></span>
<span id="cb26-2"><a href="introduction-to-r-for-biologists.html#cb26-2" tabindex="-1"></a>x</span></code></pre></div>
<pre><code>##   [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
##  [19]  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
##  [37]  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
##  [55]  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
##  [73]  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
##  [91]  91  92  93  94  95  96  97  98  99 100</code></pre>
<p>In this case, <code>[1]</code> denotes the first position of each element. Understanding indexing helps when working with large datasets.</p>
</div>
</div>
<div id="performing-vectorized-calculations" class="section level3" number="5.2.4">
<h3><span class="header-section-number">5.2.4</span> Performing Vectorized Calculations</h3>
<p>You can perform various calculations on vectors in R:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="introduction-to-r-for-biologists.html#cb28-1" tabindex="-1"></a>number_vec <span class="sc">*</span> <span class="dv">2</span></span></code></pre></div>
<pre><code>## [1] 2 4 6</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="introduction-to-r-for-biologists.html#cb30-1" tabindex="-1"></a><span class="co"># Output: [1] 2 4 6</span></span>
<span id="cb30-2"><a href="introduction-to-r-for-biologists.html#cb30-2" tabindex="-1"></a></span>
<span id="cb30-3"><a href="introduction-to-r-for-biologists.html#cb30-3" tabindex="-1"></a>number_vec <span class="sc">/</span> <span class="dv">2</span></span></code></pre></div>
<pre><code>## [1] 0.5 1.0 1.5</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="introduction-to-r-for-biologists.html#cb32-1" tabindex="-1"></a><span class="co"># Output: [1] 0.5 1.0 1.5</span></span></code></pre></div>
<p>Remember, all these calculations are vectorized, making R a powerful tool for data manipulation.</p>
</div>
<div id="operations-with-character-vectors" class="section level3" number="5.2.5">
<h3><span class="header-section-number">5.2.5</span> Operations with character vectors</h3>
<p>Character vectors in R offer a wide range of operations for text manipulation and analysis. Letâ€™s explore some of the essential operations:</p>
<div id="concatenation" class="section level4" number="5.2.5.1">
<h4><span class="header-section-number">5.2.5.1</span> Concatenation</h4>
<p>You can combine character vectors using the c function:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="introduction-to-r-for-biologists.html#cb33-1" tabindex="-1"></a>new_bases <span class="ot">&lt;-</span> <span class="fu">c</span>(dna_base, <span class="st">&quot;N&quot;</span>)</span>
<span id="cb33-2"><a href="introduction-to-r-for-biologists.html#cb33-2" tabindex="-1"></a>new_bases</span></code></pre></div>
<pre><code>## [1] &quot;A&quot; &quot;T&quot; &quot;C&quot; &quot;G&quot; &quot;N&quot;</code></pre>
<p>This operation is useful for extending or combining character vectors.</p>
</div>
<div id="changing-case" class="section level4" number="5.2.5.2">
<h4><span class="header-section-number">5.2.5.2</span> Changing Case</h4>
<p>Transforming the case of characters is straightforward in R. You can convert characters to uppercase or lowercase:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="introduction-to-r-for-biologists.html#cb35-1" tabindex="-1"></a><span class="fu">toupper</span>(dna_base)</span></code></pre></div>
<pre><code>## [1] &quot;A&quot; &quot;T&quot; &quot;C&quot; &quot;G&quot;</code></pre>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="introduction-to-r-for-biologists.html#cb37-1" tabindex="-1"></a><span class="fu">tolower</span>(dna_base)</span></code></pre></div>
<pre><code>## [1] &quot;a&quot; &quot;t&quot; &quot;c&quot; &quot;g&quot;</code></pre>
<p>This is handy when you need consistent formatting.</p>
</div>
</div>
<div id="logical-vectors" class="section level3" number="5.2.6">
<h3><span class="header-section-number">5.2.6</span> Logical Vectors</h3>
<p>Character vectors also allow for logical operations, which can be incredibly powerful</p>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/DtiatInAJ-Q" width="533" height="300" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
<div id="finding-matches" class="section level4" number="5.2.6.1">
<h4><span class="header-section-number">5.2.6.1</span> Finding Matches</h4>
<p>To check which elements in a character vector meet certain criteria, use the %in% operator:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="introduction-to-r-for-biologists.html#cb39-1" tabindex="-1"></a>dna_base <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;T&quot;</span>)</span></code></pre></div>
<pre><code>## [1]  TRUE  TRUE  TRUE FALSE</code></pre>
<p>This produces a logical vector where <code>TRUE</code> indicates a match. This is because the vector dna_base contains <code>A</code>, <code>T</code>, <code>C</code>, <code>G</code> and <code>G</code> does not match any element in the vector created by <code>c("A", "C", "T")</code>.</p>
</div>
<div id="saving-a-logical-vector" class="section level4" number="5.2.6.2">
<h4><span class="header-section-number">5.2.6.2</span> Saving a Logical Vector</h4>
<p>Save the resulting logical vector to a new variable for future use:</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="introduction-to-r-for-biologists.html#cb41-1" tabindex="-1"></a>logical_vec <span class="ot">&lt;-</span> dna_base <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;T&quot;</span>)</span>
<span id="cb41-2"><a href="introduction-to-r-for-biologists.html#cb41-2" tabindex="-1"></a>logical_vec</span></code></pre></div>
<pre><code>## [1]  TRUE  TRUE  TRUE FALSE</code></pre>
<p>The length of the logical vector matches the original character vector.</p>
</div>
<div id="negating-a-logical-vector" class="section level4" number="5.2.6.3">
<h4><span class="header-section-number">5.2.6.3</span> Negating a Logical Vector</h4>
<p>You can negate a logical vector using the <code>!</code> operator:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="introduction-to-r-for-biologists.html#cb43-1" tabindex="-1"></a><span class="sc">!</span>logical_vec</span></code></pre></div>
<pre><code>## [1] FALSE FALSE FALSE  TRUE</code></pre>
<p>Now, <code>TRUE</code> represents elements that do not match the criteria.</p>
</div>
<div id="subsetting-with-logical-vectors" class="section level4" number="5.2.6.4">
<h4><span class="header-section-number">5.2.6.4</span> Subsetting with Logical Vectors</h4>
<p>Using a logical vector for subsetting another vector is a common operation. It allows you to filter and extract specific elements:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="introduction-to-r-for-biologists.html#cb45-1" tabindex="-1"></a><span class="co"># Subsetting elements that meet the criteria</span></span>
<span id="cb45-2"><a href="introduction-to-r-for-biologists.html#cb45-2" tabindex="-1"></a>dna_base[logical_vec]</span></code></pre></div>
<pre><code>## [1] &quot;A&quot; &quot;T&quot; &quot;C&quot;</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="introduction-to-r-for-biologists.html#cb47-1" tabindex="-1"></a><span class="co"># Subsetting elements that do not meet the criteria</span></span>
<span id="cb47-2"><a href="introduction-to-r-for-biologists.html#cb47-2" tabindex="-1"></a>dna_base[<span class="sc">!</span>logical_vec]</span></code></pre></div>
<pre><code>## [1] &quot;G&quot;</code></pre>
<p>This powerful technique helps you extract and manipulate data efficiently based on specified conditions.</p>
</div>
</div>
<div id="conclusion" class="section level3" number="5.2.7">
<h3><span class="header-section-number">5.2.7</span> Conclusion</h3>
<p>Youâ€™ve learned the fundamental operations that can be performed on both numeric and character vectors. These essential skills will serve as a strong foundation as you delve deeper into the world of data analysis and manipulation using R.</p>
<p>Remember that vectors are the building blocks of R, and they are used extensively in various data analysis tasks. Whether youâ€™re combining elements, changing case, or using logical operations to filter and extract data, you have now gained valuable insights into how vectors can be harnessed to accomplish your data analysis goals.</p>
<p>As you continue your journey in R programming, youâ€™ll encounter more complex data structures and operations, but the understanding of vectors will remain a cornerstone of your proficiency.</p>
</div>
</div>
<div id="subsetting-and-indexing" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> Subsetting and Indexing</h2>
<p>In this guide, we will explain one of the fundamental concepts for dealing with vectors: indexing and slicing. Understanding how R handles these operations is crucial as it differs from some other programming languages, like Python.</p>
<div id="indexing-in-r" class="section level3" number="5.3.1">
<h3><span class="header-section-number">5.3.1</span> Indexing in R</h3>
<p>In R, unlike Python where indexing starts at 0, it begins at 1. This means that the first element in a sequence is located at position 1, the second at position 2, and so on. Letâ€™s dive into some practical examples to illustrate this concept.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="introduction-to-r-for-biologists.html#cb49-1" tabindex="-1"></a><span class="co"># Let&#39;s create a vector of DNA bases</span></span>
<span id="cb49-2"><a href="introduction-to-r-for-biologists.html#cb49-2" tabindex="-1"></a>dna_base <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;G&quot;</span>)</span>
<span id="cb49-3"><a href="introduction-to-r-for-biologists.html#cb49-3" tabindex="-1"></a></span>
<span id="cb49-4"><a href="introduction-to-r-for-biologists.html#cb49-4" tabindex="-1"></a><span class="co"># Accessing the second element (T) using indexing</span></span>
<span id="cb49-5"><a href="introduction-to-r-for-biologists.html#cb49-5" tabindex="-1"></a>dna_base[<span class="dv">2</span>]</span></code></pre></div>
<pre><code>## [1] &quot;T&quot;</code></pre>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="introduction-to-r-for-biologists.html#cb51-1" tabindex="-1"></a><span class="co"># second and fourth element</span></span>
<span id="cb51-2"><a href="introduction-to-r-for-biologists.html#cb51-2" tabindex="-1"></a>dna_base[<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>)]</span></code></pre></div>
<pre><code>## [1] &quot;T&quot; &quot;G&quot;</code></pre>
<p>Here, we use square brackets <code>[ ]</code> to access elements by their position in the dna_base vector. So, <code>dna_base[2]</code> returns the second element, which is â€œTâ€.</p>
</div>
<div id="slicing-in-r" class="section level3" number="5.3.2">
<h3><span class="header-section-number">5.3.2</span> Slicing in R</h3>
<p>Slicing in R allows you to extract a chunk of elements from a vector or sequence. You can specify the start and end positions within the square brackets to define the slice.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="introduction-to-r-for-biologists.html#cb53-1" tabindex="-1"></a><span class="co"># Slicing the first two elements (A and T) from the dna_base vector</span></span>
<span id="cb53-2"><a href="introduction-to-r-for-biologists.html#cb53-2" tabindex="-1"></a>dna_base[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span></code></pre></div>
<pre><code>## [1] &quot;A&quot; &quot;T&quot;</code></pre>
<p>In this example, <code>dna_base[1:2]</code> retrieves elements from position 1 to 2, giving us â€œAâ€ and â€œTâ€.</p>
</div>
<div id="negative-indexing" class="section level3" number="5.3.3">
<h3><span class="header-section-number">5.3.3</span> Negative Indexing</h3>
<p>R also allows negative indexing to remove that element at that position:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="introduction-to-r-for-biologists.html#cb55-1" tabindex="-1"></a><span class="co"># remove first element by negative indexing</span></span>
<span id="cb55-2"><a href="introduction-to-r-for-biologists.html#cb55-2" tabindex="-1"></a>remove_first_element <span class="ot">&lt;-</span> dna_base[<span class="sc">-</span><span class="dv">1</span>]</span></code></pre></div>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="introduction-to-r-for-biologists.html#cb56-1" tabindex="-1"></a><span class="co"># remove second and fourth</span></span>
<span id="cb56-2"><a href="introduction-to-r-for-biologists.html#cb56-2" tabindex="-1"></a>dna_base[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>)]</span></code></pre></div>
<pre><code>## [1] &quot;A&quot; &quot;C&quot;</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="introduction-to-r-for-biologists.html#cb58-1" tabindex="-1"></a><span class="co"># Output: [1] &quot;A&quot; &quot;C&quot;</span></span></code></pre></div>
</div>
<div id="conclusion-1" class="section level3" number="5.3.4">
<h3><span class="header-section-number">5.3.4</span> Conclusion</h3>
<p>Remember that R starts indexing at 1, not 0, and you can use square brackets <code>[ ]</code> to access elements and slices within vectors and sequences. This is essential for working with data and performing various operations in R.</p>
</div>
</div>
<div id="understanding-and-manipulating-matrices" class="section level2" number="5.4">
<h2><span class="header-section-number">5.4</span> Understanding and Manipulating Matrices</h2>
<p>Matrices are essential for organizing and processing data, especially when dealing with gene expression data from technologies like RNA sequencing (RNAseq). In this tutorial, we will explore how to create, manipulate, and extract information from matrices using the R programming language. We will cover topics ranging from basic matrix operations to more advanced tasks like normalization for RNAseq data analysis.</p>
<div id="creating-a-matrix" class="section level3" number="5.4.1">
<h3><span class="header-section-number">5.4.1</span> Creating a Matrix</h3>
<p>To create a matrix in R, you can use the matrix() function. A matrix is essentially a 2-dimensional table for storing numerical data. Letâ€™s start by creating a simple matrix:</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="introduction-to-r-for-biologists.html#cb59-1" tabindex="-1"></a>expression_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="at">nrow =</span> <span class="dv">3</span>, <span class="at">ncol =</span> <span class="dv">4</span>)</span>
<span id="cb59-2"><a href="introduction-to-r-for-biologists.html#cb59-2" tabindex="-1"></a>expression_mat</span></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4]
## [1,]    1    4    7   10
## [2,]    2    5    8   11
## [3,]    3    6    9   12</code></pre>
<p>Here, <code>expression_mat</code> is a dummy gene expression matrix with 3 rows (genes) and 4 columns (samples), where the entries represent counts for each gene in each sample.</p>
</div>
<div id="adding-row-and-column-names" class="section level3" number="5.4.2">
<h3><span class="header-section-number">5.4.2</span> Adding Row and Column Names</h3>
<p>You can enhance the clarity of your matrix by adding row and column names. This is particularly useful when dealing with real biological data. For example:</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="introduction-to-r-for-biologists.html#cb61-1" tabindex="-1"></a><span class="fu">rownames</span>(expression_mat) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;gene1&quot;</span>, <span class="st">&quot;gene2&quot;</span>, <span class="st">&quot;gene3&quot;</span>)</span>
<span id="cb61-2"><a href="introduction-to-r-for-biologists.html#cb61-2" tabindex="-1"></a><span class="fu">colnames</span>(expression_mat) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;sample1&quot;</span>, <span class="st">&quot;sample2&quot;</span>, <span class="st">&quot;sample3&quot;</span>, <span class="st">&quot;sample4&quot;</span>)</span>
<span id="cb61-3"><a href="introduction-to-r-for-biologists.html#cb61-3" tabindex="-1"></a>expression_mat</span></code></pre></div>
<pre><code>##       sample1 sample2 sample3 sample4
## gene1       1       4       7      10
## gene2       2       5       8      11
## gene3       3       6       9      12</code></pre>
<p>Now, instead of numerical indices, your matrix displays gene and sample names, making it easier to interpret.</p>
</div>
<div id="subsetting-a-matrix" class="section level3" number="5.4.3">
<h3><span class="header-section-number">5.4.3</span> Subsetting a Matrix</h3>
<p>Subsetting allows you to extract specific rows and columns from a matrix. You can use numerical indices, row/column names, or logical vectors. Remember, R is 1-based. Indices start at 1 while Python starts at 0.</p>
<p><img src="images/matrix.png" /></p>
<div id="subsetting-using-numerical-indices" class="section level4" number="5.4.3.1">
<h4><span class="header-section-number">5.4.3.1</span> Subsetting using numerical indices</h4>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="introduction-to-r-for-biologists.html#cb63-1" tabindex="-1"></a><span class="co"># Accessing a single element</span></span>
<span id="cb63-2"><a href="introduction-to-r-for-biologists.html#cb63-2" tabindex="-1"></a>expression_mat[<span class="dv">1</span>, <span class="dv">2</span>]</span></code></pre></div>
<pre><code>## [1] 4</code></pre>
<p>slice a chunk</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="introduction-to-r-for-biologists.html#cb65-1" tabindex="-1"></a>expression_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span></code></pre></div>
<pre><code>##       sample1 sample2
## gene1       1       4
## gene2       2       5</code></pre>
<p>If you leave either the row index blank or the column index, it will subset all the rows or columns. Subset the first two rows and all columns:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="introduction-to-r-for-biologists.html#cb67-1" tabindex="-1"></a>expression_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>,]</span></code></pre></div>
<pre><code>##       sample1 sample2 sample3 sample4
## gene1       1       4       7      10
## gene2       2       5       8      11</code></pre>
<p>Subset the columns 2 and 4 and all rows</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="introduction-to-r-for-biologists.html#cb69-1" tabindex="-1"></a>expression_mat[, <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>)]</span></code></pre></div>
<pre><code>##       sample2 sample4
## gene1       4      10
## gene2       5      11
## gene3       6      12</code></pre>
</div>
<div id="subsetting-using-row-names" class="section level4" number="5.4.3.2">
<h4><span class="header-section-number">5.4.3.2</span> Subsetting using row names</h4>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="introduction-to-r-for-biologists.html#cb71-1" tabindex="-1"></a><span class="co"># Accessing a specific gene&#39;s data</span></span>
<span id="cb71-2"><a href="introduction-to-r-for-biologists.html#cb71-2" tabindex="-1"></a>expression_mat[<span class="st">&quot;gene3&quot;</span>, ]</span></code></pre></div>
<pre><code>## sample1 sample2 sample3 sample4 
##       3       6       9      12</code></pre>
<p>When only one row or one column is left after subsetting, R returns a vector instead of a matrix. To return a single row or column matrix, add <code>drop=FALSE</code>.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="introduction-to-r-for-biologists.html#cb73-1" tabindex="-1"></a>expression_mat[<span class="st">&quot;gene3&quot;</span>, , drop<span class="ot">=</span><span class="cn">FALSE</span>]</span></code></pre></div>
<pre><code>##       sample1 sample2 sample3 sample4
## gene3       3       6       9      12</code></pre>
</div>
<div id="subsetting-using-column-names" class="section level4" number="5.4.3.3">
<h4><span class="header-section-number">5.4.3.3</span> Subsetting using column names</h4>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="introduction-to-r-for-biologists.html#cb75-1" tabindex="-1"></a><span class="co"># Using predefined gene and sample names</span></span>
<span id="cb75-2"><a href="introduction-to-r-for-biologists.html#cb75-2" tabindex="-1"></a>differential_genes<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;gene3&quot;</span>, <span class="st">&quot;gene1&quot;</span>)</span>
<span id="cb75-3"><a href="introduction-to-r-for-biologists.html#cb75-3" tabindex="-1"></a>expression_mat[differential_genes, <span class="fu">c</span>(<span class="st">&quot;sample1&quot;</span>, <span class="st">&quot;sample2&quot;</span>)]</span></code></pre></div>
<pre><code>##       sample1 sample2
## gene3       3       6
## gene1       1       4</code></pre>
<p>You see how the matrix is subsetted and the row names are ordered as in <code>differential_genes</code>.</p>
</div>
<div id="subsetting-using-logical-vectors" class="section level4" number="5.4.3.4">
<h4><span class="header-section-number">5.4.3.4</span> Subsetting using logical vectors</h4>
<p>We have a matrix called <code>expression_mat</code> that contains gene expression data, and you want to subset it to include only the rows corresponding to certain â€œdifferential genes.â€ Hereâ€™s how you can do it:</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="introduction-to-r-for-biologists.html#cb77-1" tabindex="-1"></a>logical_vec_genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(expression_mat) <span class="sc">%in%</span> differential_genes</span>
<span id="cb77-2"><a href="introduction-to-r-for-biologists.html#cb77-2" tabindex="-1"></a>expression_mat[logical_vec_genes,]</span></code></pre></div>
<pre><code>##       sample1 sample2 sample3 sample4
## gene1       1       4       7      10
## gene3       3       6       9      12</code></pre>
</div>
</div>
<div id="calculations-with-matrices" class="section level3" number="5.4.4">
<h3><span class="header-section-number">5.4.4</span> Calculations with Matrices</h3>
<p>You can perform various calculations on matrices, such as calculating the sum of counts for each sample (column level) or gene (row level):</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="introduction-to-r-for-biologists.html#cb79-1" tabindex="-1"></a><span class="fu">colSums</span>(expression_mat)</span></code></pre></div>
<pre><code>## sample1 sample2 sample3 sample4 
##       6      15      24      33</code></pre>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="introduction-to-r-for-biologists.html#cb81-1" tabindex="-1"></a><span class="fu">rowSums</span>(expression_mat)</span></code></pre></div>
<pre><code>## gene1 gene2 gene3 
##    22    26    30</code></pre>
</div>
<div id="normalization" class="section level3" number="5.4.5">
<h3><span class="header-section-number">5.4.5</span> Normalization</h3>
<p>Normalization is crucial in RNAseq data analysis to account for differences in sequencing depth and gene length. Two common methods for normalization are RPKM (Reads Per Kilobase per Million) and TPM (Transcripts Per Million). Watch this video to understand better of their differences</p>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/TTUrtCY2k-w" width="533" height="300" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
<blockquote>
<p>RPKM (Reads Per Kilobase per Million) and TPM (Transcripts Per Million) are two widely-used methods for normalizing gene expression data in RNAseq, with RPKM considering gene length and total reads per sample and TPM further accounting for differences in sequencing depth between samples.</p>
</blockquote>
<p>Hereâ€™s how you can normalize a matrix to RPKM and TPM:</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="introduction-to-r-for-biologists.html#cb83-1" tabindex="-1"></a><span class="co"># Gene lengths (in kilobases)</span></span>
<span id="cb83-2"><a href="introduction-to-r-for-biologists.html#cb83-2" tabindex="-1"></a>gene_lengths <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1000</span>, <span class="dv">2000</span>, <span class="dv">3000</span>)</span>
<span id="cb83-3"><a href="introduction-to-r-for-biologists.html#cb83-3" tabindex="-1"></a></span>
<span id="cb83-4"><a href="introduction-to-r-for-biologists.html#cb83-4" tabindex="-1"></a><span class="co"># Normalizing to RPKM</span></span>
<span id="cb83-5"><a href="introduction-to-r-for-biologists.html#cb83-5" tabindex="-1"></a>rpkm_normalized <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">t</span>(expression_mat)<span class="sc">/</span><span class="fu">colSums</span>(expression_mat))<span class="sc">/</span>gene_lengths <span class="sc">*</span> <span class="fl">1e6</span></span>
<span id="cb83-6"><a href="introduction-to-r-for-biologists.html#cb83-6" tabindex="-1"></a>rpkm_normalized</span></code></pre></div>
<pre><code>##        sample1  sample2  sample3  sample4
## gene1 166.6667 266.6667 291.6667 303.0303
## gene2 166.6667 166.6667 166.6667 166.6667
## gene3 166.6667 133.3333 125.0000 121.2121</code></pre>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="introduction-to-r-for-biologists.html#cb85-1" tabindex="-1"></a><span class="co"># Normalizing to TPM</span></span>
<span id="cb85-2"><a href="introduction-to-r-for-biologists.html#cb85-2" tabindex="-1"></a>tpm_normalized <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">t</span>(expression_mat<span class="sc">/</span>gene_lengths)<span class="sc">/</span> <span class="fu">colSums</span>((expression_mat<span class="sc">/</span>gene_lengths))) <span class="sc">*</span> <span class="fl">1e6</span></span>
<span id="cb85-3"><a href="introduction-to-r-for-biologists.html#cb85-3" tabindex="-1"></a></span>
<span id="cb85-4"><a href="introduction-to-r-for-biologists.html#cb85-4" tabindex="-1"></a>tpm_normalized</span></code></pre></div>
<pre><code>##        sample1  sample2  sample3  sample4
## gene1 333333.3 470588.2 500000.0 512820.5
## gene2 333333.3 294117.6 285714.3 282051.3
## gene3 333333.3 235294.1 214285.7 205128.2</code></pre>
<p>Note, when you divide a matrix by a vector, the operation is row-wise.</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="introduction-to-r-for-biologists.html#cb87-1" tabindex="-1"></a>expression_mat</span></code></pre></div>
<pre><code>##       sample1 sample2 sample3 sample4
## gene1       1       4       7      10
## gene2       2       5       8      11
## gene3       3       6       9      12</code></pre>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="introduction-to-r-for-biologists.html#cb89-1" tabindex="-1"></a>gene_lengths</span></code></pre></div>
<pre><code>## [1] 1000 2000 3000</code></pre>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="introduction-to-r-for-biologists.html#cb91-1" tabindex="-1"></a>expression_mat<span class="sc">/</span>gene_lengths</span></code></pre></div>
<pre><code>##       sample1 sample2 sample3 sample4
## gene1   0.001  0.0040   0.007  0.0100
## gene2   0.001  0.0025   0.004  0.0055
## gene3   0.001  0.0020   0.003  0.0040</code></pre>
<p>Thatâ€™s why if we want to divide the matrix by the column sum, we use the <code>t()</code> to transpose the matrix first.</p>
</div>
<div id="conclusion-2" class="section level3" number="5.4.6">
<h3><span class="header-section-number">5.4.6</span> Conclusion</h3>
<p>Understanding matrices and their manipulation is fundamental when working with biological data in R. Whether youâ€™re analyzing gene expression or any other numerical data, these matrix operations are essential tools for data exploration and analysis in the field of bioinformatics.</p>
</div>
</div>
<div id="essential-functions-in-r" class="section level2" number="5.5">
<h2><span class="header-section-number">5.5</span> Essential Functions in R</h2>
<p>In this foundational lesson, we will explore several fundamental R functions that are indispensable for your daily bioinformatics tasks. As a beginner in programming, mastering these basic building blocks will provide you with a sturdy groundwork upon which to build your bioinformatics skills.</p>
<div id="length-function" class="section level3" number="5.5.1">
<h3><span class="header-section-number">5.5.1</span> <code>length</code> Function</h3>
<p>The <code>length</code> function is your go-to tool for determining the size of a vector or list in R. Itâ€™s immensely useful when working with gene expression data, as it allows you to gauge the number of elements in your dataset. Letâ€™s dive into an example:</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="introduction-to-r-for-biologists.html#cb93-1" tabindex="-1"></a>expression_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">25</span>, <span class="dv">30</span>, <span class="dv">12</span>, <span class="dv">20</span>)</span>
<span id="cb93-2"><a href="introduction-to-r-for-biologists.html#cb93-2" tabindex="-1"></a><span class="fu">names</span>(expression_vec) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;gene1&quot;</span>, <span class="st">&quot;gene2&quot;</span>, <span class="st">&quot;gene3&quot;</span>, <span class="st">&quot;gene4&quot;</span>, <span class="st">&quot;gene5&quot;</span>)</span>
<span id="cb93-3"><a href="introduction-to-r-for-biologists.html#cb93-3" tabindex="-1"></a></span>
<span id="cb93-4"><a href="introduction-to-r-for-biologists.html#cb93-4" tabindex="-1"></a><span class="fu">length</span>(expression_vec)</span></code></pre></div>
<pre><code>## [1] 5</code></pre>
<p>In this case, our <code>expression_vec</code> contains gene expression values, and the length function tells us that it comprises five elements. This straightforward function provides a crucial dimension for managing your data.</p>
</div>
<div id="unique-function" class="section level3" number="5.5.2">
<h3><span class="header-section-number">5.5.2</span> <code>unique</code> Function</h3>
<p>When working with genomics data, youâ€™ll often encounter lists of genes or sequences. The unique function is a valuable asset for identifying and extracting unique elements from such lists. Letâ€™s illustrate with a simple example:</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="introduction-to-r-for-biologists.html#cb95-1" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;GeneC&quot;</span>, <span class="st">&quot;GeneA&quot;</span>, <span class="st">&quot;GeneB&quot;</span>, <span class="st">&quot;GeneA&quot;</span>)</span>
<span id="cb95-2"><a href="introduction-to-r-for-biologists.html#cb95-2" tabindex="-1"></a>unique_genes <span class="ot">&lt;-</span> <span class="fu">unique</span>(genes)</span>
<span id="cb95-3"><a href="introduction-to-r-for-biologists.html#cb95-3" tabindex="-1"></a></span>
<span id="cb95-4"><a href="introduction-to-r-for-biologists.html#cb95-4" tabindex="-1"></a>unique_genes</span></code></pre></div>
<pre><code>## [1] &quot;GeneC&quot; &quot;GeneA&quot; &quot;GeneB&quot;</code></pre>
<p>In this snippet, we have a list of genes, and unique helps us extract the unique gene names, which can be essential for various genomics analyses.</p>
</div>
<div id="sort-function" class="section level3" number="5.5.3">
<h3><span class="header-section-number">5.5.3</span> <code>sort</code> Function</h3>
<p>Sorting is a fundamental operation in data manipulation. The <code>sort</code> function in R allows you to arrange your data in ascending or descending order. Itâ€™s particularly handy when dealing with gene lists, as it helps you organize genes alphabetically or numerically. Letâ€™s explore some examples:</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="introduction-to-r-for-biologists.html#cb97-1" tabindex="-1"></a>unique_genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;GeneC&quot;</span>, <span class="st">&quot;GeneA&quot;</span>, <span class="st">&quot;GeneB&quot;</span>)</span>
<span id="cb97-2"><a href="introduction-to-r-for-biologists.html#cb97-2" tabindex="-1"></a></span>
<span id="cb97-3"><a href="introduction-to-r-for-biologists.html#cb97-3" tabindex="-1"></a><span class="co"># Sort alphabetically</span></span>
<span id="cb97-4"><a href="introduction-to-r-for-biologists.html#cb97-4" tabindex="-1"></a><span class="fu">sort</span>(unique_genes)</span></code></pre></div>
<pre><code>## [1] &quot;GeneA&quot; &quot;GeneB&quot; &quot;GeneC&quot;</code></pre>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="introduction-to-r-for-biologists.html#cb99-1" tabindex="-1"></a><span class="co"># Sort alphabetically in descending order</span></span>
<span id="cb99-2"><a href="introduction-to-r-for-biologists.html#cb99-2" tabindex="-1"></a><span class="fu">sort</span>(unique_genes, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## [1] &quot;GeneC&quot; &quot;GeneB&quot; &quot;GeneA&quot;</code></pre>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="introduction-to-r-for-biologists.html#cb101-1" tabindex="-1"></a><span class="co"># Sort numeric values</span></span>
<span id="cb101-2"><a href="introduction-to-r-for-biologists.html#cb101-2" tabindex="-1"></a><span class="fu">sort</span>(expression_vec)</span></code></pre></div>
<pre><code>## gene1 gene4 gene5 gene2 gene3 
##    10    12    20    25    30</code></pre>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="introduction-to-r-for-biologists.html#cb103-1" tabindex="-1"></a><span class="co"># Sort numeric values in descending order</span></span>
<span id="cb103-2"><a href="introduction-to-r-for-biologists.html#cb103-2" tabindex="-1"></a><span class="fu">sort</span>(expression_vec, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## gene3 gene2 gene5 gene4 gene1 
##    30    25    20    12    10</code></pre>
<p>These examples demonstrate how the <code>sort</code> function can be applied to both character and numeric data. Sorting can be particularly useful when you need to organize and prioritize genes or data for downstream analyses.</p>
</div>
<div id="cor-function" class="section level3" number="5.5.4">
<h3><span class="header-section-number">5.5.4</span> <code>cor</code> Function</h3>
<p>The cor function is indispensable for bioinformatics tasks that involve assessing the relationships between variables, such as gene expression levels. It calculates the correlation coefficient, which measures the degree of association between two variables. Letâ€™s explore how it works:</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="introduction-to-r-for-biologists.html#cb105-1" tabindex="-1"></a>gene1 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">15</span>, <span class="dv">20</span>, <span class="dv">25</span>)</span>
<span id="cb105-2"><a href="introduction-to-r-for-biologists.html#cb105-2" tabindex="-1"></a>gene2 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">8</span>, <span class="dv">12</span>, <span class="dv">18</span>, <span class="dv">22</span>)</span>
<span id="cb105-3"><a href="introduction-to-r-for-biologists.html#cb105-3" tabindex="-1"></a></span>
<span id="cb105-4"><a href="introduction-to-r-for-biologists.html#cb105-4" tabindex="-1"></a>correlation_coefficient <span class="ot">&lt;-</span> <span class="fu">cor</span>(gene1, gene2)</span>
<span id="cb105-5"><a href="introduction-to-r-for-biologists.html#cb105-5" tabindex="-1"></a></span>
<span id="cb105-6"><a href="introduction-to-r-for-biologists.html#cb105-6" tabindex="-1"></a>correlation_coefficient</span></code></pre></div>
<pre><code>## [1] 0.9965458</code></pre>
<p>In this example, we calculate the correlation between the expression levels of two genes, <code>gene1</code> and <code>gene2</code>. The resulting correlation coefficient value provides insights into the similarity of their expression patterns.</p>
<p>Itâ€™s worth noting that the cor function supports various correlation methods, such as Pearson and Spearman. Understanding these correlations is crucial for deciphering gene interactions and conducting network analyses.</p>
</div>
</div>
<div id="functions-organizing-your-code-for-reusability" class="section level2" number="5.6">
<h2><span class="header-section-number">5.6</span> Functions: Organizing Your Code for Reusability</h2>
<p>In this section, we will delve into the concept of functions in Râ€”a fundamental building block for creating organized and reusable code. Functions serve as a means to encapsulate specific tasks within your code and can significantly enhance your programming capabilities. In essence, a function is like a black box that takes input, processes it, and produces output, shielding you from the inner workings of the logic. We will explore how to create functions, define input arguments, and ensure they provide meaningful output.</p>
<div id="what-is-a-function" class="section level3" number="5.6.1">
<h3><span class="header-section-number">5.6.1</span> What is a Function?</h3>
<p>A function in R is a way to bundle code that accomplishes a specific task or computation. It comprises defined input arguments and a code block, and you can invoke the function whenever you need to execute that particular logic.</p>
<p>Letâ€™s begin by creating a simple function to calculate the mean of a numeric vector.</p>
</div>
<div id="creating-a-basic-function-calculating-the-mean" class="section level3" number="5.6.2">
<h3><span class="header-section-number">5.6.2</span> Creating a Basic Function: Calculating the Mean</h3>
<p>To create a function, you should follow these steps:</p>
<ol style="list-style-type: decimal">
<li><p>Name Your Function: Give your function a meaningful name. In our example, weâ€™ll call it <code>mean_customer</code>.</p></li>
<li><p>Use the function Keyword: Begin your function with the function keyword, followed by parentheses. Inside the parentheses, define your input arguments. You can have multiple arguments for a function.</p></li>
<li><p>Body of the Function: The actual code of the function is enclosed within curly braces {}.</p></li>
</ol>
<p>Letâ€™s create a mean_customer function to compute the mean of a numeric vector:</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="introduction-to-r-for-biologists.html#cb107-1" tabindex="-1"></a>mean_customer <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb107-2"><a href="introduction-to-r-for-biologists.html#cb107-2" tabindex="-1"></a>  total <span class="ot">&lt;-</span> <span class="fu">sum</span>(x)</span>
<span id="cb107-3"><a href="introduction-to-r-for-biologists.html#cb107-3" tabindex="-1"></a>  mean_value <span class="ot">&lt;-</span> total <span class="sc">/</span> <span class="fu">length</span>(x)</span>
<span id="cb107-4"><a href="introduction-to-r-for-biologists.html#cb107-4" tabindex="-1"></a>}</span></code></pre></div>
<p>In this function, we first calculate the total sum of the input vector x using the built-in <code>sum</code> function. Then, we divide this sum by the length of the vector to obtain the mean.</p>
</div>
<div id="using-the-custom-function" class="section level3" number="5.6.3">
<h3><span class="header-section-number">5.6.3</span> Using the Custom Function</h3>
<p>Now that we have defined our <code>mean_customer</code> function, letâ€™s use it with an example vector:</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="introduction-to-r-for-biologists.html#cb108-1" tabindex="-1"></a>input_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>)</span>
<span id="cb108-2"><a href="introduction-to-r-for-biologists.html#cb108-2" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">mean_customer</span>(input_vec)</span>
<span id="cb108-3"><a href="introduction-to-r-for-biologists.html#cb108-3" tabindex="-1"></a></span>
<span id="cb108-4"><a href="introduction-to-r-for-biologists.html#cb108-4" tabindex="-1"></a>result</span></code></pre></div>
<pre><code>## [1] 2.5</code></pre>
<p>You might have noticed that our initial function did not print anything to the console. To make a function display an output, you need a <code>return</code> statement. Letâ€™s add it to our function.</p>
</div>
<div id="returning-a-value-from-the-function" class="section level3" number="5.6.4">
<h3><span class="header-section-number">5.6.4</span> Returning a Value from the Function</h3>
<p>Returning a value in a function in a programming language like R is a fundamental concept that determines what the function does with its computed result. When you include a return statement within a function, you are specifying the value that the function should provide as output when itâ€™s called elsewhere in your code.</p>
<ol style="list-style-type: decimal">
<li><p>Calculation or Operation: Inside the function, there is a block of code that performs some calculation, operation, or task based on the input arguments provided to the function.</p></li>
<li><p>return Statement: When you include a return statement, it signifies that the function should terminate its execution at that point and immediately provide the value specified after return as its output. This value can be a single variable, an expression, or even a complex data structure.</p></li>
<li><p>Function Execution: When you call the function in your code, it starts executing from the beginning and proceeds until it encounters the return statement. At that point, the function stops executing further code within its body and exits, returning the value specified in the return statement.</p></li>
<li><p>Assigning to a Variable: Typically, you capture the returned value by assigning it to a variable when you call the function. This allows you to store and use the result elsewhere in your code.</p></li>
</ol>
<p>For example, in R, following the previous example</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="introduction-to-r-for-biologists.html#cb110-1" tabindex="-1"></a>mean_customer <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb110-2"><a href="introduction-to-r-for-biologists.html#cb110-2" tabindex="-1"></a>  total <span class="ot">&lt;-</span> <span class="fu">sum</span>(x)</span>
<span id="cb110-3"><a href="introduction-to-r-for-biologists.html#cb110-3" tabindex="-1"></a>  mean_value <span class="ot">&lt;-</span> total <span class="sc">/</span> <span class="fu">length</span>(x)</span>
<span id="cb110-4"><a href="introduction-to-r-for-biologists.html#cb110-4" tabindex="-1"></a>  <span class="fu">return</span>(mean_value)</span>
<span id="cb110-5"><a href="introduction-to-r-for-biologists.html#cb110-5" tabindex="-1"></a>}</span></code></pre></div>
<p>Now, when you use <code>mean_customer(input_vec)</code>, it will correctly display the mean value.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="introduction-to-r-for-biologists.html#cb111-1" tabindex="-1"></a><span class="co"># Calculate the mean of a numeric vector</span></span>
<span id="cb111-2"><a href="introduction-to-r-for-biologists.html#cb111-2" tabindex="-1"></a>input_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>)</span>
<span id="cb111-3"><a href="introduction-to-r-for-biologists.html#cb111-3" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">mean_customer</span>(input_vec)</span>
<span id="cb111-4"><a href="introduction-to-r-for-biologists.html#cb111-4" tabindex="-1"></a>result</span></code></pre></div>
<pre><code>## [1] 2.5</code></pre>
</div>
<div id="optional-omitting-the-return-statement" class="section level3" number="5.6.5">
<h3><span class="header-section-number">5.6.5</span> Optional: Omitting the return Statement</h3>
<p>You can also omit the return statement. By default, the last expression in the function will be returned as the output. Hereâ€™s the updated function:</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="introduction-to-r-for-biologists.html#cb113-1" tabindex="-1"></a>mean_customer <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb113-2"><a href="introduction-to-r-for-biologists.html#cb113-2" tabindex="-1"></a>  total <span class="ot">&lt;-</span> <span class="fu">sum</span>(x)</span>
<span id="cb113-3"><a href="introduction-to-r-for-biologists.html#cb113-3" tabindex="-1"></a>  mean_value <span class="ot">&lt;-</span> total <span class="sc">/</span> <span class="fu">length</span>(x)</span>
<span id="cb113-4"><a href="introduction-to-r-for-biologists.html#cb113-4" tabindex="-1"></a>  mean_value</span>
<span id="cb113-5"><a href="introduction-to-r-for-biologists.html#cb113-5" tabindex="-1"></a>}</span></code></pre></div>
<p>The behavior remains the same as before when you use <code>mean_customer(input_vec)</code>.</p>
</div>
<div id="using-more-than-one-argument" class="section level3" number="5.6.6">
<h3><span class="header-section-number">5.6.6</span> Using more than one argument</h3>
<p>Missing data is a common occurrence in real-world biological datasets, and learning how to handle it is crucial for robust data analysis in R. Letâ€™s create a custom function to calculate the mean of a numeric vector while accommodating missing values (NAs). Weâ€™ll introduce the concept of the na.rm argument and illustrate how it allows us to decide whether or not to remove NAs before performing calculations.</p>
</div>
<div id="understanding-missing-values-nas-in-r" class="section level3" number="5.6.7">
<h3><span class="header-section-number">5.6.7</span> Understanding Missing Values (NAs) in R</h3>
<p>In R, â€œNAâ€ stands for â€œNot Available,â€ and it is used to represent missing or undefined values in your data. Letâ€™s start by creating a couple of example vectors:</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="introduction-to-r-for-biologists.html#cb114-1" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;TP53&quot;</span>, <span class="cn">NA</span>, <span class="st">&quot;MYC&quot;</span>)</span>
<span id="cb114-2"><a href="introduction-to-r-for-biologists.html#cb114-2" tabindex="-1"></a>NA_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="cn">NA</span>)</span></code></pre></div>
<p>As you can see, our genes vector contains a missing value (<code>NA</code>). To identify and handle <code>NAs</code>, we can use the <code>is.na()</code> function, which returns a logical vector indicating which elements are <code>NAs</code>:</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="introduction-to-r-for-biologists.html#cb115-1" tabindex="-1"></a><span class="fu">is.na</span>(genes)</span></code></pre></div>
<pre><code>## [1] FALSE  TRUE FALSE</code></pre>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="introduction-to-r-for-biologists.html#cb117-1" tabindex="-1"></a><span class="fu">is.na</span>(NA_vec)</span></code></pre></div>
<pre><code>## [1] FALSE FALSE FALSE  TRUE</code></pre>
</div>
<div id="initial-attempt-a-function-without-handling-nas" class="section level3" number="5.6.8">
<h3><span class="header-section-number">5.6.8</span> Initial Attempt: A Function Without Handling NAs</h3>
<p>Letâ€™s start by creating a custom function to calculate the mean of a numeric vector. However, if the vector contains <code>NAs</code>, our initial function doesnâ€™t handle them correctly:</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="introduction-to-r-for-biologists.html#cb119-1" tabindex="-1"></a>mean_customer <span class="ot">&lt;-</span> <span class="cf">function</span>(x){</span>
<span id="cb119-2"><a href="introduction-to-r-for-biologists.html#cb119-2" tabindex="-1"></a>    total <span class="ot">&lt;-</span> <span class="fu">sum</span>(x)</span>
<span id="cb119-3"><a href="introduction-to-r-for-biologists.html#cb119-3" tabindex="-1"></a>    mean_value <span class="ot">&lt;-</span> total <span class="sc">/</span> <span class="fu">length</span>(x)</span>
<span id="cb119-4"><a href="introduction-to-r-for-biologists.html#cb119-4" tabindex="-1"></a>    <span class="fu">return</span>(mean_value)</span>
<span id="cb119-5"><a href="introduction-to-r-for-biologists.html#cb119-5" tabindex="-1"></a>}</span></code></pre></div>
<p>When we try to calculate the mean of <code>NA_vec</code>, it returns <code>NA</code>:</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="introduction-to-r-for-biologists.html#cb120-1" tabindex="-1"></a><span class="fu">mean_customer</span>(NA_vec)</span></code></pre></div>
<pre><code>## [1] NA</code></pre>
<p>This outcome is not ideal, especially when we want to calculate the average of the non-missing values.</p>
</div>
<div id="adding-the-remove_na-argument" class="section level3" number="5.6.9">
<h3><span class="header-section-number">5.6.9</span> Adding the <code>remove_na</code> Argument</h3>
<p>To address this issue, we can enhance our function by introducing a new argument called <code>remove_na</code>, which allows us to control whether <code>NAs</code> should be removed before performing calculations. By default, we set <code>remove_na</code> to TRUE, indicating that NAs should be removed:</p>
<blockquote>
<p>Youâ€™ll see that we declare the remove_na argument with an =. That means that if no value is provided, by default it will take the value TRUE.</p>
</blockquote>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="introduction-to-r-for-biologists.html#cb122-1" tabindex="-1"></a>mean_customer_NA <span class="ot">&lt;-</span> <span class="cf">function</span>(x, <span class="at">remove_na =</span> <span class="cn">TRUE</span>){</span>
<span id="cb122-2"><a href="introduction-to-r-for-biologists.html#cb122-2" tabindex="-1"></a>    <span class="cf">if</span> (remove_na){</span>
<span id="cb122-3"><a href="introduction-to-r-for-biologists.html#cb122-3" tabindex="-1"></a>        x <span class="ot">&lt;-</span> x[<span class="sc">!</span><span class="fu">is.na</span>(x)]</span>
<span id="cb122-4"><a href="introduction-to-r-for-biologists.html#cb122-4" tabindex="-1"></a>    }</span>
<span id="cb122-5"><a href="introduction-to-r-for-biologists.html#cb122-5" tabindex="-1"></a>    total <span class="ot">&lt;-</span> <span class="fu">sum</span>(x)</span>
<span id="cb122-6"><a href="introduction-to-r-for-biologists.html#cb122-6" tabindex="-1"></a>    mean_value <span class="ot">&lt;-</span> total <span class="sc">/</span> <span class="fu">length</span>(x)</span>
<span id="cb122-7"><a href="introduction-to-r-for-biologists.html#cb122-7" tabindex="-1"></a>    <span class="fu">return</span>(mean_value)</span>
<span id="cb122-8"><a href="introduction-to-r-for-biologists.html#cb122-8" tabindex="-1"></a>}</span></code></pre></div>
<p>Now, our function behaves differently based on the value of <code>remove_na</code>. If set to <code>TRUE</code>, it removes <code>NAs</code> from the vector before calculating the mean; if set to <code>FALSE</code>, it includes <code>NAs</code> in the calculation.</p>
</div>
<div id="practical-application" class="section level3" number="5.6.10">
<h3><span class="header-section-number">5.6.10</span> Practical Application</h3>
<p>Letâ€™s see how this enhanced function works with our example vector:</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="introduction-to-r-for-biologists.html#cb123-1" tabindex="-1"></a><span class="fu">mean_customer_NA</span>(NA_vec)  <span class="co"># Default behavior (remove_na = TRUE)</span></span></code></pre></div>
<pre><code>## [1] 2</code></pre>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="introduction-to-r-for-biologists.html#cb125-1" tabindex="-1"></a><span class="fu">mean_customer_NA</span>(NA_vec, <span class="at">remove_na =</span> <span class="cn">TRUE</span>)  <span class="co"># Explicitly removing NAs</span></span></code></pre></div>
<pre><code>## [1] 2</code></pre>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="introduction-to-r-for-biologists.html#cb127-1" tabindex="-1"></a><span class="fu">mean_customer_NA</span>(NA_vec, <span class="at">remove_na =</span> <span class="cn">FALSE</span>)  <span class="co"># Including NAs</span></span></code></pre></div>
<pre><code>## [1] NA</code></pre>
<p>In the first two calls, we get a result of 2 by removing the <code>NAs</code>, while in the last call, we receive <code>NA</code> since we chose not to remove them.</p>
</div>
<div id="getting-help-with-functions" class="section level3" number="5.6.11">
<h3><span class="header-section-number">5.6.11</span> Getting help with functions</h3>
<div id="using-the-operator-for-documentation." class="section level4" number="5.6.11.1">
<h4><span class="header-section-number">5.6.11.1</span> Using the <code>?</code> Operator for Documentation.</h4>
<p>The <code>?</code> operator is a quick and convenient way to access documentation directly within the R environment. Simply type in the console <code>?</code> followed by the name of the function you want to learn more about. For example:</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="introduction-to-r-for-biologists.html#cb129-1" tabindex="-1"></a>?mean</span></code></pre></div>
<p>This command will open the documentation for the <code>mean()</code> function, providing details on its usage, arguments, examples, and related functions.</p>
</div>
<div id="accessing-documentation-via-the-help-function" class="section level4" number="5.6.11.2">
<h4><span class="header-section-number">5.6.11.2</span> Accessing Documentation via the <code>help()</code> Function</h4>
<p>Alternatively, you can use the <code>help()</code> function to retrieve documentation for a specific function. Syntax:</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="introduction-to-r-for-biologists.html#cb130-1" tabindex="-1"></a><span class="fu">help</span>(mean)</span></code></pre></div>
<p>Executing this command will display the documentation for the <code>mean()</code> function in the Help pane of R Studio.</p>
</div>
<div id="utilizing-the-help.search-function" class="section level4" number="5.6.11.3">
<h4><span class="header-section-number">5.6.11.3</span> Utilizing the <code>help.search()</code> Function</h4>
<p>f you are unsure about the exact function name or wish to search for functions related to a certain topic, you can use the help.search() function. This function allows you to search for keywords across all available documentation. Syntax:</p>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="introduction-to-r-for-biologists.html#cb131-1" tabindex="-1"></a><span class="fu">help.search</span>(<span class="st">&quot;keyword&quot;</span>)</span></code></pre></div>
<p>For example:</p>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="introduction-to-r-for-biologists.html#cb132-1" tabindex="-1"></a><span class="fu">help.search</span>(<span class="st">&quot;linear regression&quot;</span>)</span></code></pre></div>
<p>This command will return a list of relevant documentation entries containing the specified keyword, assisting you in finding relevant functions and packages.</p>
</div>
<div id="exploring-online-resources-and-community-forums" class="section level4" number="5.6.11.4">
<h4><span class="header-section-number">5.6.11.4</span> Exploring Online Resources and Community Forums</h4>
<p>In addition to built-in documentation, online resources such as the official R documentation website (<a href="https://www.rdocumentation.org/" class="uri">https://www.rdocumentation.org/</a>) and community forums like Stack Overflow (<a href="https://stackoverflow.com/" class="uri">https://stackoverflow.com/</a>) are valuable sources of information and support. You can search for specific functions, read user discussions, and even ask questions if needed.</p>
</div>
</div>
<div id="a-real-world-example-1" class="section level3" number="5.6.12">
<h3><span class="header-section-number">5.6.12</span> A Real World Example</h3>
<p>In the field of bioinformatics, analyzing gene expression data is a fundamental task. One common analysis involves identifying the most highly variable genes within a gene expression matrix. Letâ€™s create a custom R function for this purpose.</p>
<blockquote>
<p>Before diving into the creation our function, itâ€™s essential to understand some key pre-concepts. Firstly, gene expression data typically consists of a matrix where rows represent genes and columns represent samples or conditions. Second, the variability of gene expression across samples is a crucial metric, often measured by variance or standard deviation. Highly variable genes can provide valuable insights into biological processes.</p>
</blockquote>
<p>Letâ€™s create a custom function, <code>findVariableGenes</code>, which identifies the top N most highly variable genes in a gene expression matrix. Below is the code for the function, with explanations:</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="introduction-to-r-for-biologists.html#cb133-1" tabindex="-1"></a><span class="co"># Define the custom function</span></span>
<span id="cb133-2"><a href="introduction-to-r-for-biologists.html#cb133-2" tabindex="-1"></a>findVariableGenes <span class="ot">&lt;-</span> <span class="cf">function</span>(expr_matrix, <span class="at">n =</span> <span class="dv">10</span>) {</span>
<span id="cb133-3"><a href="introduction-to-r-for-biologists.html#cb133-3" tabindex="-1"></a>  <span class="co"># Calculate variances for each gene</span></span>
<span id="cb133-4"><a href="introduction-to-r-for-biologists.html#cb133-4" tabindex="-1"></a>  gene_variances <span class="ot">&lt;-</span> <span class="fu">apply</span>(expr_matrix, <span class="at">MARGIN =</span> <span class="dv">1</span>, var)</span>
<span id="cb133-5"><a href="introduction-to-r-for-biologists.html#cb133-5" tabindex="-1"></a>  </span>
<span id="cb133-6"><a href="introduction-to-r-for-biologists.html#cb133-6" tabindex="-1"></a>  <span class="co"># Sort genes by variance in descending order</span></span>
<span id="cb133-7"><a href="introduction-to-r-for-biologists.html#cb133-7" tabindex="-1"></a>  sorted_genes <span class="ot">&lt;-</span> <span class="fu">sort</span>(gene_variances, <span class="at">decreasing=</span><span class="cn">TRUE</span>) </span>
<span id="cb133-8"><a href="introduction-to-r-for-biologists.html#cb133-8" tabindex="-1"></a>  </span>
<span id="cb133-9"><a href="introduction-to-r-for-biologists.html#cb133-9" tabindex="-1"></a>  <span class="co"># Select the top N variable genes</span></span>
<span id="cb133-10"><a href="introduction-to-r-for-biologists.html#cb133-10" tabindex="-1"></a>  top_n <span class="ot">&lt;-</span> <span class="fu">names</span>(sorted_genes[<span class="dv">1</span><span class="sc">:</span>n])</span>
<span id="cb133-11"><a href="introduction-to-r-for-biologists.html#cb133-11" tabindex="-1"></a>  </span>
<span id="cb133-12"><a href="introduction-to-r-for-biologists.html#cb133-12" tabindex="-1"></a>  <span class="co"># Return the names of top variable genes</span></span>
<span id="cb133-13"><a href="introduction-to-r-for-biologists.html#cb133-13" tabindex="-1"></a>  <span class="fu">return</span>(top_n)</span>
<span id="cb133-14"><a href="introduction-to-r-for-biologists.html#cb133-14" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li><p>expr_matrix: This is the gene expression matrix you want to analyze.</p></li>
<li><p>n: The number of top variable genes you want to identify (default is 10).</p></li>
<li><p>apply function applies the var function to the matrix for rows (<code>MARGIN =1</code>). if you want to apply a function for columns, use <code>MARGIN =2</code>.</p></li>
</ul>
<p>Now, letâ€™s put our custom function to the test using some randomly generated gene expression data.</p>
<div id="generating-random-data" class="section level4" number="5.6.12.1">
<h4><span class="header-section-number">5.6.12.1</span> Generating Random Data</h4>
<p>Weâ€™ll create a gene expression matrix with 25 genes (rows) and 4 samples (columns) using normally distributed random data. To ensure reproducibility, weâ€™ll set a seed value (123). Hereâ€™s the code and the resulting data matrix:</p>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="introduction-to-r-for-biologists.html#cb134-1" tabindex="-1"></a><span class="co"># Set a seed for reproducibility (ensure all get the same random data)</span></span>
<span id="cb134-2"><a href="introduction-to-r-for-biologists.html#cb134-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)  <span class="co"># This sets a random seed to ensure that the random data generated below is reproducible.</span></span>
<span id="cb134-3"><a href="introduction-to-r-for-biologists.html#cb134-3" tabindex="-1"></a></span>
<span id="cb134-4"><a href="introduction-to-r-for-biologists.html#cb134-4" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(<span class="dv">100</span>), <span class="at">ncol =</span> <span class="dv">4</span>)  <span class="co"># This line generates a matrix with 100 random numbers from a standard normal distribution and organizes them into 4 columns.</span></span>
<span id="cb134-5"><a href="introduction-to-r-for-biologists.html#cb134-5" tabindex="-1"></a><span class="fu">rownames</span>(data) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>)  <span class="co"># This line assigns row names to the matrix, labeling each row as &quot;gene1&quot;, &quot;gene2&quot;, and so on, up to &quot;gene25&quot;.</span></span>
<span id="cb134-6"><a href="introduction-to-r-for-biologists.html#cb134-6" tabindex="-1"></a></span>
<span id="cb134-7"><a href="introduction-to-r-for-biologists.html#cb134-7" tabindex="-1"></a><span class="co"># Display the generated data</span></span>
<span id="cb134-8"><a href="introduction-to-r-for-biologists.html#cb134-8" tabindex="-1"></a><span class="fu">print</span>(data)</span></code></pre></div>
<pre><code>##               [,1]        [,2]        [,3]         [,4]
## gene1  -0.56047565 -1.68669331  0.25331851  1.025571370
## gene2  -0.23017749  0.83778704 -0.02854676 -0.284773007
## gene3   1.55870831  0.15337312 -0.04287046 -1.220717712
## gene4   0.07050839 -1.13813694  1.36860228  0.181303480
## gene5   0.12928774  1.25381492 -0.22577099 -0.138891362
## gene6   1.71506499  0.42646422  1.51647060  0.005764186
## gene7   0.46091621 -0.29507148 -1.54875280  0.385280401
## gene8  -1.26506123  0.89512566  0.58461375 -0.370660032
## gene9  -0.68685285  0.87813349  0.12385424  0.644376549
## gene10 -0.44566197  0.82158108  0.21594157 -0.220486562
## gene11  1.22408180  0.68864025  0.37963948  0.331781964
## gene12  0.35981383  0.55391765 -0.50232345  1.096839013
## gene13  0.40077145 -0.06191171 -0.33320738  0.435181491
## gene14  0.11068272 -0.30596266 -1.01857538 -0.325931586
## gene15 -0.55584113 -0.38047100 -1.07179123  1.148807618
## gene16  1.78691314 -0.69470698  0.30352864  0.993503856
## gene17  0.49785048 -0.20791728  0.44820978  0.548396960
## gene18 -1.96661716 -1.26539635  0.05300423  0.238731735
## gene19  0.70135590  2.16895597  0.92226747 -0.627906076
## gene20 -0.47279141  1.20796200  2.05008469  1.360652449
## gene21 -1.06782371 -1.12310858 -0.49103117 -0.600259587
## gene22 -0.21797491 -0.40288484 -2.30916888  2.187332993
## gene23 -1.02600445 -0.46665535  1.00573852  1.532610626
## gene24 -0.72889123  0.77996512 -0.70920076 -0.235700359
## gene25 -0.62503927 -0.08336907 -0.68800862 -1.026420900</code></pre>
</div>
<div id="using-the-custom-function-1" class="section level4" number="5.6.12.2">
<h4><span class="header-section-number">5.6.12.2</span> Using the Custom Function</h4>
<p>Now that we have our gene expression data, letâ€™s apply our <code>findVariableGenes</code> function to identify the top 10 most highly variable genes:</p>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="introduction-to-r-for-biologists.html#cb136-1" tabindex="-1"></a><span class="co"># Use the custom function to find highly variable genes</span></span>
<span id="cb136-2"><a href="introduction-to-r-for-biologists.html#cb136-2" tabindex="-1"></a>highly_variable <span class="ot">&lt;-</span> <span class="fu">findVariableGenes</span>(data, <span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb136-3"><a href="introduction-to-r-for-biologists.html#cb136-3" tabindex="-1"></a></span>
<span id="cb136-4"><a href="introduction-to-r-for-biologists.html#cb136-4" tabindex="-1"></a><span class="co"># Display the list of highly variable genes</span></span>
<span id="cb136-5"><a href="introduction-to-r-for-biologists.html#cb136-5" tabindex="-1"></a><span class="fu">print</span>(highly_variable)</span></code></pre></div>
<pre><code>##  [1] &quot;gene22&quot; &quot;gene23&quot; &quot;gene1&quot;  &quot;gene19&quot; &quot;gene3&quot;  &quot;gene20&quot; &quot;gene18&quot; &quot;gene16&quot;
##  [9] &quot;gene4&quot;  &quot;gene8&quot;</code></pre>
<p>Encapsulating logic into functions makes our code more organized, reusable, and scalable.Functions make code more organized, reusable, and scalable. As you code more in R, youâ€™ll want to encapsulate logic into functions just as shown here.</p>
</div>
</div>
<div id="conclusion-3" class="section level3" number="5.6.13">
<h3><span class="header-section-number">5.6.13</span> Conclusion</h3>
<p>Understanding functions and their ability to encapsulate code and return specific values is crucial in R programming. Functions enhance code organization, maintainability, and reusability, making them a valuable tool for any data analyst or scientist. The flexibility and efficiency they offer become increasingly evident as you tackle more complex data analysis tasks in R.</p>
</div>
</div>
<div id="common-mistakes-to-avoid" class="section level2" number="5.7">
<h2><span class="header-section-number">5.7</span> Common Mistakes to avoid</h2>
<p>In this lesson we will discuss some common mistakes that absolute beginners often make when learning R. Learning a new programming language can be challenging, and itâ€™s natural to encounter stumbling blocks along the way. By understanding these common mistakes, you can avoid them and become a more proficient R programmer.</p>
<div id="mixing-data-types" class="section level3" number="5.7.1">
<h3><span class="header-section-number">5.7.1</span> Mixing Data Types</h3>
<p>One common mistake is mixing different data types in operations. For instance, trying to add a number to a string or perform mathematical operations on non-numeric data types.</p>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="introduction-to-r-for-biologists.html#cb138-1" tabindex="-1"></a><span class="co"># Example:</span></span>
<span id="cb138-2"><a href="introduction-to-r-for-biologists.html#cb138-2" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="st">&quot;5&quot;</span></span>
<span id="cb138-3"><a href="introduction-to-r-for-biologists.html#cb138-3" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb138-4"><a href="introduction-to-r-for-biologists.html#cb138-4" tabindex="-1"></a>z <span class="ot">&lt;-</span> x <span class="sc">+</span> y  <span class="co"># This will result in an error because you cannot add a string and a number.</span></span></code></pre></div>
</div>
<div id="forgetting-function-parentheses" class="section level3" number="5.7.2">
<h3><span class="header-section-number">5.7.2</span> Forgetting Function Parentheses</h3>
<p>Another mistake is forgetting to include parentheses when calling functions. In R, functions typically require parentheses, even if they donâ€™t have any arguments.</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="introduction-to-r-for-biologists.html#cb139-1" tabindex="-1"></a><span class="co"># Incorrect:</span></span>
<span id="cb139-2"><a href="introduction-to-r-for-biologists.html#cb139-2" tabindex="-1"></a>print <span class="st">&quot;Hello, World!&quot;</span></span>
<span id="cb139-3"><a href="introduction-to-r-for-biologists.html#cb139-3" tabindex="-1"></a></span>
<span id="cb139-4"><a href="introduction-to-r-for-biologists.html#cb139-4" tabindex="-1"></a><span class="co"># Correct:</span></span>
<span id="cb139-5"><a href="introduction-to-r-for-biologists.html#cb139-5" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&quot;Hello, World!&quot;</span>)</span></code></pre></div>
</div>
<div id="overwriting-built-in-functions" class="section level3" number="5.7.3">
<h3><span class="header-section-number">5.7.3</span> Overwriting Built-in Functions</h3>
<p>Sometimes you might unintentionally overwrite built-in functions or variable names, causing unexpected behavior in your code.</p>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="introduction-to-r-for-biologists.html#cb140-1" tabindex="-1"></a><span class="co"># Example:</span></span>
<span id="cb140-2"><a href="introduction-to-r-for-biologists.html#cb140-2" tabindex="-1"></a><span class="co"># Incorrect:</span></span>
<span id="cb140-3"><a href="introduction-to-r-for-biologists.html#cb140-3" tabindex="-1"></a>mean <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb140-4"><a href="introduction-to-r-for-biologists.html#cb140-4" tabindex="-1"></a>  <span class="fu">sum</span>(x) <span class="sc">/</span> <span class="fu">length</span>(x)</span>
<span id="cb140-5"><a href="introduction-to-r-for-biologists.html#cb140-5" tabindex="-1"></a>}</span>
<span id="cb140-6"><a href="introduction-to-r-for-biologists.html#cb140-6" tabindex="-1"></a><span class="co"># Now, mean function is overwritten and will not work as expected.</span></span>
<span id="cb140-7"><a href="introduction-to-r-for-biologists.html#cb140-7" tabindex="-1"></a><span class="co"># e.g, it can not handel na.rm argument </span></span></code></pre></div>
</div>
<div id="misunderstanding-variable-scoping" class="section level3" number="5.7.4">
<h3><span class="header-section-number">5.7.4</span> Misunderstanding Variable Scoping</h3>
<p>Variable scoping in R defines where a variable can be used in a program. If a variable is defined inside a function, itâ€™s only accessible within that function (local scope). Variables defined outside functions, usually at the start of a script or in the main program, can be used anywhere (global scope).</p>
<p>Understanding variable scoping is crucial for avoiding errors and writing maintainable code. One common mistake is assuming that variables defined in one part of the program will be accessible from another part.</p>
<p>Letâ€™s look at an example:</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="introduction-to-r-for-biologists.html#cb141-1" tabindex="-1"></a><span class="co"># Example 1: Incorrect variable scoping</span></span>
<span id="cb141-2"><a href="introduction-to-r-for-biologists.html#cb141-2" tabindex="-1"></a>calculate_sum <span class="ot">&lt;-</span> <span class="cf">function</span>(a, b) {</span>
<span id="cb141-3"><a href="introduction-to-r-for-biologists.html#cb141-3" tabindex="-1"></a>  result <span class="ot">&lt;-</span> a <span class="sc">+</span> b</span>
<span id="cb141-4"><a href="introduction-to-r-for-biologists.html#cb141-4" tabindex="-1"></a>}</span>
<span id="cb141-5"><a href="introduction-to-r-for-biologists.html#cb141-5" tabindex="-1"></a></span>
<span id="cb141-6"><a href="introduction-to-r-for-biologists.html#cb141-6" tabindex="-1"></a><span class="co"># Trying to access &#39;result&#39; outside the function will result in an error.</span></span>
<span id="cb141-7"><a href="introduction-to-r-for-biologists.html#cb141-7" tabindex="-1"></a><span class="fu">print</span>(result)</span></code></pre></div>
<pre><code>## [1] 2.5</code></pre>
</div>
<div id="best-practices-for-variable-scoping-include" class="section level3" number="5.7.5">
<h3><span class="header-section-number">5.7.5</span> Best practices for variable scoping include</h3>
<ol style="list-style-type: decimal">
<li>Explicitly Pass Variables: When variables are needed in different parts of the program, itâ€™s better to explicitly pass them as arguments to functions rather than relying on global variables.</li>
</ol>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="introduction-to-r-for-biologists.html#cb143-1" tabindex="-1"></a><span class="co"># Example 2: Explicitly passing variables</span></span>
<span id="cb143-2"><a href="introduction-to-r-for-biologists.html#cb143-2" tabindex="-1"></a>calculate_sum <span class="ot">&lt;-</span> <span class="cf">function</span>(a, b) {</span>
<span id="cb143-3"><a href="introduction-to-r-for-biologists.html#cb143-3" tabindex="-1"></a>  result <span class="ot">&lt;-</span> a <span class="sc">+</span> b</span>
<span id="cb143-4"><a href="introduction-to-r-for-biologists.html#cb143-4" tabindex="-1"></a>  <span class="fu">return</span>(result)</span>
<span id="cb143-5"><a href="introduction-to-r-for-biologists.html#cb143-5" tabindex="-1"></a>}</span>
<span id="cb143-6"><a href="introduction-to-r-for-biologists.html#cb143-6" tabindex="-1"></a></span>
<span id="cb143-7"><a href="introduction-to-r-for-biologists.html#cb143-7" tabindex="-1"></a><span class="co"># Call the function with arguments</span></span>
<span id="cb143-8"><a href="introduction-to-r-for-biologists.html#cb143-8" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">calculate_sum</span>(<span class="dv">3</span>, <span class="dv">5</span>)</span>
<span id="cb143-9"><a href="introduction-to-r-for-biologists.html#cb143-9" tabindex="-1"></a><span class="fu">print</span>(result) </span></code></pre></div>
<pre><code>## [1] 8</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Use Meaningful Variable Names: Clear and meaningful variable names can help avoid confusion about variable scope and improve code readability.</li>
</ol>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="introduction-to-r-for-biologists.html#cb145-1" tabindex="-1"></a><span class="co"># Example 3: Clear variable names</span></span>
<span id="cb145-2"><a href="introduction-to-r-for-biologists.html#cb145-2" tabindex="-1"></a>calculate_area <span class="ot">&lt;-</span> <span class="cf">function</span>(length, width) {</span>
<span id="cb145-3"><a href="introduction-to-r-for-biologists.html#cb145-3" tabindex="-1"></a>  area <span class="ot">&lt;-</span> length <span class="sc">*</span> width</span>
<span id="cb145-4"><a href="introduction-to-r-for-biologists.html#cb145-4" tabindex="-1"></a>  <span class="fu">return</span>(area)</span>
<span id="cb145-5"><a href="introduction-to-r-for-biologists.html#cb145-5" tabindex="-1"></a>}</span>
<span id="cb145-6"><a href="introduction-to-r-for-biologists.html#cb145-6" tabindex="-1"></a></span>
<span id="cb145-7"><a href="introduction-to-r-for-biologists.html#cb145-7" tabindex="-1"></a><span class="co"># Call the function with arguments</span></span>
<span id="cb145-8"><a href="introduction-to-r-for-biologists.html#cb145-8" tabindex="-1"></a>area <span class="ot">&lt;-</span> <span class="fu">calculate_area</span>(<span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb145-9"><a href="introduction-to-r-for-biologists.html#cb145-9" tabindex="-1"></a><span class="fu">print</span>(area)  </span></code></pre></div>
<pre><code>## [1] 20</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>Avoid Modifying Global Variables Within Functions: Modifying global variables within functions can lead to unexpected behavior and make code harder to understand and debug. Instead, prefer returning values from functions.</li>
</ol>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="introduction-to-r-for-biologists.html#cb147-1" tabindex="-1"></a><span class="co"># Example 4: Avoid modifying global variables</span></span>
<span id="cb147-2"><a href="introduction-to-r-for-biologists.html#cb147-2" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb147-3"><a href="introduction-to-r-for-biologists.html#cb147-3" tabindex="-1"></a></span>
<span id="cb147-4"><a href="introduction-to-r-for-biologists.html#cb147-4" tabindex="-1"></a>modify_variable <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb147-5"><a href="introduction-to-r-for-biologists.html#cb147-5" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="dv">20</span>  <span class="co"># This creates a new local variable &#39;x&#39;, it does not modify the global &#39;x&#39;</span></span>
<span id="cb147-6"><a href="introduction-to-r-for-biologists.html#cb147-6" tabindex="-1"></a>}</span>
<span id="cb147-7"><a href="introduction-to-r-for-biologists.html#cb147-7" tabindex="-1"></a></span>
<span id="cb147-8"><a href="introduction-to-r-for-biologists.html#cb147-8" tabindex="-1"></a><span class="fu">modify_variable</span>()</span>
<span id="cb147-9"><a href="introduction-to-r-for-biologists.html#cb147-9" tabindex="-1"></a><span class="fu">print</span>(x)  <span class="co"># Output: 10 (global &#39;x&#39; remains unchanged)</span></span></code></pre></div>
<pre><code>## [1] 10</code></pre>
</div>
<div id="conclusion-4" class="section level3" number="5.7.6">
<h3><span class="header-section-number">5.7.6</span> Conclusion</h3>
<p>Learning R can be a rewarding experience, but itâ€™s common to encounter challenges along the way. By being aware of these common mistakes and understanding the allowed operations in R, you can avoid many pitfalls and become a more proficient R programmer. Remember to practice regularly and donâ€™t hesitate to seek help from resources like documentation, online tutorials, and community forums. Happy coding!</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="getting-started-with-r-and-rstudio.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="controlling-the-flow-of-our-programs.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/04-Introduction_to_R_for_biologists.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
