<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 9 Introduction to the tidyverse ecosystem | Master Real-world Bioinformatics analysis in R</title>
  <meta name="description" content="This is an R course teaching you skills to do real-world bioinformatics" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 9 Introduction to the tidyverse ecosystem | Master Real-world Bioinformatics analysis in R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is an R course teaching you skills to do real-world bioinformatics" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 9 Introduction to the tidyverse ecosystem | Master Real-world Bioinformatics analysis in R" />
  
  <meta name="twitter:description" content="This is an R course teaching you skills to do real-world bioinformatics" />
  

<meta name="author" content="Ming Tommy Tang" />


<meta name="date" content="2024-05-25" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="fundamental-data-structures-in-r.html"/>
<link rel="next" href="introduction-to-bioconductor.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<link href="libs/vembedr-0.1.5/css/vembedr.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li><a href="index.html#preface" id="toc-preface"><span class="toc-section-number">1</span> Preface</a></li>
<li><a href="intro.html#intro" id="toc-intro"><span class="toc-section-number">2</span> Introduction</a>
<ul>
<li><a href="intro.html#meet-your-instructor" id="toc-meet-your-instructor"><span class="toc-section-number">2.1</span> Meet your Instructor</a></li>
<li><a href="intro.html#leveraging-our-online-community" id="toc-leveraging-our-online-community"><span class="toc-section-number">2.2</span> Leveraging our online community</a></li>
<li><a href="intro.html#install-r-and-r-studio" id="toc-install-r-and-r-studio"><span class="toc-section-number">2.3</span> Install R and R studio</a></li>
<li><a href="intro.html#the-role-of-programming-in-biology" id="toc-the-role-of-programming-in-biology"><span class="toc-section-number">2.4</span> The role of programming in Biology</a>
<ul>
<li><a href="intro.html#what-are-programming-languages" id="toc-what-are-programming-languages"><span class="toc-section-number">2.4.1</span> What are programming languages?</a></li>
<li><a href="intro.html#why-programming-is-important-in-biology" id="toc-why-programming-is-important-in-biology"><span class="toc-section-number">2.4.2</span> Why programming is important in biology?</a></li>
<li><a href="intro.html#what-are-the-most-used-programming-languages-within-biology" id="toc-what-are-the-most-used-programming-languages-within-biology"><span class="toc-section-number">2.4.3</span> What are the most used programming languages within biology?</a></li>
</ul></li>
</ul></li>
<li><a href="introduction-to-programming.html#introduction-to-programming" id="toc-introduction-to-programming"><span class="toc-section-number">3</span> Introduction to programming</a>
<ul>
<li><a href="introduction-to-programming.html#what-is-an-algorithm" id="toc-what-is-an-algorithm"><span class="toc-section-number">3.1</span> What is an algorithm?</a></li>
<li><a href="introduction-to-programming.html#why-are-algorithms-important-in-programming" id="toc-why-are-algorithms-important-in-programming"><span class="toc-section-number">3.2</span> Why are Algorithms Important in Programming?</a></li>
<li><a href="introduction-to-programming.html#a-real-world-example" id="toc-a-real-world-example"><span class="toc-section-number">3.3</span> A Real World Example:</a></li>
<li><a href="introduction-to-programming.html#variables-data-types-and-expressions" id="toc-variables-data-types-and-expressions"><span class="toc-section-number">3.4</span> Variables, Data Types, and Expressions</a>
<ul>
<li><a href="introduction-to-programming.html#variables" id="toc-variables"><span class="toc-section-number">3.4.1</span> Variables</a></li>
<li><a href="introduction-to-programming.html#data-types" id="toc-data-types"><span class="toc-section-number">3.4.2</span> Data Types</a></li>
<li><a href="introduction-to-programming.html#expressions" id="toc-expressions"><span class="toc-section-number">3.4.3</span> Expressions</a></li>
<li><a href="introduction-to-programming.html#example-finding-the-average-number-of-bases-appearances-in-a-set-of-dna-strings" id="toc-example-finding-the-average-number-of-bases-appearances-in-a-set-of-dna-strings"><span class="toc-section-number">3.4.4</span> Example: Finding the average number of bases appearances in a set of DNA strings</a></li>
</ul></li>
</ul></li>
<li><a href="getting-started-with-r-and-rstudio.html#getting-started-with-r-and-rstudio" id="toc-getting-started-with-r-and-rstudio"><span class="toc-section-number">4</span> Getting started with R and RStudio</a>
<ul>
<li><a href="getting-started-with-r-and-rstudio.html#what-is-r-and-why-is-it-used-in-biology" id="toc-what-is-r-and-why-is-it-used-in-biology"><span class="toc-section-number">4.1</span> What is R and why is it used in Biology?</a></li>
<li><a href="getting-started-with-r-and-rstudio.html#what-is-r" id="toc-what-is-r"><span class="toc-section-number">4.2</span> What is R?</a></li>
<li><a href="getting-started-with-r-and-rstudio.html#why-is-r-used-in-biology" id="toc-why-is-r-used-in-biology"><span class="toc-section-number">4.3</span> Why is R Used in Biology?</a></li>
<li><a href="getting-started-with-r-and-rstudio.html#introduction-to-r-studio" id="toc-introduction-to-r-studio"><span class="toc-section-number">4.4</span> Introduction to R-Studio</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#introduction-to-r-for-biologists" id="toc-introduction-to-r-for-biologists"><span class="toc-section-number">5</span> Introduction to R for biologists</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#basic-data-types-in-r" id="toc-basic-data-types-in-r"><span class="toc-section-number">5.1</span> Basic Data Types in R</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#before-starting" id="toc-before-starting"><span class="toc-section-number">5.1.1</span> Before Starting</a></li>
<li><a href="introduction-to-r-for-biologists.html#numeric" id="toc-numeric"><span class="toc-section-number">5.1.2</span> 1. Numeric</a></li>
<li><a href="introduction-to-r-for-biologists.html#character-string" id="toc-character-string"><span class="toc-section-number">5.1.3</span> 2. Character (String)</a></li>
<li><a href="introduction-to-r-for-biologists.html#integer" id="toc-integer"><span class="toc-section-number">5.1.4</span> 3. Integer</a></li>
<li><a href="introduction-to-r-for-biologists.html#logical-boolean" id="toc-logical-boolean"><span class="toc-section-number">5.1.5</span> 4. Logical (Boolean)</a></li>
<li><a href="introduction-to-r-for-biologists.html#factor" id="toc-factor"><span class="toc-section-number">5.1.6</span> 5. Factor</a></li>
<li><a href="introduction-to-r-for-biologists.html#date-and-time" id="toc-date-and-time"><span class="toc-section-number">5.1.7</span> 6. Date and Time</a></li>
<li><a href="introduction-to-r-for-biologists.html#complex" id="toc-complex"><span class="toc-section-number">5.1.8</span> 7. Complex</a></li>
<li><a href="introduction-to-r-for-biologists.html#missing-values-na" id="toc-missing-values-na"><span class="toc-section-number">5.1.9</span> 8. Missing Values (NA)</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#the-key-concept-of-r-vectors" id="toc-the-key-concept-of-r-vectors"><span class="toc-section-number">5.2</span> The key concept of R: Vectors</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#creating-numeric-and-character-vectors" id="toc-creating-numeric-and-character-vectors"><span class="toc-section-number">5.2.1</span> Creating Numeric and Character Vectors</a></li>
<li><a href="introduction-to-r-for-biologists.html#saving-vectors-to-variables" id="toc-saving-vectors-to-variables"><span class="toc-section-number">5.2.2</span> Saving Vectors to Variables</a></li>
<li><a href="introduction-to-r-for-biologists.html#vectorized-calculations" id="toc-vectorized-calculations"><span class="toc-section-number">5.2.3</span> Vectorized Calculations</a></li>
<li><a href="introduction-to-r-for-biologists.html#performing-vectorized-calculations" id="toc-performing-vectorized-calculations"><span class="toc-section-number">5.2.4</span> Performing Vectorized Calculations</a></li>
<li><a href="introduction-to-r-for-biologists.html#operations-with-character-vectors" id="toc-operations-with-character-vectors"><span class="toc-section-number">5.2.5</span> Operations with character vectors</a></li>
<li><a href="introduction-to-r-for-biologists.html#logical-vectors" id="toc-logical-vectors"><span class="toc-section-number">5.2.6</span> Logical Vectors</a></li>
<li><a href="introduction-to-r-for-biologists.html#conclusion" id="toc-conclusion"><span class="toc-section-number">5.2.7</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#subsetting-and-indexing" id="toc-subsetting-and-indexing"><span class="toc-section-number">5.3</span> Subsetting and Indexing</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#indexing-in-r" id="toc-indexing-in-r"><span class="toc-section-number">5.3.1</span> Indexing in R</a></li>
<li><a href="introduction-to-r-for-biologists.html#slicing-in-r" id="toc-slicing-in-r"><span class="toc-section-number">5.3.2</span> Slicing in R</a></li>
<li><a href="introduction-to-r-for-biologists.html#negative-indexing" id="toc-negative-indexing"><span class="toc-section-number">5.3.3</span> Negative Indexing</a></li>
<li><a href="introduction-to-r-for-biologists.html#conclusion-1" id="toc-conclusion-1"><span class="toc-section-number">5.3.4</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#understanding-and-manipulating-matrices" id="toc-understanding-and-manipulating-matrices"><span class="toc-section-number">5.4</span> Understanding and Manipulating Matrices</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#creating-a-matrix" id="toc-creating-a-matrix"><span class="toc-section-number">5.4.1</span> Creating a Matrix</a></li>
<li><a href="introduction-to-r-for-biologists.html#adding-row-and-column-names" id="toc-adding-row-and-column-names"><span class="toc-section-number">5.4.2</span> Adding Row and Column Names</a></li>
<li><a href="introduction-to-r-for-biologists.html#subsetting-a-matrix" id="toc-subsetting-a-matrix"><span class="toc-section-number">5.4.3</span> Subsetting a Matrix</a></li>
<li><a href="introduction-to-r-for-biologists.html#calculations-with-matrices" id="toc-calculations-with-matrices"><span class="toc-section-number">5.4.4</span> Calculations with Matrices</a></li>
<li><a href="introduction-to-r-for-biologists.html#normalization" id="toc-normalization"><span class="toc-section-number">5.4.5</span> Normalization</a></li>
<li><a href="introduction-to-r-for-biologists.html#conclusion-2" id="toc-conclusion-2"><span class="toc-section-number">5.4.6</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#essential-functions-in-r" id="toc-essential-functions-in-r"><span class="toc-section-number">5.5</span> Essential Functions in R</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#length-function" id="toc-length-function"><span class="toc-section-number">5.5.1</span> <code>length</code> Function</a></li>
<li><a href="introduction-to-r-for-biologists.html#unique-function" id="toc-unique-function"><span class="toc-section-number">5.5.2</span> <code>unique</code> Function</a></li>
<li><a href="introduction-to-r-for-biologists.html#sort-function" id="toc-sort-function"><span class="toc-section-number">5.5.3</span> <code>sort</code> Function</a></li>
<li><a href="introduction-to-r-for-biologists.html#cor-function" id="toc-cor-function"><span class="toc-section-number">5.5.4</span> <code>cor</code> Function</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#functions-organizing-your-code-for-reusability" id="toc-functions-organizing-your-code-for-reusability"><span class="toc-section-number">5.6</span> Functions: Organizing Your Code for Reusability</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#what-is-a-function" id="toc-what-is-a-function"><span class="toc-section-number">5.6.1</span> What is a Function?</a></li>
<li><a href="introduction-to-r-for-biologists.html#creating-a-basic-function-calculating-the-mean" id="toc-creating-a-basic-function-calculating-the-mean"><span class="toc-section-number">5.6.2</span> Creating a Basic Function: Calculating the Mean</a></li>
<li><a href="introduction-to-r-for-biologists.html#using-the-custom-function" id="toc-using-the-custom-function"><span class="toc-section-number">5.6.3</span> Using the Custom Function</a></li>
<li><a href="introduction-to-r-for-biologists.html#returning-a-value-from-the-function" id="toc-returning-a-value-from-the-function"><span class="toc-section-number">5.6.4</span> Returning a Value from the Function</a></li>
<li><a href="introduction-to-r-for-biologists.html#optional-omitting-the-return-statement" id="toc-optional-omitting-the-return-statement"><span class="toc-section-number">5.6.5</span> Optional: Omitting the return Statement</a></li>
<li><a href="introduction-to-r-for-biologists.html#using-more-than-one-argument" id="toc-using-more-than-one-argument"><span class="toc-section-number">5.6.6</span> Using more than one argument</a></li>
<li><a href="introduction-to-r-for-biologists.html#understanding-missing-values-nas-in-r" id="toc-understanding-missing-values-nas-in-r"><span class="toc-section-number">5.6.7</span> Understanding Missing Values (NAs) in R</a></li>
<li><a href="introduction-to-r-for-biologists.html#initial-attempt-a-function-without-handling-nas" id="toc-initial-attempt-a-function-without-handling-nas"><span class="toc-section-number">5.6.8</span> Initial Attempt: A Function Without Handling NAs</a></li>
<li><a href="introduction-to-r-for-biologists.html#adding-the-remove_na-argument" id="toc-adding-the-remove_na-argument"><span class="toc-section-number">5.6.9</span> Adding the <code>remove_na</code> Argument</a></li>
<li><a href="introduction-to-r-for-biologists.html#practical-application" id="toc-practical-application"><span class="toc-section-number">5.6.10</span> Practical Application</a></li>
<li><a href="introduction-to-r-for-biologists.html#getting-help-with-functions" id="toc-getting-help-with-functions"><span class="toc-section-number">5.6.11</span> Getting help with functions</a></li>
<li><a href="introduction-to-r-for-biologists.html#a-real-world-example-1" id="toc-a-real-world-example-1"><span class="toc-section-number">5.6.12</span> A Real World Example</a></li>
<li><a href="introduction-to-r-for-biologists.html#conclusion-3" id="toc-conclusion-3"><span class="toc-section-number">5.6.13</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-r-for-biologists.html#common-mistakes-to-avoid" id="toc-common-mistakes-to-avoid"><span class="toc-section-number">5.7</span> Common Mistakes to avoid</a>
<ul>
<li><a href="introduction-to-r-for-biologists.html#mixing-data-types" id="toc-mixing-data-types"><span class="toc-section-number">5.7.1</span> Mixing Data Types</a></li>
<li><a href="introduction-to-r-for-biologists.html#forgetting-function-parentheses" id="toc-forgetting-function-parentheses"><span class="toc-section-number">5.7.2</span> Forgetting Function Parentheses</a></li>
<li><a href="introduction-to-r-for-biologists.html#overwriting-built-in-functions" id="toc-overwriting-built-in-functions"><span class="toc-section-number">5.7.3</span> Overwriting Built-in Functions</a></li>
<li><a href="introduction-to-r-for-biologists.html#misunderstanding-variable-scoping" id="toc-misunderstanding-variable-scoping"><span class="toc-section-number">5.7.4</span> Misunderstanding Variable Scoping</a></li>
<li><a href="introduction-to-r-for-biologists.html#best-practices-for-variable-scoping-include" id="toc-best-practices-for-variable-scoping-include"><span class="toc-section-number">5.7.5</span> Best practices for variable scoping include</a></li>
<li><a href="introduction-to-r-for-biologists.html#conclusion-4" id="toc-conclusion-4"><span class="toc-section-number">5.7.6</span> Conclusion</a></li>
</ul></li>
</ul></li>
<li><a href="controlling-the-flow-of-our-programs.html#controlling-the-flow-of-our-programs" id="toc-controlling-the-flow-of-our-programs"><span class="toc-section-number">6</span> Controlling the flow of our programs</a>
<ul>
<li><a href="controlling-the-flow-of-our-programs.html#boolean-operators" id="toc-boolean-operators"><span class="toc-section-number">6.1</span> Boolean Operators</a>
<ul>
<li><a href="controlling-the-flow-of-our-programs.html#comparison-operators" id="toc-comparison-operators"><span class="toc-section-number">6.1.1</span> Comparison Operators</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#logical-operators" id="toc-logical-operators"><span class="toc-section-number">6.1.2</span> Logical Operators</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#conclusion-5" id="toc-conclusion-5"><span class="toc-section-number">6.1.3</span> Conclusion</a></li>
</ul></li>
<li><a href="controlling-the-flow-of-our-programs.html#conditional-statements-if-else" id="toc-conditional-statements-if-else"><span class="toc-section-number">6.2</span> Conditional statements (<code>if</code>, <code>else</code>)</a>
<ul>
<li><a href="controlling-the-flow-of-our-programs.html#the-if-statement" id="toc-the-if-statement"><span class="toc-section-number">6.2.1</span> The <code>if</code> Statement</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#else-statement" id="toc-else-statement"><span class="toc-section-number">6.2.2</span> <code>else</code> statement</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#else-if" id="toc-else-if"><span class="toc-section-number">6.2.3</span> <code>else if</code></a></li>
<li><a href="controlling-the-flow-of-our-programs.html#exercise" id="toc-exercise"><span class="toc-section-number">6.2.4</span> Exercise</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#conclusion-6" id="toc-conclusion-6"><span class="toc-section-number">6.2.5</span> Conclusion</a></li>
</ul></li>
<li><a href="controlling-the-flow-of-our-programs.html#loops" id="toc-loops"><span class="toc-section-number">6.3</span> Loops</a>
<ul>
<li><a href="controlling-the-flow-of-our-programs.html#the-while-loop" id="toc-the-while-loop"><span class="toc-section-number">6.3.1</span> The <code>while</code> Loop</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#while-loops-in-real-life" id="toc-while-loops-in-real-life"><span class="toc-section-number">6.3.2</span> While Loops in Real Life</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#the-for-loop" id="toc-the-for-loop"><span class="toc-section-number">6.3.3</span> The <code>for</code> Loop</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#conclusion-7" id="toc-conclusion-7"><span class="toc-section-number">6.3.4</span> Conclusion</a></li>
</ul></li>
<li><a href="controlling-the-flow-of-our-programs.html#gene-expression-annotation-using-loops-and-control-structures" id="toc-gene-expression-annotation-using-loops-and-control-structures"><span class="toc-section-number">6.4</span> Gene Expression Annotation using Loops and Control Structures</a>
<ul>
<li><a href="controlling-the-flow-of-our-programs.html#the-data" id="toc-the-data"><span class="toc-section-number">6.4.1</span> The Data</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#creating-annotations" id="toc-creating-annotations"><span class="toc-section-number">6.4.2</span> Creating Annotations</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#the-loop" id="toc-the-loop"><span class="toc-section-number">6.4.3</span> The Loop</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#output" id="toc-output"><span class="toc-section-number">6.4.4</span> Output</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#putting-it-all-together" id="toc-putting-it-all-together"><span class="toc-section-number">6.4.5</span> Putting it all together</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#conclusion-8" id="toc-conclusion-8"><span class="toc-section-number">6.4.6</span> Conclusion</a></li>
</ul></li>
<li><a href="controlling-the-flow-of-our-programs.html#lets-solve-a-challenge" id="toc-lets-solve-a-challenge"><span class="toc-section-number">6.5</span> Let’s solve a Challenge</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#solution" id="toc-solution"><span class="toc-section-number">6.6</span> Solution</a>
<ul>
<li><a href="controlling-the-flow-of-our-programs.html#using-a-for-loop" id="toc-using-a-for-loop"><span class="toc-section-number">6.6.1</span> using a for loop</a></li>
<li><a href="controlling-the-flow-of-our-programs.html#vectorized-solution" id="toc-vectorized-solution"><span class="toc-section-number">6.6.2</span> vectorized solution</a></li>
</ul></li>
<li><a href="controlling-the-flow-of-our-programs.html#section-complete" id="toc-section-complete"><span class="toc-section-number">6.7</span> Section complete</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#going-more-in-depth-with-r" id="toc-going-more-in-depth-with-r"><span class="toc-section-number">7</span> Going more in-depth with R</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#handling-missing-values" id="toc-handling-missing-values"><span class="toc-section-number">7.1</span> Handling Missing Values</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#understanding-na-in-r" id="toc-understanding-na-in-r"><span class="toc-section-number">7.1.1</span> Understanding NA in R</a></li>
<li><a href="going-more-in-depth-with-r.html#identifying-missing-values" id="toc-identifying-missing-values"><span class="toc-section-number">7.1.2</span> Identifying Missing Values</a></li>
<li><a href="going-more-in-depth-with-r.html#remove-missing-values" id="toc-remove-missing-values"><span class="toc-section-number">7.1.3</span> remove missing values</a></li>
<li><a href="going-more-in-depth-with-r.html#real-life-note" id="toc-real-life-note"><span class="toc-section-number">7.1.4</span> real-life note</a></li>
<li><a href="going-more-in-depth-with-r.html#conclusion-9" id="toc-conclusion-9"><span class="toc-section-number">7.1.5</span> Conclusion</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#introduction-to-statistical-tests-and-p-values" id="toc-introduction-to-statistical-tests-and-p-values"><span class="toc-section-number">7.2</span> Introduction to Statistical Tests and P-Values</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#practical-application-of-statistical-test-with-r" id="toc-practical-application-of-statistical-test-with-r"><span class="toc-section-number">7.2.1</span> Practical Application of statistical test with R</a></li>
<li><a href="going-more-in-depth-with-r.html#conclusion-10" id="toc-conclusion-10"><span class="toc-section-number">7.2.2</span> Conclusion</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#understanding-r-packages" id="toc-understanding-r-packages"><span class="toc-section-number">7.3</span> Understanding R Packages</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#what-is-an-r-package" id="toc-what-is-an-r-package"><span class="toc-section-number">7.3.1</span> What is an R Package?</a></li>
<li><a href="going-more-in-depth-with-r.html#why-use-r-packages" id="toc-why-use-r-packages"><span class="toc-section-number">7.3.2</span> Why Use R Packages?</a></li>
<li><a href="going-more-in-depth-with-r.html#how-to-use-a-package" id="toc-how-to-use-a-package"><span class="toc-section-number">7.3.3</span> How to use a package?</a></li>
<li><a href="going-more-in-depth-with-r.html#embracing-errors" id="toc-embracing-errors"><span class="toc-section-number">7.3.4</span> Embracing Errors</a></li>
<li><a href="going-more-in-depth-with-r.html#how-to-look-for-package-docs-within-r-studio" id="toc-how-to-look-for-package-docs-within-r-studio"><span class="toc-section-number">7.3.5</span> How to look for package docs within R studio?</a></li>
<li><a href="going-more-in-depth-with-r.html#exploring-package-vignettes" id="toc-exploring-package-vignettes"><span class="toc-section-number">7.3.6</span> Exploring Package Vignettes</a></li>
<li><a href="going-more-in-depth-with-r.html#practical-tip" id="toc-practical-tip"><span class="toc-section-number">7.3.7</span> Practical Tip</a></li>
<li><a href="going-more-in-depth-with-r.html#conclusion-11" id="toc-conclusion-11"><span class="toc-section-number">7.3.8</span> Conclusion</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#exploring-functions-in-different-packages-avoiding-collisions-and-accessing-them" id="toc-exploring-functions-in-different-packages-avoiding-collisions-and-accessing-them"><span class="toc-section-number">7.4</span> Exploring Functions in Different Packages: Avoiding Collisions and Accessing Them</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#using-double-colon" id="toc-using-double-colon"><span class="toc-section-number">7.4.1</span> Using Double Colon (<code>::</code>)</a></li>
<li><a href="going-more-in-depth-with-r.html#conclusion-12" id="toc-conclusion-12"><span class="toc-section-number">7.4.2</span> Conclusion</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#writing-custom-scripts" id="toc-writing-custom-scripts"><span class="toc-section-number">7.5</span> Writing Custom Scripts</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#your-working-directory" id="toc-your-working-directory"><span class="toc-section-number">7.5.1</span> Your Working Directory</a></li>
<li><a href="going-more-in-depth-with-r.html#writing-functions-in-scripts" id="toc-writing-functions-in-scripts"><span class="toc-section-number">7.5.2</span> Writing Functions in Scripts</a></li>
<li><a href="going-more-in-depth-with-r.html#importing-functions-from-scripts" id="toc-importing-functions-from-scripts"><span class="toc-section-number">7.5.3</span> Importing Functions from Scripts</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#data-io" id="toc-data-io"><span class="toc-section-number">7.6</span> Data I/O</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#built-in-functions" id="toc-built-in-functions"><span class="toc-section-number">7.6.1</span> Built-In Functions</a></li>
<li><a href="going-more-in-depth-with-r.html#real-world-applications" id="toc-real-world-applications"><span class="toc-section-number">7.6.2</span> Real-World Applications</a></li>
<li><a href="going-more-in-depth-with-r.html#conclusion-13" id="toc-conclusion-13"><span class="toc-section-number">7.6.3</span> Conclusion</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#best-practices-in-modular-programming-and-project-management" id="toc-best-practices-in-modular-programming-and-project-management"><span class="toc-section-number">7.7</span> Best Practices in Modular Programming and Project Management</a>
<ul>
<li><a href="going-more-in-depth-with-r.html#introduction-to-modular-programming" id="toc-introduction-to-modular-programming"><span class="toc-section-number">7.7.1</span> Introduction to Modular Programming</a></li>
<li><a href="going-more-in-depth-with-r.html#breaking-down-your-project" id="toc-breaking-down-your-project"><span class="toc-section-number">7.7.2</span> Breaking Down Your Project</a></li>
<li><a href="going-more-in-depth-with-r.html#conclusion-14" id="toc-conclusion-14"><span class="toc-section-number">7.7.3</span> Conclusion</a></li>
</ul></li>
<li><a href="going-more-in-depth-with-r.html#section-complete-1" id="toc-section-complete-1"><span class="toc-section-number">7.8</span> Section complete</a></li>
</ul></li>
<li><a href="fundamental-data-structures-in-r.html#fundamental-data-structures-in-r" id="toc-fundamental-data-structures-in-r"><span class="toc-section-number">8</span> Fundamental Data Structures in R</a>
<ul>
<li><a href="fundamental-data-structures-in-r.html#named-vectors" id="toc-named-vectors"><span class="toc-section-number">8.1</span> Named Vectors</a>
<ul>
<li><a href="fundamental-data-structures-in-r.html#what-is-a-named-vector" id="toc-what-is-a-named-vector"><span class="toc-section-number">8.1.1</span> What is a Named Vector?</a></li>
<li><a href="fundamental-data-structures-in-r.html#using-names-to-subset-a-vector" id="toc-using-names-to-subset-a-vector"><span class="toc-section-number">8.1.2</span> Using Names to Subset a Vector</a></li>
<li><a href="fundamental-data-structures-in-r.html#named-vectors-as-dictionary-like-structures" id="toc-named-vectors-as-dictionary-like-structures"><span class="toc-section-number">8.1.3</span> Named Vectors as Dictionary-Like Structures</a></li>
<li><a href="fundamental-data-structures-in-r.html#real-life-example" id="toc-real-life-example"><span class="toc-section-number">8.1.4</span> Real-life example</a></li>
<li><a href="fundamental-data-structures-in-r.html#reordering-and-subsetting-with-names" id="toc-reordering-and-subsetting-with-names"><span class="toc-section-number">8.1.5</span> Reordering and Subsetting with Names</a></li>
<li><a href="fundamental-data-structures-in-r.html#conclusion-15" id="toc-conclusion-15"><span class="toc-section-number">8.1.6</span> Conclusion</a></li>
</ul></li>
<li><a href="fundamental-data-structures-in-r.html#lists" id="toc-lists"><span class="toc-section-number">8.2</span> Lists</a>
<ul>
<li><a href="fundamental-data-structures-in-r.html#introduction-to-lists" id="toc-introduction-to-lists"><span class="toc-section-number">8.2.1</span> Introduction to Lists</a></li>
<li><a href="fundamental-data-structures-in-r.html#creating-a-list" id="toc-creating-a-list"><span class="toc-section-number">8.2.2</span> Creating a List</a></li>
<li><a href="fundamental-data-structures-in-r.html#accessing-list-elements" id="toc-accessing-list-elements"><span class="toc-section-number">8.2.3</span> Accessing List Elements</a></li>
<li><a href="fundamental-data-structures-in-r.html#conclusions" id="toc-conclusions"><span class="toc-section-number">8.2.4</span> Conclusions</a></li>
</ul></li>
<li><a href="fundamental-data-structures-in-r.html#dataframes" id="toc-dataframes"><span class="toc-section-number">8.3</span> Dataframes</a>
<ul>
<li><a href="fundamental-data-structures-in-r.html#what-is-a-data-frame" id="toc-what-is-a-data-frame"><span class="toc-section-number">8.3.1</span> What is a Data Frame?</a></li>
<li><a href="fundamental-data-structures-in-r.html#creating-a-data-frame" id="toc-creating-a-data-frame"><span class="toc-section-number">8.3.2</span> Creating a Data Frame</a></li>
<li><a href="fundamental-data-structures-in-r.html#accessing-columns" id="toc-accessing-columns"><span class="toc-section-number">8.3.3</span> Accessing Columns</a></li>
<li><a href="fundamental-data-structures-in-r.html#subsetting-a-data-frame" id="toc-subsetting-a-data-frame"><span class="toc-section-number">8.3.4</span> Subsetting a Data Frame</a></li>
<li><a href="fundamental-data-structures-in-r.html#conclusion-16" id="toc-conclusion-16"><span class="toc-section-number">8.3.5</span> Conclusion</a></li>
</ul></li>
<li><a href="fundamental-data-structures-in-r.html#how-to-represent-categorical-data-in-r-understanding-factors" id="toc-how-to-represent-categorical-data-in-r-understanding-factors"><span class="toc-section-number">8.4</span> How to Represent Categorical Data in R? Understanding Factors</a>
<ul>
<li><a href="fundamental-data-structures-in-r.html#creating-factors" id="toc-creating-factors"><span class="toc-section-number">8.4.1</span> Creating Factors</a></li>
<li><a href="fundamental-data-structures-in-r.html#changing-levels" id="toc-changing-levels"><span class="toc-section-number">8.4.2</span> Changing levels</a></li>
<li><a href="fundamental-data-structures-in-r.html#converting-factors" id="toc-converting-factors"><span class="toc-section-number">8.4.3</span> Converting Factors</a></li>
<li><a href="fundamental-data-structures-in-r.html#customizing-factor-levels" id="toc-customizing-factor-levels"><span class="toc-section-number">8.4.4</span> Customizing Factor Levels</a></li>
<li><a href="fundamental-data-structures-in-r.html#creating-factors-from-character-vectors" id="toc-creating-factors-from-character-vectors"><span class="toc-section-number">8.4.5</span> Creating Factors from Character Vectors</a></li>
<li><a href="fundamental-data-structures-in-r.html#subsetting-factors" id="toc-subsetting-factors"><span class="toc-section-number">8.4.6</span> Subsetting Factors</a></li>
<li><a href="fundamental-data-structures-in-r.html#removing-unused-levels" id="toc-removing-unused-levels"><span class="toc-section-number">8.4.7</span> Removing Unused Levels</a></li>
<li><a href="fundamental-data-structures-in-r.html#conclusion-17" id="toc-conclusion-17"><span class="toc-section-number">8.4.8</span> Conclusion</a></li>
</ul></li>
<li><a href="fundamental-data-structures-in-r.html#section-complete-2" id="toc-section-complete-2"><span class="toc-section-number">8.5</span> Section Complete</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#introduction-to-the-tidyverse-ecosystem" id="toc-introduction-to-the-tidyverse-ecosystem"><span class="toc-section-number">9</span> Introduction to the tidyverse ecosystem</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#what-is-the-tidyverse" id="toc-what-is-the-tidyverse"><span class="toc-section-number">9.1</span> What is the Tidyverse?</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#key-tidyverse-packages" id="toc-key-tidyverse-packages"><span class="toc-section-number">9.2</span> Key Tidyverse Packages</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#dplyr" id="toc-dplyr"><span class="toc-section-number">9.2.1</span> dplyr</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#a-note-on-the-pipe" id="toc-a-note-on-the-pipe"><span class="toc-section-number">9.2.2</span> A note on the pipe</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#ggplot2" id="toc-ggplot2"><span class="toc-section-number">9.2.3</span> ggplot2</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#tidyr" id="toc-tidyr"><span class="toc-section-number">9.2.4</span> tidyr</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#readr" id="toc-readr"><span class="toc-section-number">9.2.5</span> readr</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#tibble" id="toc-tibble"><span class="toc-section-number">9.2.6</span> tibble</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#stringr" id="toc-stringr"><span class="toc-section-number">9.2.7</span> stringr</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#forcats" id="toc-forcats"><span class="toc-section-number">9.2.8</span> forcats</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#real-world-applications-1" id="toc-real-world-applications-1"><span class="toc-section-number">9.2.9</span> Real-World Applications</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#tibble-and-readr---modern-data-structures-in-r" id="toc-tibble-and-readr---modern-data-structures-in-r"><span class="toc-section-number">9.3</span> Tibble and Readr - Modern Data Structures in R</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#tibbles-a-better-way-to-store-data" id="toc-tibbles-a-better-way-to-store-data"><span class="toc-section-number">9.3.1</span> Tibbles: A Better Way to Store Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#rownames-and-tibbles" id="toc-rownames-and-tibbles"><span class="toc-section-number">9.3.2</span> Rownames and Tibbles</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#reading-data-with-readr" id="toc-reading-data-with-readr"><span class="toc-section-number">9.3.3</span> Reading Data with readr</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#using-tibbles-with-imported-data" id="toc-using-tibbles-with-imported-data"><span class="toc-section-number">9.3.4</span> Using Tibbles with Imported Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#note-on-reading-excel-files" id="toc-note-on-reading-excel-files"><span class="toc-section-number">9.3.5</span> Note on reading Excel files</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-18" id="toc-conclusion-18"><span class="toc-section-number">9.3.6</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#the-tidy-data-format" id="toc-the-tidy-data-format"><span class="toc-section-number">9.4</span> The tidy data format</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#the-importance-of-tidy-data" id="toc-the-importance-of-tidy-data"><span class="toc-section-number">9.4.1</span> The Importance of Tidy Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#long-format-vs.-wide-format" id="toc-long-format-vs.-wide-format"><span class="toc-section-number">9.4.2</span> Long Format vs. Wide Format</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#introducing-dplyr-your-data-wrangling-toolkit" id="toc-introducing-dplyr-your-data-wrangling-toolkit"><span class="toc-section-number">9.5</span> Introducing dplyr: Your Data Wrangling Toolkit</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#selecting-specific-columns" id="toc-selecting-specific-columns"><span class="toc-section-number">9.5.1</span> Selecting Specific Columns</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#adding-new-columns" id="toc-adding-new-columns"><span class="toc-section-number">9.5.2</span> Adding New Columns</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#reordering-columns" id="toc-reordering-columns"><span class="toc-section-number">9.5.3</span> Reordering Columns</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#filtering-data" id="toc-filtering-data"><span class="toc-section-number">9.5.4</span> Filtering Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#summarizing-data" id="toc-summarizing-data"><span class="toc-section-number">9.5.5</span> Summarizing Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#sorting-data" id="toc-sorting-data"><span class="toc-section-number">9.5.6</span> Sorting Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-19" id="toc-conclusion-19"><span class="toc-section-number">9.5.7</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#stringr-your-essential-toolkit-to-manipulate-strings" id="toc-stringr-your-essential-toolkit-to-manipulate-strings"><span class="toc-section-number">9.6</span> stringr: your essential toolkit to manipulate strings</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#purrr-ditch-your-for-loops" id="toc-purrr-ditch-your-for-loops"><span class="toc-section-number">9.7</span> purrr: ditch your for loops</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#nesting-data-with-nest" id="toc-nesting-data-with-nest"><span class="toc-section-number">9.7.1</span> Nesting Data with nest()</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#map-and-its-variants" id="toc-map-and-its-variants"><span class="toc-section-number">9.7.2</span> <code>map()</code> and Its Variants</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#note-on-anonymous-functions" id="toc-note-on-anonymous-functions"><span class="toc-section-number">9.7.3</span> Note on Anonymous functions</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-20" id="toc-conclusion-20"><span class="toc-section-number">9.7.4</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#tidying-metadata-from-geo." id="toc-tidying-metadata-from-geo."><span class="toc-section-number">9.8</span> Tidying metadata from GEO.</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#prerequisites" id="toc-prerequisites"><span class="toc-section-number">9.8.1</span> Prerequisites</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#getting-started" id="toc-getting-started"><span class="toc-section-number">9.8.2</span> Getting Started</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#tidying-the-metadata" id="toc-tidying-the-metadata"><span class="toc-section-number">9.8.3</span> Tidying the Metadata</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#converting-to-a-tidy-data-frame" id="toc-converting-to-a-tidy-data-frame"><span class="toc-section-number">9.8.4</span> Converting to a Tidy Data Frame</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#bonus-reading-multiple-tsv-files" id="toc-bonus-reading-multiple-tsv-files"><span class="toc-section-number">9.8.5</span> Bonus: Reading Multiple TSV Files</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#section-complete-3" id="toc-section-complete-3"><span class="toc-section-number">9.9</span> Section Complete</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-21" id="toc-conclusion-21"><span class="toc-section-number">9.9.1</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#data-visualization-with-ggplot2" id="toc-data-visualization-with-ggplot2"><span class="toc-section-number">9.10</span> Data visualization with ggplot2</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#creating-scatter-plots" id="toc-creating-scatter-plots"><span class="toc-section-number">9.11</span> Creating Scatter Plots</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#prerequisites-1" id="toc-prerequisites-1"><span class="toc-section-number">9.11.1</span> Prerequisites</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#loading-and-exploring-the-data" id="toc-loading-and-exploring-the-data"><span class="toc-section-number">9.11.2</span> Loading and Exploring the Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#filtering-the-data" id="toc-filtering-the-data"><span class="toc-section-number">9.11.3</span> Filtering the Data</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#creating-a-scatter-plot" id="toc-creating-a-scatter-plot"><span class="toc-section-number">9.11.4</span> Creating a Scatter Plot</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#calculating-correlation" id="toc-calculating-correlation"><span class="toc-section-number">9.11.5</span> Calculating Correlation</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#hypothesis-testing" id="toc-hypothesis-testing"><span class="toc-section-number">9.11.6</span> Hypothesis Testing</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#adding-a-regression-line" id="toc-adding-a-regression-line"><span class="toc-section-number">9.11.7</span> Adding a Regression Line</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-22" id="toc-conclusion-22"><span class="toc-section-number">9.11.8</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#understanding-distributions-with-histograms" id="toc-understanding-distributions-with-histograms"><span class="toc-section-number">9.12</span> Understanding Distributions with Histograms</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#creating-a-basic-histogram" id="toc-creating-a-basic-histogram"><span class="toc-section-number">9.12.1</span> Creating a Basic Histogram</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#customizing-histogram-bins" id="toc-customizing-histogram-bins"><span class="toc-section-number">9.12.2</span> Customizing Histogram Bins</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-23" id="toc-conclusion-23"><span class="toc-section-number">9.12.3</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#visualizing-data-distribution-with-boxplots-and-violin-plots" id="toc-visualizing-data-distribution-with-boxplots-and-violin-plots"><span class="toc-section-number">9.13</span> Visualizing Data Distribution with Boxplots and Violin Plots</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#creating-a-basic-boxplot" id="toc-creating-a-basic-boxplot"><span class="toc-section-number">9.13.1</span> Creating a Basic Boxplot</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#rotating-x-axis-labels" id="toc-rotating-x-axis-labels"><span class="toc-section-number">9.13.2</span> Rotating X-axis Labels</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#adding-color-to-boxplots" id="toc-adding-color-to-boxplots"><span class="toc-section-number">9.13.3</span> Adding Color to Boxplots</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#customizing-color-palettes" id="toc-customizing-color-palettes"><span class="toc-section-number">9.13.4</span> Customizing Color Palettes</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#applying-custom-color-palette" id="toc-applying-custom-color-palette"><span class="toc-section-number">9.13.5</span> Applying Custom Color Palette</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#reordering-boxplots" id="toc-reordering-boxplots"><span class="toc-section-number">9.13.6</span> Reordering Boxplots</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#using-violin-plots" id="toc-using-violin-plots"><span class="toc-section-number">9.13.7</span> Using Violin Plots</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-24" id="toc-conclusion-24"><span class="toc-section-number">9.13.8</span> Conclusion</a></li>
</ul></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#creating-bar-plots-to-visualize-median-epcam-levels-by-cancer-type" id="toc-creating-bar-plots-to-visualize-median-epcam-levels-by-cancer-type"><span class="toc-section-number">9.14</span> Creating Bar Plots to Visualize Median EPCAM Levels by Cancer Type</a>
<ul>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#calculating-median-epcam-levels-by-cancer-type" id="toc-calculating-median-epcam-levels-by-cancer-type"><span class="toc-section-number">9.14.1</span> Calculating Median EPCAM Levels by Cancer Type</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#creating-a-basic-bar-plot" id="toc-creating-a-basic-bar-plot"><span class="toc-section-number">9.14.2</span> Creating a Basic Bar Plot</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#reordering-the-bars" id="toc-reordering-the-bars"><span class="toc-section-number">9.14.3</span> Reordering the Bars</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#reversing-the-order" id="toc-reversing-the-order"><span class="toc-section-number">9.14.4</span> Reversing the Order</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#note-on-factor-levels" id="toc-note-on-factor-levels"><span class="toc-section-number">9.14.5</span> Note on factor levels</a></li>
<li><a href="introduction-to-the-tidyverse-ecosystem.html#conclusion-25" id="toc-conclusion-25"><span class="toc-section-number">9.14.6</span> Conclusion</a></li>
</ul></li>
</ul></li>
<li><a href="introduction-to-bioconductor.html#introduction-to-bioconductor" id="toc-introduction-to-bioconductor"><span class="toc-section-number">10</span> Introduction to BioConductor</a>
<ul>
<li><a href="introduction-to-bioconductor.html#introduction-to-biocmanager" id="toc-introduction-to-biocmanager"><span class="toc-section-number">10.1</span> Introduction to BiocManager</a>
<ul>
<li><a href="introduction-to-bioconductor.html#why-use-biocmanager" id="toc-why-use-biocmanager"><span class="toc-section-number">10.1.1</span> Why Use BiocManager?</a></li>
<li><a href="introduction-to-bioconductor.html#installing-biocmanager" id="toc-installing-biocmanager"><span class="toc-section-number">10.1.2</span> Installing BiocManager</a></li>
<li><a href="introduction-to-bioconductor.html#installing-bioconductor-packages" id="toc-installing-bioconductor-packages"><span class="toc-section-number">10.1.3</span> Installing Bioconductor Packages</a></li>
<li><a href="introduction-to-bioconductor.html#updating-bioconductor-packages" id="toc-updating-bioconductor-packages"><span class="toc-section-number">10.1.4</span> Updating Bioconductor Packages</a></li>
<li><a href="introduction-to-bioconductor.html#checking-for-valid-bioconductor-versions" id="toc-checking-for-valid-bioconductor-versions"><span class="toc-section-number">10.1.5</span> Checking for Valid Bioconductor Versions</a></li>
<li><a href="introduction-to-bioconductor.html#conclusion-26" id="toc-conclusion-26"><span class="toc-section-number">10.1.6</span> Conclusion</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Master Real-world Bioinformatics analysis in R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-the-tidyverse-ecosystem" class="section level1" number="9">
<h1><span class="header-section-number">Chapter 9</span> Introduction to the tidyverse ecosystem</h1>
<div id="what-is-the-tidyverse" class="section level2" number="9.1">
<h2><span class="header-section-number">9.1</span> What is the Tidyverse?</h2>
<p>In this lesson, we will explore the Tidyverse, a powerful collection of R packages designed to simplify and streamline the data science workflow. The Tidyverse is particularly well-suited for beginners and professionals alike, as it encourages a structured and intuitive approach to data manipulation and visualization.</p>
<p><img src="images/tidyverse.png" /></p>
<p>The Tidyverse is a comprehensive ecosystem of R packages that share common principles for data manipulation and visualization. It encourages the use of tidy data, a specific data structure that makes data analysis more straightforward. Tidy data arranges observations in rows and variables in columns, making it easier to work with and visualize.</p>
</div>
<div id="key-tidyverse-packages" class="section level2" number="9.2">
<h2><span class="header-section-number">9.2</span> Key Tidyverse Packages</h2>
<p>Let’s dive into some of the essential packages within the Tidyverse and understand how they can be beneficial in data analysis.</p>
<div id="dplyr" class="section level3" number="9.2.1">
<h3><span class="header-section-number">9.2.1</span> dplyr</h3>
<p>Explore dplyr docs <a href="https://dplyr.tidyverse.org/">here</a>.</p>
<p>The core of the Tidyverse is the <code>dplyr</code> package. It provides a set of functions for data manipulation, including filtering, summarizing, and transforming data frames. Here’s a simple example of how you might use <code>dplyr</code> to filter data:</p>
<div class="sourceCode" id="cb320"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb320-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb320-1" tabindex="-1"></a><span class="co"># install.packages(&quot;dplyr&quot;)</span></span>
<span id="cb320-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb320-2" tabindex="-1"></a><span class="co"># Load the dplyr package</span></span>
<span id="cb320-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb320-3" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb320-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb320-4" tabindex="-1"></a></span>
<span id="cb320-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb320-5" tabindex="-1"></a><span class="co"># Create a data frame</span></span>
<span id="cb320-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb320-6" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb320-7"><a href="introduction-to-the-tidyverse-ecosystem.html#cb320-7" tabindex="-1"></a>  <span class="at">Name =</span> <span class="fu">c</span>(<span class="st">&quot;Alice&quot;</span>, <span class="st">&quot;Bob&quot;</span>, <span class="st">&quot;Charlie&quot;</span>),</span>
<span id="cb320-8"><a href="introduction-to-the-tidyverse-ecosystem.html#cb320-8" tabindex="-1"></a>  <span class="at">Age =</span> <span class="fu">c</span>(<span class="dv">25</span>, <span class="dv">30</span>, <span class="dv">22</span>)</span>
<span id="cb320-9"><a href="introduction-to-the-tidyverse-ecosystem.html#cb320-9" tabindex="-1"></a>)</span>
<span id="cb320-10"><a href="introduction-to-the-tidyverse-ecosystem.html#cb320-10" tabindex="-1"></a></span>
<span id="cb320-11"><a href="introduction-to-the-tidyverse-ecosystem.html#cb320-11" tabindex="-1"></a>data </span></code></pre></div>
<pre><code>##      Name Age
## 1   Alice  25
## 2     Bob  30
## 3 Charlie  22</code></pre>
<div class="sourceCode" id="cb322"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb322-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb322-1" tabindex="-1"></a><span class="co"># Filter the data to select individuals older than 25</span></span>
<span id="cb322-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb322-2" tabindex="-1"></a>filtered_data <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span></span>
<span id="cb322-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb322-3" tabindex="-1"></a>  <span class="fu">filter</span>(Age <span class="sc">&gt;</span> <span class="dv">25</span>)</span>
<span id="cb322-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb322-4" tabindex="-1"></a></span>
<span id="cb322-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb322-5" tabindex="-1"></a><span class="co"># View the filtered data</span></span>
<span id="cb322-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb322-6" tabindex="-1"></a>filtered_data</span></code></pre></div>
<pre><code>##   Name Age
## 1  Bob  30</code></pre>
<p>In this example, we used <code>filter()</code> to select rows where the “Age” column is greater than 25.</p>
</div>
<div id="a-note-on-the-pipe" class="section level3" number="9.2.2">
<h3><span class="header-section-number">9.2.2</span> A note on the pipe</h3>
<p>You just saw the <code>%&gt;%</code> operator. It is also called a pipe.</p>
<p><code>%&gt;%</code> is from the <a href="https://cran.r-project.org/web/packages/magrittr/vignettes/magrittr.html">magrittr</a> package and when you load the <code>dplyr</code> package, the <code>%&gt;%</code> will be available for you to use.</p>
<p>The R language has a new, built-in pipe operator as of R version 4.1: <code>|&gt;</code>.</p>
<div class="sourceCode" id="cb324"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb324-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb324-1" tabindex="-1"></a>mtcars <span class="sc">%&gt;%</span></span>
<span id="cb324-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb324-2" tabindex="-1"></a>  <span class="fu">head</span>()</span></code></pre></div>
<pre><code>##                    mpg cyl disp  hp drat    wt  qsec vs am gear carb
## Mazda RX4         21.0   6  160 110 3.90 2.620 16.46  0  1    4    4
## Mazda RX4 Wag     21.0   6  160 110 3.90 2.875 17.02  0  1    4    4
## Datsun 710        22.8   4  108  93 3.85 2.320 18.61  1  1    4    1
## Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
## Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0    3    2
## Valiant           18.1   6  225 105 2.76 3.460 20.22  1  0    3    1</code></pre>
<div class="sourceCode" id="cb326"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb326-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb326-1" tabindex="-1"></a>mtcars <span class="sc">|&gt;</span></span>
<span id="cb326-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb326-2" tabindex="-1"></a>  <span class="fu">head</span>()</span></code></pre></div>
<pre><code>##                    mpg cyl disp  hp drat    wt  qsec vs am gear carb
## Mazda RX4         21.0   6  160 110 3.90 2.620 16.46  0  1    4    4
## Mazda RX4 Wag     21.0   6  160 110 3.90 2.875 17.02  0  1    4    4
## Datsun 710        22.8   4  108  93 3.85 2.320 18.61  1  1    4    1
## Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
## Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0    3    2
## Valiant           18.1   6  225 105 2.76 3.460 20.22  1  0    3    1</code></pre>
<p>Think of <code>%&gt;%</code> and <code>|&gt;</code> in R the same as <code>|</code> in the unix commands: the output of the previous command is the input of the next command.</p>
<p>The pipe operator <code>%&gt;%</code> is a very useful tool for writing efficient, easy-to-read code in R. You can use as many pipes as you want. I usually build the pipes one by one by looking at the output.</p>
<div class="sourceCode" id="cb328"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb328-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb328-1" tabindex="-1"></a>mtcars <span class="sc">%&gt;%</span> </span>
<span id="cb328-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb328-2" tabindex="-1"></a>  <span class="fu">head</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb328-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb328-3" tabindex="-1"></a>  <span class="fu">tail</span>(<span class="at">n=</span><span class="dv">3</span>)</span></code></pre></div>
<pre><code>##                    mpg cyl disp  hp drat    wt  qsec vs am gear carb
## Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
## Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0    3    2
## Valiant           18.1   6  225 105 2.76 3.460 20.22  1  0    3    1</code></pre>
<p>This means print out the first 6 rows and then take the last 3 rows.</p>
</div>
<div id="ggplot2" class="section level3" number="9.2.3">
<h3><span class="header-section-number">9.2.3</span> ggplot2</h3>
<p>Explore <code>ggplot2</code> docs <a href="https://ggplot2.tidyverse.org/index.html">here</a>.</p>
<p>The <code>ggplot2</code> package is the go-to choice for data visualization in the Tidyverse. It follows the “grammar of graphics” approach, allowing you to build complex plots layer by layer. Here’s a basic example:</p>
<div class="sourceCode" id="cb330"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb330-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb330-1" tabindex="-1"></a><span class="co"># install.packages(&quot;ggplot2&quot;)</span></span>
<span id="cb330-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb330-2" tabindex="-1"></a><span class="co"># Load the ggplot2 package</span></span>
<span id="cb330-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb330-3" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb330-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb330-4" tabindex="-1"></a></span>
<span id="cb330-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb330-5" tabindex="-1"></a><span class="co"># Create a scatter plot</span></span>
<span id="cb330-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb330-6" tabindex="-1"></a><span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> Age, <span class="at">y =</span> Name)) <span class="sc">+</span></span>
<span id="cb330-7"><a href="introduction-to-the-tidyverse-ecosystem.html#cb330-7" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-170-1.png" width="672" /></p>
<p>n this code, we used <code>ggplot()</code> to set up the plot and <code>geom_point()</code> to add scatterplot points to the canvas.</p>
</div>
<div id="tidyr" class="section level3" number="9.2.4">
<h3><span class="header-section-number">9.2.4</span> tidyr</h3>
<p>Explore tidyr docs <a href="https://tidyr.tidyverse.org/index.html">here</a>.</p>
<p>The <code>tidyr</code> package assists in reshaping data between <strong>wide</strong> and <strong>long</strong> formats, which is often necessary for analysis and visualization. Suppose you have a dataset with columns representing different years and want to convert it to a long format for easier analysis. <code>tidyr</code> can help with this transformation.</p>
<p><img src="images/wide.png" width="100%" /></p>
</div>
<div id="readr" class="section level3" number="9.2.5">
<h3><span class="header-section-number">9.2.5</span> readr</h3>
<p>Explore readr docs <a href="https://readr.tidyverse.org/index.html">here</a>.</p>
<p><code>readr</code> is a fast and efficient package for reading <strong>tabular</strong> data into R. It’s faster than the base R functions like <code>read.csv()</code>. When working with large datasets, this can significantly speed up your data loading process.</p>
<p>Tabular data are rectangular:</p>
<p><img src="images/tabular.jpeg" /></p>
</div>
<div id="tibble" class="section level3" number="9.2.6">
<h3><span class="header-section-number">9.2.6</span> tibble</h3>
<p>Explore tibble docs <a href="https://tibble.tidyverse.org/index.html">here</a>.</p>
<p>the rectangular data will be read into R as a dataframe. <code>tibble</code> provides an improved data frame structure that addresses some issues with traditional R data frames. It displays data more neatly and provides better compatibility with the Tidyverse functions.</p>
</div>
<div id="stringr" class="section level3" number="9.2.7">
<h3><span class="header-section-number">9.2.7</span> stringr</h3>
<p>Explore stringr docs <a href="https://stringr.tidyverse.org/index.html">here</a>.</p>
<p>The <code>stringr</code> package offers a range of string manipulation functions that are easier to use and more consistent than the base R functions. It’s particularly handy when dealing with text data, such as cleaning and formatting strings.</p>
</div>
<div id="forcats" class="section level3" number="9.2.8">
<h3><span class="header-section-number">9.2.8</span> forcats</h3>
<p>Explore forcats docs <a href="https://forcats.tidyverse.org/index.html">here</a>.</p>
<p><code>forcats</code> is designed for handling factor variables, which are categorical variables with predefined levels. It provides tools for changing the order of levels and combining levels when necessary.</p>
</div>
<div id="real-world-applications-1" class="section level3" number="9.2.9">
<h3><span class="header-section-number">9.2.9</span> Real-World Applications</h3>
<p>To illustrate the practical utility of the Tidyverse, consider the following scenarios:</p>
<ol style="list-style-type: decimal">
<li><p>Data Import: <code>readr</code> helps you efficiently import data from various sources, such as CSV files, Excel spreadsheets, or even web-based data.</p></li>
<li><p>Data Cleaning: When working with messy data, <code>tidyr</code> and <code>stringr</code> can assist in reshaping and cleaning data for analysis.</p></li>
<li><p>Data Exploration: When you receive a dataset, you can use <code>dplyr</code> to quickly filter, summarize, and explore the data, helping you understand its characteristics.</p></li>
<li><p>Data Visualization: <code>ggplot2</code> allows you to create stunning visualizations, making it easier to convey insights to others. For instance, you can create bar charts, scatter plots, and histograms.</p></li>
<li><p>Working with Factors: <code>forcats</code> simplifies tasks like reordering factor levels for better visual representation.</p></li>
</ol>
<p>By mastering the Tidyverse, you’ll have a powerful set of tools at your disposal for all stages of data analysis, from data preprocessing to visualization and modeling. Happy coding!</p>
</div>
</div>
<div id="tibble-and-readr---modern-data-structures-in-r" class="section level2" number="9.3">
<h2><span class="header-section-number">9.3</span> Tibble and Readr - Modern Data Structures in R</h2>
<p>In this lesson, we will explore two essential concepts in data manipulation with R: tibbles and the readr package. These tools are designed to make working with data more efficient and user-friendly, especially when dealing with large datasets. We’ll see why tibbles are preferred over traditional data frames and how to read and manipulate data using the readr package.</p>
<div id="tibbles-a-better-way-to-store-data" class="section level3" number="9.3.1">
<h3><span class="header-section-number">9.3.1</span> Tibbles: A Better Way to Store Data</h3>
<p>Traditional Data Frames vs. Tibbles:
<img src="images/tibble.png" /></p>
<p>In R, data frames are widely used for storing and manipulating data. However, tibbles offer several advantages over traditional data frames:</p>
<ol style="list-style-type: decimal">
<li><p>Cleaner Printing: Tibbles offer a more organized way to present your data. When you print a tibble, it only displays the first 10 rows and as many columns as can fit on your screen. This ensures a concise overview of your data, which is particularly helpful for large datasets. In contrast, traditional data frames tend to print all rows and columns by default, leading to overwhelming output.</p></li>
<li><p>Structured Indexing: When you subset a tibble using [ ], it returns another tibble that retains the original data’s structure. This means you still have clear column names and data types. In contrast, data frames return vectors, which provide less information about the structure of the data subset. You need to specify drop = FALSE to retain as a data frame and we see examples in previous lectures.</p></li>
<li><p>Preservation of Variable Types: Tibbles respect the original data types of your variables, even after subsetting. This is important because it ensures that your data remains consistent. In some cases, data frames may convert variables to factors or matrices when you subset them, potentially causing unexpected issues.</p></li>
<li><p>String Handling: In tibbles, character vectors remain as characters, maintaining the integrity of your data. However, data frames may automatically convert character vectors to factors by default. This behavior can lead to unexpected changes in your data and, subsequently, your analysis.</p></li>
<li><p>List-Columns: One unique feature of tibbles is their ability to directly contain list-columns. List-columns allow you to store lists within your tibble, providing a convenient way to work with complex data structures. Data frames do not offer this feature, which can limit your ability to represent certain types of data effectively.</p></li>
<li><p>Enhanced Output: Tibbles enhance the print display by showing data types, highlighting missing values, and truncating output for better readability. This additional information helps you understand your data at a glance, making it easier to spot potential issues or trends.</p></li>
</ol>
</div>
<div id="rownames-and-tibbles" class="section level3" number="9.3.2">
<h3><span class="header-section-number">9.3.2</span> Rownames and Tibbles</h3>
<p>In regular data frames, you can assign and work with rownames, which are helpful for labeling and referencing rows. However, tibbles do not support rownames, so you want to add another column that contains the row ids.</p>
</div>
<div id="reading-data-with-readr" class="section level3" number="9.3.3">
<h3><span class="header-section-number">9.3.3</span> Reading Data with readr</h3>
<p>To work with data effectively, you first need to import it into R. The <code>readr</code> package provides a powerful toolset for reading various data file formats. Let’s see how to read data using readr.</p>
<div id="reading-csv-data" class="section level4" number="9.3.3.1">
<h4><span class="header-section-number">9.3.3.1</span> Reading CSV Data</h4>
<p>Download the TCGA gene expression data (CSV file) to your working directory or specify the correct file path.</p>
<blockquote>
<p>Download the file at <a href="https://osf.io/yeun5" class="uri">https://osf.io/yeun5</a></p>
</blockquote>
<p>Click the <code>Download</code>:</p>
<p><img src="images/osf.png" /></p>
<p>Suppose we have a CSV file named “TCGA_cancer_genes_expression.csv” in the Downloads folder, to read it using <code>readr::read_csv</code>, follow these steps:</p>
<div class="sourceCode" id="cb331"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb331-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb331-1" tabindex="-1"></a><span class="co"># Load the readr package (if not already loaded)</span></span>
<span id="cb331-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb331-2" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb331-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb331-3" tabindex="-1"></a></span>
<span id="cb331-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb331-4" tabindex="-1"></a><span class="co"># Read the CSV data into a tibble</span></span>
<span id="cb331-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb331-5" tabindex="-1"></a>tcga_data <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="st">&quot;~/Downloads/TCGA_cancer_genes_expression.csv&quot;</span>)</span>
<span id="cb331-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb331-6" tabindex="-1"></a></span>
<span id="cb331-7"><a href="introduction-to-the-tidyverse-ecosystem.html#cb331-7" tabindex="-1"></a><span class="co"># Display the first few rows of the tibble</span></span>
<span id="cb331-8"><a href="introduction-to-the-tidyverse-ecosystem.html#cb331-8" tabindex="-1"></a><span class="fu">head</span>(tcga_data)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 25
##   ...1      TACSTD2  VTCN1   MUC1 NECTIN4 FOLH1  FOLR1 CD276   MSLN  CLDN6 ERBB2
##   &lt;chr&gt;       &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;
## 1 43e715bf…   0.704 0      0.675   0.0862  7.21 0       52.8 0.0667 0.0970  1.88
## 2 1a5db9fc…  25.4   0      2.02    0.0728 23.6  0.122   78.8 0.956  0.255   7.78
## 3 93b382e4…   1.58  0      0.908   0.699   2.85 1.01   146.  0.0456 0.257   2.91
## 4 1f39dadd…   0.270 0.0910 0.0429  0.0165  1.16 0.279   48.5 0.0315 0.247   4.91
## 5 8c8c09b9…   0.412 0      0.115   0.0317  2.41 0.0492  42.3 0.270  0.126   1.49
## 6 85a86b91…   4.55  4.86   0.0421  0.0683  1.01 0.0225  20.6 0.0134 0.0182 13.5 
## # ℹ 14 more variables: MUC16 &lt;dbl&gt;, DLL3 &lt;dbl&gt;, CEACAM5 &lt;dbl&gt;, PVR &lt;dbl&gt;,
## #   EPCAM &lt;dbl&gt;, PROM1 &lt;dbl&gt;, CD24 &lt;dbl&gt;, EGFR &lt;dbl&gt;, MET &lt;dbl&gt;,
## #   TNFRSF10B &lt;dbl&gt;, tcga.tcga_barcode &lt;chr&gt;,
## #   tcga.cgc_sample_sample_type &lt;chr&gt;, study &lt;chr&gt;, sample_type &lt;chr&gt;</code></pre>
<p>Here, we load the <code>readr</code> package, use <code>read_csv</code> to read the CSV file, and store the data in the <code>tcga_data</code> tibble. Finally, we use <code>head</code> to display the first few rows of the tibble.</p>
<p>note that the first column’s name changes to “…1,” which is a default behavior when column names are missing in the source data.</p>
</div>
<div id="understanding-the-output" class="section level4" number="9.3.3.2">
<h4><span class="header-section-number">9.3.3.2</span> Understanding the Output</h4>
<p>The output shows the data in a neat tabular format. Column names are displayed at the top, followed by rows of data. Each column’s data type is indicated (e.g., <code>dbl</code> for double-precision floating point numbers or <code>chr</code> for character).</p>
</div>
</div>
<div id="using-tibbles-with-imported-data" class="section level3" number="9.3.4">
<h3><span class="header-section-number">9.3.4</span> Using Tibbles with Imported Data</h3>
<p>When working with tibbles, imported data maintains its format and advantages over data frames. Here’s how you can convert the imported data into a tibble:</p>
<div class="sourceCode" id="cb333"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb333-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb333-1" tabindex="-1"></a><span class="co"># Load the dplyr package for tibble conversion</span></span>
<span id="cb333-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb333-2" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb333-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb333-3" tabindex="-1"></a></span>
<span id="cb333-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb333-4" tabindex="-1"></a><span class="co"># read in the data using built-in read.csv</span></span>
<span id="cb333-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb333-5" tabindex="-1"></a>tcga_data<span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;~/Downloads/TCGA_cancer_genes_expression.csv&quot;</span>)</span>
<span id="cb333-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb333-6" tabindex="-1"></a></span>
<span id="cb333-7"><a href="introduction-to-the-tidyverse-ecosystem.html#cb333-7" tabindex="-1"></a><span class="co"># regular dataframe, and the first column name is X if it is empty</span></span>
<span id="cb333-8"><a href="introduction-to-the-tidyverse-ecosystem.html#cb333-8" tabindex="-1"></a><span class="fu">head</span>(tcga_data)</span></code></pre></div>
<pre><code>##                                      X    TACSTD2      VTCN1       MUC1
## 1 43e715bf-28d9-4b5e-b762-8cd1b69a430e  0.7035937 0.00000000 0.67502205
## 2 1a5db9fc-2abd-4e1b-b5ef-b1cf5e5f3872 25.4360736 0.00000000 2.01525394
## 3 93b382e4-9c9a-43f5-bd3b-502cc260b886  1.5756197 0.00000000 0.90784666
## 4 1f39dadd-3655-474e-ba4c-a5bd32c97a8b  0.2702156 0.09099681 0.04293345
## 5 8c8c09b9-ec83-45ec-bc4c-0ba92de60acb  0.4122814 0.00000000 0.11484380
## 6 85a86b91-4f24-4e77-ae2d-520f8e205efc  4.5469193 4.85973690 0.04208195
##      NECTIN4     FOLH1      FOLR1     CD276       MSLN      CLDN6     ERBB2
## 1 0.08620727  7.213342 0.00000000  52.75981 0.06674445 0.09704962  1.879518
## 2 0.07279804 23.552286 0.12154673  78.78551 0.95554610 0.25458796  7.777976
## 3 0.69905270  2.853812 1.01000271 145.84399 0.04563568 0.25701910  2.905926
## 4 0.01652257  1.157070 0.27942068  48.45022 0.03154912 0.24746913  4.914280
## 5 0.03168398  2.408137 0.04922458  42.25592 0.26968788 0.12576720  1.494744
## 6 0.06828305  1.010411 0.02248965  20.63795 0.01336404 0.01823883 13.474689
##          MUC16       DLL3 CEACAM5      PVR     EPCAM       PROM1       CD24
## 1 0.0011479879 0.49589978       0 52.08113  4.521984 0.025311008 0.55036003
## 2 0.0008049670 2.52244014       0 40.87926  9.530414 0.023576862 9.67272890
## 3 0.0026190288 0.77074712       0 33.26727 42.358567 0.000000000 0.06939934
## 4 0.0051705741 0.10636402       0 28.26457 16.316524 0.007783431 0.84522244
## 5 0.0004894306 0.04483123       0 41.66776 12.529742 0.019204339 0.21369023
## 6 0.0000000000 0.01184285       0 30.18711  2.430109 0.043719865 4.95506593
##        EGFR        MET TNFRSF10B            tcga.tcga_barcode
## 1  1.286481  0.9320235  12.80547 TCGA-OR-A5KU-01A-11R-A29S-07
## 2  5.373307  8.0610999  31.46289 TCGA-P6-A5OG-01A-22R-A29S-07
## 3  4.600918  0.1295387  65.57967 TCGA-OR-A5K5-01A-11R-A29S-07
## 4  3.010374  2.9728030  24.31636 TCGA-OR-A5K4-01A-11R-A29S-07
## 5 16.476552 19.7360055  21.11014 TCGA-OR-A5LP-01A-11R-A29S-07
## 6  2.010338  8.6087283  37.91574 TCGA-PK-A5H9-01A-11R-A29S-07
##   tcga.cgc_sample_sample_type study sample_type
## 1               Primary Tumor   ACC      cancer
## 2               Primary Tumor   ACC      cancer
## 3               Primary Tumor   ACC      cancer
## 4               Primary Tumor   ACC      cancer
## 5               Primary Tumor   ACC      cancer
## 6               Primary Tumor   ACC      cancer</code></pre>
<div class="sourceCode" id="cb335"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb335-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb335-1" tabindex="-1"></a><span class="co"># Convert the data to a tibble</span></span>
<span id="cb335-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb335-2" tabindex="-1"></a>tcga_data <span class="ot">&lt;-</span> tcga_data <span class="sc">%&gt;%</span> </span>
<span id="cb335-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb335-3" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">as_tibble</span>()</span>
<span id="cb335-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb335-4" tabindex="-1"></a></span>
<span id="cb335-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb335-5" tabindex="-1"></a><span class="co"># Display the first few rows of the tibble</span></span>
<span id="cb335-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb335-6" tabindex="-1"></a><span class="fu">head</span>(tcga_data)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 25
##   X         TACSTD2  VTCN1   MUC1 NECTIN4 FOLH1  FOLR1 CD276   MSLN  CLDN6 ERBB2
##   &lt;chr&gt;       &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;
## 1 43e715bf…   0.704 0      0.675   0.0862  7.21 0       52.8 0.0667 0.0970  1.88
## 2 1a5db9fc…  25.4   0      2.02    0.0728 23.6  0.122   78.8 0.956  0.255   7.78
## 3 93b382e4…   1.58  0      0.908   0.699   2.85 1.01   146.  0.0456 0.257   2.91
## 4 1f39dadd…   0.270 0.0910 0.0429  0.0165  1.16 0.279   48.5 0.0315 0.247   4.91
## 5 8c8c09b9…   0.412 0      0.115   0.0317  2.41 0.0492  42.3 0.270  0.126   1.49
## 6 85a86b91…   4.55  4.86   0.0421  0.0683  1.01 0.0225  20.6 0.0134 0.0182 13.5 
## # ℹ 14 more variables: MUC16 &lt;dbl&gt;, DLL3 &lt;dbl&gt;, CEACAM5 &lt;dbl&gt;, PVR &lt;dbl&gt;,
## #   EPCAM &lt;dbl&gt;, PROM1 &lt;dbl&gt;, CD24 &lt;dbl&gt;, EGFR &lt;dbl&gt;, MET &lt;dbl&gt;,
## #   TNFRSF10B &lt;dbl&gt;, tcga.tcga_barcode &lt;chr&gt;,
## #   tcga.cgc_sample_sample_type &lt;chr&gt;, study &lt;chr&gt;, sample_type &lt;chr&gt;</code></pre>
<p>By using the <code>%&gt;%</code> pipe operator and <code>tibble::as_tibble()</code>, we convert the data into a tibble while preserving its structure and advantages.</p>
<p>The output remains in a similar format as before, with clear column names and data types.</p>
</div>
<div id="note-on-reading-excel-files" class="section level3" number="9.3.5">
<h3><span class="header-section-number">9.3.5</span> Note on reading Excel files</h3>
<p>We still deal with a lot of spreadsheets, to read in Excel files, take a look at the <a href="https://readxl.tidyverse.org/"><code>readxl</code></a> package.</p>
</div>
<div id="conclusion-18" class="section level3" number="9.3.6">
<h3><span class="header-section-number">9.3.6</span> Conclusion</h3>
<p>In this lesson, we learned about <code>tibbles</code>, a modern data structure in R that offers several advantages over traditional data frames. We also explored how to read and manipulate data using the readr package, converting imported data into tibbles for efficient data analysis. Tibbles and readr are valuable tools for data scientists and analysts working with R, making data manipulation and exploration more user-friendly and efficient.</p>
</div>
</div>
<div id="the-tidy-data-format" class="section level2" number="9.4">
<h2><span class="header-section-number">9.4</span> The tidy data format</h2>
<p>In this lesson, we will delve into the essential data cleaning and tidying operations in R using the powerful dplyr package. Tidying data is a crucial step in data analysis, especially when dealing with real-world data that can be messy and inconsistent. We will explore the concept of tidy data and learn about the distinction between long and wide dataset formats.</p>
<div id="the-importance-of-tidy-data" class="section level3" number="9.4.1">
<h3><span class="header-section-number">9.4.1</span> The Importance of Tidy Data</h3>
<p>Real-world data is often messy, scattered across various files, missing metadata, and inputted in inconsistent formats. Before we can effectively analyze this data, we need to bring it all together into one structured table and resolve any issues. This process of data ingestion and standardization is known as data tidying.</p>
<p><img src="images/tidy-data.png" /></p>
<p>The tidyverse packages, including <code>dplyr</code>, work seamlessly with tidy data. As defined by Hadley Wickham in <a href="https://r4ds.had.co.nz/">R for Data Science</a>, tidy data adheres to three interrelated rules:</p>
<ol style="list-style-type: decimal">
<li><p>Each variable must have its own column.</p></li>
<li><p>Each observation must have its own row.</p></li>
<li><p>Each value must have its own cell.</p></li>
</ol>
<p>Why should we make our data tidy? There are both general and specific advantages:</p>
<ul>
<li><p>General Advantage: Using a consistent data structure makes it easier to learn and work with data manipulation tools because they follow a uniform pattern.</p></li>
<li><p>Specific Advantage: Placing variables in columns allows R’s vectorized nature to shine. Most built-in R functions work efficiently with vectors of values. This natural alignment between tidy data and R’s capabilities makes data transformation feel particularly intuitive.</p></li>
</ul>
</div>
<div id="long-format-vs.-wide-format" class="section level3" number="9.4.2">
<h3><span class="header-section-number">9.4.2</span> Long Format vs. Wide Format</h3>
<p>Before we dive into the practical aspects of tidying data, let’s understand the difference between long format and wide format for a dataframe. We’ll use an example with the dplyr package:</p>
<div class="sourceCode" id="cb337"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb337-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb337-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb337-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb337-2" tabindex="-1"></a></span>
<span id="cb337-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb337-3" tabindex="-1"></a><span class="fu">head</span>(table4a)</span></code></pre></div>
<pre><code>## # A tibble: 3 × 3
##   country     `1999` `2000`
##   &lt;chr&gt;        &lt;dbl&gt;  &lt;dbl&gt;
## 1 Afghanistan    745   2666
## 2 Brazil       37737  80488
## 3 China       212258 213766</code></pre>
<p>The <code>table4a</code> dataframe is in the wide format, which is a common way to enter data into software like Excel. In this format, each column often represents a year, and the values are spread across columns for each country.</p>
<p>To convert this data into the long format, we’ll use <code>tidyr::pivot_longer()</code>. This function reshapes the data, making it easier to work with:</p>
<div class="sourceCode" id="cb339"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb339-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb339-1" tabindex="-1"></a>table4a <span class="sc">%&gt;%</span> </span>
<span id="cb339-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb339-2" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="fu">c</span>(<span class="st">`</span><span class="at">1999</span><span class="st">`</span>, <span class="st">`</span><span class="at">2000</span><span class="st">`</span>), <span class="at">names_to =</span> <span class="st">&quot;year&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;cases&quot;</span>)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 3
##   country     year   cases
##   &lt;chr&gt;       &lt;chr&gt;  &lt;dbl&gt;
## 1 Afghanistan 1999     745
## 2 Afghanistan 2000    2666
## 3 Brazil      1999   37737
## 4 Brazil      2000   80488
## 5 China       1999  212258
## 6 China       2000  213766</code></pre>
<p>Now, the data fulfills the three rules for tidy data:</p>
<ol style="list-style-type: decimal">
<li><p>Each variable (country and year) has its own column.</p></li>
<li><p>Each observation (combinations of country and year) has its own row.</p></li>
<li><p>Each value (the number of cases) has its own cell.</p></li>
</ol>
<p>Understanding the difference between long and wide formats is essential because it determines how we structure our data for analysis. Once you grasp these concepts, reshaping and tidying data becomes a more comfortable and intuitive process.</p>
<p>Take your time to study and understand these data formats. It will significantly boost your confidence in reshaping and tidying data for your analytical tasks.</p>
</div>
</div>
<div id="introducing-dplyr-your-data-wrangling-toolkit" class="section level2" number="9.5">
<h2><span class="header-section-number">9.5</span> Introducing dplyr: Your Data Wrangling Toolkit</h2>
<p>In this lesson, we’ll delve into the world of the dplyr package in R, which offers a powerful and concise set of tools for working with data frames. As a biology student new to programming, dplyr will be your trusty companion for effortlessly managing, filtering, and summarizing biological data.</p>
<p>To master dplyr, it’s crucial to grasp its five core functions: <code>mutate()</code>, <code>select()</code>, <code>filter()</code>, <code>summarise()</code>, and <code>arrange()</code>. Each function serves a specific purpose in data manipulation.</p>
<div id="selecting-specific-columns" class="section level3" number="9.5.1">
<h3><span class="header-section-number">9.5.1</span> Selecting Specific Columns</h3>
<p>We will be working with the same tibble called tcga_data that has numerous columns, but you’re interested in extracting the “EPCAM” column along with other metadata columns. <code>EPCAM</code>, short for epithelial cellular adhesion molecule, is highly relevant in the context of epithelial cancer cells.</p>
<div class="sourceCode" id="cb341"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb341-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb341-1" tabindex="-1"></a>tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb341-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb341-2" tabindex="-1"></a>  <span class="fu">select</span>(EPCAM, tcga.tcga_barcode<span class="sc">:</span>sample_type)</span></code></pre></div>
<pre><code>## # A tibble: 11,348 × 5
##    EPCAM tcga.tcga_barcode            tcga.cgc_sample_sample…¹ study sample_type
##    &lt;dbl&gt; &lt;chr&gt;                        &lt;chr&gt;                    &lt;chr&gt; &lt;chr&gt;      
##  1  4.52 TCGA-OR-A5KU-01A-11R-A29S-07 Primary Tumor            ACC   cancer     
##  2  9.53 TCGA-P6-A5OG-01A-22R-A29S-07 Primary Tumor            ACC   cancer     
##  3 42.4  TCGA-OR-A5K5-01A-11R-A29S-07 Primary Tumor            ACC   cancer     
##  4 16.3  TCGA-OR-A5K4-01A-11R-A29S-07 Primary Tumor            ACC   cancer     
##  5 12.5  TCGA-OR-A5LP-01A-11R-A29S-07 Primary Tumor            ACC   cancer     
##  6  2.43 TCGA-PK-A5H9-01A-11R-A29S-07 Primary Tumor            ACC   cancer     
##  7  3.74 TCGA-OR-A5LD-01A-11R-A29S-07 Primary Tumor            ACC   cancer     
##  8  4.08 TCGA-OR-A5JX-01A-11R-A29S-07 Primary Tumor            ACC   cancer     
##  9  2.84 TCGA-PK-A5H8-01A-11R-A29S-07 Primary Tumor            ACC   cancer     
## 10  3.61 TCGA-OR-A5J3-01A-11R-A29S-07 Primary Tumor            ACC   cancer     
## # ℹ 11,338 more rows
## # ℹ abbreviated name: ¹​tcga.cgc_sample_sample_type</code></pre>
<p>This code utilizes the <code>select()</code> function to pick specific columns by their names, creating a tidy subset of your data. We used <code>:</code> to select a range of columns. Note that the columns are reordered
putting <code>EPCAM</code> to the first column.</p>
<p>You can also use index to select the columns</p>
<div class="sourceCode" id="cb343"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb343-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb343-1" tabindex="-1"></a><span class="fu">colnames</span>(tcga_data)</span></code></pre></div>
<pre><code>##  [1] &quot;X&quot;                           &quot;TACSTD2&quot;                    
##  [3] &quot;VTCN1&quot;                       &quot;MUC1&quot;                       
##  [5] &quot;NECTIN4&quot;                     &quot;FOLH1&quot;                      
##  [7] &quot;FOLR1&quot;                       &quot;CD276&quot;                      
##  [9] &quot;MSLN&quot;                        &quot;CLDN6&quot;                      
## [11] &quot;ERBB2&quot;                       &quot;MUC16&quot;                      
## [13] &quot;DLL3&quot;                        &quot;CEACAM5&quot;                    
## [15] &quot;PVR&quot;                         &quot;EPCAM&quot;                      
## [17] &quot;PROM1&quot;                       &quot;CD24&quot;                       
## [19] &quot;EGFR&quot;                        &quot;MET&quot;                        
## [21] &quot;TNFRSF10B&quot;                   &quot;tcga.tcga_barcode&quot;          
## [23] &quot;tcga.cgc_sample_sample_type&quot; &quot;study&quot;                      
## [25] &quot;sample_type&quot;</code></pre>
<p>The <code>EPCAM</code> column is the 16th column; <code>tcga.tcga_barcode</code> is the 23rd column;
<code>sample_type</code> is 25th column:</p>
<div class="sourceCode" id="cb345"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb345-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb345-1" tabindex="-1"></a>tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb345-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb345-2" tabindex="-1"></a>  <span class="fu">select</span>(<span class="dv">16</span>, <span class="dv">23</span><span class="sc">:</span><span class="dv">25</span>)</span></code></pre></div>
<pre><code>## # A tibble: 11,348 × 4
##    EPCAM tcga.cgc_sample_sample_type study sample_type
##    &lt;dbl&gt; &lt;chr&gt;                       &lt;chr&gt; &lt;chr&gt;      
##  1  4.52 Primary Tumor               ACC   cancer     
##  2  9.53 Primary Tumor               ACC   cancer     
##  3 42.4  Primary Tumor               ACC   cancer     
##  4 16.3  Primary Tumor               ACC   cancer     
##  5 12.5  Primary Tumor               ACC   cancer     
##  6  2.43 Primary Tumor               ACC   cancer     
##  7  3.74 Primary Tumor               ACC   cancer     
##  8  4.08 Primary Tumor               ACC   cancer     
##  9  2.84 Primary Tumor               ACC   cancer     
## 10  3.61 Primary Tumor               ACC   cancer     
## # ℹ 11,338 more rows</code></pre>
<p>We can even mix the index with column names:</p>
<div class="sourceCode" id="cb347"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb347-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb347-1" tabindex="-1"></a>tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb347-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb347-2" tabindex="-1"></a>  <span class="fu">select</span>(EPCAM, <span class="dv">23</span><span class="sc">:</span><span class="dv">25</span>)</span></code></pre></div>
<pre><code>## # A tibble: 11,348 × 4
##    EPCAM tcga.cgc_sample_sample_type study sample_type
##    &lt;dbl&gt; &lt;chr&gt;                       &lt;chr&gt; &lt;chr&gt;      
##  1  4.52 Primary Tumor               ACC   cancer     
##  2  9.53 Primary Tumor               ACC   cancer     
##  3 42.4  Primary Tumor               ACC   cancer     
##  4 16.3  Primary Tumor               ACC   cancer     
##  5 12.5  Primary Tumor               ACC   cancer     
##  6  2.43 Primary Tumor               ACC   cancer     
##  7  3.74 Primary Tumor               ACC   cancer     
##  8  4.08 Primary Tumor               ACC   cancer     
##  9  2.84 Primary Tumor               ACC   cancer     
## 10  3.61 Primary Tumor               ACC   cancer     
## # ℹ 11,338 more rows</code></pre>
<div class="sourceCode" id="cb349"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb349-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb349-1" tabindex="-1"></a><span class="co"># save it to a new variable</span></span>
<span id="cb349-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb349-2" tabindex="-1"></a></span>
<span id="cb349-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb349-3" tabindex="-1"></a>tcga_data<span class="ot">&lt;-</span> tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb349-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb349-4" tabindex="-1"></a>  <span class="fu">select</span>(EPCAM, <span class="dv">23</span><span class="sc">:</span><span class="dv">25</span>)</span></code></pre></div>
</div>
<div id="adding-new-columns" class="section level3" number="9.5.2">
<h3><span class="header-section-number">9.5.2</span> Adding New Columns</h3>
<p>The <code>mutate()</code> function allows you to introduce new variables that are derived from existing ones. For instance, you can create a new column, “log2EPCAM,” containing the logarithm base-2 of the EPCAM values:</p>
<div class="sourceCode" id="cb350"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb350-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb350-1" tabindex="-1"></a>tcga_data<span class="ot">&lt;-</span> tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb350-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb350-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">log2EPCAM =</span> <span class="fu">log2</span>(EPCAM))</span>
<span id="cb350-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb350-3" tabindex="-1"></a></span>
<span id="cb350-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb350-4" tabindex="-1"></a>tcga_data</span></code></pre></div>
<pre><code>## # A tibble: 11,348 × 5
##    EPCAM tcga.cgc_sample_sample_type study sample_type log2EPCAM
##    &lt;dbl&gt; &lt;chr&gt;                       &lt;chr&gt; &lt;chr&gt;           &lt;dbl&gt;
##  1  4.52 Primary Tumor               ACC   cancer           2.18
##  2  9.53 Primary Tumor               ACC   cancer           3.25
##  3 42.4  Primary Tumor               ACC   cancer           5.40
##  4 16.3  Primary Tumor               ACC   cancer           4.03
##  5 12.5  Primary Tumor               ACC   cancer           3.65
##  6  2.43 Primary Tumor               ACC   cancer           1.28
##  7  3.74 Primary Tumor               ACC   cancer           1.90
##  8  4.08 Primary Tumor               ACC   cancer           2.03
##  9  2.84 Primary Tumor               ACC   cancer           1.51
## 10  3.61 Primary Tumor               ACC   cancer           1.85
## # ℹ 11,338 more rows</code></pre>
</div>
<div id="reordering-columns" class="section level3" number="9.5.3">
<h3><span class="header-section-number">9.5.3</span> Reordering Columns</h3>
<p>If you want to rearrange the order of columns, you can again use <code>select()</code>. Here, we move the “log2EPCAM” column to the front while keeping all other columns intact:</p>
<div class="sourceCode" id="cb352"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb352-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb352-1" tabindex="-1"></a>tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb352-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb352-2" tabindex="-1"></a>  <span class="fu">select</span>(EPCAM, log2EPCAM, <span class="fu">everything</span>())</span></code></pre></div>
<pre><code>## # A tibble: 11,348 × 5
##    EPCAM log2EPCAM tcga.cgc_sample_sample_type study sample_type
##    &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;                       &lt;chr&gt; &lt;chr&gt;      
##  1  4.52      2.18 Primary Tumor               ACC   cancer     
##  2  9.53      3.25 Primary Tumor               ACC   cancer     
##  3 42.4       5.40 Primary Tumor               ACC   cancer     
##  4 16.3       4.03 Primary Tumor               ACC   cancer     
##  5 12.5       3.65 Primary Tumor               ACC   cancer     
##  6  2.43      1.28 Primary Tumor               ACC   cancer     
##  7  3.74      1.90 Primary Tumor               ACC   cancer     
##  8  4.08      2.03 Primary Tumor               ACC   cancer     
##  9  2.84      1.51 Primary Tumor               ACC   cancer     
## 10  3.61      1.85 Primary Tumor               ACC   cancer     
## # ℹ 11,338 more rows</code></pre>
<p>The <code>everything()</code> helper function denotes all other columns, ensuring your numeric columns are at the front.</p>
</div>
<div id="filtering-data" class="section level3" number="9.5.4">
<h3><span class="header-section-number">9.5.4</span> Filtering Data</h3>
<p><code>filter()</code> is your go-to function for extracting observations that meet specific criteria. To isolate data only related to glioblastoma (GBM), you can apply the following filter:</p>
<div class="sourceCode" id="cb354"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb354-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb354-1" tabindex="-1"></a>tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb354-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb354-2" tabindex="-1"></a>  <span class="fu">filter</span>(study <span class="sc">==</span> <span class="st">&quot;GBM&quot;</span>)</span></code></pre></div>
<pre><code>## # A tibble: 175 × 5
##     EPCAM tcga.cgc_sample_sample_type study sample_type log2EPCAM
##     &lt;dbl&gt; &lt;chr&gt;                       &lt;chr&gt; &lt;chr&gt;           &lt;dbl&gt;
##  1 0.329  Primary Tumor               GBM   cancer         -1.60 
##  2 0.152  Primary Tumor               GBM   cancer         -2.72 
##  3 0.0814 Primary Tumor               GBM   cancer         -3.62 
##  4 0.367  Recurrent Tumor             GBM   cancer         -1.45 
##  5 0.0614 Primary Tumor               GBM   cancer         -4.02 
##  6 0.350  Primary Tumor               GBM   cancer         -1.51 
##  7 0.165  Primary Tumor               GBM   cancer         -2.60 
##  8 0.0989 Primary Tumor               GBM   cancer         -3.34 
##  9 0.466  Primary Tumor               GBM   cancer         -1.10 
## 10 0.707  Recurrent Tumor             GBM   cancer         -0.500
## # ℹ 165 more rows</code></pre>
<p>This code snippet retains only the rows corresponding to GBM in your dataset.</p>
</div>
<div id="summarizing-data" class="section level3" number="9.5.5">
<h3><span class="header-section-number">9.5.5</span> Summarizing Data</h3>
<p>Suppose you want to calculate the average EPCAM expression for each cancer type in your dataset. You can utilize <code>summarise()</code> in conjunction with <code>group_by()</code>:</p>
<div class="sourceCode" id="cb356"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb356-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb356-1" tabindex="-1"></a>tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb356-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb356-2" tabindex="-1"></a>  <span class="fu">group_by</span>(study) <span class="sc">%&gt;%</span></span>
<span id="cb356-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb356-3" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">average_EPCAM =</span> <span class="fu">mean</span>(EPCAM))</span></code></pre></div>
<pre><code>## # A tibble: 33 × 2
##    study average_EPCAM
##    &lt;chr&gt;         &lt;dbl&gt;
##  1 ACC          11.2  
##  2 BLCA        102.   
##  3 BRCA        177.   
##  4 CESC        166.   
##  5 CHOL        223.   
##  6 COAD        792.   
##  7 DLBC          2.61 
##  8 ESCA        273.   
##  9 GBM           0.623
## 10 HNSC         50.3  
## # ℹ 23 more rows</code></pre>
<p>Here, the data is grouped by the “study” variable, and the <code>summarise()</code> function calculates the mean <code>EPCAM</code> value for each group.</p>
</div>
<div id="sorting-data" class="section level3" number="9.5.6">
<h3><span class="header-section-number">9.5.6</span> Sorting Data</h3>
<p>To sort your data frame based on a specific column, employ <code>arrange()</code>. For instance, you can order your dataset by the median level of EPCAM:</p>
<div class="sourceCode" id="cb358"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb358-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb358-1" tabindex="-1"></a>tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb358-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb358-2" tabindex="-1"></a>  <span class="fu">group_by</span>(study) <span class="sc">%&gt;%</span></span>
<span id="cb358-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb358-3" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">average_EPCAM =</span> <span class="fu">mean</span>(EPCAM),</span>
<span id="cb358-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb358-4" tabindex="-1"></a>            <span class="at">median_EPCAM =</span> <span class="fu">median</span>(EPCAM)) <span class="sc">%&gt;%</span></span>
<span id="cb358-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb358-5" tabindex="-1"></a>  <span class="fu">arrange</span>(median_EPCAM)</span></code></pre></div>
<pre><code>## # A tibble: 33 × 3
##    study average_EPCAM median_EPCAM
##    &lt;chr&gt;         &lt;dbl&gt;        &lt;dbl&gt;
##  1 UVM           0.765       0.0583
##  2 SKCM          0.980       0.133 
##  3 SARC          2.87        0.224 
##  4 GBM           0.623       0.323 
##  5 DLBC          2.61        0.578 
##  6 LAML          3.14        0.595 
##  7 LGG           0.906       0.681 
##  8 MESO         11.3         1.14  
##  9 LIHC         31.2         1.22  
## 10 ACC          11.2         4.83  
## # ℹ 23 more rows</code></pre>
<p>The default is arrange from the smallest to the biggest. Let’s reverse it by using the helper descending function <code>desc</code>:</p>
<div class="sourceCode" id="cb360"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb360-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb360-1" tabindex="-1"></a>tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb360-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb360-2" tabindex="-1"></a>  <span class="fu">group_by</span>(study) <span class="sc">%&gt;%</span></span>
<span id="cb360-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb360-3" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">average_EPCAM =</span> <span class="fu">mean</span>(EPCAM),</span>
<span id="cb360-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb360-4" tabindex="-1"></a>            <span class="at">median_EPCAM =</span> <span class="fu">median</span>(EPCAM)) <span class="sc">%&gt;%</span></span>
<span id="cb360-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb360-5" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(median_EPCAM))</span></code></pre></div>
<pre><code>## # A tibble: 33 × 3
##    study average_EPCAM median_EPCAM
##    &lt;chr&gt;         &lt;dbl&gt;        &lt;dbl&gt;
##  1 READ           834.         808.
##  2 COAD           792.         787.
##  3 THCA           350.         345.
##  4 UCEC           351.         337.
##  5 STAD           335.         306.
##  6 LUAD           307.         289.
##  7 PAAD           278.         265.
##  8 CHOL           223.         236.
##  9 OV             228.         207.
## 10 PRAD           199.         182.
## # ℹ 23 more rows</code></pre>
<p>We see <code>READ</code> and <code>COAD</code> colon cancers have the highest <code>EPCAM</code> expression.</p>
<p>This code sorts the dataset from the smallest to the largest median EPCAM values.</p>
</div>
<div id="conclusion-19" class="section level3" number="9.5.7">
<h3><span class="header-section-number">9.5.7</span> Conclusion</h3>
<p>In summary:</p>
<ul>
<li><p>mutate() adds new columns.</p></li>
<li><p>filter() extracts specific observations.</p></li>
<li><p>select() picks/reorders columns.</p></li>
<li><p>summarise() reduces multiple values to summaries.</p></li>
<li><p>arrange() reorders rows.</p></li>
</ul>
<p>These four fundamental functions empower you to efficiently manipulate, analyze, and summarize biological data frames, providing a more concise and readable approach compared to traditional R methods. As your programming skills grow, <code>dplyr</code> will remain an indispensable tool in your data science toolkit.</p>
</div>
</div>
<div id="stringr-your-essential-toolkit-to-manipulate-strings" class="section level2" number="9.6">
<h2><span class="header-section-number">9.6</span> stringr: your essential toolkit to manipulate strings</h2>
<p>xxxx</p>
</div>
<div id="purrr-ditch-your-for-loops" class="section level2" number="9.7">
<h2><span class="header-section-number">9.7</span> purrr: ditch your for loops</h2>
<p>In this lesson, we’ll learn about the <code>purrr</code> package in R. <code>Purrr</code> provides a set of tools for working with lists and other recursive data structures in a functional programming style.</p>
<blockquote>
<p>A recursive data structure in R refers to a data object that can contain other objects of the same type as its components. For example, a list in R can be recursive because it can contain other lists within it, creating a nested or hierarchical structure.</p>
</blockquote>
<p>As a biology student, you’ll likely need to apply the same operation to multiple data sets or columns. That’s where <code>purrr</code> becomes really handy! The key functions we’ll cover are:</p>
<ul>
<li><code>map()</code> and its variants <code>map_chr()</code>, <code>map_dbl()</code> - Applies a function to each element of a list or vector. For example, you could calculate the mean of every column in a data frame:</li>
</ul>
<p>In the previous section, we learned about loops. There is nothing wrong with for loops. However, with <code>purrr::map()</code>, I find myself writing less and less for loops.</p>
<div id="nesting-data-with-nest" class="section level3" number="9.7.1">
<h3><span class="header-section-number">9.7.1</span> Nesting Data with nest()</h3>
<p>The <code>nest()</code> function in R, when used with tibbles, groups your data based on a specific variable and creates a new column containing nested data frames for each group. It’s like putting similar data into separate containers, making it easier to work with and analyze them as a whole or individually.</p>
<p>Imagine you have a large table of information about different types of cancer samples. You want to organize this data in a way that groups all the information related to each type of cancer separately. One way to do this is by using the <code>tidyr::nest()</code> function along with the purrr package in R.</p>
<p>Here’s how you can achieve this:</p>
<div class="sourceCode" id="cb362"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb362-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb362-1" tabindex="-1"></a><span class="co"># read in the data again</span></span>
<span id="cb362-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb362-2" tabindex="-1"></a>tcga_data <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="st">&quot;~/Downloads/TCGA_cancer_genes_expression.csv&quot;</span>)</span>
<span id="cb362-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb362-3" tabindex="-1"></a></span>
<span id="cb362-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb362-4" tabindex="-1"></a><span class="co"># Group and nest the data</span></span>
<span id="cb362-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb362-5" tabindex="-1"></a>tcga_nest <span class="ot">&lt;-</span> tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb362-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb362-6" tabindex="-1"></a>  <span class="fu">filter</span>(sample_type <span class="sc">==</span> <span class="st">&quot;cancer&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb362-7"><a href="introduction-to-the-tidyverse-ecosystem.html#cb362-7" tabindex="-1"></a>  <span class="fu">select</span>(EPCAM, tcga.tcga_barcode<span class="sc">:</span>sample_type) <span class="sc">%&gt;%</span></span>
<span id="cb362-8"><a href="introduction-to-the-tidyverse-ecosystem.html#cb362-8" tabindex="-1"></a>  <span class="fu">group_by</span>(study) <span class="sc">%&gt;%</span></span>
<span id="cb362-9"><a href="introduction-to-the-tidyverse-ecosystem.html#cb362-9" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">nest</span>()</span>
<span id="cb362-10"><a href="introduction-to-the-tidyverse-ecosystem.html#cb362-10" tabindex="-1"></a></span>
<span id="cb362-11"><a href="introduction-to-the-tidyverse-ecosystem.html#cb362-11" tabindex="-1"></a>tcga_nest</span></code></pre></div>
<pre><code>## # A tibble: 32 × 2
## # Groups:   study [32]
##    study data                
##    &lt;chr&gt; &lt;list&gt;              
##  1 ACC   &lt;tibble [79 × 4]&gt;   
##  2 BLCA  &lt;tibble [414 × 4]&gt;  
##  3 BRCA  &lt;tibble [1,127 × 4]&gt;
##  4 CESC  &lt;tibble [304 × 4]&gt;  
##  5 CHOL  &lt;tibble [36 × 4]&gt;   
##  6 COAD  &lt;tibble [504 × 4]&gt;  
##  7 DLBC  &lt;tibble [48 × 4]&gt;   
##  8 ESCA  &lt;tibble [184 × 4]&gt;  
##  9 GBM   &lt;tibble [170 × 4]&gt;  
## 10 HNSC  &lt;tibble [502 × 4]&gt;  
## # ℹ 22 more rows</code></pre>
<p>The <code>tidyr::nest()</code> function creates a list-column within your tibble, where each element of the list is a nested data frame. This is a powerful feature of tibbles, as they can contain tables within the table.</p>
<p>You can access the nested data using the <code>$</code> sign, just like you would with a regular data frame, and use the double bracket to access the element. For example:</p>
<div class="sourceCode" id="cb364"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb364-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb364-1" tabindex="-1"></a><span class="co"># Accessing the first nested data frame</span></span>
<span id="cb364-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb364-2" tabindex="-1"></a>first_nested_df <span class="ot">&lt;-</span> tcga_nest<span class="sc">$</span>data[[<span class="dv">1</span>]]</span>
<span id="cb364-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb364-3" tabindex="-1"></a></span>
<span id="cb364-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb364-4" tabindex="-1"></a>first_nested_df</span></code></pre></div>
<pre><code>## # A tibble: 79 × 4
##    EPCAM tcga.tcga_barcode            tcga.cgc_sample_sample_type sample_type
##    &lt;dbl&gt; &lt;chr&gt;                        &lt;chr&gt;                       &lt;chr&gt;      
##  1  4.52 TCGA-OR-A5KU-01A-11R-A29S-07 Primary Tumor               cancer     
##  2  9.53 TCGA-P6-A5OG-01A-22R-A29S-07 Primary Tumor               cancer     
##  3 42.4  TCGA-OR-A5K5-01A-11R-A29S-07 Primary Tumor               cancer     
##  4 16.3  TCGA-OR-A5K4-01A-11R-A29S-07 Primary Tumor               cancer     
##  5 12.5  TCGA-OR-A5LP-01A-11R-A29S-07 Primary Tumor               cancer     
##  6  2.43 TCGA-PK-A5H9-01A-11R-A29S-07 Primary Tumor               cancer     
##  7  3.74 TCGA-OR-A5LD-01A-11R-A29S-07 Primary Tumor               cancer     
##  8  4.08 TCGA-OR-A5JX-01A-11R-A29S-07 Primary Tumor               cancer     
##  9  2.84 TCGA-PK-A5H8-01A-11R-A29S-07 Primary Tumor               cancer     
## 10  3.61 TCGA-OR-A5J3-01A-11R-A29S-07 Primary Tumor               cancer     
## # ℹ 69 more rows</code></pre>
<p>In this example, first_nested_df contains the first nested data frame, which corresponds to one of the “study” groups.</p>
<p>You can add the names to the list column, and now you can access it by cancer type:</p>
<div class="sourceCode" id="cb366"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb366-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb366-1" tabindex="-1"></a><span class="fu">names</span>(tcga_nest<span class="sc">$</span>data)<span class="ot">&lt;-</span> tcga_nest<span class="sc">$</span>study</span>
<span id="cb366-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb366-2" tabindex="-1"></a></span>
<span id="cb366-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb366-3" tabindex="-1"></a>tcga_nest</span></code></pre></div>
<pre><code>## # A tibble: 32 × 2
## # Groups:   study [32]
##    study data                
##    &lt;chr&gt; &lt;named list&gt;        
##  1 ACC   &lt;tibble [79 × 4]&gt;   
##  2 BLCA  &lt;tibble [414 × 4]&gt;  
##  3 BRCA  &lt;tibble [1,127 × 4]&gt;
##  4 CESC  &lt;tibble [304 × 4]&gt;  
##  5 CHOL  &lt;tibble [36 × 4]&gt;   
##  6 COAD  &lt;tibble [504 × 4]&gt;  
##  7 DLBC  &lt;tibble [48 × 4]&gt;   
##  8 ESCA  &lt;tibble [184 × 4]&gt;  
##  9 GBM   &lt;tibble [170 × 4]&gt;  
## 10 HNSC  &lt;tibble [502 × 4]&gt;  
## # ℹ 22 more rows</code></pre>
<div class="sourceCode" id="cb368"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb368-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb368-1" tabindex="-1"></a>tcga_nest<span class="sc">$</span>data[[<span class="st">&quot;ACC&quot;</span>]]</span></code></pre></div>
<pre><code>## # A tibble: 79 × 4
##    EPCAM tcga.tcga_barcode            tcga.cgc_sample_sample_type sample_type
##    &lt;dbl&gt; &lt;chr&gt;                        &lt;chr&gt;                       &lt;chr&gt;      
##  1  4.52 TCGA-OR-A5KU-01A-11R-A29S-07 Primary Tumor               cancer     
##  2  9.53 TCGA-P6-A5OG-01A-22R-A29S-07 Primary Tumor               cancer     
##  3 42.4  TCGA-OR-A5K5-01A-11R-A29S-07 Primary Tumor               cancer     
##  4 16.3  TCGA-OR-A5K4-01A-11R-A29S-07 Primary Tumor               cancer     
##  5 12.5  TCGA-OR-A5LP-01A-11R-A29S-07 Primary Tumor               cancer     
##  6  2.43 TCGA-PK-A5H9-01A-11R-A29S-07 Primary Tumor               cancer     
##  7  3.74 TCGA-OR-A5LD-01A-11R-A29S-07 Primary Tumor               cancer     
##  8  4.08 TCGA-OR-A5JX-01A-11R-A29S-07 Primary Tumor               cancer     
##  9  2.84 TCGA-PK-A5H8-01A-11R-A29S-07 Primary Tumor               cancer     
## 10  3.61 TCGA-OR-A5J3-01A-11R-A29S-07 Primary Tumor               cancer     
## # ℹ 69 more rows</code></pre>
</div>
<div id="map-and-its-variants" class="section level3" number="9.7.2">
<h3><span class="header-section-number">9.7.2</span> <code>map()</code> and Its Variants</h3>
<p>Let’s calculate the median value of EPCAM for each cancer type using <code>map()</code>.</p>
<p><code>map()</code> takes in a vector or a list, and a function to be applied to every element of the vector or the list.</p>
<div class="sourceCode" id="cb370"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb370-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb370-1" tabindex="-1"></a><span class="fu">map</span>(tcga_nest<span class="sc">$</span>data, <span class="cf">function</span>(x) (<span class="fu">median</span>(x<span class="sc">$</span>EPCAM)))</span></code></pre></div>
<pre><code>## $ACC
## [1] 4.830195
## 
## $BLCA
## [1] 81.1488
## 
## $BRCA
## [1] 161.9358
## 
## $CESC
## [1] 75.93449
## 
## $CHOL
## [1] 251.4888
## 
## $COAD
## [1] 777.5484
## 
## $DLBC
## [1] 0.5783636
## 
## $ESCA
## [1] 189.3882
## 
## $GBM
## [1] 0.3073375
## 
## $HNSC
## [1] 24.70459
## 
## $KICH
## [1] 93.78075
## 
## $KIRC
## [1] 24.10957
## 
## $KIRP
## [1] 71.4589
## 
## $LGG
## [1] 0.6812875
## 
## $LIHC
## [1] 0.7269132
## 
## $LUAD
## [1] 305.5941
## 
## $LUSC
## [1] 138.6225
## 
## $MESO
## [1] 1.144048
## 
## $OV
## [1] 206.6447
## 
## $PAAD
## [1] 267.3574
## 
## $PCPG
## [1] 6.188397
## 
## $PRAD
## [1] 194.4213
## 
## $READ
## [1] 808.5985
## 
## $SARC
## [1] 0.2179763
## 
## $SKCM
## [1] 0.2021555
## 
## $STAD
## [1] 320.1896
## 
## $TGCT
## [1] 116.1016
## 
## $THCA
## [1] 342.4736
## 
## $THYM
## [1] 13.63511
## 
## $UCEC
## [1] 348.4112
## 
## $UCS
## [1] 94.77784
## 
## $UVM
## [1] 0.05832845</code></pre>
<p>In this example, the function takes each element of the list of the data frame and return the median of the EPCAM.</p>
</div>
<div id="note-on-anonymous-functions" class="section level3" number="9.7.3">
<h3><span class="header-section-number">9.7.3</span> Note on Anonymous functions</h3>
<p>There are three ways to specify a function</p>
<div class="sourceCode" id="cb372"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb372-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb372-1" tabindex="-1"></a><span class="co"># full function with a function name</span></span>
<span id="cb372-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb372-2" tabindex="-1"></a>calculate_median<span class="ot">&lt;-</span> <span class="cf">function</span>(x){</span>
<span id="cb372-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb372-3" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">median</span>(x<span class="sc">$</span>EPCAM))</span>
<span id="cb372-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb372-4" tabindex="-1"></a>}</span></code></pre></div>
<div class="sourceCode" id="cb373"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb373-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb373-1" tabindex="-1"></a><span class="co"># base R anonymous function</span></span>
<span id="cb373-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb373-2" tabindex="-1"></a><span class="cf">function</span>(x) (<span class="fu">median</span>(x<span class="sc">$</span>EPCAM))</span></code></pre></div>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb374-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb374-1" tabindex="-1"></a><span class="co"># purrr anonymouse function using formula ~, note you use .x instead of x</span></span>
<span id="cb374-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb374-2" tabindex="-1"></a><span class="sc">~</span> <span class="fu">median</span>(.x<span class="sc">$</span>EPCAM)</span></code></pre></div>
<p>The following will have the same results</p>
<div class="sourceCode" id="cb375"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb375-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb375-1" tabindex="-1"></a><span class="fu">map</span>(tcga_nest<span class="sc">$</span>data, calculate_median)</span>
<span id="cb375-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb375-2" tabindex="-1"></a><span class="fu">map</span>(tcga_nest<span class="sc">$</span>data, <span class="cf">function</span>(x) (<span class="fu">median</span>(x<span class="sc">$</span>EPCAM)))</span>
<span id="cb375-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb375-3" tabindex="-1"></a><span class="fu">map</span>(tcga_nest<span class="sc">$</span>data, <span class="sc">~</span> <span class="fu">median</span>(.x<span class="sc">$</span>EPCAM))</span></code></pre></div>
<p>read more at <a href="https://jennybc.github.io/purrr-tutorial/ls03_map-function-syntax.html#anonymous_function,_conventional" class="uri">https://jennybc.github.io/purrr-tutorial/ls03_map-function-syntax.html#anonymous_function,_conventional</a>.</p>
<p>map always returns a list, it returns a list of median values in this case. If you want to return a vector, use <code>map_dbl</code>:</p>
<div class="sourceCode" id="cb376"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb376-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb376-1" tabindex="-1"></a><span class="fu">map_dbl</span>(tcga_nest<span class="sc">$</span>data, <span class="cf">function</span>(x) (<span class="fu">median</span>(x<span class="sc">$</span>EPCAM)))</span></code></pre></div>
<pre><code>##          ACC         BLCA         BRCA         CESC         CHOL         COAD 
##   4.83019522  81.14880204 161.93580869  75.93448764 251.48875310 777.54836210 
##         DLBC         ESCA          GBM         HNSC         KICH         KIRC 
##   0.57836358 189.38816785   0.30733745  24.70459340  93.78075352  24.10956585 
##         KIRP          LGG         LIHC         LUAD         LUSC         MESO 
##  71.45889552   0.68128748   0.72691319 305.59410291 138.62247973   1.14404760 
##           OV         PAAD         PCPG         PRAD         READ         SARC 
## 206.64472617 267.35742947   6.18839726 194.42130740 808.59850470   0.21797629 
##         SKCM         STAD         TGCT         THCA         THYM         UCEC 
##   0.20215548 320.18957706 116.10162785 342.47358682  13.63510740 348.41123728 
##          UCS          UVM 
##  94.77783740   0.05832845</code></pre>
<p>Let’s save the output to a new variable</p>
<div class="sourceCode" id="cb378"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb378-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb378-1" tabindex="-1"></a>median_EPCAM<span class="ot">&lt;-</span> <span class="fu">map_dbl</span>(tcga_nest<span class="sc">$</span>data, <span class="cf">function</span>(x) (<span class="fu">median</span>(x<span class="sc">$</span>EPCAM)))</span>
<span id="cb378-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb378-2" tabindex="-1"></a></span>
<span id="cb378-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb378-3" tabindex="-1"></a><span class="co"># returns a list of log2 values</span></span>
<span id="cb378-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb378-4" tabindex="-1"></a><span class="fu">map</span>(median_EPCAM, <span class="cf">function</span>(x) <span class="fu">log2</span>(x))</span></code></pre></div>
<pre><code>## $ACC
## [1] 2.272081
## 
## $BLCA
## [1] 6.342498
## 
## $BRCA
## [1] 7.339278
## 
## $CESC
## [1] 6.246683
## 
## $CHOL
## [1] 7.97435
## 
## $COAD
## [1] 9.602789
## 
## $DLBC
## [1] -0.7899514
## 
## $ESCA
## [1] 7.565202
## 
## $GBM
## [1] -1.702105
## 
## $HNSC
## [1] 4.626707
## 
## $KICH
## [1] 6.55122
## 
## $KIRC
## [1] 4.591534
## 
## $KIRP
## [1] 6.159042
## 
## $LGG
## [1] -0.5536644
## 
## $LIHC
## [1] -0.460145
## 
## $LUAD
## [1] 8.255473
## 
## $LUSC
## [1] 7.115017
## 
## $MESO
## [1] 0.1941471
## 
## $OV
## [1] 7.691009
## 
## $PAAD
## [1] 8.062626
## 
## $PCPG
## [1] 2.629566
## 
## $PRAD
## [1] 7.603043
## 
## $READ
## [1] 9.65928
## 
## $SARC
## [1] -2.197757
## 
## $SKCM
## [1] -2.306463
## 
## $STAD
## [1] 8.322783
## 
## $TGCT
## [1] 6.859244
## 
## $THCA
## [1] 8.419849
## 
## $THYM
## [1] 3.769254
## 
## $UCEC
## [1] 8.444647
## 
## $UCS
## [1] 6.566478
## 
## $UVM
## [1] -4.099656</code></pre>
<div class="sourceCode" id="cb380"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb380-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb380-1" tabindex="-1"></a><span class="co"># returns a vector</span></span>
<span id="cb380-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb380-2" tabindex="-1"></a><span class="fu">map_dbl</span>(median_EPCAM, <span class="cf">function</span>(x) <span class="fu">log2</span>(x))</span></code></pre></div>
<pre><code>##        ACC       BLCA       BRCA       CESC       CHOL       COAD       DLBC 
##  2.2720815  6.3424979  7.3392782  6.2466834  7.9743501  9.6027886 -0.7899514 
##       ESCA        GBM       HNSC       KICH       KIRC       KIRP        LGG 
##  7.5652024 -1.7021045  4.6267074  6.5512200  4.5915338  6.1590417 -0.5536644 
##       LIHC       LUAD       LUSC       MESO         OV       PAAD       PCPG 
## -0.4601450  8.2554729  7.1150174  0.1941471  7.6910087  8.0626260  2.6295658 
##       PRAD       READ       SARC       SKCM       STAD       TGCT       THCA 
##  7.6030425  9.6592797 -2.1977569 -2.3064627  8.3227825  6.8592444  8.4198489 
##       THYM       UCEC        UCS        UVM 
##  3.7692542  8.4446473  6.5664778 -4.0996564</code></pre>
<p>We can stay in the original tibble by just adding the median to a new column with <code>mutate()</code>:</p>
<div class="sourceCode" id="cb382"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb382-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb382-1" tabindex="-1"></a>tcga_nest <span class="sc">%&gt;%</span></span>
<span id="cb382-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb382-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">median_EPCAM =</span> <span class="fu">map_dbl</span>(data, <span class="cf">function</span>(x) (<span class="fu">median</span>(x<span class="sc">$</span>EPCAM))))</span></code></pre></div>
<pre><code>## # A tibble: 32 × 3
## # Groups:   study [32]
##    study data                 median_EPCAM
##    &lt;chr&gt; &lt;named list&gt;                &lt;dbl&gt;
##  1 ACC   &lt;tibble [79 × 4]&gt;           4.83 
##  2 BLCA  &lt;tibble [414 × 4]&gt;         81.1  
##  3 BRCA  &lt;tibble [1,127 × 4]&gt;      162.   
##  4 CESC  &lt;tibble [304 × 4]&gt;         75.9  
##  5 CHOL  &lt;tibble [36 × 4]&gt;         251.   
##  6 COAD  &lt;tibble [504 × 4]&gt;        778.   
##  7 DLBC  &lt;tibble [48 × 4]&gt;           0.578
##  8 ESCA  &lt;tibble [184 × 4]&gt;        189.   
##  9 GBM   &lt;tibble [170 × 4]&gt;          0.307
## 10 HNSC  &lt;tibble [502 × 4]&gt;         24.7  
## # ℹ 22 more rows</code></pre>
<p>of course, we can use <code>group_by</code> followed by summarise as shown above to get the same thing, but it demonstrates how we can combine <code>map()</code> function and list column to do powerful data analysis within a data frame.</p>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb384-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb384-1" tabindex="-1"></a>tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb384-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb384-2" tabindex="-1"></a>  <span class="fu">filter</span>(sample_type <span class="sc">==</span> <span class="st">&quot;cancer&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb384-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb384-3" tabindex="-1"></a>  <span class="fu">select</span>(EPCAM, tcga.tcga_barcode<span class="sc">:</span>sample_type) <span class="sc">%&gt;%</span></span>
<span id="cb384-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb384-4" tabindex="-1"></a>  <span class="fu">group_by</span>(study) <span class="sc">%&gt;%</span></span>
<span id="cb384-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb384-5" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">median_EPCAM =</span> <span class="fu">median</span>(EPCAM))</span></code></pre></div>
<pre><code>## # A tibble: 32 × 2
##    study median_EPCAM
##    &lt;chr&gt;        &lt;dbl&gt;
##  1 ACC          4.83 
##  2 BLCA        81.1  
##  3 BRCA       162.   
##  4 CESC        75.9  
##  5 CHOL       251.   
##  6 COAD       778.   
##  7 DLBC         0.578
##  8 ESCA       189.   
##  9 GBM          0.307
## 10 HNSC        24.7  
## # ℹ 22 more rows</code></pre>
<p>Read this <a href="https://dplyr.tidyverse.org/reference/summarise.html" class="uri">https://dplyr.tidyverse.org/reference/summarise.html</a> for more examples using group_by with summarise.</p>
<p>You can even nest by two columns:</p>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb386-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb386-1" tabindex="-1"></a>tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb386-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb386-2" tabindex="-1"></a>  <span class="fu">select</span>(EPCAM, tcga.tcga_barcode<span class="sc">:</span>sample_type) <span class="sc">%&gt;%</span></span>
<span id="cb386-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb386-3" tabindex="-1"></a>  <span class="fu">group_by</span>(study, sample_type) <span class="sc">%&gt;%</span></span>
<span id="cb386-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb386-4" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">nest</span>()</span></code></pre></div>
<pre><code>## # A tibble: 73 × 3
## # Groups:   study, sample_type [73]
##    study sample_type data                
##    &lt;chr&gt; &lt;chr&gt;       &lt;list&gt;              
##  1 ACC   cancer      &lt;tibble [79 × 3]&gt;   
##  2 BLCA  cancer      &lt;tibble [414 × 3]&gt;  
##  3 BLCA  normal      &lt;tibble [19 × 3]&gt;   
##  4 BRCA  cancer      &lt;tibble [1,127 × 3]&gt;
##  5 BRCA  normal      &lt;tibble [112 × 3]&gt;  
##  6 BRCA  metastatic  &lt;tibble [7 × 3]&gt;    
##  7 BRCA  &lt;NA&gt;        &lt;tibble [10 × 3]&gt;   
##  8 CESC  cancer      &lt;tibble [304 × 3]&gt;  
##  9 CESC  normal      &lt;tibble [3 × 3]&gt;    
## 10 CESC  metastatic  &lt;tibble [2 × 3]&gt;    
## # ℹ 63 more rows</code></pre>
<p>You can easily see some cancer types have normal and some have metastatic samples and you can do everything within a dataframe.</p>
<p>The other way to check the information is to use the table function:</p>
<div class="sourceCode" id="cb388"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb388-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb388-1" tabindex="-1"></a><span class="fu">table</span>(tcga_data<span class="sc">$</span>sample_type, tcga_data<span class="sc">$</span>study)</span></code></pre></div>
<pre><code>##             
##               ACC BLCA BRCA CESC CHOL COAD DLBC ESCA  GBM HNSC KICH KIRC KIRP
##   cancer       79  414 1127  304   36  504   48  184  170  502   66  544  291
##   metastatic    0    0    7    2    0    1    0    1    0    2    0    0    0
##   normal        0   19  112    3    9   41    0   13    5   44   25   72   32
##             
##              LAML  LGG LIHC LUAD LUSC MESO   OV PAAD PCPG PRAD READ SARC SKCM
##   cancer        0  532  374  542  504   87  429  178  182  505  167  262  103
##   metastatic    0    0    0    0    0    0    0    1    2    1    0    1  368
##   normal        0    0   50   59   51    0    0    4    3   52   10    2    1
##             
##              STAD TGCT THCA THYM UCEC  UCS  UVM
##   cancer      415  156  505  120  554   57   80
##   metastatic    0    0    8    0    0    0    0
##   normal       37    0   59    2   35    0    0</code></pre>
<p>The key takeaways is:</p>
<ul>
<li>map() applies a function over each element of a list/vector</li>
</ul>
</div>
<div id="conclusion-20" class="section level3" number="9.7.4">
<h3><span class="header-section-number">9.7.4</span> Conclusion</h3>
<p>With purrr, you get a powerful toolkit for iterating over biological data in a functional programming style. As you advance in bioinformatics, purrr will continue to make your code more clear and more concise. This is introduction is just the tip of the iceberg. To learn more about purrr, read this <a href="https://jennybc.github.io/purrr-tutorial/" class="uri">https://jennybc.github.io/purrr-tutorial/</a></p>
<p>Some key benefits of the tidyverse include:</p>
<ol style="list-style-type: decimal">
<li><p>Consistent language and grammar across packages like piping (%&gt;%) and verb functions (filter(), mutate()).</p></li>
<li><p>Works well with pipe workflows for transforming and visualizing data.</p></li>
<li><p>Enhances exploratory data analysis and makes it easy to go from raw data to insights.</p></li>
<li><p>Large community providing learning resources and support.</p></li>
</ol>
<p>The tidyverse packages work seamlessly together, allowing you to conduct complete data science projects in R. From import to wrangling to visualization, the tidyverse provides a set of tools that follow common principles. While we can dedicate a full chapter on each package, we only focused on <code>dplyr</code>, <code>tidyr</code> and <code>purrr</code>.</p>
</div>
</div>
<div id="tidying-metadata-from-geo." class="section level2" number="9.8">
<h2><span class="header-section-number">9.8</span> Tidying metadata from GEO.</h2>
<p>In this lesson, we will learn how to tidy metadata obtained from the <a href="https://www.ncbi.nlm.nih.gov/geo/"><code>Gene Expression Omnibus (GEO)</code></a> database using the Tidyverse package in R. Tidying metadata is an essential step in preparing data for analysis. We will use a real-world example from GEO to demonstrate the process step-by-step.</p>
<div id="prerequisites" class="section level3" number="9.8.1">
<h3><span class="header-section-number">9.8.1</span> Prerequisites</h3>
<p>Before we begin, make sure you have R and the necessary packages installed. You can install the required packages using the following commands:</p>
<div class="sourceCode" id="cb390"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb390-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb390-1" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;GEOquery&quot;</span>)</span>
<span id="cb390-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb390-2" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;tidyverse&quot;</span>)</span></code></pre></div>
</div>
<div id="getting-started" class="section level3" number="9.8.2">
<h3><span class="header-section-number">9.8.2</span> Getting Started</h3>
<p>We’ll start by loading the necessary libraries and fetching metadata from a GEO dataset called <code>GSE176021</code>. GEO is a repository for gene expression data and other omics data, and it often contains metadata in a wide format, which is not suitable for analysis.</p>
<div class="sourceCode" id="cb391"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb391-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb391-1" tabindex="-1"></a><span class="co"># Load the necessary libraries</span></span>
<span id="cb391-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb391-2" tabindex="-1"></a><span class="fu">library</span>(GEOquery)</span>
<span id="cb391-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb391-3" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb391-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb391-4" tabindex="-1"></a></span>
<span id="cb391-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb391-5" tabindex="-1"></a><span class="co"># Fetch metadata from GEO</span></span>
<span id="cb391-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb391-6" tabindex="-1"></a>GSE176021_meta <span class="ot">&lt;-</span> <span class="fu">getGEO</span>(<span class="at">GEO =</span> <span class="st">&quot;GSE176021&quot;</span>, <span class="at">GSEMatrix =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Let’s use the <code>str</code> structure command to inspect the <a href="mailto:GSE176021_meta@gsms" class="email">GSE176021_meta@gsms</a> object</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb392-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb392-1" tabindex="-1"></a><span class="fu">str</span>(GSE176021_meta<span class="sc">@</span>gsms, <span class="at">max.level =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>## List of 110
##  $ GSM5352886:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352887:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352888:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352889:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352890:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352891:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352892:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352893:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352894:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352895:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352896:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352897:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352898:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352899:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352900:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352901:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352902:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352903:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352904:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352905:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352906:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352907:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352908:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352909:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352910:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352911:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352912:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352913:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352914:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352915:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352916:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352917:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352918:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352919:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352920:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352921:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352922:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352923:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352924:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352925:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352926:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352927:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352928:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352929:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352930:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352931:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352932:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352933:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352934:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352935:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352936:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352937:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352938:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352939:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352940:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352941:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352942:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352943:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352944:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352945:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352946:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352947:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352948:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352949:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352950:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352951:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352952:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352953:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352954:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352955:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352956:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352957:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352958:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352959:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352960:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352961:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352962:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352963:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352964:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352965:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352966:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352967:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352968:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352969:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352970:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352971:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352972:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352973:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352974:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352975:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352976:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352977:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352978:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352979:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352980:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352981:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352982:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352983:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##  $ GSM5352984:Formal class &#39;GSM&#39; [package &quot;GEOquery&quot;] with 2 slots
##   [list output truncated]</code></pre>
<p>So it is a list of 110 GSM objects. Let’s take a look at the first element of the list</p>
<div class="sourceCode" id="cb394"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb394-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb394-1" tabindex="-1"></a>GSE176021_meta<span class="sc">@</span>gsms[[<span class="dv">1</span>]]<span class="sc">@</span>header </span></code></pre></div>
<pre><code>## $channel_count
## [1] &quot;1&quot;
## 
## $characteristics_ch1
## [1] &quot;patientid: MD01-024&quot;      &quot;cell type: lymphocytes&quot;  
## [3] &quot;response status: Non-MPR&quot;
## 
## $contact_address
## [1] &quot;1650 Orleans Street, CRB1, 4M&quot;
## 
## $contact_city
## [1] &quot;Baltimore&quot;
## 
## $contact_country
## [1] &quot;USA&quot;
## 
## $contact_email
## [1] &quot;ksmit228@jhmi.edu&quot;
## 
## $contact_institute
## [1] &quot;Johns Hopkins University&quot;
## 
## $contact_name
## [1] &quot;Kellie,Nicole,Smith&quot;
## 
## $contact_state
## [1] &quot;MD&quot;
## 
## $`contact_zip/postal_code`
## [1] &quot;21287&quot;
## 
## $data_processing
## [1] &quot;Cell Ranger v3.1.0 was used to demultiplex the FASTQ reads, align them to the GRCh38 human transcriptome, and extract their “cell” and “UMI” barcodes. The output of this pipeline is a digital gene expression (DGE) matrix for each sample, which records the number of UMIs for each gene that are associated with each cell barcode.&quot;
## [2] &quot;Supplementary_files_format_and_content: Cell ranger output&quot;                                                                                                                                                                                                                                                                              
## [3] &quot;Supplementary_files_format_and_content: RDS files include metadata with cell type annotations&quot;                                                                                                                                                                                                                                           
## 
## $data_row_count
## [1] &quot;0&quot;
## 
## $extract_protocol_ch1
## [1] &quot;Cryobanked T cells were thawed and washed twice with pre-warmed RPMI with 20% FBS and gentamicin. Cells were resuspended in PBS and stained with a viability marker (LIVE/DEAD™ Fixable Near-IR; ThermoFisher) for 15mins at RT in the dark. Cells were the incubated with FC block for 15 mins on ice and stained with antibody against CD3 (BV421, clone SK7) for 30mins on ice. After staining, highly viable CD3+ T cells were sorted into 0.04% BSA in PBS using a BD FACSAria II Cell Sorter. Sorted cells were manually counted using a hemocytometer and prepared at the desired cell concentration (1000 cells/ul), when possible&quot;                                                                                                                                                      
## [2] &quot;The Single Cell 5’ V(D)J and 5’ DGE kits (10X Genomics) were used to capture immune repertoire information and gene expression from the same cell in an emulsion-based protocol at the single cell level. Cells and barcoded gel beads were partitioned into nanoliter scale droplets using the 10X Genomics Chromium platform to partition up to 10,000 cells per sample followed by RNA capture and cell-barcoded cDNA synthesis using the manufacturer’s standard protocols. Libraries were generated and sequenced on an Illumina NovaSeq instrument using 2x150bp paired end sequencing. 5’ VDJ libraries were sequenced to a depth of ~5,000 reads per cell, for a total of 5 million to 25 million reads. The 5’ DGE libraries were sequenced to a target depth of ~50,000 reads per cell&quot;
## 
## $geo_accession
## [1] &quot;GSM5352886&quot;
## 
## $instrument_model
## [1] &quot;Illumina NovaSeq 6000&quot;
## 
## $last_update_date
## [1] &quot;Feb 20 2024&quot;
## 
## $library_selection
## [1] &quot;cDNA&quot;
## 
## $library_source
## [1] &quot;transcriptomic&quot;
## 
## $library_strategy
## [1] &quot;RNA-Seq&quot;
## 
## $molecule_ch1
## [1] &quot;total RNA&quot;
## 
## $organism_ch1
## [1] &quot;Homo sapiens&quot;
## 
## $platform_id
## [1] &quot;GPL24676&quot;
## 
## $relation
## [1] &quot;BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN19514461&quot;
## 
## $series_id
## [1] &quot;GSE173351&quot; &quot;GSE176021&quot;
## 
## $source_name_ch1
## [1] &quot;tumor&quot;
## 
## $status
## [1] &quot;Public on Jul 21 2021&quot;
## 
## $submission_date
## [1] &quot;Jun 02 2021&quot;
## 
## $supplementary_file_1
## [1] &quot;ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5352nnn/GSM5352886/suppl/GSM5352886_MD01-024_tumor_1.tar.gz&quot;
## 
## $supplementary_file_2
## [1] &quot;ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5352nnn/GSM5352886/suppl/GSM5352886_MD01-024_tumor_1.vdj.tar.gz&quot;
## 
## $taxid_ch1
## [1] &quot;9606&quot;
## 
## $title
## [1] &quot;MD01-024_tumor_1&quot;
## 
## $type
## [1] &quot;SRA&quot;</code></pre>
<p>It will print out a long text. We only need the <code>characteristics_ch1</code></p>
<div class="sourceCode" id="cb396"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb396-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb396-1" tabindex="-1"></a>GSE176021_meta<span class="sc">@</span>gsms[[<span class="dv">1</span>]]<span class="sc">@</span>header<span class="sc">$</span>characteristics_ch1</span></code></pre></div>
<pre><code>## [1] &quot;patientid: MD01-024&quot;      &quot;cell type: lymphocytes&quot;  
## [3] &quot;response status: Non-MPR&quot;</code></pre>
<p>This contains the metadata we need.</p>
<p>Now, let’s extract all the metadata for all samples and bind them to a dataframe.</p>
<div class="sourceCode" id="cb398"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb398-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb398-1" tabindex="-1"></a>GSE176021_meta <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map</span>(GSE176021_meta<span class="sc">@</span>gsms, </span>
<span id="cb398-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb398-2" tabindex="-1"></a>                             <span class="cf">function</span>(x) x<span class="sc">@</span>header<span class="sc">$</span>characteristics_ch1) <span class="sc">%&gt;%</span></span>
<span id="cb398-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb398-3" tabindex="-1"></a>  <span class="fu">bind_rows</span>() <span class="sc">%&gt;%</span></span>
<span id="cb398-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb398-4" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</span>
<span id="cb398-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb398-5" tabindex="-1"></a></span>
<span id="cb398-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb398-6" tabindex="-1"></a>GSE176021_meta</span></code></pre></div>
<pre><code>## # A tibble: 2 × 110
##   GSM5352886   GSM5352887 GSM5352888 GSM5352889 GSM5352890 GSM5352891 GSM5352892
##   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;     
## 1 patientid: … patientid… patientid… patientid… patientid… patientid… patientid…
## 2 response st… response … response … response … response … response … response …
## # ℹ 103 more variables: GSM5352893 &lt;chr&gt;, GSM5352894 &lt;chr&gt;, GSM5352895 &lt;chr&gt;,
## #   GSM5352896 &lt;chr&gt;, GSM5352897 &lt;chr&gt;, GSM5352898 &lt;chr&gt;, GSM5352899 &lt;chr&gt;,
## #   GSM5352900 &lt;chr&gt;, GSM5352901 &lt;chr&gt;, GSM5352902 &lt;chr&gt;, GSM5352903 &lt;chr&gt;,
## #   GSM5352904 &lt;chr&gt;, GSM5352905 &lt;chr&gt;, GSM5352906 &lt;chr&gt;, GSM5352907 &lt;chr&gt;,
## #   GSM5352908 &lt;chr&gt;, GSM5352909 &lt;chr&gt;, GSM5352910 &lt;chr&gt;, GSM5352911 &lt;chr&gt;,
## #   GSM5352912 &lt;chr&gt;, GSM5352913 &lt;chr&gt;, GSM5352914 &lt;chr&gt;, GSM5352915 &lt;chr&gt;,
## #   GSM5352916 &lt;chr&gt;, GSM5352917 &lt;chr&gt;, GSM5352918 &lt;chr&gt;, GSM5352919 &lt;chr&gt;, …</code></pre>
<p>In this code:</p>
<ol style="list-style-type: decimal">
<li><p>We load the <code>GEOquery</code> library to access functions related to the Gene Expression Omnibus (GEO) database.</p></li>
<li><p>We fetch metadata from the GEO dataset “GSE176021” using the <code>getGEO</code> function. The <code>GSEMatrix = FALSE</code> argument ensures that we retrieve metadata rather than expression data.</p></li>
<li><p>We use <code>purrr::map</code> to extract the “characteristics_ch1” information from each sample within the GEO dataset. This information typically contains details about the samples, such as patient identifiers and response statuses.</p></li>
<li><p>Next, we use <code>bind_rows()</code> to combine these extracted characteristics into a single data frame.</p></li>
<li><p>We use <code>dplyr::slice(c(1, 3))</code> to select only the first and third rows of the resulting data frame, essentially keeping a subset of the metadata for demonstration purposes.</p></li>
</ol>
<p>Now, let’s take a look at the wide-format metadata:</p>
<div class="sourceCode" id="cb400"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb400-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb400-1" tabindex="-1"></a><span class="co"># Display the first two rows and first ten columns of the wide-format metadata</span></span>
<span id="cb400-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb400-2" tabindex="-1"></a>GSE176021_meta[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>## # A tibble: 2 × 10
##   GSM5352886   GSM5352887 GSM5352888 GSM5352889 GSM5352890 GSM5352891 GSM5352892
##   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;     
## 1 patientid: … patientid… patientid… patientid… patientid… patientid… patientid…
## 2 response st… response … response … response … response … response … response …
## # ℹ 3 more variables: GSM5352893 &lt;chr&gt;, GSM5352894 &lt;chr&gt;, GSM5352895 &lt;chr&gt;</code></pre>
<p>The wide-format metadata is not suitable for analysis. It has many columns, and we need to tidy it before proceeding.</p>
</div>
<div id="tidying-the-metadata" class="section level3" number="9.8.3">
<h3><span class="header-section-number">9.8.3</span> Tidying the Metadata</h3>
<p>We will follow these steps to tidy the metadata:</p>
<ol style="list-style-type: decimal">
<li><p>Add a “meta” column with labels.</p></li>
<li><p>Reorder the columns to have the “meta” column first.</p></li>
</ol>
<div class="sourceCode" id="cb402"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb402-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb402-1" tabindex="-1"></a><span class="co"># Add a &quot;meta&quot; column and reorder the columns</span></span>
<span id="cb402-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb402-2" tabindex="-1"></a>GSE176021_meta <span class="ot">&lt;-</span> GSE176021_meta <span class="sc">%&gt;%</span></span>
<span id="cb402-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb402-3" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">meta =</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>, <span class="st">&quot;response&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb402-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb402-4" tabindex="-1"></a>  <span class="fu">select</span>(meta, <span class="fu">everything</span>())</span></code></pre></div>
<p>Now, let’s see how it looks like:</p>
<div class="sourceCode" id="cb403"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb403-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb403-1" tabindex="-1"></a><span class="co"># Display the first two rows and first ten columns of the tidied metadata</span></span>
<span id="cb403-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb403-2" tabindex="-1"></a>GSE176021_meta[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>## # A tibble: 2 × 10
##   meta     GSM5352886     GSM5352887 GSM5352888 GSM5352889 GSM5352890 GSM5352891
##   &lt;chr&gt;    &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;     
## 1 id       patientid: MD… patientid… patientid… patientid… patientid… patientid…
## 2 response response stat… response … response … response … response … response …
## # ℹ 3 more variables: GSM5352892 &lt;chr&gt;, GSM5352893 &lt;chr&gt;, GSM5352894 &lt;chr&gt;</code></pre>
<p>The metadata is ready to be shaped into a long format.</p>
</div>
<div id="converting-to-a-tidy-data-frame" class="section level3" number="9.8.4">
<h3><span class="header-section-number">9.8.4</span> Converting to a Tidy Data Frame</h3>
<p>To fulfill the three tidy data rules, we will convert the metadata into a tidy data frame using the <code>pivot_longer</code> and <code>pivot_wider</code> functions:</p>
<div class="sourceCode" id="cb405"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb405-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb405-1" tabindex="-1"></a><span class="co"># pivot it to long format</span></span>
<span id="cb405-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb405-2" tabindex="-1"></a>GSE176021_meta <span class="sc">%&gt;%</span></span>
<span id="cb405-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb405-3" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>meta)</span></code></pre></div>
<pre><code>## # A tibble: 220 × 3
##    meta  name       value               
##    &lt;chr&gt; &lt;chr&gt;      &lt;chr&gt;               
##  1 id    GSM5352886 patientid: MD01-024 
##  2 id    GSM5352887 patientid: MD01-010 
##  3 id    GSM5352888 patientid: MD01-010 
##  4 id    GSM5352889 patientid: MD01-004 
##  5 id    GSM5352890 patientid: MD01-004 
##  6 id    GSM5352891 patientid: MD01-004 
##  7 id    GSM5352892 patientid: MD01-004 
##  8 id    GSM5352893 patientid: MD043-011
##  9 id    GSM5352894 patientid: MD043-011
## 10 id    GSM5352895 patientid: MD043-011
## # ℹ 210 more rows</code></pre>
<div class="sourceCode" id="cb407"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb407-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb407-1" tabindex="-1"></a><span class="co"># pivot to wide format</span></span>
<span id="cb407-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb407-2" tabindex="-1"></a>GSE176021_meta <span class="sc">%&gt;%</span></span>
<span id="cb407-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb407-3" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>meta) <span class="sc">%&gt;%</span></span>
<span id="cb407-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb407-4" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> meta, <span class="at">values_from =</span> value)</span></code></pre></div>
<pre><code>## # A tibble: 110 × 3
##    name       id                   response                
##    &lt;chr&gt;      &lt;chr&gt;                &lt;chr&gt;                   
##  1 GSM5352886 patientid: MD01-024  response status: Non-MPR
##  2 GSM5352887 patientid: MD01-010  response status: MPR    
##  3 GSM5352888 patientid: MD01-010  response status: MPR    
##  4 GSM5352889 patientid: MD01-004  response status: Non-MPR
##  5 GSM5352890 patientid: MD01-004  response status: Non-MPR
##  6 GSM5352891 patientid: MD01-004  response status: Non-MPR
##  7 GSM5352892 patientid: MD01-004  response status: Non-MPR
##  8 GSM5352893 patientid: MD043-011 response status: Non-MPR
##  9 GSM5352894 patientid: MD043-011 response status: Non-MPR
## 10 GSM5352895 patientid: MD043-011 response status: Non-MPR
## # ℹ 100 more rows</code></pre>
<div class="sourceCode" id="cb409"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb409-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb409-1" tabindex="-1"></a><span class="co"># put it together</span></span>
<span id="cb409-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb409-2" tabindex="-1"></a>tidy_data <span class="ot">&lt;-</span> GSE176021_meta <span class="sc">%&gt;%</span></span>
<span id="cb409-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb409-3" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>meta) <span class="sc">%&gt;%</span></span>
<span id="cb409-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb409-4" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> meta, <span class="at">values_from =</span> value)</span></code></pre></div>
<p>Let’s take a look at the resulting tidy data frame:</p>
<div class="sourceCode" id="cb410"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb410-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb410-1" tabindex="-1"></a><span class="co"># Display the first 10 rows of the tidy data frame</span></span>
<span id="cb410-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb410-2" tabindex="-1"></a><span class="fu">head</span>(tidy_data, <span class="dv">10</span>)</span></code></pre></div>
<pre><code>## # A tibble: 10 × 3
##    name       id                   response                
##    &lt;chr&gt;      &lt;chr&gt;                &lt;chr&gt;                   
##  1 GSM5352886 patientid: MD01-024  response status: Non-MPR
##  2 GSM5352887 patientid: MD01-010  response status: MPR    
##  3 GSM5352888 patientid: MD01-010  response status: MPR    
##  4 GSM5352889 patientid: MD01-004  response status: Non-MPR
##  5 GSM5352890 patientid: MD01-004  response status: Non-MPR
##  6 GSM5352891 patientid: MD01-004  response status: Non-MPR
##  7 GSM5352892 patientid: MD01-004  response status: Non-MPR
##  8 GSM5352893 patientid: MD043-011 response status: Non-MPR
##  9 GSM5352894 patientid: MD043-011 response status: Non-MPR
## 10 GSM5352895 patientid: MD043-011 response status: Non-MPR</code></pre>
<p>Now, our data fulfills the three tidy data rules, making it suitable for analysis.</p>
</div>
<div id="bonus-reading-multiple-tsv-files" class="section level3" number="9.8.5">
<h3><span class="header-section-number">9.8.5</span> Bonus: Reading Multiple TSV Files</h3>
<p>If you have multiple TSV (Tab-Separated Values) files in a directory that you want to read into R and combine, here’s how you can do it:</p>
<div class="sourceCode" id="cb412"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb412-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb412-1" tabindex="-1"></a><span class="co"># List all TSV files in the current directory</span></span>
<span id="cb412-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb412-2" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">as.list</span>(<span class="fu">dir</span>(<span class="st">&quot;.&quot;</span>, <span class="at">pattern =</span> <span class="st">&quot;.tsv&quot;</span>))</span>
<span id="cb412-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb412-3" tabindex="-1"></a></span>
<span id="cb412-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb412-4" tabindex="-1"></a><span class="co"># Read and combine all TSV files into a single data frame</span></span>
<span id="cb412-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb412-5" tabindex="-1"></a>datlist <span class="ot">&lt;-</span> <span class="fu">lapply</span>(files, <span class="cf">function</span>(f) {</span>
<span id="cb412-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb412-6" tabindex="-1"></a>  dat <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(f, <span class="at">col_names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb412-7"><a href="introduction-to-the-tidyverse-ecosystem.html#cb412-7" tabindex="-1"></a>  dat<span class="sc">$</span>sample <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;.tsv&quot;</span>, <span class="st">&quot;&quot;</span>, f)</span>
<span id="cb412-8"><a href="introduction-to-the-tidyverse-ecosystem.html#cb412-8" tabindex="-1"></a>  <span class="fu">return</span>(dat)</span>
<span id="cb412-9"><a href="introduction-to-the-tidyverse-ecosystem.html#cb412-9" tabindex="-1"></a>})</span>
<span id="cb412-10"><a href="introduction-to-the-tidyverse-ecosystem.html#cb412-10" tabindex="-1"></a></span>
<span id="cb412-11"><a href="introduction-to-the-tidyverse-ecosystem.html#cb412-11" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, datlist)</span></code></pre></div>
<p>Alternatively, you can use the <code>bind_rows</code> function from the dplyr package:</p>
<div class="sourceCode" id="cb413"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb413-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb413-1" tabindex="-1"></a><span class="co"># Read and combine all TSV files into a single data frame using bind_rows</span></span>
<span id="cb413-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb413-2" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(datlist, <span class="at">.id =</span> <span class="st">&quot;sample&quot;</span>)</span></code></pre></div>
<p>If your files have a common column (e.g., “GeneID”), and you want to create a single data frame with gene IDs and raw counts, you can use the reduce function from the purrr package:</p>
<div class="sourceCode" id="cb414"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb414-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb414-1" tabindex="-1"></a><span class="co"># Combine data frames using reduce and left_join</span></span>
<span id="cb414-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb414-2" tabindex="-1"></a>CCLE_counts <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">reduce</span>(datlist, left_join, <span class="at">by =</span> <span class="st">&quot;GeneID&quot;</span>)</span></code></pre></div>
<p>Watch this video:</p>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/gIQe5kGEiIA" width="533" height="300" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
</div>
</div>
<div id="section-complete-3" class="section level2" number="9.9">
<h2><span class="header-section-number">9.9</span> Section Complete</h2>
<p>Congratulations on completing this segment of the course!</p>
<p>We’ve covered the Tidyverse, data manipulation, and analysis techniques, including handling GEO metadata and using purrr for functional programming. These skills are crucial for efficient data analysis in R, simplifying workflows and cleaning up complex datasets.</p>
<p>As you move forward in the course, remember to utilize the Q&amp;A section and comments for support. Engage actively for any clarifications or assistance you need.</p>
<div id="conclusion-21" class="section level3" number="9.9.1">
<h3><span class="header-section-number">9.9.1</span> Conclusion</h3>
<p>In this lesson, we learned how to tidy metadata from GEO using Tidyverse in R. Tidying data is a crucial step in data preparation for analysis. We also explored how to read and combine multiple TSV files from a directory, which is a common task when dealing with large datasets.</p>
<p>Remember that these skills are valuable in many data analysis and bioinformatics tasks, allowing you to work efficiently with real-world data.</p>

</div>
</div>
<div id="data-visualization-with-ggplot2" class="section level2" number="9.10">
<h2><span class="header-section-number">9.10</span> Data visualization with ggplot2</h2>
<p>In this lesson, we’ll learn how to create basic data visualizations in R using ggplot2. As a biology student, developing your data visualization skills is crucial for exploring trends and communicating findings.</p>
<p>If you read genomics papers, most of the figures are fell into several categories:</p>
<ul>
<li><p>A bar plot</p></li>
<li><p>A line plot</p></li>
<li><p>A scatter plot</p></li>
<li><p>A boxplot or violin plot</p></li>
<li><p>A histogram</p></li>
<li><p>A heatmap</p></li>
</ul>
<p>If you master those six types of figures, you can reproduce 90% of the figures in any genomics paper. Watch this:</p>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/l0xueHaMY3A" width="533" height="300" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
<p>ggplot2 is a powerful R package for flexible and professional-quality graphics. The key is thinking about visualization in layers:</p>
<ul>
<li><p>The data layer - the data frame you want to plot</p></li>
<li><p>The aesthetic mapping - how data columns map to visual properties like x/y position, color, size etc.</p></li>
<li><p>Geometric objects like points, lines, boxes - the type of plot</p></li>
<li><p>Facets - panels for subgrouping the data</p></li>
<li><p>Stats transformations like group means, regressions etc.</p></li>
<li><p>Themes to refine the final look</p></li>
</ul>
<p>Let’s walk through examples of common plot types.</p>
</div>
<div id="creating-scatter-plots" class="section level2" number="9.11">
<h2><span class="header-section-number">9.11</span> Creating Scatter Plots</h2>
<p>In this lesson, we will delve into data visualization and statistical analysis using R. We will work with a real-world dataset related to gene expression from The Cancer Genome Atlas (TCGA) and learn how to create scatter plots, calculate correlation coefficients, and visualize regression lines. These skills are fundamental for understanding relationships between variables in your data, which can be crucial for making data-driven conclusions.</p>
<div id="prerequisites-1" class="section level3" number="9.11.1">
<h3><span class="header-section-number">9.11.1</span> Prerequisites</h3>
<p>Before we begin, make sure you have R installed on your computer, along with the following R packages: readr, dplyr, and ggplot2. You can install these packages using the <code>install.packages()</code> function if you haven’t already.</p>
<div class="sourceCode" id="cb415"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb415-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb415-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;readr&quot;</span>)</span>
<span id="cb415-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb415-2" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;dplyr&quot;</span>)</span>
<span id="cb415-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb415-3" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;ggplot2&quot;</span>)</span></code></pre></div>
<p>Also, download the same TCGA gene expression data (CSV file) to your working directory or specify the correct file path.</p>
<blockquote>
<p>Download the file at <a href="https://osf.io/yeun5" class="uri">https://osf.io/yeun5</a></p>
</blockquote>
</div>
<div id="loading-and-exploring-the-data" class="section level3" number="9.11.2">
<h3><span class="header-section-number">9.11.2</span> Loading and Exploring the Data</h3>
<p>We’ll start by loading the TCGA gene expression data into R using the read_csv function from the readr package. This dataset contains information about various genes’ expression levels across different samples, including cancer, metastatic, and normal samples.</p>
<div class="sourceCode" id="cb416"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb416-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb416-1" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb416-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb416-2" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb416-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb416-3" tabindex="-1"></a></span>
<span id="cb416-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb416-4" tabindex="-1"></a><span class="co"># Load the TCGA gene expression data</span></span>
<span id="cb416-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb416-5" tabindex="-1"></a>tcga_data <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;~/Downloads/TCGA_cancer_genes_expression.csv&quot;</span>)</span></code></pre></div>
<p>Let’s take a look at the first few rows of the dataset to understand its structure.</p>
<div class="sourceCode" id="cb417"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb417-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb417-1" tabindex="-1"></a><span class="fu">head</span>(tcga_data)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 25
##   ...1      TACSTD2  VTCN1   MUC1 NECTIN4 FOLH1  FOLR1 CD276   MSLN  CLDN6 ERBB2
##   &lt;chr&gt;       &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;
## 1 43e715bf…   0.704 0      0.675   0.0862  7.21 0       52.8 0.0667 0.0970  1.88
## 2 1a5db9fc…  25.4   0      2.02    0.0728 23.6  0.122   78.8 0.956  0.255   7.78
## 3 93b382e4…   1.58  0      0.908   0.699   2.85 1.01   146.  0.0456 0.257   2.91
## 4 1f39dadd…   0.270 0.0910 0.0429  0.0165  1.16 0.279   48.5 0.0315 0.247   4.91
## 5 8c8c09b9…   0.412 0      0.115   0.0317  2.41 0.0492  42.3 0.270  0.126   1.49
## 6 85a86b91…   4.55  4.86   0.0421  0.0683  1.01 0.0225  20.6 0.0134 0.0182 13.5 
## # ℹ 14 more variables: MUC16 &lt;dbl&gt;, DLL3 &lt;dbl&gt;, CEACAM5 &lt;dbl&gt;, PVR &lt;dbl&gt;,
## #   EPCAM &lt;dbl&gt;, PROM1 &lt;dbl&gt;, CD24 &lt;dbl&gt;, EGFR &lt;dbl&gt;, MET &lt;dbl&gt;,
## #   TNFRSF10B &lt;dbl&gt;, tcga.tcga_barcode &lt;chr&gt;,
## #   tcga.cgc_sample_sample_type &lt;chr&gt;, study &lt;chr&gt;, sample_type &lt;chr&gt;</code></pre>
<p>Now, you should see a table with multiple columns, where each row represents a sample. The columns represent different genes’ expression levels, and we also have columns indicating the sample type, study, and more.</p>
</div>
<div id="filtering-the-data" class="section level3" number="9.11.3">
<h3><span class="header-section-number">9.11.3</span> Filtering the Data</h3>
<p>For our analysis, we want to focus on cancer samples only. Let’s filter the dataset to include only these samples and exclude normal and metastatic samples.</p>
<div class="sourceCode" id="cb419"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb419-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb419-1" tabindex="-1"></a><span class="fu">table</span>(tcga_data<span class="sc">$</span>sample_type)</span></code></pre></div>
<pre><code>## 
##     cancer metastatic     normal 
##      10021        394        740</code></pre>
<div class="sourceCode" id="cb421"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb421-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb421-1" tabindex="-1"></a><span class="co"># Filter the data to include only cancer samples</span></span>
<span id="cb421-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb421-2" tabindex="-1"></a>tcga_cancer <span class="ot">&lt;-</span> tcga_data <span class="sc">%&gt;%</span></span>
<span id="cb421-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb421-3" tabindex="-1"></a>  <span class="fu">filter</span>(sample_type <span class="sc">==</span> <span class="st">&quot;cancer&quot;</span>)</span></code></pre></div>
<p>We use the <code>%&gt;%</code> operator to chain commands together. Here, we first filter the dataset to include only rows where sample_type is “cancer.”</p>
</div>
<div id="creating-a-scatter-plot" class="section level3" number="9.11.4">
<h3><span class="header-section-number">9.11.4</span> Creating a Scatter Plot</h3>
<blockquote>
<p>For docs check: <a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="uri">https://ggplot2.tidyverse.org/reference/ggplot.html</a></p>
</blockquote>
<p>Now, let’s create a scatter plot to visualize the relationship between two gene expression levels. We’ll plot the expression of the <code>EPCAM</code> gene on the x-axis and the <code>CEACAM5</code> gene (CEA) on the y-axis.</p>
<blockquote>
<p>Docs on geom_point: <a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="uri">https://ggplot2.tidyverse.org/reference/geom_point.html</a></p>
</blockquote>
<div class="sourceCode" id="cb422"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb422-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb422-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb422-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb422-2" tabindex="-1"></a></span>
<span id="cb422-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb422-3" tabindex="-1"></a><span class="co"># Create a scatter plot</span></span>
<span id="cb422-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb422-4" tabindex="-1"></a><span class="fu">ggplot</span>(tcga_cancer, <span class="fu">aes</span>(<span class="at">x =</span> EPCAM, <span class="at">y =</span> CEACAM5)) <span class="sc">+</span></span>
<span id="cb422-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb422-5" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-220-1.png" width="672" /></p>
<p>In the code above:</p>
<ul>
<li><p>We use ggplot() to initialize a new ggplot2 plot.</p></li>
<li><p>Inside aes(), we specify the aesthetic mappings, where x represents the <code>EPCAM</code> gene expression and y represents the <code>CEACAM5</code> gene expression.</p></li>
<li><p>We add `<code>geom_point()</code> to plot the data points on the graph.</p></li>
</ul>
<div id="interpretation" class="section level4" number="9.11.4.1">
<h4><span class="header-section-number">9.11.4.1</span> Interpretation</h4>
<p>The scatter plot visually represents the relationship between the <code>EPCAM</code> and <code>CEACAM5</code> gene expression levels. Each point on the plot corresponds to a cancer sample, with its EPCAM expression on the x-axis and CEACAM5 expression on the y-axis. If the points tend to fall on the diagonal line, it suggests a relationship between the two gene expressions.</p>
</div>
</div>
<div id="calculating-correlation" class="section level3" number="9.11.5">
<h3><span class="header-section-number">9.11.5</span> Calculating Correlation</h3>
<p>To quantify the relationship between <code>EPCAM</code> and <code>CEACAM5</code> gene expressions, we can calculate the Pearson correlation coefficient. This coefficient measures the strength and direction of the linear relationship between two variables.</p>
<div class="sourceCode" id="cb423"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb423-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb423-1" tabindex="-1"></a><span class="co"># Calculate the Pearson correlation coefficient</span></span>
<span id="cb423-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb423-2" tabindex="-1"></a>correlation <span class="ot">&lt;-</span> <span class="fu">cor</span>(tcga_cancer<span class="sc">$</span>EPCAM, tcga_cancer<span class="sc">$</span>CEACAM5)</span>
<span id="cb423-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb423-3" tabindex="-1"></a>correlation</span></code></pre></div>
<pre><code>## [1] 0.6324328</code></pre>
<p>The output will be a value between -1 and 1, where:</p>
<p><img src="images/cor.png" /></p>
<ul>
<li><p>A positive value (closer to 1) indicates a positive correlation (both variables increase together).</p></li>
<li><p>A negative value (closer to -1) indicates a negative correlation (one variable increases as the other decreases).</p></li>
<li><p>A value close to 0 indicates little to no linear correlation.</p></li>
</ul>
<p>In our example, the correlation coefficient (Pearson’s r) is approximately <code>0.6324</code>, which suggests a moderately positive correlation between <code>EPCAM</code> and <code>CEACAM5</code> gene expressions among cancer samples.</p>
</div>
<div id="hypothesis-testing" class="section level3" number="9.11.6">
<h3><span class="header-section-number">9.11.6</span> Hypothesis Testing</h3>
<blockquote>
<p>Check docs here: <a href="https://rdrr.io/r/stats/cor.test.html" class="uri">https://rdrr.io/r/stats/cor.test.html</a></p>
</blockquote>
<p>To determine if this correlation is statistically significant, we can perform a hypothesis test. In our case, we’re interested in testing whether the correlation is significantly different from zero.</p>
<div class="sourceCode" id="cb425"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb425-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb425-1" tabindex="-1"></a><span class="co"># Perform a correlation hypothesis test</span></span>
<span id="cb425-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb425-2" tabindex="-1"></a>cor_test_result <span class="ot">&lt;-</span> <span class="fu">cor.test</span>(tcga_cancer<span class="sc">$</span>EPCAM, tcga_cancer<span class="sc">$</span>CEACAM5)</span>
<span id="cb425-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb425-3" tabindex="-1"></a>cor_test_result</span></code></pre></div>
<pre><code>## 
##  Pearson&#39;s product-moment correlation
## 
## data:  tcga_cancer$EPCAM and tcga_cancer$CEACAM5
## t = 81.722, df = 10019, p-value &lt; 2.2e-16
## alternative hypothesis: true correlation is not equal to 0
## 95 percent confidence interval:
##  0.6205372 0.6440374
## sample estimates:
##       cor 
## 0.6324328</code></pre>
<p>The output will provide various statistics, including the t-value, degrees of freedom, and the p-value.</p>
<div id="interpretation-1" class="section level4" number="9.11.6.1">
<h4><span class="header-section-number">9.11.6.1</span> Interpretation</h4>
<p>In the results, you’ll see:</p>
<ul>
<li><p>The t-value, which measures the number of standard errors the correlation coefficient is away from zero.</p></li>
<li><p>The degrees of freedom (df), which are related to the sample size.</p></li>
<li><p>The p-value, which tells us whether the correlation is statistically significant.</p></li>
</ul>
<p>In our case, the p-value is very close to zero (<code>p-value &lt; 2.2e-16</code>), indicating strong evidence against the null hypothesis (the true correlation is zero). Therefore, we can conclude that the correlation between EPCAM and CEACAM5 gene expressions is statistically significant. You need to keep in mind that in genomics data analysis, typically you have thousands of samples and you will inherently get tiny p values. In our case, focus on the effect size (in this case, the coefficient value which is 0.63).</p>
</div>
</div>
<div id="adding-a-regression-line" class="section level3" number="9.11.7">
<h3><span class="header-section-number">9.11.7</span> Adding a Regression Line</h3>
<blockquote>
<p>Check docs on geom_smooth here: <a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="uri">https://ggplot2.tidyverse.org/reference/geom_smooth.html</a></p>
</blockquote>
<p>To further explore the relationship between the two gene expressions, we can add a linear regression line to our scatter plot using <code>geom_smooth()</code>.</p>
<div class="sourceCode" id="cb427"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb427-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb427-1" tabindex="-1"></a><span class="co"># Create a scatter plot with a regression line</span></span>
<span id="cb427-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb427-2" tabindex="-1"></a><span class="fu">ggplot</span>(tcga_cancer, <span class="fu">aes</span>(<span class="at">x =</span> EPCAM, <span class="at">y =</span> CEACAM5)) <span class="sc">+</span></span>
<span id="cb427-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb427-3" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb427-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb427-4" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-223-1.png" width="672" /></p>
<p>The <code>geom_smooth()</code> function with <code>method = "lm"</code> fits a linear regression line to the data points, helping us visualize the trend more clearly.</p>
<p>The regression line provides a visual representation of how one gene’s expression (EPCAM) changes concerning the other (CEACAM5). If the line slopes upward, it suggests a positive correlation, while a downward slope indicates a negative correlation.</p>
</div>
<div id="conclusion-22" class="section level3" number="9.11.8">
<h3><span class="header-section-number">9.11.8</span> Conclusion</h3>
<p>In this lesson, we’ve covered the basics of creating scatter plots, calculating correlation coefficients, and performing hypothesis tests using R. These skills are essential for exploring relationships between variables in your data, whether you’re analyzing gene expressions, financial data, or any other dataset. Remember that correlation does not imply causation, so it’s essential to interpret your findings carefully and consider the context of your analysis.</p>
</div>
</div>
<div id="understanding-distributions-with-histograms" class="section level2" number="9.12">
<h2><span class="header-section-number">9.12</span> Understanding Distributions with Histograms</h2>
<p>In this section, we will explore another powerful data visualization tool: histograms. Histograms are especially useful for understanding the distribution of a single numerical variable. We will use R and the ggplot2 package to create histograms and customize them to gain insights into the distribution of gene expression levels.</p>
<p>For this course we’re using the data we generated in previous lesson.</p>
<div id="creating-a-basic-histogram" class="section level3" number="9.12.1">
<h3><span class="header-section-number">9.12.1</span> Creating a Basic Histogram</h3>
<p>Let’s start by creating a basic histogram to visualize the distribution of the EPCAM gene expression levels across all cancer samples.</p>
<div class="sourceCode" id="cb428"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb428-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb428-1" tabindex="-1"></a><span class="co"># Create a basic histogram for EPCAM gene expression</span></span>
<span id="cb428-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb428-2" tabindex="-1"></a><span class="fu">ggplot</span>(tcga_cancer, <span class="fu">aes</span>(<span class="at">x =</span> EPCAM)) <span class="sc">+</span></span>
<span id="cb428-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb428-3" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-224-1.png" width="672" /></p>
<p>In this code:</p>
<ul>
<li><p>We use <code>ggplot()</code> to initialize a new ggplot2 plot.</p></li>
<li><p>Inside <code>aes()</code>, we specify that we want to map the <code>EPCAM</code> gene expression values to the x-axis.</p></li>
<li><p>We add <code>geom_histogram()</code> to create the histogram.</p></li>
</ul>
<p>The resulting plot will display the EPCAM gene expression levels on the x-axis and the count of samples falling into each “bin” on the y-axis. A bin is a range of values, and the height of each bar represents how many samples have expression levels within that range.</p>
</div>
<div id="customizing-histogram-bins" class="section level3" number="9.12.2">
<h3><span class="header-section-number">9.12.2</span> Customizing Histogram Bins</h3>
<p>You can customize the granularity of your histogram by changing the number of bins or specifying the bin size. This allows you to get a more detailed or broader view of the data distribution.</p>
<div id="changing-the-number-of-bins" class="section level4" number="9.12.2.1">
<h4><span class="header-section-number">9.12.2.1</span> Changing the Number of Bins</h4>
<p>To change the number of bins, you can use the bins parameter within geom_histogram(). Increasing the number of bins provides more detail.</p>
<div class="sourceCode" id="cb429"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb429-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb429-1" tabindex="-1"></a><span class="co"># Create a histogram with 50 bins</span></span>
<span id="cb429-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb429-2" tabindex="-1"></a><span class="fu">ggplot</span>(tcga_cancer, <span class="fu">aes</span>(<span class="at">x =</span> EPCAM)) <span class="sc">+</span></span>
<span id="cb429-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb429-3" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-225-1.png" width="672" /></p>
</div>
<div id="customizing-bin-borders" class="section level4" number="9.12.2.2">
<h4><span class="header-section-number">9.12.2.2</span> Customizing Bin Borders</h4>
<p>By default, the borders of the bins in the histogram are not very visible. You can change the border color to white to make them more distinct.</p>
<div class="sourceCode" id="cb430"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb430-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb430-1" tabindex="-1"></a><span class="co"># Create a histogram with white bin borders</span></span>
<span id="cb430-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb430-2" tabindex="-1"></a><span class="fu">ggplot</span>(tcga_cancer, <span class="fu">aes</span>(<span class="at">x =</span> EPCAM)) <span class="sc">+</span></span>
<span id="cb430-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb430-3" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-226-1.png" width="672" /></p>
<p>Changing the bin border color to white makes it easier to distinguish between adjacent bins.</p>
</div>
</div>
<div id="conclusion-23" class="section level3" number="9.12.3">
<h3><span class="header-section-number">9.12.3</span> Conclusion</h3>
<p>Histograms are valuable tools for visualizing the distribution of a single numerical variable, helping you understand the underlying data structure. By customizing the number of bins, bin sizes, and bin borders, you can tailor your histograms to reveal the level of detail you need. Whether you are analyzing gene expression data or any other quantitative data, histograms are an essential part of your data exploration toolkit.</p>
</div>
</div>
<div id="visualizing-data-distribution-with-boxplots-and-violin-plots" class="section level2" number="9.13">
<h2><span class="header-section-number">9.13</span> Visualizing Data Distribution with Boxplots and Violin Plots</h2>
<p>In this lesson, we will explore how to visualize the distribution of gene expression data across different cancer types using boxplots and violin plots in R. These graphical tools are invaluable for gaining insights into the spread and central tendency of data within different categories.</p>
<p>We will continue to work with the TCGA gene expression dataset, specifically focusing on cancer samples (tcga_cancer). This dataset contains various gene expression measurements across different cancer types.</p>
<div id="creating-a-basic-boxplot" class="section level3" number="9.13.1">
<h3><span class="header-section-number">9.13.1</span> Creating a Basic Boxplot</h3>
<p>Let’s start by creating a basic boxplot to visualize the distribution of the EPCAM gene expression across different cancer types.</p>
<div class="sourceCode" id="cb431"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb431-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb431-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb431-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb431-2" tabindex="-1"></a></span>
<span id="cb431-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb431-3" tabindex="-1"></a><span class="fu">ggplot</span>(tcga_cancer, <span class="fu">aes</span>(<span class="at">x =</span> study, <span class="at">y =</span> EPCAM)) <span class="sc">+</span></span>
<span id="cb431-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb431-4" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-227-1.png" width="672" /></p>
<p>In this code:</p>
<ul>
<li><p>We use <code>ggplot()</code> to initialize the plot.</p></li>
<li><p>We map the x-axis to study (representing cancer types) and the y-axis to EPCAM gene expression.</p></li>
<li><p>We add <code>geom_boxplot()</code> to create the boxplots.</p></li>
</ul>
</div>
<div id="rotating-x-axis-labels" class="section level3" number="9.13.2">
<h3><span class="header-section-number">9.13.2</span> Rotating X-axis Labels</h3>
<p>You may notice that the x-axis labels (cancer types) overlap. To make the plot more readable, we can rotate the x-axis labels by 90 degrees and remove the x-axis label using the <code>theme</code> function.</p>
<div class="sourceCode" id="cb432"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb432-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb432-1" tabindex="-1"></a><span class="fu">ggplot</span>(tcga_cancer, <span class="fu">aes</span>(<span class="at">x =</span> study, <span class="at">y =</span> EPCAM)) <span class="sc">+</span></span>
<span id="cb432-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb432-2" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb432-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb432-3" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb432-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb432-4" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/test-1.png" width="672" />
Now, the x-axis labels are more legible.</p>
</div>
<div id="adding-color-to-boxplots" class="section level3" number="9.13.3">
<h3><span class="header-section-number">9.13.3</span> Adding Color to Boxplots</h3>
<p>To distinguish between cancer types more effectively, let’s add color to the boxplots by mapping the color aesthetic to the study.</p>
<div class="sourceCode" id="cb433"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb433-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb433-1" tabindex="-1"></a><span class="fu">ggplot</span>(tcga_cancer, <span class="fu">aes</span>(<span class="at">x =</span> study, <span class="at">y =</span> EPCAM)) <span class="sc">+</span></span>
<span id="cb433-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb433-2" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">color =</span> study)) <span class="sc">+</span></span>
<span id="cb433-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb433-3" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb433-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb433-4" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-228-1.png" width="672" /></p>
<p>Alternatively, you can use <code>fill</code> to color the boxes instead of the outlines:</p>
<div class="sourceCode" id="cb434"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb434-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb434-1" tabindex="-1"></a><span class="fu">ggplot</span>(tcga_cancer, <span class="fu">aes</span>(<span class="at">x =</span> study, <span class="at">y =</span> EPCAM)) <span class="sc">+</span></span>
<span id="cb434-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb434-2" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">fill =</span> study)) <span class="sc">+</span></span>
<span id="cb434-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb434-3" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb434-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb434-4" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-229-1.png" width="672" /></p>
<p>By default, ggplot2 uses a default color palette, but you can specify colors manually if needed.</p>
</div>
<div id="customizing-color-palettes" class="section level3" number="9.13.4">
<h3><span class="header-section-number">9.13.4</span> Customizing Color Palettes</h3>
<p>In case you want to define your color palette for the cancer types, you can use the Polychrome package.</p>
<div class="sourceCode" id="cb435"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb435-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb435-1" tabindex="-1"></a><span class="co"># there are total 32 cancer types</span></span>
<span id="cb435-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb435-2" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(tcga_cancer<span class="sc">$</span>study))</span></code></pre></div>
<pre><code>## [1] 32</code></pre>
<p>Here’s how to create a custom color palette for 32 cancer types:</p>
<div class="sourceCode" id="cb437"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb437-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb437-1" tabindex="-1"></a><span class="co"># install.packages(&quot;Polychrome&quot;)</span></span>
<span id="cb437-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb437-2" tabindex="-1"></a><span class="fu">library</span>(Polychrome)</span>
<span id="cb437-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb437-3" tabindex="-1"></a></span>
<span id="cb437-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb437-4" tabindex="-1"></a><span class="co"># There are a total of 32 cancer types</span></span>
<span id="cb437-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb437-5" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(tcga_cancer<span class="sc">$</span>study))</span></code></pre></div>
<pre><code>## [1] 32</code></pre>
<div class="sourceCode" id="cb439"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb439-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb439-1" tabindex="-1"></a><span class="co"># Create a custom color palette with Polychrome</span></span>
<span id="cb439-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb439-2" tabindex="-1"></a>P32 <span class="ot">&lt;-</span> <span class="fu">createPalette</span>(<span class="dv">32</span>, <span class="fu">c</span>(<span class="st">&quot;#FF0000&quot;</span>, <span class="st">&quot;#00FF00&quot;</span>, <span class="st">&quot;#0000FF&quot;</span>), <span class="at">range =</span> <span class="fu">c</span>(<span class="dv">30</span>, <span class="dv">80</span>))</span>
<span id="cb439-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb439-3" tabindex="-1"></a></span>
<span id="cb439-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb439-4" tabindex="-1"></a>P32</span></code></pre></div>
<pre><code>##       NC1       NC2       NC3       NC4       NC5       NC6       NC7       NC8 
## &quot;#FE162A&quot; &quot;#2EDE1C&quot; &quot;#0000FE&quot; &quot;#6A3B3D&quot; &quot;#FC00E1&quot; &quot;#0DD7F6&quot; &quot;#FFB100&quot; &quot;#D1A1F7&quot; 
##       NC9      NC10      NC11      NC12      NC13      NC14      NC15      NC16 
## &quot;#FE659A&quot; &quot;#9ED1A2&quot; &quot;#164D75&quot; &quot;#BED000&quot; &quot;#C500FB&quot; &quot;#9D0D81&quot; &quot;#169BF9&quot; &quot;#EE6616&quot; 
##      NC17      NC18      NC19      NC20      NC21      NC22      NC23      NC24 
## &quot;#FDAF9F&quot; &quot;#C0B2C3&quot; &quot;#867716&quot; &quot;#165638&quot; &quot;#94350D&quot; &quot;#532EAB&quot; &quot;#FFA4D7&quot; &quot;#0D911C&quot; 
##      NC25      NC26      NC27      NC28      NC29      NC30      NC31      NC32 
## &quot;#1CDCC6&quot; &quot;#FE007E&quot; &quot;#FA16B2&quot; &quot;#F97FF0&quot; &quot;#A0BFF0&quot; &quot;#764088&quot; &quot;#B5002A&quot; &quot;#569EA3&quot;</code></pre>
<p>You see the <a href="https://en.wikipedia.org/wiki/Web_colors">hex code</a> for the colors.</p>
<p>Now, we have a custom color palette P32 with 32 unique colors. You can visualize the colors using <code>swatch(P32)</code> in the polychrom package.</p>
<div class="sourceCode" id="cb441"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb441-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb441-1" tabindex="-1"></a><span class="fu">swatch</span>(P32)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-232-1.png" width="672" /></p>
</div>
<div id="applying-custom-color-palette" class="section level3" number="9.13.5">
<h3><span class="header-section-number">9.13.5</span> Applying Custom Color Palette</h3>
<p>To use the custom color palette with your boxplot, use <code>scale_fill_manual()</code> or <code>scale_color_manual()</code> to map the colors manually to the study variable.</p>
<div class="sourceCode" id="cb442"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb442-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb442-1" tabindex="-1"></a><span class="fu">ggplot</span>(tcga_cancer, <span class="fu">aes</span>(<span class="at">x =</span> study, <span class="at">y =</span> EPCAM)) <span class="sc">+</span></span>
<span id="cb442-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb442-2" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">fill =</span> study)) <span class="sc">+</span></span>
<span id="cb442-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb442-3" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">unname</span>(P32)) <span class="sc">+</span></span>
<span id="cb442-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb442-4" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb442-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb442-5" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-233-1.png" width="672" />
Note, you need to remove the names of the named color vector using <code>unname</code>.</p>
<p>Now, the boxplot colors are based on your custom palette, making it more visually appealing and distinct.</p>
</div>
<div id="reordering-boxplots" class="section level3" number="9.13.6">
<h3><span class="header-section-number">9.13.6</span> Reordering Boxplots</h3>
<p>To reorder the boxes according to the median level of EPCAM expression, you can use the <code>forcats::fct_reorder()</code> function within your <code>aes()</code> mapping.</p>
<div class="sourceCode" id="cb443"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb443-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb443-1" tabindex="-1"></a><span class="fu">ggplot</span>(tcga_cancer, <span class="fu">aes</span>(<span class="at">x =</span> study <span class="sc">%&gt;%</span></span>
<span id="cb443-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb443-2" tabindex="-1"></a>                          forcats<span class="sc">::</span><span class="fu">fct_reorder</span>(EPCAM, median), </span>
<span id="cb443-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb443-3" tabindex="-1"></a>                        <span class="at">y =</span> EPCAM)) <span class="sc">+</span></span>
<span id="cb443-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb443-4" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">fill =</span> study)) <span class="sc">+</span></span>
<span id="cb443-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb443-5" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">unname</span>(P32)) <span class="sc">+</span></span>
<span id="cb443-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb443-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb443-7"><a href="introduction-to-the-tidyverse-ecosystem.html#cb443-7" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-234-1.png" width="672" /></p>
<p>The <code>forcats::fct_reorder()</code> function takes two arguments:</p>
<ul>
<li><p>The first argument (study) is the factor variable whose levels we want to reorder.</p></li>
<li><p>The second argument (EPCAM) is the numeric variable based on which the reordering will be done (in this case, the median expression levels of EPCAM).</p></li>
</ul>
<p>As a result of this reordering, the levels of the study factor will be rearranged from low to high median EPCAM expression levels. Now, the boxes are arranged from low to high EPCAM expression levels, providing a clearer view of the data distribution.</p>
</div>
<div id="using-violin-plots" class="section level3" number="9.13.7">
<h3><span class="header-section-number">9.13.7</span> Using Violin Plots</h3>
<p>Violin plots combine a boxplot with a kernel density plot, allowing you to visualize both the summary statistics and the entire data distribution.</p>
<div class="sourceCode" id="cb444"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb444-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb444-1" tabindex="-1"></a><span class="fu">ggplot</span>(tcga_cancer, <span class="fu">aes</span>(<span class="at">x =</span> study <span class="sc">%&gt;%</span></span>
<span id="cb444-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb444-2" tabindex="-1"></a>                          forcats<span class="sc">::</span><span class="fu">fct_reorder</span>(EPCAM, median), </span>
<span id="cb444-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb444-3" tabindex="-1"></a>                        <span class="at">y =</span> EPCAM)) <span class="sc">+</span></span>
<span id="cb444-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb444-4" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="fu">aes</span>(<span class="at">fill =</span> study), <span class="at">scale =</span> <span class="st">&quot;width&quot;</span>) <span class="sc">+</span></span>
<span id="cb444-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb444-5" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">unname</span>(P32)) <span class="sc">+</span></span>
<span id="cb444-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb444-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb444-7"><a href="introduction-to-the-tidyverse-ecosystem.html#cb444-7" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-235-1.png" width="672" /></p>
<p>In this code, we replaced <code>geom_boxplot</code> with <code>geom_violin</code> to create a violin plot. The width of the violin plots is scaled for better visibility.</p>
</div>
<div id="conclusion-24" class="section level3" number="9.13.8">
<h3><span class="header-section-number">9.13.8</span> Conclusion</h3>
<p>In this lesson, we’ve explored how to visualize data distribution using boxplots and violin plots in R. These techniques are essential for understanding the spread and central tendencies of data within different categories or groups. By customizing colors, reordering boxes, and using violin plots, you can effectively communicate insights from your data to others.</p>
</div>
</div>
<div id="creating-bar-plots-to-visualize-median-epcam-levels-by-cancer-type" class="section level2" number="9.14">
<h2><span class="header-section-number">9.14</span> Creating Bar Plots to Visualize Median EPCAM Levels by Cancer Type</h2>
<p>In this section, we will learn how to create informative bar plots using R. Specifically, we’ll use bar plots to display the median levels of the EPCAM gene expression for different types of cancer. This visualization will help us understand how EPCAM expression varies across various cancer studies. We will also explore techniques to reorder and customize our bar plots for better data interpretation.</p>
<div id="calculating-median-epcam-levels-by-cancer-type" class="section level3" number="9.14.1">
<h3><span class="header-section-number">9.14.1</span> Calculating Median EPCAM Levels by Cancer Type</h3>
<p>The median EPCAM expression levels represent the middle value of EPCAM gene expression measurements for different types of cancer studies, providing a central measure to understand how this gene behaves in each type of cancer.</p>
<p>To begin, we’ll calculate the median EPCAM expression levels for each cancer study type. We will create a summary table that includes the cancer study names and their corresponding median EPCAM values.</p>
<div class="sourceCode" id="cb445"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb445-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb445-1" tabindex="-1"></a><span class="co"># Calculate median EPCAM levels by cancer type</span></span>
<span id="cb445-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb445-2" tabindex="-1"></a>EPCAM_median <span class="ot">&lt;-</span> tcga_cancer <span class="sc">%&gt;%</span></span>
<span id="cb445-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb445-3" tabindex="-1"></a>  <span class="fu">group_by</span>(study) <span class="sc">%&gt;%</span></span>
<span id="cb445-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb445-4" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">median_EPCAM =</span> <span class="fu">median</span>(EPCAM))</span>
<span id="cb445-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb445-5" tabindex="-1"></a></span>
<span id="cb445-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb445-6" tabindex="-1"></a><span class="co"># Display the summary table</span></span>
<span id="cb445-7"><a href="introduction-to-the-tidyverse-ecosystem.html#cb445-7" tabindex="-1"></a>EPCAM_median</span></code></pre></div>
<pre><code>## # A tibble: 32 × 2
##    study median_EPCAM
##    &lt;chr&gt;        &lt;dbl&gt;
##  1 ACC          4.83 
##  2 BLCA        81.1  
##  3 BRCA       162.   
##  4 CESC        75.9  
##  5 CHOL       251.   
##  6 COAD       778.   
##  7 DLBC         0.578
##  8 ESCA       189.   
##  9 GBM          0.307
## 10 HNSC        24.7  
## # ℹ 22 more rows</code></pre>
<p>This table now shows the median EPCAM expression levels for each cancer study type, making it easier to compare across different types of cancer.</p>
</div>
<div id="creating-a-basic-bar-plot" class="section level3" number="9.14.2">
<h3><span class="header-section-number">9.14.2</span> Creating a Basic Bar Plot</h3>
<p>Now, let’s create a simple bar plot to visualize these median EPCAM levels. We will use <code>geom_bar()</code> to create bars representing the median values.</p>
<div class="sourceCode" id="cb447"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb447-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb447-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb447-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb447-2" tabindex="-1"></a></span>
<span id="cb447-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb447-3" tabindex="-1"></a><span class="co"># Create a basic bar plot</span></span>
<span id="cb447-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb447-4" tabindex="-1"></a><span class="fu">ggplot</span>(EPCAM_median, <span class="fu">aes</span>(<span class="at">x =</span> study, <span class="at">y =</span> median_EPCAM)) <span class="sc">+</span></span>
<span id="cb447-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb447-5" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">&quot;identity&quot;</span>) <span class="sc">+</span></span>
<span id="cb447-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb447-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb447-7"><a href="introduction-to-the-tidyverse-ecosystem.html#cb447-7" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-237-1.png" width="672" /></p>
<p>In this code:</p>
<ul>
<li><p>We use <code>ggplot()</code> to initialize the plot.</p></li>
<li><p><code>aes()</code> specifies the aesthetic mappings, where <code>x</code> represents the cancer study names, and <code>y</code> represents the median EPCAM values.</p></li>
<li><p><code>geom_bar(stat = "identity")</code> creates bars where the height of each bar corresponds to the median EPCAM value.</p></li>
<li><p><code>theme()</code> is used to customize the appearance of the plot, including rotating the x-axis labels for better readability.</p></li>
</ul>
<p>The basic bar plot displays the median EPCAM levels for each cancer study type, but the bars are not ordered based on the values. We can improve the plot by reordering the bars according to the median EPCAM levels.</p>
</div>
<div id="reordering-the-bars" class="section level3" number="9.14.3">
<h3><span class="header-section-number">9.14.3</span> Reordering the Bars</h3>
<p>To make our bar plot more informative, let’s reorder the bars based on the median EPCAM values. We can use the <code>forcats::fct_reorder()</code> function to achieve this.</p>
<div class="sourceCode" id="cb448"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb448-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb448-1" tabindex="-1"></a><span class="co"># Reorder the bars based on median EPCAM values</span></span>
<span id="cb448-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb448-2" tabindex="-1"></a><span class="fu">ggplot</span>(EPCAM_median, <span class="fu">aes</span>(<span class="at">x =</span> study <span class="sc">%&gt;%</span></span>
<span id="cb448-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb448-3" tabindex="-1"></a>                           forcats<span class="sc">::</span><span class="fu">fct_reorder</span>(median_EPCAM), </span>
<span id="cb448-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb448-4" tabindex="-1"></a>                         <span class="at">y =</span> median_EPCAM)) <span class="sc">+</span></span>
<span id="cb448-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb448-5" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">&quot;identity&quot;</span>) <span class="sc">+</span></span>
<span id="cb448-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb448-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb448-7"><a href="introduction-to-the-tidyverse-ecosystem.html#cb448-7" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-238-1.png" width="672" /></p>
<p>In this code:</p>
<p>We use <code>forcats::fct_reorder()</code> within <code>aes()</code> to reorder the study variable based on the <code>median_EPCAM</code> values.</p>
<p>The <code>median_EPCAM</code> values determine the order of the bars from lowest to highest.</p>
<p>Now, the bars are reordered based on the median EPCAM levels, allowing for easier comparison between different cancer study types. This visualization provides insights into which studies exhibit higher or lower median EPCAM expression levels.</p>
</div>
<div id="reversing-the-order" class="section level3" number="9.14.4">
<h3><span class="header-section-number">9.14.4</span> Reversing the Order</h3>
<p>If you want to reverse the order of the bars, you can use the .desc = TRUE argument with <code>forcats::fct_reorder()</code>.</p>
<div class="sourceCode" id="cb449"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb449-1"><a href="introduction-to-the-tidyverse-ecosystem.html#cb449-1" tabindex="-1"></a><span class="co"># Reverse the order of the bars</span></span>
<span id="cb449-2"><a href="introduction-to-the-tidyverse-ecosystem.html#cb449-2" tabindex="-1"></a><span class="fu">ggplot</span>(EPCAM_median, <span class="fu">aes</span>(<span class="at">x =</span> study <span class="sc">%&gt;%</span></span>
<span id="cb449-3"><a href="introduction-to-the-tidyverse-ecosystem.html#cb449-3" tabindex="-1"></a>                           forcats<span class="sc">::</span><span class="fu">fct_reorder</span>(median_EPCAM, <span class="at">.desc =</span> <span class="cn">TRUE</span>), </span>
<span id="cb449-4"><a href="introduction-to-the-tidyverse-ecosystem.html#cb449-4" tabindex="-1"></a>                         <span class="at">y =</span> median_EPCAM)) <span class="sc">+</span></span>
<span id="cb449-5"><a href="introduction-to-the-tidyverse-ecosystem.html#cb449-5" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">&quot;identity&quot;</span>) <span class="sc">+</span></span>
<span id="cb449-6"><a href="introduction-to-the-tidyverse-ecosystem.html#cb449-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb449-7"><a href="introduction-to-the-tidyverse-ecosystem.html#cb449-7" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-239-1.png" width="672" /></p>
<p>By specifying <code>.desc = TRUE</code>, the bars will be reordered in descending order based on the median EPCAM values.</p>
</div>
<div id="note-on-factor-levels" class="section level3" number="9.14.5">
<h3><span class="header-section-number">9.14.5</span> Note on factor levels</h3>
<p>xxxx</p>
</div>
<div id="conclusion-25" class="section level3" number="9.14.6">
<h3><span class="header-section-number">9.14.6</span> Conclusion</h3>
<p>In this lesson, we have learned how to create bar plots to visualize median EPCAM expression levels by different cancer study types. We explored the process of reordering bars based on values, allowing us to better compare and interpret the data. The order of the bar depends on the factor levels of the study and we used fct_reorder to accomplish that. Remember that effective data visualization can greatly enhance your ability to communicate insights and make data-driven decisions.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="fundamental-data-structures-in-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="introduction-to-bioconductor.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/08_Introduction_to_the_tidyverse_ecosystem.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
